
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Metadata &#8212; PySLiM manual</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Migrating from previous versions of pyslim" href="previous_versions.html" />
    <link rel="prev" title="Time units" href="time_units.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">PySLiM manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction.html">
   Introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Using pyslim
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_space.html">
   Vignette: A spatial simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_continuing.html">
   Vignette: Following up with more coalescent simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_coalescent_diversity.html">
   Vignette: Starting with diversity generated by coalescent simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_parallel_phylo.html">
   Vignette: Parallelizing SLiM simulations in a phylogenetic tree
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="time_units.html">
   Time units
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Metadata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="previous_versions.html">
   Migrating from previous versions of pyslim
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  pyslim reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="python_api.html">
   Python API
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Development
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  pyslim 1.0.1
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="_sources/metadata.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/metadata.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/tskit-dev/pyslim"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/tskit-dev/pyslim/issues/new?title=Issue%20on%20page%20%2Fmetadata.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/tskit-dev/pyslim/edit/main/docs/metadata.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/tskit-dev/pyslim/main?urlpath=tree/docs/metadata.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#metadata-tools">
   Metadata tools
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modifying-slim-metadata">
   Modifying SLiM metadata
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#top-level-metadata">
     Top-level metadata
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#modifying-slim-metadata-in-tables">
     Modifying SLiM metadata in tables
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#technical-details">
   Technical details
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#metadata-entries">
     Metadata entries
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Metadata</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#metadata-tools">
   Metadata tools
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modifying-slim-metadata">
   Modifying SLiM metadata
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#top-level-metadata">
     Top-level metadata
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#modifying-slim-metadata-in-tables">
     Modifying SLiM metadata in tables
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#technical-details">
   Technical details
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#metadata-entries">
     Metadata entries
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="metadata">
<span id="sec-metadata"></span><h1>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<span id="sec-metdata-overview"></span><h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>SLiM puts SLiM-specific information into the <em>metadata</em> for the tree sequence,
as well as for each populations, individuals, nodes and mutations.
Here is a quick reference to what information is available:
see the SLiM manual for the more technical writeup.
A good way to get a generic metadata example is with <a class="reference internal" href="#pyslim.default_slim_metadata" title="pyslim.default_slim_metadata"><code class="xref py py-func docutils literal notranslate"><span class="pre">default_slim_metadata()</span></code></a>.</p>
<p><strong>Top-level:</strong>
If <code class="docutils literal notranslate"><span class="pre">ts</span></code> is your tree sequence, then <code class="docutils literal notranslate"><span class="pre">ts.metadata</span></code> is a dict,
and <code class="docutils literal notranslate"><span class="pre">ts.metadata[&quot;SLiM&quot;]</span></code> contains information about the simulation:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">file_version</span></code>: the version of the SLiM tree sequence file format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tick</span></code>: the value of <code class="docutils literal notranslate"><span class="pre">community.tick</span></code> within SLiM when the file was written out</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cycle</span></code>: the value of <code class="docutils literal notranslate"><span class="pre">sim.cycle</span></code> within SLiM when the file was written out</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">model_type</span></code>: either <code class="docutils literal notranslate"><span class="pre">&quot;WF&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;nonWF&quot;</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">nucleotide_based</span></code>: whether this is a nucleotide-based simulation</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">separate_sexes</span></code>: whether the simulation has separate sexes or not</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spatial_dimensionality</span></code>: for instance, <code class="docutils literal notranslate"><span class="pre">&quot;&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;x&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;xy&quot;</span></code> (etcetera)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spatial_periodicity</span></code>: whether space wraps around in some directions (same format as dimensionality)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">stage</span></code>: the <em>stage</em> of the life cycle at which the file was written out (either <code class="docutils literal notranslate"><span class="pre">&quot;first&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;early&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;late&quot;</span></code>)</p></li>
</ul>
<p><strong>Populations:</strong>
Information about each SLiM-produced population is written to metatadata.
The format uses JSON and is extensible, so other keys may be present
and some keys may be missing (for instance, there are no spatial bounds
in a nonspatial simulation). The metadata may be <code class="docutils literal notranslate"><span class="pre">None</span></code> for populations
that SLiM did not use. The keys that SLiM uses are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">slim_id</span></code>: the ID of this population in SLiM</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">name</span></code>: the name of the population (by default, <code class="docutils literal notranslate"><span class="pre">p0</span></code>, <code class="docutils literal notranslate"><span class="pre">p1</span></code>, etcetera)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">description</span></code>: a string describing the population</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selfing_fraction</span></code>, <code class="docutils literal notranslate"><span class="pre">female_cloning_fraction</span></code>, <code class="docutils literal notranslate"><span class="pre">male_cloning_fraction</span></code>, and <code class="docutils literal notranslate"><span class="pre">sex_ratio</span></code>: only present when applicable (e.g., in WF simulations)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bounds_x0</span></code>, <code class="docutils literal notranslate"><span class="pre">bounds_x1</span></code>, <code class="docutils literal notranslate"><span class="pre">bounds_y0</span></code>, <code class="docutils literal notranslate"><span class="pre">bounds_y1</span></code>, <code class="docutils literal notranslate"><span class="pre">bounds_z0</span></code>, and <code class="docutils literal notranslate"><span class="pre">bounds_z1</span></code>: the spatial bounds, when applicable</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">migration_records</span></code>: A <em>list</em> of entries decribing migration between populations in a WF model.</p></li>
</ul>
<p><strong>Individuals:</strong>
Each individual produced by SLiM contains the following metadata:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pedigree_id</span></code>: the “pedigree ID”, unique within the SLiM simulation</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pedigree_p1</span></code>, <code class="docutils literal notranslate"><span class="pre">pedigree_p2</span></code>: the pedigree IDs of the individuals’ two
parents (they may be equal in the case of selfing, or <code class="docutils literal notranslate"><span class="pre">-1</span></code> to indicate no
parent, in the case of the initial generation or for cloning)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">age</span></code>: the <code class="docutils literal notranslate"><span class="pre">.age</span></code> property within SLiM at the time the file was written out</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subpopulation</span></code>: the subpopulation within SLiM the individual was in at the time the file was written out</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sex</span></code>: the sex of the individual (either <a class="reference internal" href="python_api.html#pyslim.INDIVIDUAL_TYPE_FEMALE" title="pyslim.INDIVIDUAL_TYPE_FEMALE"><code class="xref py py-data docutils literal notranslate"><span class="pre">INDIVIDUAL_TYPE_FEMALE</span></code></a>, <a class="reference internal" href="python_api.html#pyslim.INDIVIDUAL_TYPE_MALE" title="pyslim.INDIVIDUAL_TYPE_MALE"><code class="xref py py-data docutils literal notranslate"><span class="pre">INDIVIDUAL_TYPE_MALE</span></code></a>, or <a class="reference internal" href="python_api.html#pyslim.INDIVIDUAL_TYPE_HERMAPHRODITE" title="pyslim.INDIVIDUAL_TYPE_HERMAPHRODITE"><code class="xref py py-data docutils literal notranslate"><span class="pre">INDIVIDUAL_TYPE_HERMAPHRODITE</span></code></a>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">flags</span></code>: additional information; currently only recording whether the individual was a “migrant” or not (see the SLiM manual)</p></li>
</ul>
<p><strong>Nodes:</strong>
Each “node” produced by SLiM (i.e., “genome” within SLiM) has:</p>
<ul class="simple">
<li><p>‘slim_id’: the unique ID associated with the genome by SLiM</p></li>
<li><p>‘is_null’: whether the genome is a “null” genome (in which case it isn’t
really there, so shouldn’t have any mutations or relationships in the tree
sequence!)</p></li>
<li><p>‘genome_type’: the ‘type’ of this genome (0 for autosome, 1 for X, 2 for Y)</p></li>
</ul>
<p><strong>Mutations:</strong>
Each mutation’s metadata is a dictionary with a single key, <code class="docutils literal notranslate"><span class="pre">&quot;mutation_list&quot;</span></code>,
whose entry is a <em>list</em> of metadata dictionaries corresponding to the mutations that are “stacked”,
i.e., all present, in all genomes inheriting from this (tskit) mutation.
So, <code class="docutils literal notranslate"><span class="pre">ts.mutation(12).metadata[&quot;mutation_list&quot;]</span></code> is a list, each of whose entries contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">mutation_type</span></code>: the numeric ID of the <code class="docutils literal notranslate"><span class="pre">MutationType</span></code> within SLiM</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selection_coeff</span></code>: the selection coefficient</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subpopulation</span></code>: the numeric ID of the subpopulation the mutation occurred in</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">slim_time</span></code>: the value of <code class="docutils literal notranslate"><span class="pre">community.tick</span></code> when the mutation occurred</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">nucleotide</span></code>: either <code class="docutils literal notranslate"><span class="pre">-1</span></code> if there is no associated nucleotide, or the numeric code for the nucleotide (see <a class="reference internal" href="python_api.html#pyslim.NUCLEOTIDES" title="pyslim.NUCLEOTIDES"><code class="xref py py-data docutils literal notranslate"><span class="pre">NUCLEOTIDES</span></code></a>)</p></li>
</ul>
</div>
<div class="section" id="metadata-tools">
<span id="sec-metadata-tools"></span><h2>Metadata tools<a class="headerlink" href="#metadata-tools" title="Permalink to this headline">¶</a></h2>
<p>The dictionaries describing the schema for these metadata entries
are available in <code class="xref py py-data docutils literal notranslate"><span class="pre">slim_metadata_schemas</span></code>.
Furthermore, this method may be useful in working with metadata:</p>
<dl class="py function">
<dt class="sig sig-object py" id="pyslim.default_slim_metadata">
<span class="sig-prename descclassname"><span class="pre">pyslim.</span></span><span class="sig-name descname"><span class="pre">default_slim_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyslim/slim_metadata.html#default_slim_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyslim.default_slim_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns default metadata of type <code class="docutils literal notranslate"><span class="pre">name</span></code>, where <code class="docutils literal notranslate"><span class="pre">name</span></code> is one of
“tree_sequence”, “edge”, “site”, “mutation”, “mutation_list_entry”,
“node”, “individual”, or “population”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – The type of metadata requested.</p>
</dd>
<dt class="field-even">Rtype dict</dt>
<dd class="field-even"><p></p></dd>
</dl>
</dd></dl>

</div>
<div class="section" id="modifying-slim-metadata">
<h2>Modifying SLiM metadata<a class="headerlink" href="#modifying-slim-metadata" title="Permalink to this headline">¶</a></h2>
<p>For more on working with metadata,
see <a class="reference external" href="https://tskit.dev/tskit/docs/stable/metadata.html#sec-metadata" title="(in Python)"><span class="xref myst">tskit’s metadata documentation</span></a>.</p>
<div class="section" id="top-level-metadata">
<h3>Top-level metadata<a class="headerlink" href="#top-level-metadata" title="Permalink to this headline">¶</a></h3>
<p>The entries of the top-level metadata dict are <em>read-only</em>.
So, you might think that
<code class="docutils literal notranslate"><span class="pre">tables.metadata[&quot;SLiM&quot;][&quot;model_type&quot;]</span> <span class="pre">=</span> <span class="pre">&quot;nonWF&quot;</span></code>
would switch the model type,
but this in fact (silently) does nothing. To modify the top-level metadata,
we must (a) work with tables (as tree sequences are immutable, and (b)
extract the metadata dict, modify the dict, and copy it back in.
Instead, you should do</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">md</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
<span class="n">md</span><span class="p">[</span><span class="s2">&quot;SLiM&quot;</span><span class="p">][</span><span class="s2">&quot;model_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;nonWF&quot;</span>
<span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">md</span>
</pre></div>
</div>
</div>
</div>
<p>Modifying the top-level metadata
could be used to set spatial bounds on an annotated msprime simulation, for instance.
(This is recorded in the population metadata.)</p>
</div>
<div class="section" id="modifying-slim-metadata-in-tables">
<h3>Modifying SLiM metadata in tables<a class="headerlink" href="#modifying-slim-metadata-in-tables" title="Permalink to this headline">¶</a></h3>
<p>To modify the metadata that <code class="docutils literal notranslate"><span class="pre">pyslim</span></code> has introduced into
the tree sequence produced by a coalescent simulation,
or the metadata in a SLiM-produced tree sequence,
we need to edit the TableCollection that forms the editable data behind the tree sequence.
For instance, to set the ages of the individuals in the tree sequence to random numbers between 1 and 4,
we will extract a copy of the underlying tables, clear it,
and then iterate over the individuals in the tree sequence,
as we go re-inserting them into the tables
after replacing their metadata with a modified version:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
<span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">individuals</span><span class="p">():</span>
    <span class="n">md</span> <span class="o">=</span> <span class="n">ind</span><span class="o">.</span><span class="n">metadata</span>
    <span class="n">md</span><span class="p">[</span><span class="s2">&quot;age&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">md</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">mod_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>

<span class="c1"># check that it worked:</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First ten ages:&quot;</span><span class="p">,</span> <span class="p">[</span><span class="n">mod_ts</span><span class="o">.</span><span class="n">individual</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;age&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)])</span>
<span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">mod_ts</span><span class="o">.</span><span class="n">individuals</span><span class="p">():</span>
    <span class="k">assert</span> <span class="n">ind</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;age&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>

<span class="c1"># save out the tree sequence</span>
<span class="n">mod_ts</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="s2">&quot;modified_ts.trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>First ten ages: [3, 1, 1, 3, 4, 4, 3, 2, 2, 3]
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="technical-details">
<h2>Technical details<a class="headerlink" href="#technical-details" title="Permalink to this headline">¶</a></h2>
<div class="section" id="metadata-entries">
<h3>Metadata entries<a class="headerlink" href="#metadata-entries" title="Permalink to this headline">¶</a></h3>
<p>SLiM records additional information in the metadata columns of Individual, Node, and Mutation tables,
in a binary format using the python <code class="docutils literal notranslate"><span class="pre">struct</span></code> module.
See <a class="reference external" href="https://tskit.dev/tskit/docs/stable/metadata.html#sec-metadata" title="(in Python)"><span class="xref myst">tskit’s metadata documentation</span></a>
for details on how this works.
Nothing besides this binary information can be stored in the metadata of these tables if the tree sequence is to be used by SLiM,
and so when <code class="docutils literal notranslate"><span class="pre">pyslim</span></code> annotates an existing tree sequence, anything in those columns is overwritten.
Population metadata is stored as JSON, however, which is more flexible.
For more detailed documentation on the contents and format of the metadata, see the SLiM manual.</p>
<p>Of particular note is that <em>nodes</em> and <em>populations</em> may have empty metadata.
SLiM will not use the metadata of nodes that are not associated with alive individuals,
so this can safely be omitted (and makes recapitation easier).
And, populations not used by SLiM will have empty metadata.
All remaining metadata are required (besides edges and sites, whose metadata is not used at all).</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="time_units.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Time units</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="previous_versions.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Migrating from previous versions of pyslim</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Tskit Developers<br/>
    
        &copy; Copyright 2021.<br/>
      <div class="extra_footer">
        pyslim 1.0.1
      </div>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>