
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vignette: Starting with diversity generated by coalescent simulation &#8212; PySLiM manual</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Vignette: Parallelizing SLiM simulations in a phylogenetic tree" href="vignette_parallel_phylo.html" />
    <link rel="prev" title="Vignette: Following up with more coalescent simulation" href="vignette_continuing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">PySLiM manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction.html">
   Introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Using pyslim
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_space.html">
   Vignette: A spatial simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_continuing.html">
   Vignette: Following up with more coalescent simulation
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Vignette: Starting with diversity generated by coalescent simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_parallel_phylo.html">
   Vignette: Parallelizing SLiM simulations in a phylogenetic tree
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="time_units.html">
   Time units
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="metadata.html">
   Metadata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="previous_versions.html">
   Migrating from previous versions of pyslim
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  pyslim reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="python_api.html">
   Python API
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Development
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  pyslim 1.0.1
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="_sources/vignette_coalescent_diversity.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/vignette_coalescent_diversity.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/tskit-dev/pyslim"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/tskit-dev/pyslim/issues/new?title=Issue%20on%20page%20%2Fvignette_coalescent_diversity.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/tskit-dev/pyslim/edit/main/docs/vignette_coalescent_diversity.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/tskit-dev/pyslim/main?urlpath=tree/docs/vignette_coalescent_diversity.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutation-and-recombination-maps">
   Mutation and recombination maps
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-coalescent-simulation">
   The coalescent simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#annotate-everyone">
   Annotate everyone
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#add-slim-mutations">
   Add SLiM mutations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-into-slim">
   Load into SLiM
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analyze-results">
   Analyze results
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#add-neutral-mutations">
   Add neutral mutations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#diversity-along-the-genome">
   Diversity along the genome
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Vignette: Starting with diversity generated by coalescent simulation</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutation-and-recombination-maps">
   Mutation and recombination maps
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-coalescent-simulation">
   The coalescent simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#annotate-everyone">
   Annotate everyone
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#add-slim-mutations">
   Add SLiM mutations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-into-slim">
   Load into SLiM
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#analyze-results">
   Analyze results
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#add-neutral-mutations">
   Add neutral mutations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#diversity-along-the-genome">
   Diversity along the genome
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="vignette-starting-with-diversity-generated-by-coalescent-simulation">
<span id="sec-vignette-coalescent-diversity"></span><h1>Vignette: Starting with diversity generated by coalescent simulation<a class="headerlink" href="#vignette-starting-with-diversity-generated-by-coalescent-simulation" title="Permalink to this headline">¶</a></h1>
<p>This vignette shows how to simulate history with msprime,
add SLiM mutations to it and assign them selection coefficients,
then run a SLiM simulation using this as a starting point.</p>
<p>Simulations of large populations with selection can be costly,
especially if we need to run a lengthy “burn-in” period to get the
genetic diversity for selection to act on.
Sometimes, the precise form of the burn-in is not important,
and so a <em>neutral</em> burn-in is acceptable - allowing us to use msprime.
For instance, suppose we’d like to simulate a lab experiment
in which we take high-diversity organisms from the wild and subject them
to selection for a few dozen generations.
Genetic diversity in the wild is certainly not neutral, but then again,
we don’t quite know what it <em>does</em> look like, so a coalescent simulation
would be better than nothing. The key attribute of reality we’d like to
approximate is the joint distribution of allele frequencies and effect
sizes. If the alleles affect a trait under stabilizing selection, we’d
expect a negative correlation between the two. On the other hand,
if the trait we’re selecting on in the lab is not under strong selection
in the wild, there might not be much of a relationship.
This is a simple example, to show how to do this:
the trait under selection is just fitness,
and there is no relationship between allele frequency and effect size.</p>
<p>The steps will be:</p>
<ol class="simple">
<li><p>Run a coalescent simulation with msprime.</p></li>
<li><p>Add SLiM metadata to the nodes, individuals, and populations.</p></li>
<li><p>Add SLiM mutations with msprime,
and edit the mutation metadata to assign selection coefficients.</p></li>
<li><p>Run the SLiM portion of the simulation.</p></li>
<li><p>Do some descriptive analysis of the results of selection.</p></li>
<li><p>Add neutral mutations to the tree sequence.</p></li>
<li><p>Do some descriptive analysis of genetic diversity along the genome.</p></li>
</ol>
<div class="section" id="mutation-and-recombination-maps">
<h2>Mutation and recombination maps<a class="headerlink" href="#mutation-and-recombination-maps" title="Permalink to this headline">¶</a></h2>
<p>In this model,
we’ll also demonstrate how to modulate the mutation and recombination rate
along the genome. We’ll do a simple example: a 9MB genome
with three equally-sized domains.
The first and last third of the genome will have high recombination
(5e-8 per generation per bp),
and the middle third will have low recombination
(0.5e-8 per generation per bp).
The mutation rate will be constant (3e-8 per generation per bp),
but a lower proportion of mutations in the middle region are under selection:
on the outside thirds, 1% of mutations are beneficial with a selection coefficient
drawn from an Exponential distribution,
and in the middle third, only 0.1% are (but with the same distribution of selection coefficients).
This implies that the mutation rate <em>of beneficial mutations</em> on the ends
is <span class="math notranslate nohighlight">\(0.01 \times 30^{-8}\)</span> = 3e-10, and in the middle is 3e-11.
We’ll simulate these first, and only add the neutral mutations after everything else,
at rates 2.97e-8 on the ends, and 2.997e-8 in the middle.</p>
</div>
<div class="section" id="the-coalescent-simulation">
<h2>The coalescent simulation<a class="headerlink" href="#the-coalescent-simulation" title="Permalink to this headline">¶</a></h2>
<p>First, we’ll use msprime
to simulate the demographic history of 2,000 smallish chromosomes
(<span class="math notranslate nohighlight">\(N_e = 1,0000\)</span> diploids)
from a population of 10,000 diploids total.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">breaks</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3000000</span><span class="p">,</span> <span class="mi">6000000</span><span class="p">,</span> <span class="mi">9000000</span><span class="p">]</span>
<span class="n">recomb_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="p">(</span>
  <span class="n">position</span> <span class="o">=</span> <span class="n">breaks</span><span class="p">,</span>
  <span class="n">rate</span> <span class="o">=</span> <span class="p">[</span><span class="mf">5e-8</span><span class="p">,</span> <span class="mf">0.5e-8</span><span class="p">,</span> <span class="mf">5e-8</span><span class="p">])</span>
<span class="n">demog_model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demog_model</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">initial_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
<span class="n">ots</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
        <span class="n">samples</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
        <span class="n">demography</span><span class="o">=</span><span class="n">demog_model</span><span class="p">,</span>
        <span class="n">random_seed</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recomb_map</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="annotate-everyone">
<h2>Annotate everyone<a class="headerlink" href="#annotate-everyone" title="Permalink to this headline">¶</a></h2>
<p>At this point, we have genealogical information: individuals,
nodes (chromosomes), and relationships between them,
but no genetic diversity; no mutations.
First, we’ll add SLiM metadata to all of these things,
a procedure we call “annotating”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ots</span> <span class="o">=</span> <span class="n">pyslim</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">ots</span><span class="p">,</span> <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;WF&quot;</span><span class="p">,</span> <span class="n">tick</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">stage</span><span class="o">=</span><span class="s2">&quot;late&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This method adds default metadata to everything that needs it:
in this case, all individuals, all nodes that are part of alive individuals,
and all populations referenced by nodes.
These default values are returned by <code class="xref py py-func docutils literal notranslate"><span class="pre">slim_default_metadata()</span></code>
(e.g., all individuals are hermaphrodite, all chromosomes are autosomal);
see <a class="reference internal" href="python_api.html#pyslim.annotate" title="pyslim.annotate"><code class="xref py py-func docutils literal notranslate"><span class="pre">annotate()</span></code></a> for more information.</p>
</div>
<div class="section" id="add-slim-mutations">
<h2>Add SLiM mutations<a class="headerlink" href="#add-slim-mutations" title="Permalink to this headline">¶</a></h2>
<p>Next, we’re going to use the <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.SLiMMutationModel" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.SLiMMutationModel</span></code></a> to add mutations
to the tree sequence. These will carry SLiM metadata, but this metadata
will say that the mutations are neutral. So, we’ll then need to modify their metadata
after the fact to have selection coefficients drawn from some distribution.
(Remember, our motivation here is that we are using msprime to obtain
a plausible level of functional standing genetic variation
on which selection can act:
imagine that the mutations were (nearly) neutral in the wild,
but then became subject to strong selection in the lab, for some reason.)
We’ll want this to be as if we’d done it in a burn-in script in SLiM:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>   <span class="n">initialize</span><span class="p">()</span> <span class="p">{</span>
      <span class="o">...</span>
      <span class="n">initializeMutationType</span><span class="p">(</span><span class="s2">&quot;m2&quot;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;e&quot;</span><span class="p">,</span> <span class="mf">0.04</span><span class="p">);</span>
      <span class="n">initializeMutationRate</span><span class="p">(</span><span class="mf">3e-10</span><span class="p">);</span>
   <span class="p">}</span>

   <span class="n">fitness</span><span class="p">(</span><span class="n">m2</span><span class="p">)</span> <span class="p">{</span>
       <span class="k">return</span> <span class="mf">1.0</span><span class="p">;</span>
   <span class="p">}</span>
</pre></div>
</div>
<p>In other words, we’d like to pull selection coefficients from an exponential distribution
with mean 0.04 (but, of course, this is a coalescent simulation, so the
dynamics of the mutations up until this point have been neutral).
Note that the dominance coefficient is <em>not</em> stored in the tree sequence:
it gets set in the SLiM recipe
because it’s a property of the mutation type, not of individual mutations, in SLiM.
Here’s how to add SLiM mutations with msprime:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mut_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="p">(</span>
           <span class="n">position</span><span class="o">=</span><span class="n">breaks</span><span class="p">,</span>
           <span class="n">rate</span><span class="o">=</span><span class="p">[</span><span class="mf">0.03e-8</span><span class="p">,</span> <span class="mf">0.003e-8</span><span class="p">,</span> <span class="mf">0.03e-8</span><span class="p">])</span>
<span class="n">mut_model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">SLiMMutationModel</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ots</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span>
            <span class="n">ots</span><span class="p">,</span>
            <span class="n">rate</span><span class="o">=</span><span class="n">mut_map</span><span class="p">,</span>
            <span class="n">model</span><span class="o">=</span><span class="n">mut_model</span><span class="p">,</span>
            <span class="n">keep</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
            <span class="n">random_seed</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The tree sequence now has </span><span class="si">{</span><span class="n">ots</span><span class="o">.</span><span class="n">num_mutations</span><span class="si">}</span><span class="s2"> mutations, at &quot;</span>
      <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ots</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> distinct sites.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The tree sequence now has 609 mutations, at 609 distinct sites.
</pre></div>
</div>
</div>
</div>
<p>Note the <code class="docutils literal notranslate"><span class="pre">type=2</span></code> argument to <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.SLiMMutationModel" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.SLiMMutationModel</span></code></a>:
this means the mutations will be of type “m2” in SLiM (and, so you must
initialize that mutation type in the recipe that loads this tree sequence in).</p>
<p>Now, we’ll assign selection coefficients.
Recall that to accomodate mutation stacking in SLiM,
a mutation metadata entry is in fact a <em>list</em> of metadata entries,
one for each of the SLiM mutations that are stacked at this position.
The SLiM IDs of these mutations are available (in the same order)
as a comma-separated list of integers in the derived state of the mutation.
So, in case some SLiM mutations appear in more than one mutation
in the tree sequence, we will build a map from SLiM ID to selection coefficient:
<code class="docutils literal notranslate"><span class="pre">mut_map[k]</span></code> will give the selection coefficient of the SLiM mutation with
SLiM mutation ID <code class="docutils literal notranslate"><span class="pre">k</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">ots</span><span class="o">.</span><span class="n">tables</span>
<span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
<span class="n">mut_map</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">ots</span><span class="o">.</span><span class="n">mutations</span><span class="p">():</span>
  <span class="n">md_list</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;mutation_list&quot;</span><span class="p">]</span>
  <span class="n">slim_ids</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">derived_state</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
  <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">slim_ids</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">md_list</span><span class="p">)</span>
  <span class="k">for</span> <span class="n">sid</span><span class="p">,</span> <span class="n">md</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">slim_ids</span><span class="p">,</span> <span class="n">md_list</span><span class="p">):</span>
     <span class="k">if</span> <span class="n">sid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mut_map</span><span class="p">:</span>
        <span class="n">mut_map</span><span class="p">[</span><span class="n">sid</span><span class="p">]</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">exponential</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="mf">0.04</span><span class="p">)</span>
     <span class="n">md</span><span class="p">[</span><span class="s2">&quot;selection_coeff&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mut_map</span><span class="p">[</span><span class="n">sid</span><span class="p">]</span>
  <span class="n">_</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
          <span class="n">m</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;mutation_list&quot;</span><span class="p">:</span> <span class="n">md_list</span><span class="p">})</span>
  <span class="p">)</span>

<span class="c1"># check we didn&#39;t mess anything up</span>
<span class="k">assert</span> <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">num_rows</span> <span class="o">==</span> <span class="n">ots</span><span class="o">.</span><span class="n">num_mutations</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The selection coefficients range from </span><span class="si">{</span><span class="nb">min</span><span class="p">(</span><span class="n">mut_map</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="si">:</span><span class="s2">0.2e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;to </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span><span class="n">mut_map</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="si">:</span><span class="s2">0.2e</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The selection coefficients range from 3.68e-06
to 3.51e-01.
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="load-into-slim">
<h2>Load into SLiM<a class="headerlink" href="#load-into-slim" title="Permalink to this headline">¶</a></h2>
<p>Before loading the tree sequence into SLiM, we should check the top-level metadata.
We can see this with <code class="docutils literal notranslate"><span class="pre">tables.metadata</span></code>:</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{
 &#39;SLiM&#39;: {
     &#39;cycle&#39;: 1,
     &#39;description&#39;: &#39;&#39;,
     &#39;file_version&#39;: &#39;0.8&#39;,
     &#39;model_type&#39;: &#39;WF&#39;,
     &#39;name&#39;: &#39;&#39;,
     &#39;nucleotide_based&#39;: False,
     &#39;separate_sexes&#39;: False,
     &#39;spatial_dimensionality&#39;: &#39;&#39;,
     &#39;spatial_periodicity&#39;: &#39;&#39;,
     &#39;stage&#39;: &#39;late&#39;,
     &#39;tick&#39;: 1
    }
}
</pre></div>
</div>
</div>
</div>
<p>We should edit this to match our planned slimulation</p>
<ul class="simple">
<li><p>particularly the <code class="docutils literal notranslate"><span class="pre">model_type</span></code> (WF or nonWF) and the <code class="docutils literal notranslate"><span class="pre">tick</span></code>.
The <code class="docutils literal notranslate"><span class="pre">tick</span></code> tells SLiM what value to set the tick counter to
once this tree sequence is loaded. In principle, it can be set to anything,
independently of the times in the tree sequence,
because the times in the tree sequence are measured in units of
“time before the end”; and the <code class="docutils literal notranslate"><span class="pre">tick</span></code> that gets
passed to SLiM sets what that “end time” is, in SLiM’s time.
However, if you change this, the <code class="docutils literal notranslate"><span class="pre">slim_time</span></code> attributes in mutation metadata
will not be accurate. This is harmless, unless you do something with mutations’
times yourself.</p></li>
</ul>
<p>The <code class="docutils literal notranslate"><span class="pre">model_type</span></code> is already Wright-Fisher, but just to demonstrate how to
edit the metadata, let’s make sure,
and then we’ll write the tree sequence to a file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts_metadata</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
<span class="n">ts_metadata</span><span class="p">[</span><span class="s2">&quot;SLiM&quot;</span><span class="p">][</span><span class="s2">&quot;model_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;WF&quot;</span>
<span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">ts_metadata</span>
<span class="n">ots</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">ots</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="s2">&quot;vignette_annotated.init.trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now for the SLiM recipe.
This simply continues selected mutations as before
(with mutation rate 1e-10 per bp per generation
and the same distribution of fitness effects).
The population size is determined by the number of individuals that were
read in from the tree sequence.
We need to make sure that the genome lengths match,
so we provide that as a constant <code class="docutils literal notranslate"><span class="pre">L</span></code>, that will be provided at run time.
To facilitate later analysis, we’ll also “Remember” the individuals
present at the <em>start</em> of the simulation,
so that they will remain in the tree sequence.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">initialize</span><span class="p">()</span>
<span class="p">{</span>
    <span class="o">//</span> <span class="n">must</span> <span class="n">define</span> <span class="n">L</span>
    <span class="n">initializeSLiMModelType</span><span class="p">(</span><span class="s2">&quot;WF&quot;</span><span class="p">);</span>
    <span class="n">initializeTreeSeq</span><span class="p">();</span>
    <span class="n">initializeMutationRate</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mf">0.03e-8</span><span class="p">,</span> <span class="mf">0.003e-8</span><span class="p">,</span> <span class="mf">0.03e-8</span><span class="p">),</span>
                           <span class="n">c</span><span class="p">(</span><span class="mi">3000000</span><span class="p">,</span> <span class="mi">6000000</span><span class="p">,</span> <span class="mi">9000000</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">);</span>
    <span class="n">initializeMutationType</span><span class="p">(</span><span class="s2">&quot;m1&quot;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">);</span>
    <span class="n">initializeMutationType</span><span class="p">(</span><span class="s2">&quot;m2&quot;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;e&quot;</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">);</span>
    <span class="n">initializeGenomicElementType</span><span class="p">(</span><span class="s2">&quot;g1&quot;</span><span class="p">,</span> <span class="n">m2</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">);</span>
    <span class="n">initializeGenomicElement</span><span class="p">(</span><span class="n">g1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">L</span><span class="o">-</span><span class="mi">1</span><span class="p">);</span>
    <span class="n">initializeRecombinationRate</span><span class="p">(</span><span class="mf">1e-8</span><span class="p">);</span>
<span class="p">}</span>

<span class="mi">1</span> <span class="n">late</span><span class="p">()</span> <span class="p">{</span> 
    <span class="n">sim</span><span class="o">.</span><span class="n">readFromPopulationFile</span><span class="p">(</span><span class="s2">&quot;vignette_annotated.init.trees&quot;</span><span class="p">);</span>
    <span class="n">sim</span><span class="o">.</span><span class="n">treeSeqRememberIndividuals</span><span class="p">(</span><span class="n">p0</span><span class="o">.</span><span class="n">individuals</span><span class="p">);</span>
<span class="p">}</span>

<span class="mi">100</span> <span class="n">late</span><span class="p">()</span> <span class="p">{</span>
    <span class="n">catn</span><span class="p">(</span><span class="s2">&quot;Done.&quot;</span><span class="p">);</span>
    <span class="n">sim</span><span class="o">.</span><span class="n">treeSeqOutput</span><span class="p">(</span><span class="s2">&quot;vignette_annotated.trees&quot;</span><span class="p">);</span>
    <span class="n">sim</span><span class="o">.</span><span class="n">simulationFinished</span><span class="p">();</span>
<span class="p">}</span>

</pre></div>
</div>
<p>Note that the simulation only has selected mutations (of type <code class="docutils literal notranslate"><span class="pre">m2</span></code>),
but as we’ll add in type <code class="docutils literal notranslate"><span class="pre">m1</span></code> mutations later,
we’ve declared them in the recipe as a placeholder.</p>
<p>We could run this on the command line as
<code class="docutils literal notranslate"><span class="pre">slim</span> <span class="pre">-d</span> <span class="pre">L=100000000</span> <span class="pre">reload_annotated.slim</span></code>,
but this time we’ll stay within python,
and obtain the sequence length programatically:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">subprocess</span>
<span class="n">msg</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span>
     <span class="p">[</span><span class="s2">&quot;slim&quot;</span><span class="p">,</span> <span class="s2">&quot;-d&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;L=</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">ots</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
      <span class="s2">&quot;-s&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;reload_annotated.slim&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">msg</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>// Initial random seed:
5

// RunInitializeCallbacks():
initializeSLiMModelType(modelType = &#39;WF&#39;);
initializeTreeSeq();
initializeMutationRate(c(3e-10, 3e-11, 3e-10), c(2999999, 5999999, 8999999));
initializeMutationType(1, 0.5, &quot;f&quot;, 0);
initializeMutationType(2, 0.5, &quot;e&quot;, 0.01);
initializeGenomicElementType(1, m2, 1);
initializeGenomicElement(g1, 0, 8999998);
initializeRecombinationRate(1e-08);

// Starting run at tick &lt;start&gt;:
1 

Done.
</pre></div>
</div>
</div>
</div>
<p>This runs quickly, since it’s only 100 generations.</p>
</div>
<div class="section" id="analyze-results">
<h2>Analyze results<a class="headerlink" href="#analyze-results" title="Permalink to this headline">¶</a></h2>
<p>First, let’s look at what mutations are present.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;vignette_annotated.trees&quot;</span><span class="p">)</span>
<span class="n">num_stacked</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;mutation_list&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">()])</span>
<span class="n">init_time</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;tick&#39;</span><span class="p">]</span>
<span class="n">old_mut</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">m</span><span class="o">.</span><span class="n">time</span> <span class="o">&gt;</span> <span class="n">init_time</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">-</span> <span class="mf">1e-12</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">()])</span>
<span class="k">assert</span> <span class="nb">sum</span><span class="p">(</span><span class="n">old_mut</span><span class="p">)</span> <span class="o">==</span> <span class="n">ots</span><span class="o">.</span><span class="n">num_mutations</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_mutations</span><span class="si">}</span><span class="s2"> present at </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> distinct sites.&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Of these, </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">num_stacked</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span><span class="si">}</span><span class="s2"> have more than one stacked mutation,&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;and </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">old_mut</span><span class="p">)</span><span class="si">}</span><span class="s2"> were produced by msprime.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 632 present at 632 distinct sites.
Of these, 0 have more than one stacked mutation,
and 609 were produced by msprime.
</pre></div>
</div>
</div>
</div>
<p>Most of the mutations were present as initial diversity,
but a few were added during the course of the simulation.
Along the way we did a consistency check, that the number of “old” mutations
matches the number of mutations we had in the tree sequence we loaded into SLiM.
Since we ran SLiM for 100 time steps, but loaded the tree sequence in during the <code class="docutils literal notranslate"><span class="pre">late()</span></code>
stage of time step 1, the “old” mutations are those
from at least 99 units of time ago
(and the 1e-12 is necessary for floating-point error).</p>
<p>A simple thing to look at next is: how did the selected mutations
change in frequency? We can do this thanks to our having
Remembered the first generation.
First, we’ll compute all allele frequencies
among both the first generation and the final generation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">times</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">individuals_time</span><span class="p">))</span>
<span class="n">times</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The times ago at which individuals in the tree sequence were born:&quot;</span><span class="p">,</span> <span class="n">times</span><span class="p">)</span>
<span class="c1"># The times ago at which individuals in the tree sequence were born: [0.0, 100.0]</span>
<span class="n">nodes_by_time</span> <span class="o">=</span> <span class="p">[</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">times</span><span class="p">]</span>

<span class="n">num_nodes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">nodes_by_time</span><span class="p">])</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">sample_count_stat</span><span class="p">(</span><span class="n">nodes_by_time</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">/</span><span class="n">num_nodes</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="s1">&#39;sites&#39;</span><span class="p">,</span>
        <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">sum</span><span class="p">([</span><span class="nb">sum</span><span class="p">([</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;selection_coeff&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">md</span> <span class="ow">in</span> <span class="n">m</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;mutation_list&quot;</span><span class="p">]])</span>
                  <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">])</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">sites</span><span class="p">()])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The times ago at which individuals in the tree sequence were born: [0.0, 99.0]
</pre></div>
</div>
</div>
</div>
<p>To do this, we used the <code class="docutils literal notranslate"><span class="pre">time=t</span></code> argument to :meth:<code class="docutils literal notranslate"><span class="pre">tskit.TreeSequence.samples</span></code>
to find the nodes alive at each of the two times (0 and 100 generations ago);
then computed an array <code class="docutils literal notranslate"><span class="pre">p</span></code> of allele frequencies, with one row per site,
the first column giving the frequency among the initial generation,
and the second giving the frequency at the end.
We also pull out <code class="docutils literal notranslate"><span class="pre">s</span></code>, the selection coefficients.
This last bit is a bit complex because each site can have more than one mutation,
and each tree sequence mutation can represent more than one SLiM mutation.
And, the way we’ve dealt with this is a bit of a hack,
so let’s look at that site with multiple mutations:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">()):</span>
  <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
     <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site </span><span class="si">{</span><span class="n">j</span><span class="si">}</span><span class="s2"> has </span><span class="si">{</span><span class="n">num_stacked</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="si">}</span><span class="s2"> stacked mutations, &quot;</span>
           <span class="sa">f</span><span class="s2">&quot;with total derived allele frequency </span><span class="si">{</span><span class="n">p</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
           <span class="sa">f</span><span class="s2">&quot;and sum of selection coefficients </span><span class="si">{</span><span class="n">s</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
     <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The allele frequencies are:&quot;</span><span class="p">)</span>
     <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">alleles</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  &#39;</span><span class="si">{</span><span class="n">a</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="nb">sum</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">genotypes</span> <span class="o">==</span> <span class="n">k</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
     <span class="nb">print</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>There were two mutations at this site, both before the SLiM portion of the simulation
started. One happened on the background of the other,
and no genomes either today or in the initial generation carry the first allele in isolation.
Their effects combine in SLiM, so treating this as a single allele is correct.</p>
<p>Now, we’ll plot the initial and final allele frequencies,
with point size and color determined by the selection coefficient:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">dp</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">p</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">p</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="o">*</span><span class="mi">800</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;frequencies&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;initial allele frequency&quot;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;final allele frequency&quot;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">dp</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;selection coefficient&#39;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/vignette_coalescent_diversity_23_0.png" src="_images/vignette_coalescent_diversity_23_0.png" />
</div>
</div>
<p>Unsurprisingly, mutations that had a large change in allele frequency seem
to be biased towards ones with higher selection coefficients,
and those that were initially present at moderate frequency but were lost
are biased towards smaller selection coefficients.</p>
</div>
<div class="section" id="add-neutral-mutations">
<h2>Add neutral mutations<a class="headerlink" href="#add-neutral-mutations" title="Permalink to this headline">¶</a></h2>
<p>In real data, of course, we don’t get to observe selection coefficients.
We haven’t added in neutral mutations until this point for efficiency -
they are just bookkeeping, and do not affect the course of the simulation
in any way. For this reason, we can add them in after the fact, in a way
that is exactly equivalent to having kept track of them as we went along.</p>
<p>Recall that out of an overall mutation rate of 3e-8,
we wanted 99% of the mutations to be neutral on the ends of the chromosome,
and 99.9% to be neutral in the middle.
So, we’ll now add mutations at these rates,
using the same model of mutation as before.
The code is nearly the same as before,
with a few changes.
We’ve changed the <code class="docutils literal notranslate"><span class="pre">type</span></code> of the mutations
(so that neutral mutations will show up in SLiM as m1,
while selected mutations above were m2),
and we’ve asked these mutations to have SLiM mutation IDs
beginning at the ID where the previous mutations left off.
(This would be important were we to read this tree sequence
back in to SLiM; mutation IDs must be unique.)
And, importantly, we’ve added <code class="docutils literal notranslate"><span class="pre">keep=True</span></code> so that existing mutations
are not discarded.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">neutral_mut_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="p">(</span>
           <span class="n">position</span><span class="o">=</span><span class="n">breaks</span><span class="p">,</span>
           <span class="n">rate</span><span class="o">=</span><span class="p">[</span><span class="mf">2.97e-8</span><span class="p">,</span> <span class="mf">2.997e-8</span><span class="p">,</span> <span class="mf">2.97e-8</span><span class="p">])</span>
<span class="n">next_id</span> <span class="o">=</span> <span class="n">pyslim</span><span class="o">.</span><span class="n">next_slim_mutation_id</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
<span class="n">neutral_mut_model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">SLiMMutationModel</span><span class="p">(</span>
                                <span class="nb">type</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                <span class="n">next_id</span><span class="o">=</span><span class="n">next_id</span><span class="p">)</span>
<span class="n">mts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span>
                <span class="n">ts</span><span class="p">,</span>
                <span class="n">rate</span><span class="o">=</span><span class="n">neutral_mut_map</span><span class="p">,</span>
                <span class="n">model</span><span class="o">=</span><span class="n">neutral_mut_model</span><span class="p">,</span>
                <span class="n">keep</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                <span class="n">random_seed</span><span class="o">=</span><span class="mi">35</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The tree sequence now has </span><span class="si">{</span><span class="n">mts</span><span class="o">.</span><span class="n">num_mutations</span><span class="si">}</span><span class="s2"> mutations,&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;at </span><span class="si">{</span><span class="n">mts</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> distinct sites.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The tree sequence now has 91218 mutations,
at 90757 distinct sites.
</pre></div>
</div>
</div>
</div>
<p>We’ve now got a lot more mutations!
And, we’ve got a lot more sites with multiple mutations:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">num_alleles</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">mts</span><span class="o">.</span><span class="n">sites</span><span class="p">()])</span>
<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">num_alleles</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
  <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="nb">sum</span><span class="p">(</span><span class="n">num_alleles</span> <span class="o">==</span> <span class="n">k</span><span class="p">)</span><span class="si">}</span><span class="s2"> sites with </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> distinct alleles.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 90297 sites with 1 distinct alleles.
There are 459 sites with 2 distinct alleles.
There are 1 sites with 3 distinct alleles.
</pre></div>
</div>
</div>
</div>
<p>To get a nice a picture of what’s happened,
we’ll pull out a tree that had a lot of mutations on it,
and print a picture of it, with mutations labeled by their type:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">mts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
 <span class="n">mt</span> <span class="o">=</span> <span class="p">[</span><span class="nb">max</span><span class="p">([</span><span class="n">u</span><span class="p">[</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">m</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;mutation_list&#39;</span><span class="p">]])</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">mutations</span><span class="p">()]</span>
 <span class="k">if</span> <span class="n">t</span><span class="o">.</span><span class="n">num_mutations</span> <span class="o">&gt;</span> <span class="mi">12</span><span class="p">:</span>
   <span class="k">break</span>

<span class="n">ml</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mtype</span><span class="p">)</span> <span class="k">for</span> <span class="n">mtype</span><span class="p">,</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">mutations</span><span class="p">())}</span>
<span class="n">SVG</span><span class="p">(</span>
    <span class="n">t</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">mutation_labels</span><span class="o">=</span><span class="n">ml</span><span class="p">,</span>
               <span class="n">node_labels</span><span class="o">=</span><span class="p">{},</span>
               <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">300</span><span class="p">))</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/vignette_coalescent_diversity_29_0.svg" src="_images/vignette_coalescent_diversity_29_0.svg" /></div>
</div>
<p>On this tree each mutation is marked by a red “x”, and labeled with its mutation type:
either “1”, for newly added mutations, or “2”, for selected mutations present during the SLiM portion.
(Note: this is a large tree, with 68,211 nodes!
But as usual, the main structure is visible
because most nodes coalesce very recently.)</p>
<p>OK, but how exactly is this working?
Can a neutral mutation be added to a site that previously had a selected mutation?
The short answer is: yes, and new alleles stack on top of
existing alleles, but existing alleles replace new alleles.
This is equivalent to including them as the simulation went along,
by the additivity property of Poisson mutations:
it turns out that the following two ways of generating mutations along the genome
are equivalent: either
(a) placing a random Poisson number with mean <span class="math notranslate nohighlight">\(\mu\)</span>,
and randomly choosing each one to be non-neutral with probability 0.01, or
(b) placing random, independent Poisson numbers of neutral and non-neutral mutations
with means <span class="math notranslate nohighlight">\(0.99\mu\)</span> and <span class="math notranslate nohighlight">\(0.01\mu\)</span> respectively.
Since the neutral ones don’t affect the simulation otherwise,
we can add them in afterwards.
Now, when the mutation algorithm in msprime puts down a new mutation
at a site with mutations already existing,
it appends the newly generated SLiM mutation ID to the previous derived state,
and adds the metadata for the new SLiM mutation to the list of metadata
from the previous mutation.
However, it doesn’t modify any existing mutations,
so their derived states (and metadata) are unchanged.
The result is that, from the point of view of SLiM,
neutral (“m1”) mutations “stack” on top of any other mutations (neutral or selected),
while selected (“m2”) mutations stack with each other, but replace any neutral mutations.
This “stacking policy” is not actually exactly implementable in SLiM,
but given that our newly added mutations are meant to be entirely neutral,
seems like a reasonable policy.
If you wanted some other arrangement (e.g., to have m1 stack on top of m2),
you could go through and modify derived states and metadata appropriately.</p>
<p>Let’s check there are any sites with stacked mutations of different types in the simulation.
There is indeed one such site:</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">mts</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
  <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
     <span class="n">types</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">([</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;mutation_type&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">md</span> <span class="ow">in</span> <span class="n">mut</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;mutation_list&quot;</span><span class="p">]])</span>
              <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">]</span>
     <span class="k">if</span> <span class="nb">max</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">len</span><span class="p">,</span> <span class="n">types</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site(id=11755, position=1163202.0, ancestral_state=&#39;&#39;, mutations=[Mutation(id=11823, site=11755, node=55331, derived_state=&#39;115&#39;, parent=-1, metadata={
    &#39;mutation_list&#39;: [{
        &#39;mutation_type&#39;: 2,
        &#39;selection_coeff&#39;: 0.05027877911925316,
        &#39;subpopulation&#39;: -1,
        &#39;slim_time&#39;: -27938,
        &#39;nucleotide&#39;: -1
       }]
   },
time=28038.93833283228, edge=340518), Mutation(id=11824, site=11755, node=19240, derived_state=&#39;115,12687&#39;, parent=11823, metadata={
    &#39;mutation_list&#39;: [{
        &#39;mutation_type&#39;: 2,
        &#39;selection_coeff&#39;: 0.05027877911925316,
        &#39;subpopulation&#39;: -1,
        &#39;slim_time&#39;: -27938,
        &#39;nucleotide&#39;: -1
       },
    {
     &#39;mutation_type&#39;: 1,
     &#39;selection_coeff&#39;: 0.0,
     &#39;subpopulation&#39;: -1,
     &#39;slim_time&#39;: -599,
     &#39;nucleotide&#39;: -1
    }]
   },
time=600.933737340145, edge=151458)], metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>Here, a neutral mutation has been put down on top of a selected mutation,
but stacked, so that any samples inheriting either of these mutations carries
the selected mutation.
For more discussion of how this works, see <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.SLiMMutationModel" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.SLiMMutationModel</span></code></a>.</p>
</div>
<div class="section" id="diversity-along-the-genome">
<h2>Diversity along the genome<a class="headerlink" href="#diversity-along-the-genome" title="Permalink to this headline">¶</a></h2>
<p>Now that we’ve correctly added neutral mutations to the tree sequence,
and lengthily digested what exactly happened,
let’s have a look at the result.
To do this, we’ll compute two standard measures of genetic diversity
in windows along the genome:
nucleotide diverstiy (also called “Tajima’s <span class="math notranslate nohighlight">\(\pi\)</span>” or “mean density of pairwise differences”),
and Tajima’s <span class="math notranslate nohighlight">\(D\)</span> (with no known aliases).
This is easy to do thanks to the (statistics methods in tskit)[https://tskit.dev/tskit/docs/stable/stats.html].</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">mts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span> <span class="mi">21</span><span class="p">)</span>
<span class="n">pi</span> <span class="o">=</span> <span class="n">mts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">mts</span><span class="o">.</span><span class="n">samples</span><span class="p">(),</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
<span class="n">taj_d</span> <span class="o">=</span> <span class="n">mts</span><span class="o">.</span><span class="n">Tajimas_D</span><span class="p">(</span><span class="n">mts</span><span class="o">.</span><span class="n">samples</span><span class="p">(),</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>


<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="n">mids</span> <span class="o">=</span> <span class="n">windows</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">windows</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;chromosome position (bp)&quot;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;pairwise diversity&quot;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">mids</span><span class="p">,</span> <span class="n">pi</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;pairwise diversity&quot;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;chromosome position (bp)&quot;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Tajima&#39;s D&quot;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">mids</span><span class="p">,</span> <span class="n">taj_d</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Tajima&#39;s D&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/vignette_coalescent_diversity_34_0.png" src="_images/vignette_coalescent_diversity_34_0.png" />
</div>
</div>
<p>The two statistics are very similar - perhaps unsurprisingly, because Tajima’s D is calculated using
pairwise diversity, and we have a very large sample size (here, the entire population).
Tajima’s D is negative across the entire genome, as the result of selection.
However, we don’t see a strong difference between the three regions,
despite the stronger action of linked selection on the ends.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="vignette_continuing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Vignette: Following up with more coalescent simulation</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="vignette_parallel_phylo.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Vignette: Parallelizing SLiM simulations in a phylogenetic tree</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Tskit Developers<br/>
    
        &copy; Copyright 2021.<br/>
      <div class="extra_footer">
        pyslim 1.0.1
      </div>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>