{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a0546c1e",
   "metadata": {},
   "source": [
    "```{currentmodule} tskit\n",
    "```\n",
    "\n",
    "(sec_python_api)=\n",
    "\n",
    "# Python API\n",
    "\n",
    "This page documents the full tskit Python API. Brief thematic summaries of common\n",
    "classes and methods are presented first. The {ref}`sec_python_api_reference` section\n",
    "at the end then contains full details which aim to be concise, precise and exhaustive.\n",
    "Note that this may not therefore be the best place to start if you are new\n",
    "to a particular piece of functionality.\n",
    "\n",
    "(sec_python_api_trees_and_tree_sequences)=\n",
    "\n",
    "## Trees and tree sequences\n",
    "\n",
    "The {class}`TreeSequence` class represents a sequence of correlated\n",
    "evolutionary trees along a genome. The {class}`Tree` class represents a\n",
    "single tree in this sequence. These classes are the interfaces used to\n",
    "interact with the trees and mutational information stored in a tree sequence,\n",
    "for example as returned from a simulation or inferred from a set of DNA\n",
    "sequences.\n",
    "\n",
    "\n",
    "(sec_python_api_tree_sequences)=\n",
    "\n",
    "### {class}`TreeSequence` API\n",
    "\n",
    "\n",
    "(sec_python_api_tree_sequences_properties)=\n",
    "\n",
    "#### General properties\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.time_units\n",
    "  TreeSequence.nbytes\n",
    "  TreeSequence.sequence_length\n",
    "  TreeSequence.max_root_time\n",
    "  TreeSequence.discrete_genome\n",
    "  TreeSequence.discrete_time\n",
    "  TreeSequence.metadata\n",
    "  TreeSequence.metadata_schema\n",
    "  TreeSequence.reference_sequence\n",
    "```\n",
    "\n",
    "#### Efficient table column access\n",
    "\n",
    "The {class}`.TreeSequence` class provides access to underlying numerical\n",
    "data defined in the {ref}`data model<sec_data_model>` in two ways:\n",
    "\n",
    "1. Via the {attr}`.TreeSequence.tables` property and the\n",
    "    {ref}`Tables API<sec_tables_api_accessing_table_data>`\n",
    "2. Via a set of properties on the ``TreeSequence`` class that provide\n",
    "   direct and efficient access to the underlying memory.\n",
    "\n",
    ":::{warning}\n",
    "Accessing table data via {attr}`.TreeSequence.tables` can be very inefficient\n",
    "at the moment because accessing the `.tables` property incurs a **full copy**\n",
    "of the data model. While we intend to implement this as a read-only view\n",
    "in the future, the engineering involved is nontrivial, and so we recommend\n",
    "using the properties listed here like ``ts.nodes_time`` in favour of\n",
    "``ts.tables.nodes.time``.\n",
    "Please see [issue #760](https://github.com/tskit-dev/tskit/issues/760)\n",
    "for more information.\n",
    ":::\n",
    "\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.individuals_flags\n",
    "  TreeSequence.nodes_time\n",
    "  TreeSequence.nodes_flags\n",
    "  TreeSequence.nodes_population\n",
    "  TreeSequence.nodes_individual\n",
    "  TreeSequence.edges_left\n",
    "  TreeSequence.edges_right\n",
    "  TreeSequence.edges_parent\n",
    "  TreeSequence.edges_child\n",
    "  TreeSequence.sites_position\n",
    "  TreeSequence.mutations_site\n",
    "  TreeSequence.mutations_node\n",
    "  TreeSequence.mutations_parent\n",
    "  TreeSequence.mutations_time\n",
    "  TreeSequence.migrations_left\n",
    "  TreeSequence.migrations_right\n",
    "  TreeSequence.migrations_right\n",
    "  TreeSequence.migrations_node\n",
    "  TreeSequence.migrations_source\n",
    "  TreeSequence.migrations_dest\n",
    "  TreeSequence.migrations_time\n",
    "  TreeSequence.indexes_edge_insertion_order\n",
    "  TreeSequence.indexes_edge_removal_order\n",
    "```\n",
    "\n",
    "(sec_python_api_tree_sequences_loading_and_saving)=\n",
    "\n",
    "#### Loading and saving\n",
    "\n",
    "There are several methods for loading data into a {class}`TreeSequence`\n",
    "instance. The simplest and most convenient is the use the {func}`tskit.load`\n",
    "function to load a {ref}`tree sequence file <sec_tree_sequence_file_format>`. For small\n",
    "scale data and debugging, it is often convenient to use the {func}`tskit.load_text`\n",
    "function to read data in the {ref}`text file format<sec_text_file_format>`.\n",
    "The {meth}`TableCollection.tree_sequence` function\n",
    "efficiently creates a {class}`TreeSequence` object from a\n",
    "{class}`collection of tables<TableCollection>`\n",
    "using the {ref}`Tables API <sec_tables_api>`.\n",
    "\n",
    "```{eval-rst}\n",
    "Load a tree sequence\n",
    "    .. autosummary::\n",
    "      load\n",
    "      load_text\n",
    "      TableCollection.tree_sequence\n",
    "\n",
    "Save a tree sequence\n",
    "    .. autosummary::\n",
    "      TreeSequence.dump\n",
    "```\n",
    "\n",
    ":::{seealso}\n",
    "Tree sequences with a single simple topology can also be created from scratch by\n",
    "{ref}`generating<sec_python_api_trees_creating>` a {class}`Tree` and accessing its\n",
    "{attr}`~Tree.tree_sequence` property.\n",
    ":::\n",
    "\n",
    "(sec_python_api_tree_sequences_obtaining_trees)=\n",
    "\n",
    "#### Obtaining trees\n",
    "\n",
    "The following properties and methods return information about the\n",
    "{class}`trees<Tree>` that are generated along a tree sequence.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "\n",
    "  TreeSequence.num_trees\n",
    "  TreeSequence.trees\n",
    "  TreeSequence.breakpoints\n",
    "  TreeSequence.coiterate\n",
    "  TreeSequence.first\n",
    "  TreeSequence.last\n",
    "  TreeSequence.aslist\n",
    "  TreeSequence.at\n",
    "  TreeSequence.at_index\n",
    "```\n",
    "\n",
    "#### Obtaining other objects\n",
    "\n",
    "(sec_python_api_tree_sequences_obtaining_other_objects)=\n",
    "\n",
    "Various components make up a tree sequence, such as nodes and edges, sites and\n",
    "mutations, and populations and individuals. These can be counted or converted into\n",
    "Python objects using the following classes, properties, and methods.\n",
    "\n",
    "##### Tree topology\n",
    "\n",
    "```{eval-rst}\n",
    "Nodes\n",
    "    .. autosummary::\n",
    "      Node\n",
    "      TreeSequence.num_nodes\n",
    "      TreeSequence.nodes\n",
    "      TreeSequence.node\n",
    "      TreeSequence.num_samples\n",
    "      TreeSequence.samples\n",
    "\n",
    "Edges\n",
    "    .. autosummary::\n",
    "      Edge\n",
    "      TreeSequence.num_edges\n",
    "      TreeSequence.edges\n",
    "      TreeSequence.edge\n",
    "```\n",
    "\n",
    "##### Genetic variation\n",
    "\n",
    "```{eval-rst}\n",
    "Sites\n",
    "    .. autosummary::\n",
    "      Site\n",
    "      TreeSequence.num_sites\n",
    "      TreeSequence.sites\n",
    "      TreeSequence.site\n",
    "      Variant\n",
    "      TreeSequence.variants\n",
    "      TreeSequence.genotype_matrix\n",
    "      TreeSequence.haplotypes\n",
    "      TreeSequence.alignments\n",
    "\n",
    "Mutations\n",
    "    .. autosummary::\n",
    "      Mutation\n",
    "      TreeSequence.num_mutations\n",
    "      TreeSequence.mutations\n",
    "      TreeSequence.mutation\n",
    "```\n",
    "\n",
    "##### Demography\n",
    "\n",
    "```{eval-rst}\n",
    "Populations\n",
    "    .. autosummary::\n",
    "      Population\n",
    "      TreeSequence.num_populations\n",
    "      TreeSequence.populations\n",
    "      TreeSequence.population\n",
    "\n",
    "Migrations\n",
    "    .. autosummary::\n",
    "      Migration\n",
    "      TreeSequence.num_migrations\n",
    "      TreeSequence.migrations\n",
    "      TreeSequence.migration\n",
    "```\n",
    "\n",
    "##### Other\n",
    "\n",
    "```{eval-rst}\n",
    "Individuals\n",
    "    .. autosummary::\n",
    "      Individual\n",
    "      TreeSequence.num_individuals\n",
    "      TreeSequence.individuals\n",
    "      TreeSequence.individual\n",
    "\n",
    "\n",
    "Provenance entries (also see :ref:`sec_python_api_provenance`)\n",
    "    .. autosummary::\n",
    "      Provenance\n",
    "      TreeSequence.num_provenances\n",
    "      TreeSequence.provenances\n",
    "      TreeSequence.provenance\n",
    "```\n",
    "\n",
    "(sec_python_api_tree_sequences_modification)=\n",
    "\n",
    "#### Tree sequence modification\n",
    "\n",
    "Although tree sequences are immutable, several methods will taken an existing tree\n",
    "sequence and return a modifed version. These are thin wrappers around the\n",
    "{ref}`identically named methods of a TableCollection<sec_tables_api_modification>`,\n",
    "which perform the same actions but modify the {class}`TableCollection` in place.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.simplify\n",
    "  TreeSequence.subset\n",
    "  TreeSequence.union\n",
    "  TreeSequence.keep_intervals\n",
    "  TreeSequence.delete_intervals\n",
    "  TreeSequence.delete_sites\n",
    "  TreeSequence.trim\n",
    "  TreeSequence.split_edges\n",
    "  TreeSequence.decapitate\n",
    "```\n",
    "\n",
    "(sec_python_api_tree_sequences_ibd)=\n",
    "\n",
    "#### Identity by descent\n",
    "\n",
    "The {meth}`.TreeSequence.ibd_segments` method allows us to compute\n",
    "identity relationships between pairs of samples. See the\n",
    "{ref}`sec_identity` section for more details and examples\n",
    "and the {ref}`sec_python_api_reference_identity` section for\n",
    "API documentation on the associated classes.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.ibd_segments\n",
    "```\n",
    "\n",
    "(sec_python_api_tree_sequences_tables)=\n",
    "\n",
    "#### Tables\n",
    "\n",
    "The underlying data in a tree sequence is stored in a\n",
    "{ref}`collection of tables<sec_tables_api>`. The following methods give access\n",
    "to tables and associated functionality. Since tables can be modified, this\n",
    "allows tree sequences to be edited: see the {ref}`sec_tables` tutorial for\n",
    "an introduction.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.tables\n",
    "  TreeSequence.dump_tables\n",
    "  TreeSequence.table_metadata_schemas\n",
    "  TreeSequence.tables_dict\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_tree_sequences_statistics)=\n",
    "\n",
    "#### Statistics\n",
    "\n",
    "```{eval-rst}\n",
    "\n",
    "Single site\n",
    "    .. autosummary::\n",
    "      TreeSequence.allele_frequency_spectrum\n",
    "      TreeSequence.divergence\n",
    "      TreeSequence.diversity\n",
    "      TreeSequence.f2\n",
    "      TreeSequence.f3\n",
    "      TreeSequence.f4\n",
    "      TreeSequence.Fst\n",
    "      TreeSequence.genealogical_nearest_neighbours\n",
    "      TreeSequence.genetic_relatedness\n",
    "      TreeSequence.general_stat\n",
    "      TreeSequence.segregating_sites\n",
    "      TreeSequence.sample_count_stat\n",
    "      TreeSequence.mean_descendants\n",
    "      TreeSequence.Tajimas_D\n",
    "      TreeSequence.trait_correlation\n",
    "      TreeSequence.trait_covariance\n",
    "      TreeSequence.trait_linear_model\n",
    "      TreeSequence.Y2\n",
    "      TreeSequence.Y3\n",
    "\n",
    "Comparative\n",
    "    .. autosummary::\n",
    "      TreeSequence.kc_distance\n",
    "```\n",
    "\n",
    "(sec_python_api_tree_sequences_topological_analysis)=\n",
    "\n",
    "#### Topological analysis\n",
    "\n",
    "The topology of a tree in a tree sequence refers to the relationship among\n",
    "samples ignoring branch lengths. Functionality as described in\n",
    "{ref}`sec_topological_analysis` is mainly provided via\n",
    "{ref}`methods on trees<sec_python_api_trees_topological_analysis>`, but more\n",
    "efficient methods sometimes exist for entire tree sequences:\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.count_topologies\n",
    "```\n",
    "\n",
    "(sec_python_api_tree_sequences_display)=\n",
    "\n",
    "#### Display\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.draw_svg\n",
    "  TreeSequence.draw_text\n",
    "  TreeSequence.__str__\n",
    "  TreeSequence._repr_html_\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_tree_sequences_export)=\n",
    "\n",
    "#### Export\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TreeSequence.as_fasta\n",
    "  TreeSequence.as_nexus\n",
    "  TreeSequence.dump_text\n",
    "  TreeSequence.to_macs\n",
    "  TreeSequence.write_fasta\n",
    "  TreeSequence.write_nexus\n",
    "  TreeSequence.write_vcf\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees)=\n",
    "\n",
    "### {class}`Tree<Tree>` API\n",
    "\n",
    "A tree is an instance of the {class}`Tree` class. These trees cannot exist\n",
    "independently of the {class}`TreeSequence` from which they are generated.\n",
    "Usually, therefore, a {class}`Tree` instance is created by\n",
    "{ref}`sec_python_api_tree_sequences_obtaining_trees` from an existing tree\n",
    "sequence (although it is also possible to generate a new instance of a\n",
    "{class}`Tree` belonging to the same tree sequence using {meth}`Tree.copy`).\n",
    "\n",
    ":::{note}\n",
    "For efficiency, each instance of a {class}`Tree` is a state-machine\n",
    "whose internal state corresponds to one of the trees in the parent tree sequence:\n",
    "{ref}`sec_python_api_trees_moving_to` in the tree sequence does not require a\n",
    "new instance to be created, but simply the internal state to be changed.\n",
    ":::\n",
    "\n",
    "(sec_python_api_trees_general_properties)=\n",
    "\n",
    "#### General properties\n",
    "\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  Tree.tree_sequence\n",
    "  Tree.total_branch_length\n",
    "  Tree.root_threshold\n",
    "  Tree.virtual_root\n",
    "  Tree.num_edges\n",
    "  Tree.num_roots\n",
    "  Tree.has_single_root\n",
    "  Tree.has_multiple_roots\n",
    "  Tree.root\n",
    "  Tree.roots\n",
    "  Tree.index\n",
    "  Tree.interval\n",
    "  Tree.span\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees_creating)=\n",
    "\n",
    "#### Creating new trees\n",
    "\n",
    "It is sometimes useful to create an entirely new tree sequence consisting\n",
    "of just a single tree (a \"one-tree sequence\"). The follow methods create such an\n",
    "object and return a {class}`Tree` instance corresponding to that tree.\n",
    "The new tree sequence to which the tree belongs is available through the\n",
    "{attr}`~Tree.tree_sequence` property.\n",
    "\n",
    "```{eval-rst}\n",
    "Creating a new tree\n",
    "    .. autosummary::\n",
    "      Tree.generate_balanced\n",
    "      Tree.generate_comb\n",
    "      Tree.generate_random_binary\n",
    "      Tree.generate_star\n",
    "\n",
    "Creating a new tree from an existing tree\n",
    "    .. autosummary::\n",
    "      Tree.split_polytomies\n",
    "```\n",
    "\n",
    ":::{seealso}\n",
    "{meth}`Tree.unrank` for creating a new one-tree sequence from its\n",
    "{ref}`topological rank<sec_python_api_trees_topological_analysis>`.\n",
    ":::\n",
    "\n",
    ":::{note}\n",
    "Several of these methods are {func}`static<python:staticmethod>`, so should\n",
    "be called e.g. as `tskit.Tree.generate_balanced(4)` rather than used on\n",
    "a specific {class}`Tree` instance.\n",
    ":::\n",
    "\n",
    "\n",
    "(sec_python_api_trees_node_measures)=\n",
    "\n",
    "#### Node measures\n",
    "\n",
    "Often it is useful to access information pertinant to a specific node or set of nodes\n",
    "but which might also change from tree to tree in the tree sequence. Examples include\n",
    "the encoding of the tree via `parent`, `left_child`, etc.\n",
    "(see {ref}`sec_data_model_tree_structure`), the number of samples under a node,\n",
    "or the most recent common ancestor (MRCA) of two nodes. This sort of information is\n",
    "available via simple and high performance {class}`Tree` methods\n",
    "\n",
    "\n",
    "(sec_python_api_trees_node_measures_simple)=\n",
    "\n",
    "##### Simple measures\n",
    "\n",
    "These return a simple number, or (usually) short list of numbers relevant to a specific\n",
    "node or limited set of nodes.\n",
    "\n",
    "```{eval-rst}\n",
    "Node information\n",
    "    .. autosummary::\n",
    "      Tree.is_sample\n",
    "      Tree.is_isolated\n",
    "      Tree.is_leaf\n",
    "      Tree.is_internal\n",
    "      Tree.parent\n",
    "      Tree.num_children\n",
    "      Tree.time\n",
    "      Tree.branch_length\n",
    "      Tree.depth\n",
    "      Tree.population\n",
    "      Tree.right_sib\n",
    "      Tree.left_sib\n",
    "      Tree.right_child\n",
    "      Tree.left_child\n",
    "      Tree.children\n",
    "      Tree.edge\n",
    "\n",
    "Descendant nodes\n",
    "    .. autosummary::\n",
    "      Tree.leaves\n",
    "      Tree.samples\n",
    "      Tree.num_samples\n",
    "      Tree.num_tracked_samples\n",
    "\n",
    "Multiple nodes\n",
    "    .. autosummary::\n",
    "      Tree.is_descendant\n",
    "      Tree.mrca\n",
    "      Tree.tmrca\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees_node_measures_array)=\n",
    "\n",
    "##### Array access\n",
    "\n",
    "These all return a numpy array whose length corresponds to\n",
    "the total number of nodes in the tree sequence. They provide direct access\n",
    "to the underlying memory structures, and are thus very efficient, providing a\n",
    "high performance interface which can be used in conjunction with the equivalent\n",
    "{ref}`traversal methods<sec_python_api_trees_traversal>`.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  Tree.parent_array\n",
    "  Tree.left_child_array\n",
    "  Tree.right_child_array\n",
    "  Tree.left_sib_array\n",
    "  Tree.right_sib_array\n",
    "  Tree.num_children_array\n",
    "  Tree.edge_array\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees_traversal)=\n",
    "\n",
    "#### Tree traversal\n",
    "\n",
    "Moving around within a tree usually involves visiting the tree nodes in some sort of\n",
    "order. Often, given a particular order, it is convenient to iterate over each node\n",
    "using the {meth}`Tree.nodes` method. However, for high performance algorithms, it\n",
    "may be more convenient to access the node indices for a particular order as\n",
    "an array, and use this, for example, to index into one of the node arrays (see\n",
    "{ref}`sec_topological_analysis_traversal`).\n",
    "\n",
    "```{eval-rst}\n",
    "Iterator access\n",
    "    .. autosummary::\n",
    "\n",
    "      Tree.nodes\n",
    "\n",
    "Array access\n",
    "    .. autosummary::\n",
    "\n",
    "      Tree.postorder\n",
    "      Tree.preorder\n",
    "      Tree.timeasc\n",
    "      Tree.timedesc\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees_topological_analysis)=\n",
    "\n",
    "#### Topological analysis\n",
    "\n",
    "The topology of a tree refers to the simple relationship among samples\n",
    "(i.e. ignoring branch lengths), see {ref}`sec_combinatorics` for more details. These\n",
    "methods provide ways to enumerate and count tree topologies.\n",
    "\n",
    "Briefly, the position of a tree in the enumeration `all_trees` can be obtained using\n",
    "the tree's {meth}`~Tree.rank` method. Inversely, a {class}`Tree` can be constructed\n",
    "from a position in the enumeration with {meth}`Tree.unrank`.\n",
    "\n",
    "\n",
    "```{eval-rst}\n",
    "Methods of a tree\n",
    "    .. autosummary::\n",
    "      Tree.rank\n",
    "      Tree.count_topologies\n",
    "\n",
    "Functions and static methods\n",
    "    .. autosummary::\n",
    "      Tree.unrank\n",
    "      all_tree_shapes\n",
    "      all_tree_labellings\n",
    "      all_trees\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees_comparing)=\n",
    "\n",
    "#### Comparing trees\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  Tree.kc_distance\n",
    "```\n",
    "\n",
    "(sec_python_api_trees_balance)=\n",
    "\n",
    "#### Balance/imbalance indices\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  Tree.colless_index\n",
    "  Tree.sackin_index\n",
    "  Tree.b1_index\n",
    "  Tree.b2_index\n",
    "```\n",
    "\n",
    "(sec_python_api_trees_sites_mutations)=\n",
    "\n",
    "#### Sites and mutations\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  Tree.sites\n",
    "  Tree.num_sites\n",
    "  Tree.mutations\n",
    "  Tree.num_mutations\n",
    "  Tree.map_mutations\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_trees_moving_to)=\n",
    "\n",
    "#### Moving to other trees\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "\n",
    "  Tree.next\n",
    "  Tree.prev\n",
    "  Tree.first\n",
    "  Tree.last\n",
    "  Tree.seek\n",
    "  Tree.seek_index\n",
    "  Tree.clear\n",
    "```\n",
    "\n",
    "#### Display\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "\n",
    "  Tree.draw_svg\n",
    "  Tree.draw_text\n",
    "  Tree.__str__\n",
    "  Tree._repr_html_\n",
    "```\n",
    "\n",
    "#### Export\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "\n",
    "  Tree.as_dict_of_dicts\n",
    "  Tree.as_newick\n",
    "```\n",
    "\n",
    "\n",
    "(sec_tables_api)=\n",
    "\n",
    "## Tables and Table Collections\n",
    "\n",
    "The information required to construct a tree sequence is stored in a collection\n",
    "of *tables*, each defining a different aspect of the structure of a tree\n",
    "sequence. These tables are described individually in\n",
    "{ref}`the next section<sec_tables_api_table>`. However, these are interrelated,\n",
    "and so many operations work\n",
    "on the entire collection of tables, known as a *table collection*.\n",
    "\n",
    "(sec_tables_api_table_collection)=\n",
    "\n",
    "### `TableCollection` API\n",
    "\n",
    "The {class}`TableCollection` and {class}`TreeSequence` classes are\n",
    "deeply related. A `TreeSequence` instance is based on the information\n",
    "encoded in a `TableCollection`. Tree sequences are **immutable**, and\n",
    "provide methods for obtaining trees from the sequence. A `TableCollection`\n",
    "is **mutable**, and does not have any methods for obtaining trees.\n",
    "The `TableCollection` class thus allows creation and modification of\n",
    "tree sequences (see the {ref}`sec_tables` tutorial).\n",
    "\n",
    "\n",
    "#### General properties\n",
    "\n",
    "Specific {ref}`tables<sec_tables_api_table>` in the {class}`TableCollection`\n",
    "are be accessed using the plural version of their name, so that, for instance, the\n",
    "individual table can be accessed using `table_collection.individuals`. A table\n",
    "collection also has other properties containing, for example, number of bytes taken\n",
    "to store it and the top-level metadata associated with the tree sequence as a whole.\n",
    "\n",
    "```{eval-rst}\n",
    "Table access\n",
    "    .. autosummary::\n",
    "      TableCollection.individuals\n",
    "      TableCollection.nodes\n",
    "      TableCollection.edges\n",
    "      TableCollection.migrations\n",
    "      TableCollection.sites\n",
    "      TableCollection.mutations\n",
    "      TableCollection.populations\n",
    "      TableCollection.provenances\n",
    "\n",
    "Other properties\n",
    "    .. autosummary::\n",
    "      TableCollection.file_uuid\n",
    "      TableCollection.indexes\n",
    "      TableCollection.nbytes\n",
    "      TableCollection.table_name_map\n",
    "      TableCollection.metadata\n",
    "      TableCollection.metadata_bytes\n",
    "      TableCollection.metadata_schema\n",
    "      TableCollection.sequence_length\n",
    "      TableCollection.time_units\n",
    "```\n",
    "\n",
    "\n",
    "(sec_tables_api_transformation)=\n",
    "\n",
    "#### Transformation\n",
    "\n",
    "These methods act in-place to transform the contents of a {class}`TableCollection`,\n",
    "either by modifying the underlying tables (removing, editing, or adding to them) or\n",
    "by adjusting the table collection so that it meets the\n",
    "{ref}`sec_valid_tree_sequence_requirements`.\n",
    "\n",
    "\n",
    "(sec_tables_api_modification)=\n",
    "\n",
    "##### Modification\n",
    "\n",
    "These methods modify the data stored in a {class}`TableCollection`. They also have\n",
    "{ref}`equivalant TreeSequence versions<sec_python_api_tree_sequences_modification>`\n",
    "(unlike the methods described below those do *not* operate in place, but rather act in\n",
    "a functional way, returning a new tree sequence while leaving the original unchanged).\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TableCollection.clear\n",
    "  TableCollection.simplify\n",
    "  TableCollection.subset\n",
    "  TableCollection.delete_intervals\n",
    "  TableCollection.keep_intervals\n",
    "  TableCollection.delete_sites\n",
    "  TableCollection.trim\n",
    "  TableCollection.union\n",
    "  TableCollection.delete_older\n",
    "```\n",
    "\n",
    "(sec_tables_api_creating_valid_tree_sequence)=\n",
    "\n",
    "##### Creating a valid tree sequence\n",
    "\n",
    "These methods can be used to help reorganise or rationalise the\n",
    "{class}`TableCollection` so that it is in the form\n",
    "{ref}`required<sec_valid_tree_sequence_requirements>` for\n",
    "it to be {meth}`converted<TableCollection.tree_sequence>`\n",
    "into a {class}`TreeSequence`. This may require sorting the tables,\n",
    "ensuring they are logically consistent, and adding {ref}`sec_table_indexes`.\n",
    "\n",
    ":::{note}\n",
    "These methods are not guaranteed to make valid a {class}`TableCollection` which is\n",
    "logically inconsistent, for example if multiple edges have the same child at a\n",
    "given position on the genome or if non-existent node IDs are referenced.\n",
    ":::\n",
    "\n",
    "```{eval-rst}\n",
    "Sorting\n",
    "    .. autosummary::\n",
    "      TableCollection.sort\n",
    "      TableCollection.sort_individuals\n",
    "      TableCollection.canonicalise\n",
    "\n",
    "Logical consistency\n",
    "    .. autosummary::\n",
    "      TableCollection.compute_mutation_parents\n",
    "      TableCollection.compute_mutation_times\n",
    "      TableCollection.deduplicate_sites\n",
    "\n",
    "Indexing\n",
    "    .. autosummary::\n",
    "      TableCollection.has_index\n",
    "      TableCollection.build_index\n",
    "      TableCollection.drop_index\n",
    "```\n",
    "\n",
    "#### Miscellaneous methods\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TableCollection.copy\n",
    "  TableCollection.equals\n",
    "  TableCollection.link_ancestors\n",
    "```\n",
    "\n",
    "#### Export\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  TableCollection.tree_sequence\n",
    "  TableCollection.dump\n",
    "```\n",
    "\n",
    "(sec_tables_api_table)=\n",
    "\n",
    "### Table APIs\n",
    "\n",
    "Here we outline the table classes and the common methods and variables available for\n",
    "each. For description and definition of each table's meaning\n",
    "and use, see {ref}`the table definitions <sec_table_definitions>`.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "\n",
    "  IndividualTable\n",
    "  NodeTable\n",
    "  EdgeTable\n",
    "  MigrationTable\n",
    "  SiteTable\n",
    "  MutationTable\n",
    "  PopulationTable\n",
    "  ProvenanceTable\n",
    "```\n",
    "\n",
    "(sec_tables_api_accessing_table_data)=\n",
    "\n",
    "#### Accessing table data\n",
    "\n",
    "The tables API provides an efficient way of working\n",
    "with and interchanging {ref}`tree sequence data <sec_data_model>`. Each table class\n",
    "(e.g, {class}`NodeTable`, {class}`EdgeTable`, {class}`SiteTable`) has a specific set\n",
    "of columns with fixed types, and a set of methods for setting and getting the data\n",
    "in these columns. The number of rows in the table `t` is given by `len(t)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8ddc928e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The table contains 2 rows\n",
      "╔══╤════╤═════╤══════╤═════╤════════╗\n",
      "║id│left│right│parent│child│metadata║\n",
      "╠══╪════╪═════╪══════╪═════╪════════╣\n",
      "║0 │   0│    1│    10│   11│        ║\n",
      "║1 │   1│    2│     9│   11│        ║\n",
      "╚══╧════╧═════╧══════╧═════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import tskit\n",
    "t = tskit.EdgeTable()\n",
    "t.add_row(left=0, right=1, parent=10, child=11)\n",
    "t.add_row(left=1, right=2, parent=9, child=11)\n",
    "print(\"The table contains\", len(t), \"rows\")\n",
    "print(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d78b49fb",
   "metadata": {},
   "source": [
    "Each table supports accessing the data either by row or column. To access the data in\n",
    "a *column*, we can use standard attribute access which will\n",
    "return a copy of the column data as a numpy array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0f21e6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0., 1.])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.left"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2495a646",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([10,  9], dtype=int32)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.parent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed2ea58d",
   "metadata": {},
   "source": [
    "To access the data in a *row*, say row number `j` in table `t`, simply use `t[j]`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "34cc0106",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EdgeTableRow(left=0.0, right=1.0, parent=10, child=11, metadata=b'')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38fdcc73",
   "metadata": {},
   "source": [
    "This also works as expected with negative `j`, counting rows from the end of the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4366187e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EdgeTableRow(left=1.0, right=2.0, parent=9, child=11, metadata=b'')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t[-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdfa8dff",
   "metadata": {},
   "source": [
    "The returned row has attributes allowing contents to be accessed by name, e.g.\n",
    "`site_table[0].position`, `site_table[0].ancestral_state`, `site_table[0].metadata`\n",
    "etc.:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dd99fd18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t[-1].right"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3542f04c",
   "metadata": {},
   "source": [
    "Row attributes cannot be modified directly. Instead, the `replace` method of a row\n",
    "object can be used to create a new row with one or more changed column\n",
    "values, which can then be used to replace the original. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eb3461fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════╤═════╤══════╤═════╤════════╗\n",
      "║id│left│right│parent│child│metadata║\n",
      "╠══╪════╪═════╪══════╪═════╪════════╣\n",
      "║0 │   0│    1│    10│   11│        ║\n",
      "║1 │   1│    3│     9│    4│        ║\n",
      "╚══╧════╧═════╧══════╧═════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t[-1] = t[-1].replace(child=4, right=3)\n",
    "print(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c75de71",
   "metadata": {},
   "source": [
    "Tables also support the {mod}`pickle` protocol, and so can be easily serialised and\n",
    "deserialised. This can be useful, for example, when performing parallel computations\n",
    "using the {mod}`multiprocessing` module (however, pickling will not be as efficient\n",
    "as storing tables in the native {ref}`format <sec_tree_sequence_file_format>`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa3056ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════╤═════╤══════╤═════╤════════╗\n",
      "║id│left│right│parent│child│metadata║\n",
      "╠══╪════╪═════╪══════╪═════╪════════╣\n",
      "║0 │   0│    1│    10│   11│        ║\n",
      "║1 │   1│    3│     9│    4│        ║\n",
      "╚══╧════╧═════╧══════╧═════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "serialised = pickle.dumps(t)\n",
    "t2 = pickle.loads(serialised)\n",
    "print(t2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9011e43f",
   "metadata": {},
   "source": [
    "Tables support the equality operator `==` based on the data\n",
    "held in the columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1532389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t == t2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bd34884c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t is t2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e0b6681c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════╤═════╤══════╤═════╤════════╗\n",
      "║id│left│right│parent│child│metadata║\n",
      "╠══╪════╪═════╪══════╪═════╪════════╣\n",
      "║0 │   0│    1│    10│   11│        ║\n",
      "║1 │   1│    3│     9│    4│        ║\n",
      "║2 │   0│    1│     2│    3│        ║\n",
      "╚══╧════╧═════╧══════╧═════╧════════╝\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t2.add_row(0, 1, 2, 3)\n",
    "print(t2)\n",
    "t == t2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3659fb48",
   "metadata": {},
   "source": [
    ":::{todo}\n",
    "Move some or all of these examples into a suitable alternative chapter.\n",
    ":::\n",
    "\n",
    "\n",
    "(sec_tables_api_text_columns)=\n",
    "\n",
    "##### Text columns\n",
    "\n",
    "As described in the {ref}`sec_encoding_ragged_columns`, working with\n",
    "variable length columns is somewhat more involved. Columns\n",
    "encoding text data store the **encoded bytes** of the flattened\n",
    "strings, and the offsets into this column in two separate\n",
    "arrays.\n",
    "\n",
    "Consider the following example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c3825ab6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════════╤═══════════════╤════════╗\n",
      "║id│position│ancestral_state│metadata║\n",
      "╠══╪════════╪═══════════════╪════════╣\n",
      "║0 │       0│              A│        ║\n",
      "║1 │       1│             BB│        ║\n",
      "║2 │       2│               │        ║\n",
      "║3 │       3│            CCC│        ║\n",
      "╚══╧════════╧═══════════════╧════════╝\n",
      "\n",
      "SiteTableRow(position=0.0, ancestral_state='A', metadata=b'')\n",
      "SiteTableRow(position=1.0, ancestral_state='BB', metadata=b'')\n",
      "SiteTableRow(position=2.0, ancestral_state='', metadata=b'')\n",
      "SiteTableRow(position=3.0, ancestral_state='CCC', metadata=b'')\n"
     ]
    }
   ],
   "source": [
    "t = tskit.SiteTable()\n",
    "t.add_row(0, \"A\")\n",
    "t.add_row(1, \"BB\")\n",
    "t.add_row(2, \"\")\n",
    "t.add_row(3, \"CCC\")\n",
    "print(t)\n",
    "print(t[0])\n",
    "print(t[1])\n",
    "print(t[2])\n",
    "print(t[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4729d0a6",
   "metadata": {},
   "source": [
    "Here we create a {class}`SiteTable` and add four rows, each with a different\n",
    "`ancestral_state`. We can then access this information from each\n",
    "row in a straightforward manner. Working with columns of text data\n",
    "is a little trickier, however:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ea36524c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[65 66 66 67 67 67]\n",
      "[0 1 3 3 6]\n"
     ]
    }
   ],
   "source": [
    "print(t.ancestral_state)\n",
    "print(t.ancestral_state_offset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c8061be0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A', 'BB', '', 'CCC']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tskit.unpack_strings(t.ancestral_state, t.ancestral_state_offset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e9b9566",
   "metadata": {},
   "source": [
    "Here, the `ancestral_state` array is the UTF8 encoded bytes of the flattened\n",
    "strings, and the `ancestral_state_offset` is the offset into this array\n",
    "for each row. The {func}`tskit.unpack_strings` function, however, is a convient\n",
    "way to recover the original strings from this encoding. We can also use the\n",
    "{func}`tskit.pack_strings` to insert data using this approach:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c205e7ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════════╤═══════════════╤════════╗\n",
      "║id│position│ancestral_state│metadata║\n",
      "╠══╪════════╪═══════════════╪════════╣\n",
      "║0 │       0│              0│        ║\n",
      "║1 │       1│             12│        ║\n",
      "║2 │       2│               │        ║\n",
      "╚══╧════════╧═══════════════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a, off = tskit.pack_strings([\"0\", \"12\", \"\"])\n",
    "t.set_columns(position=[0, 1, 2], ancestral_state=a, ancestral_state_offset=off)\n",
    "print(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e71da976",
   "metadata": {},
   "source": [
    "When inserting many rows with standard infinite sites mutations (i.e.,\n",
    "ancestral state is \"0\"), it is more efficient to construct the\n",
    "numpy arrays directly than to create a list of strings and use\n",
    "{func}`pack_strings`. When doing this, it is important to note that\n",
    "it is the **encoded** byte values that are stored; by default, we\n",
    "use UTF8 (which corresponds to ASCII for simple printable characters).:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "51239624",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════════╤═══════════════╤════════╗\n",
      "║id│position│ancestral_state│metadata║\n",
      "╠══╪════════╪═══════════════╪════════╣\n",
      "║0 │       0│              0│        ║\n",
      "║1 │       1│              0│        ║\n",
      "║2 │       2│              0│        ║\n",
      "║3 │       3│              0│        ║\n",
      "║4 │       4│              0│        ║\n",
      "║5 │       5│              0│        ║\n",
      "║6 │       6│              0│        ║\n",
      "║7 │       7│              0│        ║\n",
      "║8 │       8│              0│        ║\n",
      "║9 │       9│              0│        ║\n",
      "╚══╧════════╧═══════════════╧════════╝\n",
      "\n",
      "ancestral state data [48 48 48 48 48 48 48 48 48 48]\n",
      "ancestral state offsets [ 0  1  2  3  4  5  6  7  8  9 10]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "t_s = tskit.SiteTable()\n",
    "m = 10\n",
    "a = ord(\"0\") + np.zeros(m, dtype=np.int8)\n",
    "off = np.arange(m + 1, dtype=np.uint32)\n",
    "t_s.set_columns(position=np.arange(m), ancestral_state=a, ancestral_state_offset=off)\n",
    "print(t_s)\n",
    "print(\"ancestral state data\", t_s.ancestral_state)\n",
    "print(\"ancestral state offsets\", t_s.ancestral_state_offset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b9b56f0",
   "metadata": {},
   "source": [
    "In the mutation table, the derived state of each mutation can be handled similarly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7d0223a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤════╤════╤════╤═════════════╤══════╤════════╗\n",
      "║id│site│node│time│derived_state│parent│metadata║\n",
      "╠══╪════╪════╪════╪═════════════╪══════╪════════╣\n",
      "║0 │   0│   0│ nan│            1│    -1│        ║\n",
      "║1 │   1│   0│ nan│            1│    -1│        ║\n",
      "║2 │   2│   0│ nan│            1│    -1│        ║\n",
      "║3 │   3│   0│ nan│            1│    -1│        ║\n",
      "║4 │   4│   0│ nan│            1│    -1│        ║\n",
      "║5 │   5│   0│ nan│            1│    -1│        ║\n",
      "║6 │   6│   0│ nan│            1│    -1│        ║\n",
      "║7 │   7│   0│ nan│            1│    -1│        ║\n",
      "║8 │   8│   0│ nan│            1│    -1│        ║\n",
      "║9 │   9│   0│ nan│            1│    -1│        ║\n",
      "╚══╧════╧════╧════╧═════════════╧══════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t_m = tskit.MutationTable()\n",
    "site = np.arange(m, dtype=np.int32)\n",
    "d, off = tskit.pack_strings([\"1\"] * m)\n",
    "node = np.zeros(m, dtype=np.int32)\n",
    "t_m.set_columns(site=site, node=node, derived_state=d, derived_state_offset=off)\n",
    "print(t_m)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24870878",
   "metadata": {},
   "source": [
    ":::{todo}\n",
    "Move some or all of these examples into a suitable alternative chapter.\n",
    ":::\n",
    "\n",
    "\n",
    "(sec_tables_api_binary_columns)=\n",
    "\n",
    "##### Binary columns\n",
    "\n",
    "Columns storing binary data take the same approach as\n",
    "{ref}`sec_tables_api_text_columns` to encoding\n",
    "{ref}`variable length data <sec_encoding_ragged_columns>`.\n",
    "The difference between the two is only raw {class}`bytes` values are accepted: no\n",
    "character encoding or decoding is done on the data. Consider the following example\n",
    "where a table has no `metadata_schema` such that arbitrary bytes can be stored and\n",
    "no automatic encoding or decoding of objects is performed by the Python API and we can\n",
    "store and retrieve raw `bytes`. (See {ref}`sec_metadata` for details):\n",
    "\n",
    "Below, we add two rows to a {class}`NodeTable`, with different\n",
    "{ref}`metadata <sec_metadata_definition>`. The first row contains a simple\n",
    "byte string, and the second contains a Python dictionary serialised using\n",
    "{mod}`pickle`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7f796f88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤═════╤══════════╤══════════╤════╤════════════════════════════════════════╗\n",
      "║id│flags│population│individual│time│metadata                                ║\n",
      "╠══╪═════╪══════════╪══════════╪════╪════════════════════════════════════════╣\n",
      "║0 │    0│        -1│        -1│   0│                  b'these are raw bytes'║\n",
      "║1 │    0│        -1│        -1│   0│b'\\x80\\x04\\x95\\x11\\x00\\x00\\x00\\x00\\x0...║\n",
      "╚══╧═════╧══════════╧══════════╧════╧════════════════════════════════════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t = tskit.NodeTable()\n",
    "t.add_row(metadata=b\"these are raw bytes\")\n",
    "t.add_row(metadata=pickle.dumps({\"x\": 1.1}))\n",
    "print(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad99bfeb",
   "metadata": {},
   "source": [
    "Note that the pickled dictionary is encoded in 24 bytes containing unprintable\n",
    "characters. It appears to be unrelated to the original contents, because the binary\n",
    "data is [base64 encoded](https://en.wikipedia.org/wiki/Base64) to ensure that it is\n",
    "print-safe (and doesn't break your terminal). (See the\n",
    "{ref}`sec_metadata_definition` section for more information on the\n",
    "use of base64 encoding.).\n",
    "\n",
    "We can access the metadata in a row (e.g., `t[0].metadata`) which returns a Python\n",
    "bytes object containing precisely the bytes that were inserted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a46fa3bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b'these are raw bytes'\n",
      "b'\\x80\\x04\\x95\\x11\\x00\\x00\\x00\\x00\\x00\\x00\\x00}\\x94\\x8c\\x01x\\x94G?\\xf1\\x99\\x99\\x99\\x99\\x99\\x9as.'\n"
     ]
    }
   ],
   "source": [
    "print(t[0].metadata)\n",
    "print(t[1].metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0a9ec20",
   "metadata": {},
   "source": [
    "The metadata containing the pickled dictionary can be unpickled using\n",
    "{func}`pickle.loads`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b248da13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'x': 1.1}\n"
     ]
    }
   ],
   "source": [
    "print(pickle.loads(t[1].metadata))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06646397",
   "metadata": {},
   "source": [
    "As previously, the `replace` method can be used to change the metadata,\n",
    "by overwriting an existing row with an updated one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "91c708f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤═════╤══════════╤══════════╤════╤════════════════════════════════════════╗\n",
      "║id│flags│population│individual│time│metadata                                ║\n",
      "╠══╪═════╪══════════╪══════════╪════╪════════════════════════════════════════╣\n",
      "║0 │    0│        -1│        -1│   0│                  b'different raw bytes'║\n",
      "║1 │    0│        -1│        -1│   0│b'\\x80\\x04\\x95\\x11\\x00\\x00\\x00\\x00\\x0...║\n",
      "╚══╧═════╧══════════╧══════════╧════╧════════════════════════════════════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "t[0] = t[0].replace(metadata=b\"different raw bytes\")\n",
    "print(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "700fc2f6",
   "metadata": {},
   "source": [
    "Finally, when we print the `metadata` column, we see the raw byte values\n",
    "encoded as signed integers. As for {ref}`sec_tables_api_text_columns`,\n",
    "the `metadata_offset` column encodes the offsets into this array. So, we\n",
    "see that the first metadata value is 9 bytes long and the second is 24."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6b8d194f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 100  105  102  102  101  114  101  110  116   32  114   97  119   32\n",
      "   98  121  116  101  115 -128    4 -107   17    0    0    0    0    0\n",
      "    0    0  125 -108 -116    1  120 -108   71   63  -15 -103 -103 -103\n",
      " -103 -103 -102  115   46]\n",
      "[ 0 19 47]\n"
     ]
    }
   ],
   "source": [
    "print(t.metadata)\n",
    "print(t.metadata_offset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "130cacb2",
   "metadata": {},
   "source": [
    "The {func}`tskit.pack_bytes` and {func}`tskit.unpack_bytes` functions are\n",
    "also useful for encoding data in these columns.\n",
    "\n",
    ":::{todo}\n",
    "Move some or all of these examples into a suitable alternative chapter.\n",
    ":::\n",
    "\n",
    "\n",
    "\n",
    "#### Table functions\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "\n",
    "  parse_nodes\n",
    "  parse_edges\n",
    "  parse_sites\n",
    "  parse_mutations\n",
    "  parse_individuals\n",
    "  parse_populations\n",
    "  parse_migrations\n",
    "  pack_strings\n",
    "  unpack_strings\n",
    "  pack_bytes\n",
    "  unpack_bytes\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_metadata)=\n",
    "\n",
    "## Metadata API\n",
    "\n",
    "The `metadata` module provides validation, encoding and decoding of metadata\n",
    "using a schema. See {ref}`sec_metadata`, {ref}`sec_metadata_api_overview` and\n",
    "{ref}`sec_tutorial_metadata`.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  MetadataSchema\n",
    "  register_metadata_codec\n",
    "```\n",
    "\n",
    ":::{seealso}\n",
    "Refer to the top level metadata-related properties of TreeSequences and TableCollections,\n",
    "such as {attr}`TreeSequence.metadata` and {attr}`TreeSequence.metadata_schema`. Also the\n",
    "metadata fields of\n",
    "{ref}`objects accessed<sec_python_api_tree_sequences_obtaining_other_objects>` through\n",
    "the {class}`TreeSequence` API.\n",
    ":::\n",
    "\n",
    "\n",
    "(sec_python_api_provenance)=\n",
    "\n",
    "## Provenance\n",
    "\n",
    "We provide some preliminary support for validating JSON documents against the\n",
    "{ref}`provenance schema <sec_provenance>`. Programmatic access to provenance\n",
    "information is planned for future versions.\n",
    "\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  validate_provenance\n",
    "```\n",
    "\n",
    "(sec_utility_api)=\n",
    "\n",
    "## Utility functions\n",
    "\n",
    "Miscellaneous top-level utility functions.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autosummary::\n",
    "  is_unknown_time\n",
    "  random_nucleotides\n",
    "```\n",
    "\n",
    "\n",
    "(sec_python_api_reference)=\n",
    "\n",
    "## Reference documentation\n",
    "\n",
    "(sec_python_api_constants)=\n",
    "\n",
    "### Constants\n",
    "\n",
    "The following constants are used throughout the `tskit` API.\n",
    "\n",
    "```{eval-rst}\n",
    ".. automodule:: tskit\n",
    "   :members:\n",
    "```\n",
    "\n",
    "(sec_python_api_exceptions)=\n",
    "\n",
    "### Exceptions\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoexception:: DuplicatePositionsError\n",
    ".. autoexception:: MetadataEncodingError\n",
    ".. autoexception:: MetadataSchemaValidationError\n",
    ".. autoexception:: MetadataValidationError\n",
    ".. autoexception:: ProvenanceValidationError\n",
    "```\n",
    "\n",
    "(sec_python_api_functions)=\n",
    "\n",
    "### Top-level functions\n",
    "\n",
    "```{eval-rst}\n",
    ".. autofunction:: all_trees\n",
    ".. autofunction:: all_tree_shapes\n",
    ".. autofunction:: all_tree_labellings\n",
    ".. autofunction:: is_unknown_time\n",
    ".. autofunction:: load\n",
    ".. autofunction:: load_text\n",
    ".. autofunction:: pack_bytes\n",
    ".. autofunction:: pack_strings\n",
    ".. autofunction:: parse_edges\n",
    ".. autofunction:: parse_individuals\n",
    ".. autofunction:: parse_mutations\n",
    ".. autofunction:: parse_nodes\n",
    ".. autofunction:: parse_populations\n",
    ".. autofunction:: parse_migrations\n",
    ".. autofunction:: parse_sites\n",
    ".. autofunction:: random_nucleotides\n",
    ".. autofunction:: register_metadata_codec\n",
    ".. autofunction:: validate_provenance\n",
    ".. autofunction:: unpack_bytes\n",
    ".. autofunction:: unpack_strings\n",
    "\n",
    "```\n",
    "\n",
    "### Tree and tree sequence classes\n",
    "\n",
    "#### The {class}`Tree` class\n",
    "\n",
    "Also see the {ref}`sec_python_api_trees` summary.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Tree()\n",
    "    :members:\n",
    "    :special-members: __str__\n",
    "    :private-members: _repr_html_\n",
    "```\n",
    "\n",
    "#### The {class}`TreeSequence` class\n",
    "\n",
    "Also see the {ref}`sec_python_api_tree_sequences` summary.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: TreeSequence()\n",
    "    :members:\n",
    "    :special-members: __str__\n",
    "    :private-members: _repr_html_\n",
    "```\n",
    "\n",
    "### Simple container classes\n",
    "\n",
    "#### The {class}`Individual` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Individual()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Node` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Node()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Edge` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Edge()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Site` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Site()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Mutation` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Mutation()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Variant` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Variant()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Migration` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Migration()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Population` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Population()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Provenance` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Provenance()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Interval` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Interval()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`Rank` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: Rank()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "### TableCollection and Table classes\n",
    "\n",
    "#### The {class}`TableCollection` class\n",
    "\n",
    "Also see the {ref}`sec_tables_api_table_collection` summary.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: TableCollection\n",
    "    :inherited-members:\n",
    "    :members:\n",
    "```\n",
    "\n",
    "% Overriding the default signatures for the tables here as they will be\n",
    "% confusing to most users.\n",
    "\n",
    "\n",
    "#### {class}`IndividualTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: IndividualTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from an {class}`IndividualTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Individual` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: IndividualTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`NodeTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: NodeTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from a {class}`NodeTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Node` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: NodeTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`EdgeTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: EdgeTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from an {class}`EdgeTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Edge` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: EdgeTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`MigrationTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: MigrationTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from a {class}`MigrationTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Migration` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: MigrationTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`SiteTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: SiteTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from a {class}`SiteTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Site` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: SiteTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`MutationTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: MutationTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from a {class}`MutationTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Mutation` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: MutationTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`PopulationTable` classes\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: PopulationTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "    :special-members: __getitem__\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from a {class}`PopulationTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Population` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: PopulationTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "\n",
    "#### {class}`ProvenanceTable` classes\n",
    "\n",
    "Also see the {ref}`sec_provenance` and\n",
    "{ref}`provenance API methods<sec_python_api_provenance>`.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: ProvenanceTable()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "##### Associated row class\n",
    "\n",
    "A row returned from a {class}`ProvenanceTable` is an instance of the following\n",
    "basic class, where each attribute matches an identically named attribute in the\n",
    "{class}`Provenance` class.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: ProvenanceTableRow()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "(sec_python_api_reference_identity)=\n",
    "\n",
    "### Identity classes\n",
    "\n",
    "The classes documented in this section are associated with summarising\n",
    "identity relationships between pairs of samples. See the {ref}`sec_identity`\n",
    "section for more details and examples.\n",
    "\n",
    "#### The {class}`IdentitySegments` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: IdentitySegments()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`IdentitySegmentList` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: IdentitySegmentList()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`IdentitySegment` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: IdentitySegment()\n",
    "    :members:\n",
    "```\n",
    "\n",
    "### Miscellaneous classes\n",
    "\n",
    "#### The {class}`ReferenceSequence` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: Add a top-level summary section that we can link to from here.\n",
    "```\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: ReferenceSequence()\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "#### The {class}`MetadataSchema` class\n",
    "\n",
    "Also see the {ref}`sec_python_api_metadata` summary.\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: MetadataSchema\n",
    "    :members:\n",
    "    :inherited-members:\n",
    "```\n",
    "\n",
    "#### The {class}`TableMetadataSchemas` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: TableMetadataSchemas\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`TopologyCounter` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: TopologyCounter\n",
    "```\n",
    "\n",
    "#### The {class}`LdCalculator` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: LdCalculator\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`TableCollectionIndexes` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: TableCollectionIndexes\n",
    "    :members:\n",
    "```\n",
    "\n",
    "#### The {class}`SVGString` class\n",
    "\n",
    "```{eval-rst}\n",
    ".. autoclass:: SVGString\n",
    "    :members:\n",
    "    :private-members: _repr_svg_\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.14"
  },
  "source_map": [
   12,
   836,
   843,
   849,
   853,
   855,
   859,
   861,
   865,
   867,
   873,
   875,
   881,
   884,
   891,
   896,
   901,
   905,
   909,
   913,
   932,
   943,
   950,
   955,
   957,
   965,
   969,
   978,
   988,
   993,
   1000,
   1025,
   1030,
   1042,
   1045,
   1050,
   1052,
   1057,
   1060,
   1067,
   1070
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}