{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d6506a12",
   "metadata": {},
   "source": [
    "```{currentmodule} tskit\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da7f5850",
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c66403a",
   "metadata": {},
   "source": [
    "(sec_export)=\n",
    "\n",
    "# Data export\n",
    "\n",
    "(sec_export_vcf)=\n",
    "## Variant Call Format\n",
    "\n",
    "Tskit supports exporting data to the standard\n",
    "[Variant Call Format](http://samtools.github.io/hts-specs/VCFv4.3.pdf)\n",
    "via the `tskit vcf` {ref}`command line interface<sec_cli>` command\n",
    "and the {meth}`TreeSequence.write_vcf` method in the {ref}`sec_python_api`.\n",
    "Conversion is quite efficient, with tskit producing VCF data at several\n",
    "hundred megabytes per second (for large files), which is usually as fast as\n",
    "it can be written to storage or consumed by programs in a pipeline.\n",
    "\n",
    "::::{tip}\n",
    "If we have a tree sequence file the\n",
    "{ref}`command line interface<sec_cli>` is often the most\n",
    "convenient way to convert to VCF:\n",
    "\n",
    ":::{code-block} bash\n",
    "$ tskit vcf example.trees > example.vcf\n",
    ":::\n",
    "\n",
    "See the {ref}`sec_export_vcf_compression` section for information\n",
    "on how to compress the VCF output.\n",
    "::::\n",
    "\n",
    "For tree sequences produced by recent versions of programs such as\n",
    "``msprime``, ``SLiM``, ``fwdpy11`` or ``tsinfer``, VCF output will\n",
    "\"do the right thing\" and no further arguments are needed.\n",
    "For example, here we simulate 3 diploid individuals\n",
    "with mutations using ``msprime``, and convert to VCF."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "78468df6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##source=tskit 0.5.3\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##contig=<ID=1,length=10>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\ttsk_0\ttsk_1\ttsk_2\n",
      "1\t2\t0\tA\tT,G\t.\tPASS\t.\tGT\t0|0\t0|0\t2|2\n",
      "1\t3\t1\tC\tT\t.\tPASS\t.\tGT\t0|0\t0|1\t0|0\n",
      "1\t4\t2\tG\tC\t.\tPASS\t.\tGT\t1|0\t1|1\t0|0\n",
      "1\t6\t3\tC\tT\t.\tPASS\t.\tGT\t1|0\t1|1\t0|0\n",
      "1\t9\t4\tA\tC\t.\tPASS\t.\tGT\t0|0\t1|0\t0|0\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import msprime\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=3, ploidy=2, sequence_length=10, random_seed=2)\n",
    "ts = msprime.sim_mutations(ts, rate=0.1, random_seed=2)\n",
    "ts.write_vcf(sys.stdout)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d34556f",
   "metadata": {},
   "source": [
    "In the output VCF we have 3 diploid samples\n",
    "(see the {ref}`sec_export_vcf_terminology` section)\n",
    "corresponding to samples specified in the ancestry simulation\n",
    "with IDs ``tsk_0``, ``tsk_1`` and ``tsk_2``\n",
    "(see the {ref}`sec_export_vcf_individual_names`\n",
    "section for how to override these default labels).\n",
    "We then have a line for every row\n",
    "in the {ref}`site table<sec_site_table_definition>`, and\n",
    "the data is derived directly from the {meth}`TreeSequence.variants`\n",
    "method; e.g."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2d9cc422",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0\t0\t('A', 'T', 'G')\t[0 0 0 0 2 2]\n",
      "3.0\t1\t('C', 'T')\t[0 0 0 1 0 0]\n",
      "4.0\t2\t('G', 'C')\t[1 0 1 1 0 0]\n",
      "6.0\t3\t('C', 'T')\t[1 0 1 1 0 0]\n",
      "9.0\t4\t('A', 'C')\t[0 0 1 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "for var in ts.variants():\n",
    "    print(var.site.position, var.site.id, var.alleles, var.genotypes, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61eb5eb3",
   "metadata": {},
   "source": [
    "We can see the ``POS`` value is equal to the site's position\n",
    "(see the {ref}`sec_export_vcf_modifying_coordinates` for information\n",
    "on how we deal with continuous coordinates), the ``ID`` value\n",
    "is the site's ID, and the ``REF`` and ``ALT`` values\n",
    "are derived from the variant's ``alleles``.\n",
    "\n",
    "The ``GT`` values for the three diploid individuals are derived from the\n",
    "variant's genotypes (see the {ref}`sec_export_vcf_terminology` section).\n",
    "For this simulation, the diploid individuals correspond to\n",
    "adjacent sample nodes in order, and we can see there is a direct\n",
    "correspondence between the phased ``GT`` values and variant's genotypes.\n",
    "See the {ref}`sec_export_vcf_constructing_gt` section for\n",
    "more information on how this done in general and for options\n",
    "to control the VCF sample and ``GT`` values.\n",
    "\n",
    "::::{important}\n",
    "In these examples we write the VCF data to ``sys.stdout`` so that we can see\n",
    "the output. Usually, however, you'd write to a file:\n",
    "\n",
    ":::{code-block}\n",
    "with open(\"output.vcf\", \"w\") as vcf_file:\n",
    "    ts.write_vcf(vcf_file)\n",
    ":::\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_export_vcf_compression` section for information\n",
    "on how to compress the output or convert to BCF.\n",
    ":::\n",
    "\n",
    "::::\n",
    "\n",
    "(sec_export_vcf_terminology)=\n",
    "\n",
    "### Terminology\n",
    "\n",
    "There are some mismatches between the terminology for tskit and VCF.\n",
    "In VCF a \"sample\" is a multiploid individual, but in tskit a sample\n",
    "refers to a single **node** (monoploid genome), and an individual\n",
    "consists of one or more nodes (e.g., two nodes for a diploid).\n",
    "Similarly, in VCF a \"genotype\" refers to the observed allelic state\n",
    "for a sample **individual** at a particular site,\n",
    "whereas in tskit a genotype is the observed allelic state\n",
    "for a **node** (see {attr}`.Variant.genotypes`).\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_glossary` for more details on tskit's data model\n",
    " and terminology.\n",
    ":::\n",
    "\n",
    "(sec_export_vcf_compression)=\n",
    "\n",
    "### Compressed output\n",
    "\n",
    "The simplest way to compress the VCF output is to use the\n",
    "`tskit vcf` {ref}`command line interface<sec_cli>`\n",
    "and pipe the output to `bgzip`:\n",
    "\n",
    ":::{code-block} bash\n",
    "$ tskit vcf example.trees | bgzip -c > example.vcf.gz\n",
    ":::\n",
    "A general way to convert VCF data to various formats is to pipe the text\n",
    "produced by ``tskit`` into ``bcftools`` using the command\n",
    "line interface:\n",
    "\n",
    ":::{code-block} bash\n",
    "$ tskit vcf example.trees | bcftools view -O b > example.bcf\n",
    ":::\n",
    "\n",
    "If you need more control over the form of the output (or want to work\n",
    "directly in Python), the following recipe has the same effect:\n",
    "\n",
    ":::{code-block}\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "read_fd, write_fd = os.pipe()\n",
    "write_pipe = os.fdopen(write_fd, \"w\")\n",
    "with open(\"output.bcf\", \"w\") as bcf_file:\n",
    "    proc = subprocess.Popen(\n",
    "        [\"bcftools\", \"view\", \"-O\", \"b\"], stdin=read_fd, stdout=bcf_file\n",
    "    )\n",
    "    ts.write_vcf(write_pipe)\n",
    "    write_pipe.close()\n",
    "    os.close(read_fd)\n",
    "    proc.wait()\n",
    "    if proc.returncode != 0:\n",
    "        raise RuntimeError(\"bcftools failed with status:\", proc.returncode)\n",
    ":::\n",
    "\n",
    "\n",
    "The VCF output can also be compressed using the {mod}`gzip` Python module:\n",
    "\n",
    ":::{code-block}\n",
    "\n",
    "import gzip\n",
    "\n",
    "with gzip.open(\"output.vcf.gz\", \"wt\") as f:\n",
    "    ts.write_vcf(f)\n",
    ":::\n",
    "\n",
    "However, this gzipped VCF won't be fully compatible with downstream tools\n",
    "such as tabix, which usually require the VCF to use the specialised bgzip format.\n",
    "\n",
    "(sec_export_vcf_masking_output)=\n",
    "\n",
    "### Masking output\n",
    "\n",
    "The {meth}`TreeSequence.write_vcf` method provides the\n",
    "``site_mask`` and ``sample_mask`` arguments to\n",
    "omit or mark parts of the output as missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "318a4c5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>position</th><th>ancestral_state</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>2</td><td>A</td><td></td></tr>\n",
       "<tr><td>1</td><td>3</td><td>C</td><td></td></tr>\n",
       "<tr><td>2</td><td>4</td><td>G</td><td></td></tr>\n",
       "<tr><td>3</td><td>6</td><td>C</td><td></td></tr>\n",
       "<tr><td>4</td><td>9</td><td>A</td><td></td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n"
      ],
      "text/plain": [
       "<tskit.tables.SiteTable at 0x7efd8ac243d0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=3, ploidy=2, sequence_length=10, random_seed=2)\n",
    "ts = msprime.sim_mutations(ts, rate=0.1, random_seed=2)\n",
    "ts.tables.sites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533acb6a",
   "metadata": {},
   "source": [
    "The ``sample_mask`` argument provides a general way to mask out\n",
    "parts of the output, which can be helpful when simulating missing\n",
    "data. In this (contrived) example, we create a sample mask function\n",
    "that marks one genotype missing in each variant in a regular\n",
    "pattern:\n",
    "\n",
    ":::{code-block}\n",
    "\n",
    "def sample_mask(variant):\n",
    "    sample_mask = np.zeros(ts.num_samples, dtype=bool)\n",
    "    sample_mask[variant.site.id % ts.num_samples] = 1\n",
    "    return sample_mask\n",
    "\n",
    "\n",
    "ts.write_vcf(sys.stdout, sample_mask=sample_mask)\n",
    ":::\n",
    "\n",
    "(sec_export_vcf_constructing_gt)=\n",
    "\n",
    "### Constructing GT values\n",
    "\n",
    "The core elements of the tskit\n",
    "{ref}`data model<sec_data_model>`\n",
    "are {ref}`nodes<sec_node_table_definition>`,\n",
    "{ref}`edges<sec_node_table_definition>`,\n",
    "{ref}`sites<sec_site_table_definition>` and\n",
    "{ref}`mutations<sec_mutation_table_definition>`.\n",
    "These four tables allow us to completely describe the\n",
    "genetic ancestry of a set of sampled monoploid\n",
    "genomes and their genetic variation.\n",
    "The {ref}`individual table<sec_individual_table_definition>`\n",
    "defines a set of individual *organisms*, and it can\n",
    "be used to define the inheritance relationships between\n",
    "then (the pedigree). An individual may be associated\n",
    "with one or more nodes, and these nodes may or\n",
    "may not be samples (see the {ref}`sec_glossary`\n",
    "for clarification of these terms).\n",
    "Thus, there is some complexity in how the per-individual GT values\n",
    "are generated, which we explain in this section.\n",
    "\n",
    "#### Without individuals\n",
    "\n",
    "We start with an example in which there are no individuals\n",
    "defined (which was the default in msprime before version 1.0):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5dc95659",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\" /></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\" /><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g><g class=\"site s0\" transform=\"translate(68 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /><g class=\"mut m0\"><polyline class=\"sym\" points=\"2.5,-6.5 0,-1.5 -2.5,-6.5\" /></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 root\" transform=\"translate(80 26.8)\"><g class=\"a6 c2 node n4\" transform=\"translate(-30 47.3)\"><g class=\"a4 leaf m0 node n0 s0 sample\" transform=\"translate(-15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 15\" /><g class=\"mut m0 s0 unknown_time\" transform=\"translate(0 -23.65)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"23.65\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">0</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 leaf node n1 sample\" transform=\"translate(15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -15\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H 30\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><g class=\"a6 c2 node n5\" transform=\"translate(30 47.3)\"><g class=\"a5 leaf node n2 sample\" transform=\"translate(-15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 15\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 leaf node n3 sample\" transform=\"translate(15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -15\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H -30\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "'<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\" /></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\" /><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\" /><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g><g class=\"site s0\" transform=\"translate(68 163.2)\"><line class=\"sym\" x1=\"0\" x2=\"0\" y1=\"0\" y2=\"-10\" /><g class=\"mut m0\"><polyline class=\"sym\" points=\"2.5,-6.5 0,-1.5 -2.5,-6.5\" /></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 root\" transform=\"translate(80 26.8)\"><g class=\"a6 c2 node n4\" transform=\"translate(-30 47.3)\"><g class=\"a4 leaf m0 node n0 s0 sample\" transform=\"translate(-15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 15\" /><g class=\"mut m0 s0 unknown_time\" transform=\"translate(0 -23.65)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"23.65\" /><path class=\"sym\" d=\"M -3,-3 l 6,6 M -3,3 l 6,-6\" /><text class=\"lab lft\" transform=\"translate(-5 0)\">0</text></g><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 leaf node n1 sample\" transform=\"translate(15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -15\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H 30\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><g class=\"a6 c2 node n5\" transform=\"translate(30 47.3)\"><g class=\"a5 leaf node n2 sample\" transform=\"translate(-15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 15\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 leaf node n3 sample\" transform=\"translate(15 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -15\" /><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\" /><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H -30\" /><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\" /><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>'"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/export_9_0.svg"
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>1</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>10.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>4</td></tr>\n",
       "          <tr><td>Total Size</td><td>1.1 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>6</td>\n",
       "          <td>200 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>0</td>\n",
       "          <td>24 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>1</td>\n",
       "          <td>53 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>7</td>\n",
       "          <td>204 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>506 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>1</td>\n",
       "          <td>41 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7efdb8ca9550>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##source=tskit 0.5.3\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##contig=<ID=1,length=10>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\ttsk_0\ttsk_1\ttsk_2\ttsk_3\n",
      "1\t3\t0\tA\tT\t.\tPASS\t.\tGT\t1\t0\t0\t0\n"
     ]
    }
   ],
   "source": [
    "import tskit\n",
    "tables = tskit.Tree.generate_balanced(4, span=10).tree_sequence.dump_tables()\n",
    "tables.sites.add_row(3, ancestral_state=\"A\")\n",
    "tables.mutations.add_row(site=0, node=0, derived_state=\"T\")\n",
    "ts = tables.tree_sequence()\n",
    "display(ts.draw_svg())\n",
    "display(ts)\n",
    "ts.write_vcf(sys.stdout)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e82aeee",
   "metadata": {},
   "source": [
    "Here we define a tree sequence consisting of a single tree, which\n",
    "has a variant site at position 3 and a mutation over node 0.\n",
    "There is no information about individuals in this tree sequence,\n",
    "and so we assume that each of the nodes corresponds to a single\n",
    "haploid individual.\n",
    "\n",
    "Users of msprime simulations would often be interested in producing\n",
    "VCFs for diploid organisms. Because of the assumptions made\n",
    "by these simulations, this means arbitrarily combining the sample\n",
    "nodes into pairs. This is what the ``ploidy`` option does:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7525ef84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##source=tskit 0.5.3\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##contig=<ID=1,length=10>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\ttsk_0\ttsk_1\n",
      "1\t3\t0\tA\tT\t.\tPASS\t.\tGT\t1|0\t0|0\n"
     ]
    }
   ],
   "source": [
    "ts.write_vcf(sys.stdout, ploidy=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d6451c5",
   "metadata": {},
   "source": [
    "Thus, the ``GT`` values for the (synthetic) diploid individual ``tsk_0``\n",
    "are generated by combining nodes 0 and 1, and ``tsk_1``\n",
    "by combining nodes 2 and 3.\n",
    "\n",
    ":::{important}\n",
    "Software packages modelling multiploid individuals are encouraged to\n",
    "use the individual table to make their assumptions explicit. Recent\n",
    "versions of simulators and inference methods should all do this,\n",
    "and so the ``ploidy`` argument is really only intended to support\n",
    "legacy code. It is therefore an error to supply a value for ``ploidy``\n",
    "when individual information is present in a tree sequence.\n",
    ":::\n",
    "\n",
    "#### With individuals\n",
    "\n",
    "Extending the example in the previous section, we add some individual data\n",
    "defining a pair of diploid sibs and their parents.\n",
    "\n",
    ":::{note}\n",
    "We set the nodes for (e.g.) individual 2 to [1, 3] here to illustrate\n",
    "that nodes for a given individual are not necessarily contiguous.\n",
    ":::"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "57694e49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>flags</th><th>location</th><th>parents</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>0</td><td></td><td>-1, -1</td><td></td></tr>\n",
       "<tr><td>1</td><td>0</td><td></td><td>-1, -1</td><td></td></tr>\n",
       "<tr><td>2</td><td>0</td><td></td><td>0, 1</td><td></td></tr>\n",
       "<tr><td>3</td><td>0</td><td></td><td>0, 1</td><td></td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n"
      ],
      "text/plain": [
       "<tskit.tables.IndividualTable at 0x7efd8abda460>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>1</td><td>-1</td><td>3</td><td>0</td><td></td></tr>\n",
       "<tr><td>1</td><td>1</td><td>-1</td><td>2</td><td>0</td><td></td></tr>\n",
       "<tr><td>2</td><td>1</td><td>-1</td><td>3</td><td>0</td><td></td></tr>\n",
       "<tr><td>3</td><td>1</td><td>-1</td><td>2</td><td>0</td><td></td></tr>\n",
       "<tr><td>4</td><td>0</td><td>-1</td><td>-1</td><td>1</td><td></td></tr>\n",
       "<tr><td>5</td><td>0</td><td>-1</td><td>-1</td><td>1</td><td></td></tr>\n",
       "<tr><td>6</td><td>0</td><td>-1</td><td>-1</td><td>2</td><td></td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n"
      ],
      "text/plain": [
       "<tskit.tables.NodeTable at 0x7efd8abda520>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##source=tskit 0.5.3\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##contig=<ID=1,length=10>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\ttsk_0\ttsk_1\n",
      "1\t3\t0\tA\tT\t.\tPASS\t.\tGT\t0|0\t1|0\n"
     ]
    }
   ],
   "source": [
    "tables.individuals.add_row(parents=[-1, -1])\n",
    "tables.individuals.add_row(parents=[-1, -1])\n",
    "tables.individuals.add_row(parents=[0, 1])\n",
    "tables.individuals.add_row(parents=[0, 1])\n",
    "node_individual = tables.nodes.individual\n",
    "node_individual[[1, 3]] = 2\n",
    "node_individual[[0, 2]] = 3\n",
    "tables.nodes.individual = node_individual\n",
    "display(tables.individuals)\n",
    "display(tables.nodes)\n",
    "ts = tables.tree_sequence()\n",
    "ts.write_vcf(sys.stdout)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a839b53e",
   "metadata": {},
   "source": [
    "In this model we have four individuals defined, but only\n",
    "individuals 2 and 3 are associated with nodes (more specifically,\n",
    "**sample** nodes). Thus, we output **two** VCF sample individuals\n",
    "composed of the linked nodes.\n",
    "\n",
    ":::{note}\n",
    "Note that the labels are ``tsk_0`` and ``tsk_1`` even though\n",
    "the individual IDs are 2 and 3. See the\n",
    "{ref}`sec_export_vcf_individual_names` section for how to change the\n",
    "these default labels.\n",
    ":::\n",
    "\n",
    "(sec_export_vcf_individual_names)=\n",
    "\n",
    "### Individual names\n",
    "\n",
    "By default the VCF samples are given the labels ``tsk_0``, ``tsk_1``,\n",
    "..., ``tsk_{N - 1}``, where ``N`` is the number of individuals to\n",
    "be output (see the {ref}`sec_export_vcf_constructing_gt` section).\n",
    "\n",
    "We can change this default labelling using the ``individual_names``\n",
    "argument::"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d12de5ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##source=tskit 0.5.3\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##contig=<ID=1,length=10>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tA\tB\tC\n",
      "1\t2\t0\tA\tT,G\t.\tPASS\t.\tGT\t0|0\t0|0\t2|2\n",
      "1\t3\t1\tC\tT\t.\tPASS\t.\tGT\t0|0\t0|1\t0|0\n",
      "1\t4\t2\tG\tC\t.\tPASS\t.\tGT\t1|0\t1|1\t0|0\n",
      "1\t6\t3\tC\tT\t.\tPASS\t.\tGT\t1|0\t1|1\t0|0\n",
      "1\t9\t4\tA\tC\t.\tPASS\t.\tGT\t0|0\t1|0\t0|0\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import msprime\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=3, ploidy=2, sequence_length=10, random_seed=2)\n",
    "ts = msprime.sim_mutations(ts, rate=0.1, random_seed=2)\n",
    "ts.write_vcf(sys.stdout, individual_names=[\"A\", \"B\", \"C\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efdf2cd1",
   "metadata": {},
   "source": [
    "#### Exporting to plink\n",
    "\n",
    "The default VCF sample IDs produced by ``tskit`` do not work well\n",
    "with plink because it parses the individual\n",
    "IDs based on a particular format, and does not allow ``0`` as a valid\n",
    "identifier. We get an error like this:\n",
    "\n",
    "```\n",
    "Error: Sample ID ends with \"_0\", which induces an invalid IID of '0`.\n",
    "```\n",
    "\n",
    "This can be fixed by using the ``individual_names`` argument\n",
    "to set the names to anything where the first name doesn't end with ``_0``.\n",
    "An example implementation for diploid individuals is:\n",
    "\n",
    ":::{code-block}\n",
    "n_dip_indv = int(ts.num_samples / 2)\n",
    "indv_names = [f\"tsk_{i}indv\" for i in range(n_dip_indv)]\n",
    "with open(\"output.vcf\", \"w\") as vcf_file:\n",
    "    ts.write_vcf(vcf_file, individual_names=indv_names)\n",
    ":::\n",
    "\n",
    "Adding a second ``_`` (eg: ``tsk_0_indv``) is not recommended as\n",
    "``plink`` uses ``_`` as the default separator for separating family\n",
    "id and individual id, and two underscores will throw an error.\n",
    "\n",
    "(sec_export_vcf_modifying_coordinates)=\n",
    "\n",
    "### Modifying coordinates\n",
    "\n",
    "Tskit supports continuous genome coordinates, but VCF only supports discrete\n",
    "genome positions. Thus, when we convert a tree sequence that has sites\n",
    "at continuous positions we must discretise the coordinates in some way.\n",
    "\n",
    "The ``position_transform`` argument provides a way to flexibly translate\n",
    "the genomic location of sites in tskit to the appropriate value in VCF.\n",
    "There are two fundamental differences in the way that tskit and VCF define\n",
    "genomic coordinates. The first is that tskit uses floating point values\n",
    "to encode positions, whereas VCF uses integers. Thus, if the tree sequence\n",
    "contains positions at non-integral locations there is an information loss\n",
    "incurred by translating to VCF. By default, we round the site positions\n",
    "to the nearest integer, such that there may be several sites with the\n",
    "same integer position in the output. The second difference between VCF\n",
    "and tskit is that VCF is defined to be a 1-based coordinate system, whereas\n",
    "tskit uses 0-based. However, how coordinates are transformed depends\n",
    "on the VCF parser, and so we do **not** account for this change in\n",
    "coordinate system by default.\n",
    "\n",
    ":::{note}\n",
    "The msprime 0.x legacy API simulates using continuous coordinates. It may\n",
    "be simpler to update your code to use the msprime 1.0 API (which uses\n",
    "discrete coordinates by default) than to work out how to transform\n",
    "coordinates in a way that is suitable for your application.\n",
    ":::\n",
    "\n",
    ":::{todo}\n",
    "Provide some examples of using position transform.\n",
    ":::"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.14"
  },
  "source_map": [
   12,
   17,
   21,
   57,
   64,
   77,
   80,
   194,
   199,
   246,
   255,
   268,
   270,
   295,
   308,
   333,
   340
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}