{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "84f20ac5",
   "metadata": {},
   "source": [
    ":::{currentmodule} tskit\n",
    ":::\n",
    "\n",
    "\n",
    "(sec_glossary)=\n",
    "\n",
    "# Glossary\n",
    "\n",
    "(sec_data_model_definitions)=\n",
    "\n",
    "## Definitions\n",
    "\n",
    "Here are some definitions of some key ideas encountered in this documentation.\n",
    "\n",
    "(sec_data_model_definitions_tree)=\n",
    "\n",
    "tree\n",
    ": A \"gene tree\", i.e., the genealogical tree describing how a collection of\n",
    "  genomes (usually at the tips of the tree) are related to each other at some\n",
    "  chromosomal location. See {ref}`sec_nodes_or_individuals` for discussion\n",
    "  of what a \"genome\" is.\n",
    "\n",
    "(sec_data_model_definitions_tree_sequence)=\n",
    "\n",
    "tree sequence\n",
    ": A \"succinct tree sequence\" (or tree sequence, for brevity) is an efficient\n",
    "  encoding of a sequence of correlated trees, such as one encounters looking\n",
    "  at the gene trees along a genome. A tree sequence efficiently captures the\n",
    "  structure shared by adjacent trees, (essentially) storing only what differs\n",
    "  between them.\n",
    "\n",
    "(sec_data_model_definitions_node)=\n",
    "\n",
    "node\n",
    ": Each branching point in each tree is associated with a particular genome\n",
    "  in a particular ancestor, called a \"node\".  Since each node represents a\n",
    "  specific genome it has a unique `time`, thought of as its birth time,\n",
    "  which determines the height of any branching points it is associated with.\n",
    "  See {ref}`sec_nodes_or_individuals` for discussion of what a \"node\" is.\n",
    "\n",
    "(sec_data_model_definitions_individual)=\n",
    "\n",
    "individual\n",
    ": In certain situations we are interested in how nodes (representing\n",
    "  individual homologous genomes) are grouped together into individuals\n",
    "  (e.g. two nodes per diploid individual). For example, when we are working\n",
    "  with polyploid samples it is useful to associate metadata with a specific\n",
    "  individual rather than duplicate this information on the constituent nodes.\n",
    "  See {ref}`sec_nodes_or_individuals` for more discussion on this point.\n",
    "\n",
    "(sec_data_model_definitions_sample)=\n",
    "\n",
    "sample\n",
    ": The focal nodes of a tree sequence, usually thought of as those from which\n",
    "  we have obtained data. The specification of these affects various\n",
    "  methods: (1) {meth}`TreeSequence.variants` and\n",
    "  {meth}`TreeSequence.haplotypes` will output the genotypes of the samples,\n",
    "  and {attr}`Tree.roots` only return roots ancestral to at least one\n",
    "  sample.\n",
    "  (This can be checked with {meth}`~Node.is_sample`;\n",
    "  see the {ref}`node table definitions <sec_node_table_definition>`\n",
    "  for information on how the sample\n",
    "  status a node is encoded in the `flags` column.)\n",
    "\n",
    "(sec_data_model_definitions_edge)=\n",
    "\n",
    "edge\n",
    ": The topology of a tree sequence is defined by a set of **edges**. Each\n",
    "  edge is a tuple `(left, right, parent, child)`, which records a\n",
    "  parent-child relationship among a pair of nodes on the\n",
    "  on the half-open interval of chromosome `[left, right)`.\n",
    "\n",
    "(sec_data_model_definitions_site)=\n",
    "\n",
    "site\n",
    ": Tree sequences can define the mutational state of nodes as well as their\n",
    "  topological relationships. A **site** is thought of as some position along\n",
    "  the genome at which variation occurs. Each site is associated with\n",
    "  a unique position and ancestral state.\n",
    "\n",
    "(sec_data_model_definitions_mutation)=\n",
    "\n",
    "mutation\n",
    ": A mutation records the change of state at a particular site 'above'\n",
    "  a particular node (more precisely, along the branch between the node\n",
    "  in question and its parent). Each mutation is associated with a specific\n",
    "  site (which defines the position along the genome), a node (which defines\n",
    "  where it occurs within the tree at this position), and a derived state\n",
    "  (which defines the mutational state inherited by all nodes in the subtree\n",
    "  rooted at the focal node). In more complex situations in which we have\n",
    "  back or recurrent mutations, a mutation must also specify its 'parent'\n",
    "  mutation.\n",
    "\n",
    "(sec_data_model_definitions_migration)=\n",
    "\n",
    "migration\n",
    ": An event at which a parent and child node were born in different populations.\n",
    "\n",
    "(sec_data_model_definitions_population)=\n",
    "\n",
    "population\n",
    ": A grouping of nodes, e.g., by sampling location.\n",
    "\n",
    "(sec_data_model_definitions_provenance)=\n",
    "\n",
    "provenance\n",
    ": An entry recording the origin and history of the data encoded in a tree sequence.\n",
    "\n",
    "(sec_data_model_definitions_ID)=\n",
    "\n",
    "ID\n",
    ": In the set of interconnected tables that we define here, we refer\n",
    "  throughout to the IDs of particular entities. The ID of an\n",
    "  entity (e.g., a node) is defined by the position of the corresponding\n",
    "  row in the table. These positions are zero indexed. For example, if we\n",
    "  refer to node with ID zero, this corresponds to the node defined by the\n",
    "  first row in the node table.\n",
    "\n",
    "(sec_data_model_definitions_sequence_length)=\n",
    "\n",
    "sequence length\n",
    ": This value defines the coordinate space in which the edges and site positions\n",
    "  are defined. This is most often assumed to be equal to the largest\n",
    "  `right` coordinate in the edge table, but there are situations in which\n",
    "  we might wish to specify the sequence length explicitly.\n",
    "\n",
    "## Further discussion\n",
    "\n",
    "(sec_nodes_or_individuals)=\n",
    "\n",
    "### Nodes, Genomes, or Individuals?\n",
    "\n",
    "The natural unit of biological analysis is (usually) the *individual*. However,\n",
    "many organisms we study are diploid, and so each individual contains *two*\n",
    "homologous copies of the entire genome, separately inherited from the two\n",
    "parental individuals. Since each monoploid copy of the genome is inherited separately,\n",
    "each diploid individual lies at the end of two distinct lineages, and so will\n",
    "be represented by *two* places in any given genealogical tree. This makes it\n",
    "difficult to precisely discuss tree sequences for diploids, as we have no\n",
    "simple way to refer to the bundle of chromosomes that make up the \"copy of the\n",
    "genome inherited from one particular parent\". For this reason, in this\n",
    "documentation we use the non-descriptive term \"node\" to refer to this concept\n",
    "-- and so, a diploid individual is composed of two nodes -- although we use the\n",
    "term \"genome\" at times, for concreteness.\n",
    "\n",
    "Several properties naturally associated with individuals are in fact assigned\n",
    "to nodes in what follows: birth time and population. This is for two reasons:\n",
    "First, since coalescent simulations naturally lack a notion of polyploidy, earlier\n",
    "versions of `tskit` lacked the notion of an individual. Second, ancestral\n",
    "nodes are not naturally grouped together into individuals -- we know they must have\n",
    "existed, but have no way of inferring this grouping, so in fact many nodes in\n",
    "an empirically-derived tree sequence will not be associated with individuals,\n",
    "even though their birth times might be inferred."
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.14"
  },
  "source_map": [
   12
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}