{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6326ee53",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2267/167322575.py:7: DeprecationWarning: `set_matplotlib_formats` is deprecated since IPython 7.23, directly use `matplotlib_inline.backend_inline.set_matplotlib_formats()`\n",
      "  set_matplotlib_formats(\"svg\")\n"
     ]
    }
   ],
   "source": [
    "import msprime\n",
    "from IPython.display import SVG, set_matplotlib_formats\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "\n",
    "set_matplotlib_formats(\"svg\")\n",
    "\n",
    "\n",
    "# Note: duplicating this from the version used in the pedigrees.md file.\n",
    "def draw_pedigree(ped_ts):\n",
    "\n",
    "    G = nx.DiGraph()\n",
    "    for ind in ped_ts.individuals():\n",
    "        time = ped_ts.node(ind.nodes[0]).time\n",
    "        pop = ped_ts.node(ind.nodes[0]).population\n",
    "        G.add_node(ind.id, time=time, population=pop)\n",
    "        for p in ind.parents:\n",
    "            if p != tskit.NULL:\n",
    "                G.add_edge(ind.id, p)\n",
    "    pos = nx.multipartite_layout(G, subset_key=\"time\", align=\"horizontal\")\n",
    "    colours = plt.rcParams['axes.prop_cycle'].by_key()['color']\n",
    "    node_colours = [colours[node_attr[\"population\"]] for node_attr in G.nodes.values()]\n",
    "    nx.draw_networkx(G, pos, with_labels=True, node_color=node_colours)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13c620b7",
   "metadata": {},
   "source": [
    "(sec_ancestry)=\n",
    "\n",
    "# Ancestry simulations\n",
    "\n",
    "Msprime simulates the ancestry of sampled genomes under a number of\n",
    "different backwards-in-time population genetic\n",
    "{ref}`models <sec_ancestry_models>`. The simulated ancestry is\n",
    "represented as a [succinct tree sequence](https://tskit.dev) using\n",
    "the [tskit](https://tskit.dev/tskit) library, which provides an\n",
    "extensive suite of operations for analysing genealogical trees\n",
    "and DNA sequence data. See the {ref}`tutorials:sec_tskit_getting_started`\n",
    "tutorial for more information on how to process data using tskit.\n",
    "\n",
    "Here we run a simple simulation and show a summary of the resulting\n",
    "tree sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9886293e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>1</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>4</td></tr>\n",
       "          <tr><td>Total Size</td><td>1.7 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>6</td>\n",
       "          <td>200 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>2</td>\n",
       "          <td>80 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>7</td>\n",
       "          <td>204 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>968 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2f05654f0>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2)\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afab5567",
   "metadata": {},
   "source": [
    "The {func}`.sim_ancestry` function has many parameters to specify\n",
    "the details of the simulations. The {func}`API documentation <.sim_ancestry>`\n",
    "contains a precise description of the parameter; the sections\n",
    "in this page provide some explanation and examples of how to use these\n",
    "parameters.\n",
    "\n",
    ":::{note}\n",
    "It's important to note that {func}`.sim_ancestry` only simulates\n",
    "the ancestral trees for samples: if we want actual sequence data\n",
    "then we also need to simulate mutations on these trees.\n",
    "See the {ref}`sec_mutations` section for more information on how to do\n",
    "this.\n",
    ":::\n",
    "\n",
    "---\n",
    "\n",
    "(sec_ancestry_quickref)=\n",
    "\n",
    "## Quick reference\n",
    "\n",
    "{func}`.sim_ancestry`\n",
    ": Simulate ancestral tree sequence topology\n",
    "\n",
    "**Models**\n",
    "\n",
    "{class}`.StandardCoalescent`\n",
    ": Coalescent with recombination (\"hudson\")\n",
    "\n",
    "{class}`.SmcApproxCoalescent`\n",
    ": Sequentially Markov Coalescent (\"smc\")\n",
    "\n",
    "{class}`.SmcPrimeApproxCoalescent`\n",
    ": SMC'(\"smc_prime\")\n",
    "\n",
    "{class}`.DiscreteTimeWrightFisher`\n",
    ": Generation-by-generation Wright-Fisher\n",
    "\n",
    "{class}`.FixedPedigree`\n",
    ": Simulated conditioned on a given pedigree\n",
    "\n",
    "{class}`.BetaCoalescent`\n",
    ": Beta coalescent multiple-merger\n",
    "\n",
    "{class}`.DiracCoalescent`\n",
    ": Dirac coalescent multiple-merger\n",
    "\n",
    "{class}`.SweepGenicSelection`\n",
    ": Selective sweep at a linked locus\n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "(sec_ancestry_samples)=\n",
    "\n",
    "## Specifying samples\n",
    "\n",
    "The `samples` argument to {func}`.sim_ancestry` defines the number\n",
    "of sampled individuals whose history will be simulated. There are three different\n",
    "forms; the `samples` argument can be:\n",
    "\n",
    "- an **integer**, interpreted as the number of sampled individuals to draw in\n",
    "  a single population model;\n",
    "- a **dictionary** mapping population references (either integer IDs or\n",
    "  names) to the number of sampled individuals for that population;\n",
    "- a list of {class}`.SampleSet` objects, which provide more flexibility\n",
    "  in how groups of similar samples are drawn from populations.\n",
    "\n",
    "\n",
    "In the simplest case, provide a single integer which defines the number\n",
    "of samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "49f9ed28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>1</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>4</td></tr>\n",
       "          <tr><td>Total Size</td><td>1.7 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>6</td>\n",
       "          <td>200 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>2</td>\n",
       "          <td>80 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>7</td>\n",
       "          <td>204 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>963 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2f0565820>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, random_seed=1234)\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d7078f",
   "metadata": {},
   "source": [
    "Here we specify two sample **individuals**, which at the default\n",
    "ploidy of two, gives us four sample **nodes**.\n",
    "\n",
    ":::{warning}\n",
    "It is important to note that the number of samples\n",
    "refers to the number of *individuals* not the number of *nodes*\n",
    "(monoploid genomes). See the {ref}`sec_ancestry_ploidy`\n",
    "section for details.\n",
    ":::\n",
    "\n",
    "This integer form for the ``samples`` argument can only be used\n",
    "when the ``demography`` is a single population model.\n",
    "\n",
    "### Populations\n",
    "\n",
    "The next example illustrates one usage of the dictionary form of the `samples`\n",
    "argument, in which the keys refer to populations in the\n",
    "{ref}`demographic model<sec_demography>` and\n",
    "the values are the number of samples to draw.\n",
    "(See the {ref}`sec_ancestry_demography` section for more details on\n",
    "the ``demography`` argument to {func}`.sim_ancestry`.)\n",
    "We first create a {class}`.Demography` object representing\n",
    "a 10 deme linear stepping stone model. Then, we run the simulation\n",
    "with 3 diploid samples each drawn from the first and last demes in this\n",
    "linear habitat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b632059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>1</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>12</td></tr>\n",
       "          <tr><td>Total Size</td><td>5.5 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>22</td>\n",
       "          <td>712 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>6</td>\n",
       "          <td>192 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>23</td>\n",
       "          <td>652 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>10</td>\n",
       "          <td>593 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>3.2 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2f059f730>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N = 10\n",
    "demography = msprime.Demography.stepping_stone_model(\n",
    "    [100] * N,\n",
    "    migration_rate=0.1,\n",
    "    boundaries=True)\n",
    "ts = msprime.sim_ancestry({0: 3, N - 1: 3}, demography=demography)\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b34141",
   "metadata": {},
   "source": [
    "The keys in the dictionary can also be the string names of the\n",
    "population (see the {ref}`sec_demography_populations_identifiers` section),\n",
    "which is useful when we are simulating from empirically\n",
    "estimated models. For example, here create a {class}`.Demography` object\n",
    "based on a species tree, and then draw samples using the species names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0eaef222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>1</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>12</td></tr>\n",
       "          <tr><td>Total Size</td><td>5.0 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>22</td>\n",
       "          <td>712 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>6</td>\n",
       "          <td>192 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>23</td>\n",
       "          <td>652 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>7</td>\n",
       "          <td>481 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>2.7 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2f0565af0>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography.from_species_tree(\n",
    "    \"(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)\",\n",
    "    time_units=\"myr\",\n",
    "    initial_size=10**4,\n",
    "    generation_time=20)\n",
    "ts = msprime.sim_ancestry({\"gorilla\": 2, \"human\": 4}, demography=demography)\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd3f529d",
   "metadata": {},
   "source": [
    "When samples are drawn using this dictionary form it is always\n",
    "at the population's {ref}`sec_demography_populations_default_sampling_time`.\n",
    "If you wish to draw samples at a different time, then you must use the\n",
    "more general form of a list of {class}`.SampleSet` objects.\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_ancestry_demography` section for more information\n",
    "on how to specify demographic models in a simulation.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_samples_sampling_time)=\n",
    "\n",
    "### Sampling time\n",
    "\n",
    "By default the samples that we draw from a {class}`.Population` are\n",
    "at the population's {attr}`~.Population.default_sampling_time`.\n",
    "See the {ref}`sec_demography_populations_default_sampling_time`\n",
    "section for more information about how this is\n",
    "set for populations in a demographic model.\n",
    "\n",
    "We can manually control the time at which samples\n",
    "are drawn using list of {class}`.SampleSet` objects form for the\n",
    "samples argument.\n",
    "\n",
    "In this example we create two diploid sample individuals,\n",
    "one at the present time\n",
    "and one taken 50 generations ago, representing one modern\n",
    "and one ancient individual. By default, when we draw the\n",
    "trees, the nodes that belong to samples are drawn as squares\n",
    "whereas nodes that are not samples are drawn as circles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa5aca0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 108.704)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">50.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 95.2261)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">103.08</text></g></g><g class=\"tick\" transform=\"translate(56.8 72.9357)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">190.86</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">372.55</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.6 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(-20.8 68.4261)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 26.1739)\"><path class=\"edge\" d=\"M 0 0 V -26.1739 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(10.4 13.4776)\"><path class=\"edge\" d=\"M 0 0 V -13.4776 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -68.4261 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(20.8 46.1357)\"><g class=\"a5 i0 leaf node n1 p0 sample\" transform=\"translate(-10.4 48.4643)\"><path class=\"edge\" d=\"M 0 0 V -48.4643 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(10.4 35.768)\"><path class=\"edge\" d=\"M 0 0 V -35.768 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -46.1357 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 6,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_10_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples = [\n",
    "        msprime.SampleSet(1),\n",
    "        msprime.SampleSet(1, time=50)\n",
    "    ],\n",
    "    population_size=100,\n",
    "    random_seed=42\n",
    ")\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca9adc65",
   "metadata": {},
   "source": [
    ":::{seealso}\n",
    "See the {ref}`sec_ancestry_event_order` section for details\n",
    "of the relative order in which sampling events occur.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_samples_sampling_details)=\n",
    "\n",
    "### Sample details\n",
    "\n",
    "Sample individuals and nodes are allocated sequentially in the order that\n",
    "they are specified. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d813c1d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤═════╤════════╤═══════╤════════╗\n",
      "║id│flags│location│parents│metadata║\n",
      "╠══╪═════╪════════╪═══════╪════════╣\n",
      "║0 │    0│        │       │        ║\n",
      "║1 │    0│        │       │        ║\n",
      "║2 │    0│        │       │        ║\n",
      "╚══╧═════╧════════╧═══════╧════════╝\n",
      "\n",
      "╔══╤═════╤══════════╤══════════╤════╤════════╗\n",
      "║id│flags│population│individual│time│metadata║\n",
      "╠══╪═════╪══════════╪══════════╪════╪════════╣\n",
      "║0 │    1│         1│         0│   0│        ║\n",
      "║1 │    1│         1│         0│   0│        ║\n",
      "║2 │    1│         0│         1│   0│        ║\n",
      "║3 │    1│         0│         1│   0│        ║\n",
      "║4 │    1│         0│         2│   0│        ║\n",
      "║5 │    1│         0│         2│   0│        ║\n",
      "╚══╧═════╧══════════╧══════════╧════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "demography = msprime.Demography.island_model([10, 10], migration_rate=1)\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=[\n",
    "        msprime.SampleSet(1, population=1),\n",
    "        msprime.SampleSet(2, population=0)],\n",
    "    demography=demography,\n",
    "    end_time=0)\n",
    "print(ts.tables.individuals)\n",
    "print(ts.tables.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d372b03d",
   "metadata": {},
   "source": [
    "(Because we're only interested in the sampled nodes and individuals we\n",
    "stopped the simulation from actually doing anything by setting\n",
    "`end_time=0` --- see the {ref}`sec_ancestry_start_time` section\n",
    "for more information.) Here we define three sample individuals,\n",
    "and we therefore have three rows in the individual table.\n",
    "Because these are diploid individuals, the node table contains\n",
    "six sample nodes. If we look at the `individual` column in the\n",
    "node table we can see that the first two nodes correspond to individual\n",
    "`0`, the next two nodes individual `1`, etc. The sample configuration\n",
    "stated that the first sample should come from population `1` and\n",
    "the other two from population `0`, and we can see this reflected\n",
    "in the `population` column of the node table. (Somewhat confusingly,\n",
    "population values are associated with nodes rather than individuals;\n",
    "this is mostly for historical reasons.)\n",
    "\n",
    "The {class}`.SampleSet` class has a number of attributes which default\n",
    "to `None`. If these are set they will **override** the values\n",
    "which might be specified elsewhere. For example, we can specify\n",
    "mixed ploidy samples via the `ploidy` attribute of {class}`.SampleSet`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e4b29c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤═════╤════════╤═══════╤════════╗\n",
      "║id│flags│location│parents│metadata║\n",
      "╠══╪═════╪════════╪═══════╪════════╣\n",
      "║0 │    0│        │       │        ║\n",
      "║1 │    0│        │       │        ║\n",
      "║2 │    0│        │       │        ║\n",
      "╚══╧═════╧════════╧═══════╧════════╝\n",
      "\n",
      "╔══╤═════╤══════════╤══════════╤════╤════════╗\n",
      "║id│flags│population│individual│time│metadata║\n",
      "╠══╪═════╪══════════╪══════════╪════╪════════╣\n",
      "║0 │    1│         0│         0│   0│        ║\n",
      "║1 │    1│         0│         0│   0│        ║\n",
      "║2 │    1│         0│         0│   0│        ║\n",
      "║3 │    1│         0│         1│   0│        ║\n",
      "║4 │    1│         0│         2│   0│        ║\n",
      "╚══╧═════╧══════════╧══════════╧════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=[\n",
    "        msprime.SampleSet(1, ploidy=3),\n",
    "        msprime.SampleSet(2)],\n",
    "    ploidy=1,\n",
    "    end_time=0)\n",
    "print(ts.tables.individuals)\n",
    "print(ts.tables.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b9e27fc",
   "metadata": {},
   "source": [
    "(Again, we stop the simulation immediately because we're only interested\n",
    "in the initial samples.) Here we have three sampled individuals again\n",
    "but in this they are mixed ploidy: the first individual is triploid\n",
    "and the other two are haploid.\n",
    "\n",
    ":::{warning}\n",
    "It is vital to note that setting the `ploidy` value\n",
    "of {class}`.SampleSet` objects only affects sampling and does\n",
    "not affect the actual simulation. In this case, the simulation\n",
    "will be run on the haploid time scale. See the\n",
    "{ref}`sec_ancestry_ploidy` section for more details.\n",
    ":::\n",
    "\n",
    "If you wish to set up the node and individual IDs in some other way,\n",
    "it is possible to do so by using the ``initial_state`` parameter.\n",
    "See the  {ref}`sec_ancestry_initial_state` for more information\n",
    "on how to use this (advanced) feature.\n",
    "\n",
    "(sec_ancestry_demography)=\n",
    "\n",
    "## Demography\n",
    "\n",
    "A {ref}`demographic model<sec_demography>` is a description of a\n",
    "set of populations, the migration rates between then and the\n",
    "events that modify the populations over time. Ancestry simulations\n",
    "can take one of these models as input via the ``demography``\n",
    "parameter to {func}`.sim_ancestry`.\n",
    "\n",
    ":::{seealso}\n",
    "Please see the {ref}`sec_demography` section for details on how\n",
    "to create and debug demographic models, and the\n",
    "{ref}`sec_ancestry_samples` section for help on how to\n",
    "define the locations and times of sampled individuals.\n",
    ":::\n",
    "\n",
    "We don't need to explicitly specify a {class}`.Demography`\n",
    "for every ancestry simulation, however. By default, we assume a\n",
    "single fixed-size population. For the default\n",
    "{ref}`standard coalescent<sec_ancestry_models_hudson>` ancestry\n",
    "model we assume a population size of 1, so that time is measured\n",
    "in \"coalescent units\". This is useful for theoretical work where\n",
    "we usually work in scaled coalescent time.\n",
    "\n",
    ":::{important}\n",
    "By default msprime uses the **diploid** coalescent time scale;\n",
    "see the {ref}`sec_ancestry_ploidy` section for more details.\n",
    ":::\n",
    "\n",
    "Outside of theoretical analysis, working with scaled\n",
    "coalescent time can be confusing and it's usually better to\n",
    "explicitly set the population size so that the output\n",
    "times are directly interpretable in generations.\n",
    "The ``population_size`` parameter to {func}`.sim_ancestry`\n",
    "allows us to set the size of the single constant\n",
    "sized population in the default demography.\n",
    "\n",
    "For example, here we run the same simulation twice, one with\n",
    "the default population size and one with a population size\n",
    "of 100:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "918513c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 119.835)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.07</text></g></g><g class=\"tick\" transform=\"translate(56.8 94.7646)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1.21</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4.28</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.6 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(20.8 93.035)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -93.035 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-20.8 67.9646)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -67.9646 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 9,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_16_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, random_seed=1234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "39097cb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 119.835)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">7.09</text></g></g><g class=\"tick\" transform=\"translate(56.8 94.7646)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">120.61</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">428.37</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.6 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(20.8 93.035)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -93.035 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-20.8 67.9646)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -67.9646 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 10,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_17_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, population_size=100, random_seed=1234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ca3de4c",
   "metadata": {},
   "source": [
    "We can see that the results are identical, except the node times have\n",
    "been scaled by the population size.\n",
    "\n",
    "The {ref}`sec_ancestry_models_dtwf` model is different. Because it\n",
    "explicitly works with a population of N individuals we get an\n",
    "error if we don't specify a population size:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3e37dae2",
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "When using the DTWF model, the population size must be set explicitly, either using the population_size or demography arguments.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mmsprime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msim_ancestry\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdtwf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrandom_seed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1234\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1207\u001b[0m, in \u001b[0;36msim_ancestry\u001b[0;34m(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, num_labels, random_seed, num_replicates, replicate_index, record_provenance)\u001b[0m\n\u001b[1;32m   1184\u001b[0m     parameters \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\n\u001b[1;32m   1185\u001b[0m         command\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msim_ancestry\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   1186\u001b[0m         samples\u001b[38;5;241m=\u001b[39msamples,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1204\u001b[0m         \u001b[38;5;66;03m# replicate index is excluded as it is inserted for each replicate\u001b[39;00m\n\u001b[1;32m   1205\u001b[0m     )\n\u001b[1;32m   1206\u001b[0m     provenance_dict \u001b[38;5;241m=\u001b[39m provenance\u001b[38;5;241m.\u001b[39mget_provenance_dict(parameters)\n\u001b[0;32m-> 1207\u001b[0m sim \u001b[38;5;241m=\u001b[39m \u001b[43m_parse_sim_ancestry\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1208\u001b[0m \u001b[43m    \u001b[49m\u001b[43msamples\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msamples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1209\u001b[0m \u001b[43m    \u001b[49m\u001b[43msequence_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msequence_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1210\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecombination_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecombination_rate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1211\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1212\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1213\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdiscrete_genome\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdiscrete_genome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1214\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpopulation_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpopulation_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1215\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1216\u001b[0m \u001b[43m    \u001b[49m\u001b[43mploidy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mploidy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1217\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1218\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1219\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstart_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1220\u001b[0m \u001b[43m    \u001b[49m\u001b[43mend_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1221\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecord_migrations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_migrations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1222\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecord_full_arg\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_full_arg\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1223\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_labels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1224\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_seed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_seed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1225\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1226\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _wrap_replicates(\n\u001b[1;32m   1227\u001b[0m     sim,\n\u001b[1;32m   1228\u001b[0m     num_replicates\u001b[38;5;241m=\u001b[39mnum_replicates,\n\u001b[1;32m   1229\u001b[0m     replicate_index\u001b[38;5;241m=\u001b[39mreplicate_index,\n\u001b[1;32m   1230\u001b[0m     provenance_dict\u001b[38;5;241m=\u001b[39mprovenance_dict,\n\u001b[1;32m   1231\u001b[0m )\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:938\u001b[0m, in \u001b[0;36m_parse_sim_ancestry\u001b[0;34m(samples, sequence_length, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, discrete_genome, population_size, demography, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, num_labels, random_seed, init_for_debugger)\u001b[0m\n\u001b[1;32m    934\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_dtwf:\n\u001b[1;32m    935\u001b[0m     \u001b[38;5;66;03m# A default size of 1 isn't so smart for DTWF and almost certainly\u001b[39;00m\n\u001b[1;32m    936\u001b[0m     \u001b[38;5;66;03m# an error.\u001b[39;00m\n\u001b[1;32m    937\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m population_size \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 938\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    939\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mWhen using the DTWF model, the population size must be set \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    940\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexplicitly, either using the population_size or demography \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    941\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124marguments.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    942\u001b[0m         )\n\u001b[1;32m    943\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m initial_state \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    944\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m population_size \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mValueError\u001b[0m: When using the DTWF model, the population size must be set explicitly, either using the population_size or demography arguments."
     ]
    }
   ],
   "source": [
    "msprime.sim_ancestry(2, model=\"dtwf\", random_seed=1234)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c2254e",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "Because the DTWF models populations of individuals rather than\n",
    "using continuous\n",
    "time approximations, simulations with different population sizes\n",
    "will yield entirely different results; we are **not** simply\n",
    "rescaling time by using different population sizes. Furthermore,\n",
    "simulations with large population sizes will take relatively\n",
    "more CPU time and memory.\n",
    ":::\n",
    "\n",
    "---\n",
    "\n",
    "(sec_ancestry_ploidy)=\n",
    "\n",
    "## Ploidy\n",
    "\n",
    "The ``ploidy`` argument to {func}`.sim_ancestry` has two effects:\n",
    "\n",
    "1. It sets the default ploidy, that is, the number of nodes per sample individual.\n",
    "2. For continuous time coalescent models, it sets the time scale.\n",
    "\n",
    "Both of these are somewhat confusing, so let's look at them one at a time.\n",
    "\n",
    "(sec_ancestry_ploidy_default_num_nodes)=\n",
    "\n",
    "### Default number of nodes per sample individual\n",
    "\n",
    "As discussed in the {ref}`sec_ancestry_samples_sampling_details` section,\n",
    "the number of nodes (i.e., monoploid genomes) that we allocate per\n",
    "sample individual is determined by the ploidy. Here, we specify\n",
    "2 sample individuals with a ploidy of 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "230db326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤═════╤════════╤═══════╤════════╗\n",
      "║id│flags│location│parents│metadata║\n",
      "╠══╪═════╪════════╪═══════╪════════╣\n",
      "║0 │    0│        │       │        ║\n",
      "║1 │    0│        │       │        ║\n",
      "╚══╧═════╧════════╧═══════╧════════╝\n",
      "\n",
      "╔══╤═════╤══════════╤══════════╤════╤════════╗\n",
      "║id│flags│population│individual│time│metadata║\n",
      "╠══╪═════╪══════════╪══════════╪════╪════════╣\n",
      "║0 │    1│         0│         0│   0│        ║\n",
      "║1 │    1│         0│         0│   0│        ║\n",
      "║2 │    1│         0│         0│   0│        ║\n",
      "║3 │    1│         0│         1│   0│        ║\n",
      "║4 │    1│         0│         1│   0│        ║\n",
      "║5 │    1│         0│         1│   0│        ║\n",
      "╚══╧═════╧══════════╧══════════╧════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(samples=2, ploidy=3, end_time=0)\n",
    "print(ts.tables.individuals)\n",
    "print(ts.tables.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f08906b7",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "Because we're only interested in the initial conditions here we\n",
    "set the ``end_time`` to zero: see the {ref}`sec_ancestry_start_time`\n",
    "section for more details.\n",
    ":::\n",
    "\n",
    "Our individual table has two rows (the number of ``samples``)\n",
    "and node table has 6 rows: three for each of the sampled\n",
    "individuals.\n",
    "\n",
    "Note that this is the **default**, and can be overridden by the\n",
    "{attr}`.SampleSet.ploidy` value, as described in the\n",
    "{ref}`sec_ancestry_samples_sampling_details` section.\n",
    "\n",
    "(sec_ancestry_ploidy_coalescent_time_scales)=\n",
    "\n",
    "### Coalescent time scales\n",
    "\n",
    "The ploidy argument also\n",
    "affects the time scale on which coalescence occurs,\n",
    "since it affects the total number of genomes in the population\n",
    "(see also {ref}`population size<sec_ancestry_demography>`).\n",
    "For example, consider the following two simulations, which\n",
    "are identical except for different ``ploidy`` values\n",
    "(note we use the {class}`.SampleSet` object to make\n",
    "sure we allocate the same number of **nodes**, as discussed\n",
    "in the {ref}`sec_ancestry_ploidy_default_num_nodes` section):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "845d3052",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 119.835)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.04</text></g></g><g class=\"tick\" transform=\"translate(56.8 94.7646)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.60</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2.14</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.6 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(20.8 93.035)\"><g class=\"a4 i1 leaf node n1 p0 sample\" transform=\"translate(-10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i2 leaf node n2 p0 sample\" transform=\"translate(10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -93.035 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-20.8 67.9646)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i3 leaf node n3 p0 sample\" transform=\"translate(10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -67.9646 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 13,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_23_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=[msprime.SampleSet(4, ploidy=1)],\n",
    "    population_size=1,\n",
    "    ploidy=1,\n",
    "    random_seed=1234\n",
    ")\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cb3bff1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 119.835)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.07</text></g></g><g class=\"tick\" transform=\"translate(56.8 94.7646)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1.21</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4.28</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.6 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(20.8 93.035)\"><g class=\"a4 i1 leaf node n1 p0 sample\" transform=\"translate(-10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i2 leaf node n2 p0 sample\" transform=\"translate(10.4 1.565)\"><path class=\"edge\" d=\"M 0 0 V -1.565 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -93.035 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-20.8 67.9646)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i3 leaf node n3 p0 sample\" transform=\"translate(10.4 26.6354)\"><path class=\"edge\" d=\"M 0 0 V -26.6354 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -67.9646 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 14,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_24_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=[msprime.SampleSet(4, ploidy=1)],\n",
    "    population_size=1,\n",
    "    ploidy=2,\n",
    "    random_seed=1234\n",
    ")\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d3ae5d2",
   "metadata": {},
   "source": [
    "The resulting tree sequences are identical, except that node time differ by a\n",
    "factor of 2, because in the underlying coalescent algorithm the total number of\n",
    "genomes in the population acts as a time scaling.\n",
    "\n",
    ":::{todo}\n",
    "Add a discussion here about the Lambda coalescents an how\n",
    "things are more complicated for them.\n",
    ":::\n",
    "\n",
    ":::{important}\n",
    "The {ref}`sec_ancestry_models_dtwf` model is an explicitly diploid model\n",
    "and it is therefore\n",
    "an error to run it with any ``ploidy`` value other than 2.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_genome_properties)=\n",
    "\n",
    "## Genome properties\n",
    "\n",
    "(sec_ancestry_sequence_length)=\n",
    "\n",
    "### Sequence length\n",
    "\n",
    "There are a number of ways to specify the length of the\n",
    "chromosome that we want to simulate. In the absence of recombination\n",
    "and gene conversion, we assume a genome of length 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "058d21ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(98.75 26.8)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(26.25 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -26.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-26.25 25.2505)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(22.5 69.3495)\"><path class=\"edge\" d=\"M 0 0 V -69.3495 H -22.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-22.5 63.0417)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -63.0417 H 22.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -25.2505 H 26.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 15,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_26_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, random_seed=1)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbf8e160",
   "metadata": {},
   "source": [
    "The ``sequence_length`` parameter determines the length of the\n",
    "simulated sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4b9530aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(98.75 26.8)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(26.25 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -26.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-26.25 25.2505)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(22.5 69.3495)\"><path class=\"edge\" d=\"M 0 0 V -69.3495 H -22.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-22.5 63.0417)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H 15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(15 6.3078)\"><path class=\"edge\" d=\"M 0 0 V -6.3078 H -15\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -63.0417 H 22.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -25.2505 H 26.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 16,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_28_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, sequence_length=100, random_seed=1)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1b77fc9",
   "metadata": {},
   "source": [
    "By default there is no recombination or gene conversion, and so we\n",
    "therefore still have only one tree defining the ancestry of the\n",
    "samples. The sequence length makes little difference if we are\n",
    "considering only the ancestry of a non-recombining sequence, however,\n",
    "this will be important if we are interested in adding\n",
    "{ref}`mutations<sec_mutations>` later.\n",
    "\n",
    "(sec_ancestry_recombination)=\n",
    "\n",
    "### Recombination\n",
    "\n",
    "As we trace the ancestry of a sample of genomes backwards in time,\n",
    "recombination has the effect of creating segments of genome\n",
    "that have different (but highly correlated) genealogical trees.\n",
    "The {func}`.sim_ancestry` function has a ``recombination_rate``\n",
    "argument that allows us to flexibly specify both uniform and\n",
    "non-uniform rates of recombination along the genome.\n",
    "\n",
    "The simplest way to run simulations with recombination is to\n",
    "use the ``recombination_rate`` and ``sequence_length`` parameters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a9cd9026",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l-38,25 l0,5 l-152,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l-76,25 l0,5 l-152,0 l0,-5 l38,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l-38,25 l0,5 l-228,0 l0,-5 l76,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-228,0 l0,-5 l38,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(172 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">2</text></g></g><g class=\"tick\" transform=\"translate(324 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">4</text></g></g><g class=\"tick\" transform=\"translate(552 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">7</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(74.6875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-29.6875 93.2031)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.2031 H 29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n9 p0\" transform=\"translate(29.6875 67.5815)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-21.875 27.0185)\"><path class=\"edge\" d=\"M 0 0 V -27.0185 H 21.875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(21.875 9.95537)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-18.75 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(18.75 2.86901)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(-12.5 14.1941)\"><path class=\"edge\" d=\"M 0 0 V -14.1941 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 14.1941)\"><path class=\"edge\" d=\"M 0 0 V -14.1941 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -2.86901 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -9.95537 H -21.875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -67.5815 H -29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(87.1875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-42.1875 93.2031)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.2031 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(42.1875 12.6456)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(28.125 81.9544)\"><path class=\"edge\" d=\"M 0 0 V -81.9544 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n9 p0\" transform=\"translate(-28.125 54.9359)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 27.0185)\"><path class=\"edge\" d=\"M 0 0 V -27.0185 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(18.75 9.95537)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -9.95537 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -54.9359 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -12.6456 H -42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(101.25 26.8)\"><g class=\"a14 c2 node n13 p0\" transform=\"translate(37.5 12.6456)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 81.9544)\"><path class=\"edge\" d=\"M 0 0 V -81.9544 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n8 p0\" transform=\"translate(-18.75 64.8913)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -64.8913 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -12.6456 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><g class=\"a14 c2 node n10 p0\" transform=\"translate(-37.5 60.2032)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(18.75 34.3968)\"><path class=\"edge\" d=\"M 0 0 V -34.3968 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a10 c2 node n6 p0\" transform=\"translate(-18.75 32.9998)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -32.9998 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -60.2032 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(95 39.4456)\"><g class=\"a13 c2 node n11 p0\" transform=\"translate(31.25 15.3506)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 66.6037)\"><path class=\"edge\" d=\"M 0 0 V -66.6037 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n8 p0\" transform=\"translate(18.75 49.5406)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -49.5406 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -15.3506 H -31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-31.25 12.027)\"><g class=\"a12 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 69.9273)\"><path class=\"edge\" d=\"M 0 0 V -69.9273 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a12 c2 node n6 p0\" transform=\"translate(-18.75 68.5304)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -68.5304 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -12.027 H 31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 17,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_30_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, recombination_rate=0.1, sequence_length=10, random_seed=2)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9877a200",
   "metadata": {},
   "source": [
    "Here, we have four distinct trees along the genome,\n",
    "created by recombination events."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dc457768",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "assert 1 < ts.num_trees < 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "846143d9",
   "metadata": {},
   "source": [
    "We can also provide a {class}`.RateMap` argument to ``recombination_rate``:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e50680b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "rate_map = msprime.RateMap.uniform(sequence_length=10, rate=0.1)\n",
    "other_ts = msprime.sim_ancestry(3, recombination_rate=rate_map, random_seed=2)\n",
    "assert ts.equals(other_ts, ignore_provenance=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03d3c23e",
   "metadata": {},
   "source": [
    "There are two things to note here:\n",
    "\n",
    "1. We didn't need to specify a ``sequence_length`` argument to\n",
    "{func}`.sim_ancestry`, as this information is encapsulated by the ``rate_map``\n",
    "instance. (We can provide a value for ``sequence_length`` if we wish, but\n",
    "it must be equal to the {attr}`~.RateMap.sequence_length` property of the\n",
    "{class}`.RateMap`).\n",
    "2. The result of the simulation for a {ref}`sec_rate_maps_creating_uniform`\n",
    "rate map with the same rate and ``sequence_length`` is identical to\n",
    "the first simulation above.\n",
    "\n",
    "We can simulate non-uniform recombination by defining our {class}`.RateMap`\n",
    "appropriately. For example, here we define a map in which the\n",
    "rate over the first half of the sequence is one tenth of the second half:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "571b7a4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l228,25 l0,5 l-418,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l76,25 l0,5 l-38,0 l0,-5 l-228,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l0,25 l0,5 l-114,0 l0,-5 l-76,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-190,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(438 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">11</text></g></g><g class=\"tick\" transform=\"translate(476 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">12</text></g></g><g class=\"tick\" transform=\"translate(590 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">15</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">20</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(74.6875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-29.6875 93.0814)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.0814 H 29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n9 p0\" transform=\"translate(29.6875 68.193)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-21.875 26.407)\"><path class=\"edge\" d=\"M 0 0 V -26.407 H 21.875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(21.875 9.93916)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-18.75 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(18.75 2.75709)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(-12.5 13.7107)\"><path class=\"edge\" d=\"M 0 0 V -13.7107 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 13.7107)\"><path class=\"edge\" d=\"M 0 0 V -13.7107 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -2.75709 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -9.93916 H -21.875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -68.193 H -29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(87.1875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-42.1875 93.0814)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.0814 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(42.1875 13.747)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(28.125 80.853)\"><path class=\"edge\" d=\"M 0 0 V -80.853 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n9 p0\" transform=\"translate(-28.125 54.446)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 26.407)\"><path class=\"edge\" d=\"M 0 0 V -26.407 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(18.75 9.93916)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -9.93916 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -54.446 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -13.747 H -42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(101.25 26.8)\"><g class=\"a14 c2 node n13 p0\" transform=\"translate(37.5 13.747)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 80.853)\"><path class=\"edge\" d=\"M 0 0 V -80.853 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n8 p0\" transform=\"translate(-18.75 64.3851)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -64.3851 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -13.747 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><g class=\"a14 c2 node n10 p0\" transform=\"translate(-37.5 60.1721)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(18.75 34.4279)\"><path class=\"edge\" d=\"M 0 0 V -34.4279 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a10 c2 node n6 p0\" transform=\"translate(-18.75 32.9093)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -32.9093 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -60.1721 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(95 40.547)\"><g class=\"a13 c2 node n11 p0\" transform=\"translate(31.25 16.6876)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 64.1653)\"><path class=\"edge\" d=\"M 0 0 V -64.1653 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n8 p0\" transform=\"translate(18.75 47.6975)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 16.4678)\"><path class=\"edge\" d=\"M 0 0 V -16.4678 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -47.6975 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -16.6876 H -31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-31.25 13.0746)\"><g class=\"a12 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 67.7784)\"><path class=\"edge\" d=\"M 0 0 V -67.7784 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a12 c2 node n6 p0\" transform=\"translate(-18.75 66.2598)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.51859)\"><path class=\"edge\" d=\"M 0 0 V -1.51859 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -66.2598 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -13.0746 H 31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 20,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_36_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rate_map = msprime.RateMap(position=[0, 10, 20], rate=[0.01, 0.1])\n",
    "ts = msprime.sim_ancestry(3, recombination_rate=rate_map, random_seed=2)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70f94232",
   "metadata": {},
   "source": [
    "Because the recombination rate in from position 10 to 20 is so much\n",
    "higher than the rate from 0 to 10, we can see that all the recombinations\n",
    "have fallen in the high recombination region.\n",
    "\n",
    ":::{seealso}\n",
    "- See the {ref}`sec_rate_maps_creating` section for more examples of\n",
    "creating {class}`.RateMap` instances.\n",
    "\n",
    "- See the {meth}`.RateMap.read_hapmap` method for reading in\n",
    "genetic maps from text files.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_gene_conversion)=\n",
    "\n",
    "### Gene conversion\n",
    "\n",
    "Gene conversion events are defined by two parameters: the rate at which gene\n",
    "conversion events are initiated and the distribution of tract lengths.\n",
    "In the default case of discrete genome coordinates, tract lengths are drawn\n",
    "from a geometric distribution with mean ``gene_conversion_tract_length`` (which\n",
    "must be at least 1). Note that if we specify a tract length of 1, then all\n",
    "gene conversion tracts will have length exactly 1.\n",
    "In the following example one gene conversion event of length 1 has occurred:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2969a8af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l186.667,0 l0,138.2 l93.3333,25 l0,5 l-280,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M206.667,0 l186.667,0 l0,138.2 l-37.3333,25 l0,5 l-56,0 l0,-5 l-93.3333,-25 l0,-138.2z\"/><path d=\"M393.333,0 l186.667,0 l0,138.2 l0,25 l0,5 l-224,0 l0,-5 l37.3333,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(300 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(300 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">5</text></g></g><g class=\"tick\" transform=\"translate(356 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">6</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(81.1111 26.8)\"><g class=\"a11 c2 node n6 p0\" transform=\"translate(-36.6667 89.2217)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.2222 5.37835)\"><path class=\"edge\" d=\"M 0 0 V -5.37835 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.2222 5.37835)\"><path class=\"edge\" d=\"M 0 0 V -5.37835 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -89.2217 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a11 c2 node n10 p0\" transform=\"translate(36.6667 67.5377)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(24.4444 15.7471)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.2222 11.3152)\"><path class=\"edge\" d=\"M 0 0 V -11.3152 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 11.3152)\"><path class=\"edge\" d=\"M 0 0 V -11.3152 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -15.7471 H -24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-24.4444 11.2407)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 15.8216)\"><path class=\"edge\" d=\"M 0 0 V -15.8216 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(12.2222 15.8216)\"><path class=\"edge\" d=\"M 0 0 V -15.8216 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -11.2407 H 24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -67.5377 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t1\" transform=\"translate(206.667 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(99.4444 26.8)\"><g class=\"a11 c2 node n10 p0\" transform=\"translate(36.6667 67.5377)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(-18.3333 15.7471)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.2222 11.3152)\"><path class=\"edge\" d=\"M 0 0 V -11.3152 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 11.3152)\"><path class=\"edge\" d=\"M 0 0 V -11.3152 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -15.7471 H 18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a10 i2 leaf node n5 p0 sample\" transform=\"translate(18.3333 27.0623)\"><path class=\"edge\" d=\"M 0 0 V -27.0623 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -67.5377 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a11 c2 node n7 p0\" transform=\"translate(-36.6667 83.9282)\"><g class=\"a7 i0 leaf node n1 p0 sample\" transform=\"translate(18.3333 10.6718)\"><path class=\"edge\" d=\"M 0 0 V -10.6718 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-18.3333 5.29346)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.2222 5.37835)\"><path class=\"edge\" d=\"M 0 0 V -5.37835 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.2222 5.37835)\"><path class=\"edge\" d=\"M 0 0 V -5.37835 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.29346 H 18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -83.9282 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t2\" transform=\"translate(393.333 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(81.1111 26.8)\"><g class=\"a11 c2 node n10 p0\" transform=\"translate(36.6667 67.5377)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(24.4444 15.7471)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.2222 11.3152)\"><path class=\"edge\" d=\"M 0 0 V -11.3152 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 11.3152)\"><path class=\"edge\" d=\"M 0 0 V -11.3152 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -15.7471 H -24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-24.4444 11.2407)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 15.8216)\"><path class=\"edge\" d=\"M 0 0 V -15.8216 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(12.2222 15.8216)\"><path class=\"edge\" d=\"M 0 0 V -15.8216 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -11.2407 H 24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -67.5377 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a11 c2 node n6 p0\" transform=\"translate(-36.6667 89.2217)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.2222 5.37835)\"><path class=\"edge\" d=\"M 0 0 V -5.37835 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.2222 5.37835)\"><path class=\"edge\" d=\"M 0 0 V -5.37835 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -89.2217 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 21,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_38_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, gene_conversion_rate=0.02, gene_conversion_tract_length=1,\n",
    "    sequence_length=10, random_seed=3)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa57924b",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "We can see the specific effect of gene conversion here in this simple\n",
    "example, because the trees at either side of the event are identical.\n",
    "However, this will not be true in general.\n",
    ":::"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7df5574c",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "assert ts.num_trees == 3\n",
    "assert ts.at(0).parent_dict == ts.at(2).parent_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c15a531a",
   "metadata": {},
   "source": [
    "Continuous genomes can also be used. In this case the parameters define\n",
    "the rate at which gene conversion events are initiated per unit of sequence\n",
    "length and the mean of the exponentially distributed gene conversion tract\n",
    "lengths. The following example shows the same simulation as above but for a\n",
    "continuous genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3c2541f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l186.667,0 l0,138.2 l73.5848,25 l0,5 l-260.251,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M206.667,0 l186.667,0 l0,138.2 l-80.9051,25 l0,5 l-32.1768,0 l0,-5 l-73.5848,-25 l0,-138.2z\"/><path d=\"M393.333,0 l186.667,0 l0,138.2 l0,25 l0,5 l-267.572,0 l0,-5 l80.9051,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(300 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0.00</text></g></g><g class=\"tick\" transform=\"translate(280.251 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">4.65</text></g></g><g class=\"tick\" transform=\"translate(312.428 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">5.22</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10.00</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(81.1111 26.8)\"><g class=\"a11 c2 node n6 p0\" transform=\"translate(-36.6667 89.0063)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.2222 5.59372)\"><path class=\"edge\" d=\"M 0 0 V -5.59372 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.2222 5.59372)\"><path class=\"edge\" d=\"M 0 0 V -5.59372 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -89.0063 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a11 c2 node n10 p0\" transform=\"translate(36.6667 67.0871)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(24.4444 16.2341)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.2222 11.2788)\"><path class=\"edge\" d=\"M 0 0 V -11.2788 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 11.2788)\"><path class=\"edge\" d=\"M 0 0 V -11.2788 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -16.2341 H -24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-24.4444 11.6908)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 15.8221)\"><path class=\"edge\" d=\"M 0 0 V -15.8221 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(12.2222 15.8221)\"><path class=\"edge\" d=\"M 0 0 V -15.8221 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -11.6908 H 24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -67.0871 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t1\" transform=\"translate(206.667 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(99.4444 26.8)\"><g class=\"a11 c2 node n10 p0\" transform=\"translate(36.6667 67.0871)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(-18.3333 16.2341)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.2222 11.2788)\"><path class=\"edge\" d=\"M 0 0 V -11.2788 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 11.2788)\"><path class=\"edge\" d=\"M 0 0 V -11.2788 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -16.2341 H 18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a10 i2 leaf node n5 p0 sample\" transform=\"translate(18.3333 27.5129)\"><path class=\"edge\" d=\"M 0 0 V -27.5129 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -67.0871 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a11 c2 node n7 p0\" transform=\"translate(-36.6667 83.9904)\"><g class=\"a7 i0 leaf node n1 p0 sample\" transform=\"translate(18.3333 10.6096)\"><path class=\"edge\" d=\"M 0 0 V -10.6096 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-18.3333 5.01584)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.2222 5.59372)\"><path class=\"edge\" d=\"M 0 0 V -5.59372 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.2222 5.59372)\"><path class=\"edge\" d=\"M 0 0 V -5.59372 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.01584 H 18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -83.9904 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t2\" transform=\"translate(393.333 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(81.1111 26.8)\"><g class=\"a11 c2 node n10 p0\" transform=\"translate(36.6667 67.0871)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(24.4444 16.2341)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.2222 11.2788)\"><path class=\"edge\" d=\"M 0 0 V -11.2788 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 11.2788)\"><path class=\"edge\" d=\"M 0 0 V -11.2788 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -16.2341 H -24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-24.4444 11.6908)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 15.8221)\"><path class=\"edge\" d=\"M 0 0 V -15.8221 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(12.2222 15.8221)\"><path class=\"edge\" d=\"M 0 0 V -15.8221 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -11.6908 H 24.4444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -67.0871 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a11 c2 node n6 p0\" transform=\"translate(-36.6667 89.0063)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.2222 5.59372)\"><path class=\"edge\" d=\"M 0 0 V -5.59372 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(12.2222 5.59372)\"><path class=\"edge\" d=\"M 0 0 V -5.59372 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -89.0063 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 23,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_42_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, gene_conversion_rate=0.02, gene_conversion_tract_length=1,\n",
    "    sequence_length=10, random_seed=3, discrete_genome=False)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "59eec150",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "assert ts.num_trees == 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b390744",
   "metadata": {},
   "source": [
    "Recombination and gene conversion at constant rates can be simulated alongside.\n",
    "In the following example recombinations at site 60 and 97 have occurred in addition\n",
    "to a gene conversion event covering the tract from site 76 to site 80."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5ced0114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"1000\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l192,0 l0,138.2 l384,25 l0,5 l-576,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M212,0 l192,0 l0,138.2 l345.6,25 l0,5 l-153.6,0 l0,-5 l-384,-25 l0,-138.2z\"/><path d=\"M404,0 l192,0 l0,138.2 l201.6,25 l0,5 l-48,0 l0,-5 l-345.6,-25 l0,-138.2z\"/><path d=\"M596,0 l192,0 l0,138.2 l163.2,25 l0,5 l-153.6,0 l0,-5 l-201.6,-25 l0,-138.2z\"/><path d=\"M788,0 l192,0 l0,138.2 l0,25 l0,5 l-28.8,0 l0,-5 l-163.2,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(500 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"980\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(596 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">60</text></g></g><g class=\"tick\" transform=\"translate(749.6 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">76</text></g></g><g class=\"tick\" transform=\"translate(797.6 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">81</text></g></g><g class=\"tick\" transform=\"translate(951.2 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">97</text></g></g><g class=\"tick\" transform=\"translate(980 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n10 p0 root\" transform=\"translate(119.75 102.603)\"><g class=\"a10 i1 leaf node n2 p0 sample\" transform=\"translate(39.5833 18.7971)\"><path class=\"edge\" d=\"M 0 0 V -18.7971 H -39.5833\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-39.5833 5.07917)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(34.8333 10.3648)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(19 3.35313)\"><path class=\"edge\" d=\"M 0 0 V -3.35313 H -19\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-19 3.03802)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -3.03802 H 19\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -10.3648 H -34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-34.8333 5.42491)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.42491 H 34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -5.07917 H 39.5833\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g><g class=\"tree t1\" transform=\"translate(212 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(119.75 89.986)\"><g class=\"a11 i1 leaf node n2 p0 sample\" transform=\"translate(39.5833 31.414)\"><path class=\"edge\" d=\"M 0 0 V -31.414 H -39.5833\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a11 c2 node n9 p0\" transform=\"translate(-39.5833 17.6961)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(34.8333 10.3648)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(19 3.35313)\"><path class=\"edge\" d=\"M 0 0 V -3.35313 H -19\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-19 3.03802)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -3.03802 H 19\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -10.3648 H -34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-34.8333 5.42491)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.42491 H 34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -17.6961 H 39.5833\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t2\" transform=\"translate(404 0)\"><g class=\"plotbox\"><g class=\"c2 node n12 p0 root\" transform=\"translate(119.75 82.5253)\"><g class=\"a12 i1 leaf node n2 p0 sample\" transform=\"translate(39.5833 38.8747)\"><path class=\"edge\" d=\"M 0 0 V -38.8747 H -39.5833\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a12 c2 node n9 p0\" transform=\"translate(-39.5833 25.1568)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(34.8333 10.3648)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(19 3.35313)\"><path class=\"edge\" d=\"M 0 0 V -3.35313 H -19\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-19 3.03802)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -3.03802 H 19\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -10.3648 H -34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-34.8333 5.42491)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.42491 H 34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -25.1568 H 39.5833\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">12</text></g></g></g><g class=\"tree t3\" transform=\"translate(596 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(119.75 89.986)\"><g class=\"a11 i1 leaf node n2 p0 sample\" transform=\"translate(39.5833 31.414)\"><path class=\"edge\" d=\"M 0 0 V -31.414 H -39.5833\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a11 c2 node n9 p0\" transform=\"translate(-39.5833 17.6961)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(34.8333 10.3648)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(19 3.35313)\"><path class=\"edge\" d=\"M 0 0 V -3.35313 H -19\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-19 3.03802)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -3.03802 H 19\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -10.3648 H -34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-34.8333 5.42491)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.42491 H 34.8333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -17.6961 H 39.5833\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t4\" transform=\"translate(788 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(130.833 26.8)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(28.5 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -28.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n11 p0\" transform=\"translate(-28.5 63.186)\"><g class=\"a11 i1 leaf node n2 p0 sample\" transform=\"translate(31.6667 31.414)\"><path class=\"edge\" d=\"M 0 0 V -31.414 H -31.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a11 c2 node n9 p0\" transform=\"translate(-31.6667 17.6961)\"><g class=\"a9 c2 node n8 p0\" transform=\"translate(-25.3333 5.42491)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.6667 8.29301)\"><path class=\"edge\" d=\"M 0 0 V -8.29301 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.42491 H 25.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a9 c2 node n6 p0\" transform=\"translate(25.3333 13.4028)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H 12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.6667 0.315114)\"><path class=\"edge\" d=\"M 0 0 V -0.315114 H -12.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -13.4028 H -25.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -17.6961 H 31.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -63.186 H 28.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 25,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_45_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, sequence_length = 100, recombination_rate=0.003,\n",
    "    gene_conversion_rate=0.002, gene_conversion_tract_length=5,\n",
    "    random_seed=6)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0cbab774",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "assert 1 < ts.num_trees < 6"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b1016ed",
   "metadata": {},
   "source": [
    "Variable recombination rates and constant gene conversion rates\n",
    "can also be combined.\n",
    "\n",
    ":::{note}\n",
    "Variable rates of gene conversion breakpoint initiation are not currently\n",
    "supported, but are planned for a future release. If you are interested\n",
    "in this feature please let us know!\n",
    ":::\n",
    "\n",
    "(sec_ancestry_discrete_genome)=\n",
    "\n",
    "### Discrete or continuous?\n",
    "\n",
    "By default, we assume that the genome we are simulating is *discrete*\n",
    "so that genome coordinates are at integer positions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "327077f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l-38,25 l0,5 l-152,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l-76,25 l0,5 l-152,0 l0,-5 l38,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l-38,25 l0,5 l-228,0 l0,-5 l76,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-228,0 l0,-5 l38,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(172 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">2</text></g></g><g class=\"tick\" transform=\"translate(324 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">4</text></g></g><g class=\"tick\" transform=\"translate(552 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">7</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(74.6875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-29.6875 93.2031)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.2031 H 29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n9 p0\" transform=\"translate(29.6875 67.5815)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-21.875 27.0185)\"><path class=\"edge\" d=\"M 0 0 V -27.0185 H 21.875\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(21.875 9.95537)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-18.75 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(18.75 2.86901)\"><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(-12.5 14.1941)\"><path class=\"edge\" d=\"M 0 0 V -14.1941 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 14.1941)\"><path class=\"edge\" d=\"M 0 0 V -14.1941 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -2.86901 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -9.95537 H -21.875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -67.5815 H -29.6875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(87.1875 26.8)\"><g class=\"a14 c2 node n6 p0\" transform=\"translate(-42.1875 93.2031)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -93.2031 H 42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(42.1875 12.6456)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(28.125 81.9544)\"><path class=\"edge\" d=\"M 0 0 V -81.9544 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n9 p0\" transform=\"translate(-28.125 54.9359)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 27.0185)\"><path class=\"edge\" d=\"M 0 0 V -27.0185 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(18.75 9.95537)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -9.95537 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -54.9359 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -12.6456 H -42.1875\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(101.25 26.8)\"><g class=\"a14 c2 node n13 p0\" transform=\"translate(37.5 12.6456)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 81.9544)\"><path class=\"edge\" d=\"M 0 0 V -81.9544 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 c2 node n8 p0\" transform=\"translate(-18.75 64.8913)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -64.8913 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -12.6456 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><g class=\"a14 c2 node n10 p0\" transform=\"translate(-37.5 60.2032)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(18.75 34.3968)\"><path class=\"edge\" d=\"M 0 0 V -34.3968 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a10 c2 node n6 p0\" transform=\"translate(-18.75 32.9998)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -32.9998 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -60.2032 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(95 39.4456)\"><g class=\"a13 c2 node n11 p0\" transform=\"translate(31.25 15.3506)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 66.6037)\"><path class=\"edge\" d=\"M 0 0 V -66.6037 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c2 node n8 p0\" transform=\"translate(18.75 49.5406)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i2 leaf node n4 p0 sample\" transform=\"translate(12.5 17.0631)\"><path class=\"edge\" d=\"M 0 0 V -17.0631 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -49.5406 H -18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -15.3506 H -31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-31.25 12.027)\"><g class=\"a12 i1 leaf node n3 p0 sample\" transform=\"translate(18.75 69.9273)\"><path class=\"edge\" d=\"M 0 0 V -69.9273 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a12 c2 node n6 p0\" transform=\"translate(-18.75 68.5304)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H 12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(12.5 1.39692)\"><path class=\"edge\" d=\"M 0 0 V -1.39692 H -12.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -68.5304 H 18.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -12.027 H 31.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 27,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_48_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, recombination_rate=0.1, sequence_length=10, random_seed=2)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b6088f06",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "assert 1 < ts.num_trees < 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76036a65",
   "metadata": {},
   "source": [
    "Alternatively, we can simulate a continous genome by setting\n",
    "`discrete_genome=False`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1457bd16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l186.667,0 l0,138.2 l-86.9204,25 l0,5 l-99.7463,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M206.667,0 l186.667,0 l0,138.2 l-204.494,25 l0,5 l-69.0929,0 l0,-5 l86.9204,-25 l0,-138.2z\"/><path d=\"M393.333,0 l186.667,0 l0,138.2 l0,25 l0,5 l-391.161,0 l0,-5 l204.494,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(300 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0.00</text></g></g><g class=\"tick\" transform=\"translate(119.746 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0.18</text></g></g><g class=\"tick\" transform=\"translate(188.839 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0.30</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1.00</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n10 p0 root\" transform=\"translate(129.236 37.0296)\"><g class=\"a10 i2 leaf node n4 p0 sample\" transform=\"translate(25.2083 84.3704)\"><path class=\"edge\" d=\"M 0 0 V -84.3704 H -25.2083\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-25.2083 60.3652)\"><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(25.9722 24.0052)\"><path class=\"edge\" d=\"M 0 0 V -24.0052 H -25.9722\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-25.9722 5.89198)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(27.5 18.1132)\"><path class=\"edge\" d=\"M 0 0 V -18.1132 H -27.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-27.5 4.89431)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-18.3333 13.2189)\"><path class=\"edge\" d=\"M 0 0 V -13.2189 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(18.3333 10.5158)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 2.70312)\"><path class=\"edge\" d=\"M 0 0 V -2.70312 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 2.70312)\"><path class=\"edge\" d=\"M 0 0 V -2.70312 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -10.5158 H -18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -4.89431 H 27.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -5.89198 H 25.9722\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -60.3652 H 25.2083\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g><g class=\"tree t1\" transform=\"translate(206.667 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(129.236 26.8)\"><g class=\"a11 i2 leaf node n4 p0 sample\" transform=\"translate(25.2083 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -25.2083\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a11 c2 node n9 p0\" transform=\"translate(-25.2083 70.5948)\"><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(25.9722 24.0052)\"><path class=\"edge\" d=\"M 0 0 V -24.0052 H -25.9722\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-25.9722 5.89198)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(27.5 18.1132)\"><path class=\"edge\" d=\"M 0 0 V -18.1132 H -27.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-27.5 4.89431)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-18.3333 13.2189)\"><path class=\"edge\" d=\"M 0 0 V -13.2189 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(18.3333 10.5158)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 2.70312)\"><path class=\"edge\" d=\"M 0 0 V -2.70312 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 2.70312)\"><path class=\"edge\" d=\"M 0 0 V -2.70312 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -10.5158 H -18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -4.89431 H 27.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -5.89198 H 25.9722\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -70.5948 H 25.2083\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g><g class=\"tree t2\" transform=\"translate(393.333 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(129.236 26.8)\"><g class=\"a11 i2 leaf node n4 p0 sample\" transform=\"translate(25.2083 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -25.2083\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a11 c2 node n10 p0\" transform=\"translate(-25.2083 10.2296)\"><g class=\"a10 i2 leaf node n5 p0 sample\" transform=\"translate(25.9722 84.3704)\"><path class=\"edge\" d=\"M 0 0 V -84.3704 H -25.9722\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a10 c2 node n8 p0\" transform=\"translate(-25.9722 66.2571)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(27.5 18.1132)\"><path class=\"edge\" d=\"M 0 0 V -18.1132 H -27.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-27.5 4.89431)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-18.3333 13.2189)\"><path class=\"edge\" d=\"M 0 0 V -13.2189 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(18.3333 10.5158)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-12.2222 2.70312)\"><path class=\"edge\" d=\"M 0 0 V -2.70312 H 12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(12.2222 2.70312)\"><path class=\"edge\" d=\"M 0 0 V -2.70312 H -12.2222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -10.5158 H -18.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -4.89431 H 27.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -66.2571 H 25.9722\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -10.2296 H 25.2083\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 29,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_51_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, recombination_rate=0.25, sequence_length=1, discrete_genome=False,\n",
    "    random_seed=33)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cf85abb4",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "assert 1 < ts.num_trees < 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "121054fd",
   "metadata": {},
   "source": [
    "Here we see that the breakpoints along the genome occur at floating point\n",
    "positions. Simulating a continuous genome sequence can be useful for\n",
    "theoretical work, but we recommend using discrete coordinates for most\n",
    "purposes.\n",
    "\n",
    "\n",
    "(sec_ancestry_multiple_chromosomes)=\n",
    "\n",
    "### Multiple chromosomes\n",
    "\n",
    "#### Do I need to do this?\n",
    "\n",
    "The main purpose of simulating multiple chromosomes together (instead of\n",
    "running independent simulations for each chromosome) is to capture\n",
    "correlations between trees on distinct chromosomes. In many scenarios,\n",
    "**simulating multiple chromosomes may not be necessary**, and independent\n",
    "simulations may be preferable as they will be much more efficient.\n",
    "\n",
    "That being said, this section describes how to simulate data over\n",
    "multiple chromosomes, or more generally, over multiple regions\n",
    "with free recombination between them. \n",
    "\n",
    "#### Using a fixed pedigree\n",
    "\n",
    "The most natural way to simulate multiple regions with free recombination for\n",
    "a set of sampled individuals is by simulating through a known pedigree. The\n",
    "{ref}`fixed pedigree <sec_ancestry_models_fixed_pedigree>` simulation model\n",
    "constrains simulated genomic regions by an input pedigree, which specifies\n",
    "parent-offspring relationships. Since autosomal genetic material in an\n",
    "offspring has equal chance of being inherited from either their mother or their\n",
    "father, independent simulations through the same pedigree ensures that those\n",
    "regions are inherited from different parents or from the same parent each with\n",
    "probability 1/2. This ensures free recombination between independent\n",
    "simulations.\n",
    "\n",
    "We often do not have a full pedigree, but instead have a pedigree that\n",
    "extends some number of generations into the past. Because a recombination\n",
    "fraction of 1/2 between regions causes correlations between chromosomes to\n",
    "decay rapidly, even shallow pedigrees containing just a handful of generations\n",
    "should broadly capture expected levels of correlations. At the termination of\n",
    "the pedigree, the simulation can be completed using the DTWF, Hudson, or other\n",
    "model. See {ref}`example 1 <sec_multi_chrom_example_1>` below for a small\n",
    "example of this procedure, and more details on completing pedigree simulations\n",
    "can be found in the {ref}`Fixed pedigree <sec_ancestry_models_fixed_pedigree>`\n",
    "section.\n",
    "\n",
    "Why might simulating multiple chromosomes using a pedigree be preferable to\n",
    "other simulation options? First, there is no need to concatenate recombination\n",
    "maps for multiple chromosomes, as each region is simulated independently. This\n",
    "also results in simpler, reusable code which can be parallelized. Finally, this\n",
    "approach improves efficiency over a single larger simulation of multiple\n",
    "concatenated chromosomes, as computational costs scale as the square of the\n",
    "sequence length.\n",
    "\n",
    "#### Simulations without a fixed pedigree\n",
    "\n",
    "Without a pedigree, we can use other simulation models in msprime such as\n",
    "{ref}`DTWF <sec_ancestry_models_dtwf>` and the \n",
    "{ref}`multiple merger coalescent <sec_ancestry_models_multiple_mergers>`.\n",
    "These models do not directly support simulating multiple chromosomes\n",
    "simultaneously, but we can emulate it using a single linear genome split into\n",
    "chromosome segments. In this case, multiple chromosomes are modelled by specifying\n",
    "a recombination map with single base-pair segments with recombination\n",
    "probability 1/2 separating adjacent chromosomes.\n",
    "The probability that two chromosomes are co-inherited across {math}`t` generations\n",
    "is {math}`2^{-t}`. In the Poisson model of recombination used by msprime, the\n",
    "probability of co-inheritance across {math}`t` generations of two adjacent base\n",
    "pairs separated by a recombination rate {math}`r` per base pair is {math}`e^{-rt}`.\n",
    "To make these agree, we set the recombination rate in the base pair segments\n",
    "separating chromosomes to {math}`\\log(2)`.\n",
    "\n",
    "[Nelson et al. 2020](https://doi.org/10.1371/journal.pgen.1008619) describes\n",
    "a few cases where simulating multiple chromosomes using the DTWF is important\n",
    "for accurately modeling shared patterns of variation between samples.\n",
    "In large simulated cohorts, where many samples can be close\n",
    "relatives, we need the DTWF model and multiple chromosomes to obtain the\n",
    "correct distribution of the number and total length of shared IBD segments. We\n",
    "similarly require multiple chromosomes and the DTWF model to get the correct\n",
    "ancestry patterns for very recently admixed populations.\n",
    "\n",
    "Additionally,\n",
    "{ref}`Multiple merger coalescents <sec_ancestry_models_multiple_mergers>`\n",
    "result in positively correlated ancestries between unlinked chromosomes\n",
    "(see [Birkner et al. 2013](https://www.genetics.org/content/193/1/255)).\n",
    "This correlation does not break down in time and multiple chromosomes should\n",
    "not be simulated independently under the multiple merger models. Unlinked\n",
    "chromosomes should scatter into distinct ancestors instantaneously under\n",
    "multiple merger coalescents, not with probability 1/2 per generation\n",
    "as in discrete population models. In msprime, this waiting time of length zero\n",
    "is approximated by an exponential distribution with rate {math}`r`, i.e. the\n",
    "recombination rate at the base pair separating the chromosomes, which\n",
    "should be set to a high enough value to obtain an acceptable approximation.\n",
    "We recommend considering in particular the relative error in comparison to\n",
    "the magnitudes other waiting times which are likely to arise in the simulation.\n",
    "\n",
    ":::{important}\n",
    "Simulations of multiple chromosomes under either DTWF or the multiple merger\n",
    "models should use a {ref}`discrete genome <sec_ancestry_discrete_genome>`.\n",
    "While it is possible to set up such a simulation using a continuous genome,\n",
    "we *strongly* recommend against it, as chromosome divisions cannot be defined\n",
    "as discrete break points and the simulation will be very inefficient.\n",
    ":::\n",
    "\n",
    "(sec_multi_chrom_example_1)=\n",
    "\n",
    "#### Example 1: simulating with a pedigree\n",
    "\n",
    "In this example, we simulate genomic data for a pair of first cousins in\n",
    "a small pedigree extending 3 (?) generations into the past. See {ref}`here\n",
    "<sec_ancestry_models_fixed_pedigree>` for details on specifying, importing, and\n",
    "simulating using pedigrees.\n",
    "\n",
    "Let's first specify our pedigree:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c0b49aa2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"371.52pt\" height=\"280.512pt\" viewBox=\"0 0 371.52 280.512\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-10-17T07:24:25.123129</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 280.512 \n",
       "L 371.52 280.512 \n",
       "L 371.52 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "L 7.2 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 219.36862 242.591849 \n",
       "Q 244.787615 195.23872 269.677822 148.870673 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 266.023808 151.449079 \n",
       "L 269.677822 148.870673 \n",
       "L 269.548139 153.340928 \n",
       "L 266.023808 151.449079 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 215.27405 241.558255 \n",
       "Q 215.27405 195.236773 215.27405 150.033324 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 213.27405 154.033324 \n",
       "L 215.27405 150.033324 \n",
       "L 217.27405 154.033324 \n",
       "L 213.27405 154.033324 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 278.396719 132.628212 \n",
       "Q 303.815714 85.275084 328.705921 38.907036 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 325.051907 41.485443 \n",
       "L 328.705921 38.907036 \n",
       "L 328.576238 43.377292 \n",
       "L 325.051907 41.485443 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 274.302149 131.594619 \n",
       "Q 274.302149 85.273136 274.302149 40.069688 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 272.302149 44.069688 \n",
       "L 274.302149 40.069688 \n",
       "L 276.302149 44.069688 \n",
       "L 272.302149 44.069688 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 215.27405 131.594619 \n",
       "Q 215.27405 85.273136 215.27405 40.069688 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 213.27405 44.069688 \n",
       "L 215.27405 40.069688 \n",
       "L 217.27405 44.069688 \n",
       "L 213.27405 44.069688 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 211.17948 132.628212 \n",
       "Q 185.760484 85.275084 160.870277 38.907036 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 160.99996 43.377292 \n",
       "L 160.870277 38.907036 \n",
       "L 164.524291 41.485443 \n",
       "L 160.99996 43.377292 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 156.24595 241.558255 \n",
       "Q 156.24595 195.236773 156.24595 150.033324 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 154.24595 154.033324 \n",
       "L 156.24595 150.033324 \n",
       "L 158.24595 154.033324 \n",
       "L 154.24595 154.033324 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 152.15138 242.591849 \n",
       "Q 126.732385 195.23872 101.842178 148.870673 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 101.971861 153.340928 \n",
       "L 101.842178 148.870673 \n",
       "L 105.496192 151.449079 \n",
       "L 101.971861 153.340928 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 163.603747 135.687046 \n",
       "Q 244.788157 85.274146 325.022759 35.451047 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 320.569555 35.862112 \n",
       "L 325.022759 35.451047 \n",
       "L 322.679688 39.260249 \n",
       "L 320.569555 35.862112 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 162.583257 134.353106 \n",
       "Q 215.273675 85.274531 267.145981 36.957988 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 262.85585 38.220833 \n",
       "L 267.145981 36.957988 \n",
       "L 265.582178 41.1478 \n",
       "L 262.85585 38.220833 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 97.217851 131.594619 \n",
       "Q 97.217851 85.273136 97.217851 40.069688 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 95.217851 44.069688 \n",
       "L 97.217851 40.069688 \n",
       "L 99.217851 44.069688 \n",
       "L 95.217851 44.069688 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 93.123281 132.628212 \n",
       "Q 67.704286 85.275084 42.814079 38.907036 \n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 42.943762 43.377292 \n",
       "L 42.814079 38.907036 \n",
       "L 46.468093 41.485443 \n",
       "L 42.943762 43.377292 \n",
       "z\n",
       "\" clip-path=\"url(#pd072f5d43d)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\"/>\n",
       "    <g id=\"ytick_2\"/>\n",
       "    <g id=\"ytick_3\"/>\n",
       "    <g id=\"ytick_4\"/>\n",
       "    <g id=\"ytick_5\"/>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_1\">\n",
       "    <defs>\n",
       "     <path id=\"C0_0_65adfaac7e\" d=\"M 0 8.660254 \n",
       "C 2.296726 8.660254 4.499694 7.747755 6.123724 6.123724 \n",
       "C 7.747755 4.499694 8.660254 2.296726 8.660254 -0 \n",
       "C 8.660254 -2.296726 7.747755 -4.499694 6.123724 -6.123724 \n",
       "C 4.499694 -7.747755 2.296726 -8.660254 0 -8.660254 \n",
       "C -2.296726 -8.660254 -4.499694 -7.747755 -6.123724 -6.123724 \n",
       "C -7.747755 -4.499694 -8.660254 -2.296726 -8.660254 0 \n",
       "C -8.660254 2.296726 -7.747755 4.499694 -6.123724 6.123724 \n",
       "C -4.499694 7.747755 -2.296726 8.660254 0 8.660254 \n",
       "z\n",
       "\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"215.27405\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"274.302149\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"215.27405\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"156.24595\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"156.24595\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"97.217851\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"333.330248\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"274.302149\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"215.27405\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"156.24595\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"97.217851\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <use xlink:href=\"#C0_0_65adfaac7e\" x=\"38.189752\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 7.2 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_16\">\n",
       "    <path d=\"M 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_17\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_18\">\n",
       "    <path d=\"M 7.2 7.2 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 0 -->\n",
       "     <g transform=\"translate(211.45655 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 2 -->\n",
       "     <g transform=\"translate(270.484649 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 3 -->\n",
       "     <g transform=\"translate(211.45655 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-33\" d=\"M 2597 2516 \n",
       "Q 3050 2419 3304 2112 \n",
       "Q 3559 1806 3559 1356 \n",
       "Q 3559 666 3084 287 \n",
       "Q 2609 -91 1734 -91 \n",
       "Q 1441 -91 1130 -33 \n",
       "Q 819 25 488 141 \n",
       "L 488 750 \n",
       "Q 750 597 1062 519 \n",
       "Q 1375 441 1716 441 \n",
       "Q 2309 441 2620 675 \n",
       "Q 2931 909 2931 1356 \n",
       "Q 2931 1769 2642 2001 \n",
       "Q 2353 2234 1838 2234 \n",
       "L 1294 2234 \n",
       "L 1294 2753 \n",
       "L 1863 2753 \n",
       "Q 2328 2753 2575 2939 \n",
       "Q 2822 3125 2822 3475 \n",
       "Q 2822 3834 2567 4026 \n",
       "Q 2313 4219 1838 4219 \n",
       "Q 1578 4219 1281 4162 \n",
       "Q 984 4106 628 3988 \n",
       "L 628 4550 \n",
       "Q 988 4650 1302 4700 \n",
       "Q 1616 4750 1894 4750 \n",
       "Q 2613 4750 3031 4423 \n",
       "Q 3450 4097 3450 3541 \n",
       "Q 3450 3153 3228 2886 \n",
       "Q 3006 2619 2597 2516 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 1 -->\n",
       "     <g transform=\"translate(152.42845 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 4 -->\n",
       "     <g transform=\"translate(152.42845 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 5 -->\n",
       "     <g transform=\"translate(93.400351 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_7\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 6 -->\n",
       "     <g transform=\"translate(329.512748 33.603614)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_8\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 7 -->\n",
       "     <g transform=\"translate(270.484649 33.603614)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-37\" d=\"M 525 4666 \n",
       "L 3525 4666 \n",
       "L 3525 4397 \n",
       "L 1831 0 \n",
       "L 1172 0 \n",
       "L 2766 4134 \n",
       "L 525 4134 \n",
       "L 525 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-37\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 8 -->\n",
       "     <g transform=\"translate(211.45655 33.603614)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_10\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 9 -->\n",
       "     <g transform=\"translate(152.42845 33.603614)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-39\" d=\"M 703 97 \n",
       "L 703 672 \n",
       "Q 941 559 1184 500 \n",
       "Q 1428 441 1663 441 \n",
       "Q 2288 441 2617 861 \n",
       "Q 2947 1281 2994 2138 \n",
       "Q 2813 1869 2534 1725 \n",
       "Q 2256 1581 1919 1581 \n",
       "Q 1219 1581 811 2004 \n",
       "Q 403 2428 403 3163 \n",
       "Q 403 3881 828 4315 \n",
       "Q 1253 4750 1959 4750 \n",
       "Q 2769 4750 3195 4129 \n",
       "Q 3622 3509 3622 2328 \n",
       "Q 3622 1225 3098 567 \n",
       "Q 2575 -91 1691 -91 \n",
       "Q 1453 -91 1209 -44 \n",
       "Q 966 3 703 97 \n",
       "z\n",
       "M 1959 2075 \n",
       "Q 2384 2075 2632 2365 \n",
       "Q 2881 2656 2881 3163 \n",
       "Q 2881 3666 2632 3958 \n",
       "Q 2384 4250 1959 4250 \n",
       "Q 1534 4250 1286 3958 \n",
       "Q 1038 3666 1038 3163 \n",
       "Q 1038 2656 1286 2365 \n",
       "Q 1534 2075 1959 2075 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-39\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_11\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 10 -->\n",
       "     <g transform=\"translate(89.582851 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <g clip-path=\"url(#pd072f5d43d)\">\n",
       "     <!-- 11 -->\n",
       "     <g transform=\"translate(30.554752 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pd072f5d43d\">\n",
       "   <rect x=\"7.2\" y=\"7.2\" width=\"357.12\" height=\"266.112\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_54_0.svg"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "import tskit\n",
    "ped_txt = \"\"\"\\\n",
    "# id parent0 parent1 time is_sample\n",
    "0   2   3   0.0 1\n",
    "1   4   5   0.0 1\n",
    "2   6   7   1.0 0\n",
    "3   8   9   1.0 0\n",
    "4   6   7   1.0 0\n",
    "5   10  11  1.0 0\n",
    "6   .   .   2.0 0\n",
    "7   .   .   2.0 0\n",
    "8   .   .   2.0 0\n",
    "9   .   .   2.0 0\n",
    "10  .   .   2.0 0\n",
    "11  .   .   2.0 0\n",
    "\"\"\"\n",
    "pedigree = msprime.parse_pedigree(io.StringIO(ped_txt), sequence_length=100)\n",
    "\n",
    "draw_pedigree(pedigree.tree_sequence())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d650a015",
   "metadata": {},
   "source": [
    "Then suppose we would like to simulate three chromosomes of differing lengths\n",
    "and recombination rates. In practice, we might have recombination\n",
    "maps for each chromosome, but the general strategy remains the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7ccd6dae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome 0 has length 1000000.0 and 48 trees\n",
      "chromosome 1 has length 2000000.0 and 155 trees\n",
      "chromosome 2 has length 3000000.0 and 377 trees\n"
     ]
    }
   ],
   "source": [
    "Ls = [1000000, 2000000, 3000000]\n",
    "rs = [1e-8, 2e-8, 3e-8]\n",
    "\n",
    "ts_chroms = []\n",
    "pedigree = msprime.parse_pedigree(io.StringIO(ped_txt), sequence_length=1)\n",
    "\n",
    "for i, (L, r) in enumerate(zip(Ls, rs)):\n",
    "    pedigree.sequence_length = L\n",
    "    ped_ts = msprime.sim_ancestry(\n",
    "        initial_state=pedigree, model=\"fixed_pedigree\",\n",
    "        recombination_rate=r, random_seed=i+1)\n",
    "    ts_chroms.append(\n",
    "        msprime.sim_ancestry(\n",
    "            initial_state=ped_ts, population_size=1000,\n",
    "            recombination_rate=r, model=\"dtwf\", random_seed=i+100))\n",
    "\n",
    "for i, ts in enumerate(ts_chroms):\n",
    "    print(f\"chromosome {i} has length {ts.sequence_length} and {ts.num_trees} trees\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bfb19cf",
   "metadata": {},
   "source": [
    "This results in three chromosomes of the desired lengths.\n",
    "\n",
    "(sec_multi_chrom_example_2)=\n",
    "\n",
    "#### Example 2: simulating without a pedigree\n",
    "\n",
    "We first set up the chromosome coordinates, recombination rates, and the\n",
    "corresponding recombination map. The following defines a recombination map for\n",
    "three chromosomes each 1 cM in length:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "983d9235",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "r_chrom = 1e-8\n",
    "r_break = math.log(2)\n",
    "chrom_positions = [0, 1e6, 2e6, 3e6]\n",
    "map_positions = [\n",
    "    chrom_positions[0],\n",
    "    chrom_positions[1],\n",
    "    chrom_positions[1] + 1,\n",
    "    chrom_positions[2],\n",
    "    chrom_positions[2] + 1,\n",
    "    chrom_positions[3]\n",
    "]\n",
    "rates = [r_chrom, r_break, r_chrom, r_break, r_chrom]\n",
    "rate_map = msprime.RateMap(position=map_positions, rate=rates)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36315704",
   "metadata": {},
   "source": [
    "We cannot use the default {class}`.StandardCoalescent` model to run simulations\n",
    "of multiple chromosomes; doing so would be equivalent to running independent\n",
    "replicate simulations and **much** less efficient. To get the correct\n",
    "correlation between chromosomes, we must use a model like\n",
    "{ref}`discrete-time Wright-Fisher <sec_ancestry_models_dtwf>`.\n",
    "However, because correlations\n",
    "between chromosomes break down very rapidly, we only need to simulate using\n",
    "`dtwf` for roughly 10-20 generations, after which we can switch to the `hudson`\n",
    "model to finish the simulation more efficiently (see the\n",
    "{ref}`sec_ancestry_models_specifying` section for more information on switching\n",
    "between ancestry models).\n",
    "\n",
    "In this example, we simulate 10 sampled diploid individuals using the\n",
    "recombination map defined above for three chromosomes and switching from the\n",
    "`dtwf` to `hudson` model 20 generations ago:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5d1dbe58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>351</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>3000000.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>20</td></tr>\n",
       "          <tr><td>Total Size</td><td>60.2 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>1254</td>\n",
       "          <td>39.2 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>10</td>\n",
       "          <td>304 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>342</td>\n",
       "          <td>9.4 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>1.3 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2bde48370>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    10,\n",
    "    population_size=1000,\n",
    "    recombination_rate=rate_map,\n",
    "    model=[\n",
    "        msprime.DiscreteTimeWrightFisher(duration=20),\n",
    "        msprime.StandardCoalescent(),\n",
    "    ],\n",
    "    random_seed=1234,\n",
    ")\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05bd759a",
   "metadata": {},
   "source": [
    "To place each chromosome in its own separate tree sequence, we can trim the full tree\n",
    "sequence for each chromosome by defining the end points of each chromosome and\n",
    "using the {meth}`~tskit.TreeSequence.keep_intervals` method. It is important to\n",
    "specify `simplify=False` so that node indices remain consistent across\n",
    "chromosomes. The {meth}`~tskit.TreeSequence.trim` method is then used to trim\n",
    "the tree sequences to the focal chromosome, so that position indices begin at\n",
    "0 within each chromosome tree sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3ad9fbfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000000.0\n",
      "1000000.0\n",
      "1000000.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>102</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>1000000.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>20</td></tr>\n",
       "          <tr><td>Total Size</td><td>26.7 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>371</td>\n",
       "          <td>11.6 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>10</td>\n",
       "          <td>304 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>342</td>\n",
       "          <td>9.4 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>3</td>\n",
       "          <td>2.2 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2bde48d90>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_chroms = []\n",
    "for j in range(len(chrom_positions) - 1):\n",
    "    start, end = chrom_positions[j: j + 2]\n",
    "    chrom_ts = ts.keep_intervals([[start, end]], simplify=False).trim()\n",
    "    ts_chroms.append(chrom_ts)\n",
    "    print(chrom_ts.sequence_length)\n",
    "\n",
    "# the third chromosome\n",
    "chrom_ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50231564",
   "metadata": {},
   "source": [
    "This gives us a list of tree sequences, one for each chromosome, in the order\n",
    "that they were stitched together in the initial recombination map.\n",
    "\n",
    ":::{note}\n",
    "Many recombination maps inferred from data will contain regions at the start\n",
    "and end of the chromosome where recombination rates could not be inferred.\n",
    "These regions typically are assigned rate values of NaN. If we were to\n",
    "concatenate multiple recombination maps that have flanking NaN regions, the\n",
    "resulting maps would have regions with NaN rates in the interior of the map,\n",
    "which will produce an error when using such a map in simulations.\n",
    "\n",
    "These regions typically represent unassay-able segments of the chromosome, and\n",
    "simulated genetic material in these regions should be discarded in downstream\n",
    "analyses. Thus, the suggested approach is to modify the recombination maps to\n",
    "set rates to zero in these flanking regions, concatenate the maps as shown in\n",
    "this example, and then remove those regions of the resulting tree sequences by\n",
    "calling `ts.delete_intervals(nan_intervals)`, where `nan_intervals` is a list\n",
    "containing `[start, end]` values for each NaN flanking region.\n",
    ":::\n",
    "\n",
    "## Recording more information\n",
    "\n",
    "By default `msprime` stores the minimum amount of information required\n",
    "to represent the simulated genealogical history of the samples. Sometimes\n",
    "we are interested in more detailed information; this section gives details\n",
    "of options that allow us to do this.\n",
    "\n",
    "(sec_ancestry_full_arg)=\n",
    "\n",
    "### Ancestral recombination graph\n",
    "\n",
    "In `msprime` we usually want to simulate the coalescent with recombination\n",
    "and represent the output as efficiently as possible. As a result, we don't\n",
    "store individual recombination events, but rather their effects on the output\n",
    "tree sequence. We also do not explicitly store common ancestor events that\n",
    "do not result in marginal coalescences. For some purposes, however, we want\n",
    "to get information on the full history of the simulation, not just the minimal\n",
    "representation of its outcome. The `record_full_arg` option to\n",
    "{func}`.sim_ancestry` provides this functionality, as illustrated in the\n",
    "following example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0a1ab55b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤══════╤══════════╤══════════╤══════════╤════════╗\n",
      "║id│flags │population│individual│time      │metadata║\n",
      "╠══╪══════╪══════════╪══════════╪══════════╪════════╣\n",
      "║0 │     1│         0│         0│0.00000000│        ║\n",
      "║1 │     1│         0│         0│0.00000000│        ║\n",
      "║2 │     1│         0│         1│0.00000000│        ║\n",
      "║3 │     1│         0│         1│0.00000000│        ║\n",
      "║4 │     1│         0│         2│0.00000000│        ║\n",
      "║5 │     1│         0│         2│0.00000000│        ║\n",
      "║6 │     0│         0│        -1│0.21230674│        ║\n",
      "║7 │     0│         0│        -1│0.51485157│        ║\n",
      "║8 │     0│         0│        -1│0.71161449│        ║\n",
      "║9 │131072│         0│        -1│0.91107705│        ║\n",
      "║10│131072│         0│        -1│0.91107705│        ║\n",
      "║11│262144│         0│        -1│1.21743776│        ║\n",
      "║12│     0│         0│        -1│1.23130397│        ║\n",
      "║13│     0│         0│        -1│4.80416300│        ║\n",
      "╚══╧══════╧══════════╧══════════╧══════════╧════════╝\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"400\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l180,0 l0,138.2 l0,25 l0,5 l-180,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M200,0 l180,0 l0,138.2 l0,25 l0,5 l-180,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(200 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"380\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(200 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(380 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">2</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(101.667 26.8)\"><g class=\"a13 c2 node n8 p0\" transform=\"translate(-35 80.5874)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(-23.3333 9.832)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-11.6667 4.18059)\"><path class=\"edge\" d=\"M 0 0 V -4.18059 H 11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(11.6667 4.18059)\"><path class=\"edge\" d=\"M 0 0 V -4.18059 H -11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -9.832 H 23.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(23.3333 3.87451)\"><g class=\"a7 i1 leaf node n2 p0 sample\" transform=\"translate(-11.6667 10.1381)\"><path class=\"edge\" d=\"M 0 0 V -10.1381 H 11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(11.6667 10.1381)\"><path class=\"edge\" d=\"M 0 0 V -10.1381 H -11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -3.87451 H -23.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -80.5874 H 35\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(35 70.3541)\"><g class=\"a12 i2 leaf node n4 p0 sample\" transform=\"translate(11.6667 24.2459)\"><path class=\"edge\" d=\"M 0 0 V -24.2459 H -11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a12 c1 node n11 p0\" transform=\"translate(-11.6667 0.273043)\"><g class=\"a11 c1 node n9 p0\" transform=\"translate(0.0 6.03263)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 17.9403)\"><path class=\"edge\" d=\"M 0 0 V -17.9403 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -6.03263 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -0.273043 H 11.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -70.3541 H -35\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g><g class=\"tree t1\" transform=\"translate(200 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(101.667 26.8)\"><g class=\"a13 c2 node n8 p0\" transform=\"translate(-35 80.5874)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(-23.3333 9.832)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-11.6667 4.18059)\"><path class=\"edge\" d=\"M 0 0 V -4.18059 H 11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(11.6667 4.18059)\"><path class=\"edge\" d=\"M 0 0 V -4.18059 H -11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -9.832 H 23.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(23.3333 3.87451)\"><g class=\"a7 i1 leaf node n2 p0 sample\" transform=\"translate(-11.6667 10.1381)\"><path class=\"edge\" d=\"M 0 0 V -10.1381 H 11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(11.6667 10.1381)\"><path class=\"edge\" d=\"M 0 0 V -10.1381 H -11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -3.87451 H -23.3333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -80.5874 H 35\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(35 70.3541)\"><g class=\"a12 i2 leaf node n4 p0 sample\" transform=\"translate(11.6667 24.2459)\"><path class=\"edge\" d=\"M 0 0 V -24.2459 H -11.6667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a12 c1 node n11 p0\" transform=\"translate(-11.6667 0.273043)\"><g class=\"a11 c1 node n10 p0\" transform=\"translate(0.0 6.03263)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 17.9403)\"><path class=\"edge\" d=\"M 0 0 V -17.9403 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -6.03263 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -0.273043 H 11.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -70.3541 H -35\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 36,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_64_1.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3, recombination_rate=0.1, sequence_length=2,\n",
    "    record_full_arg=True, random_seed=42)\n",
    "print(ts.tables.nodes)\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fcb674e",
   "metadata": {},
   "source": [
    "After running the simulation we first print out the\n",
    "[node table](<https://tskit.readthedocs.io/en/stable/data-model.html#node-table>), which\n",
    "contains information on all the nodes in the tree sequence. Note that `flags`\n",
    "column contains several different values: all of the sample nodes (at time 0)\n",
    "have a flag value of `1` ({data}`tskit.NODE_IS_SAMPLE`). Most other\n",
    "nodes have a flag value of `0`, which is the standard for internal nodes\n",
    "in a coalescent simulations.\n",
    "\n",
    "Nodes 9 and 10 have flags equal to 131072 ({data}`.NODE_IS_RE_EVENT`), which\n",
    "tells us that they correspond to a recombination event in the ARG. A\n",
    "recombination event results in two extra nodes being recorded, one identifying\n",
    "the individual providing the genetic material to the left of the breakpoint and\n",
    "the other identifying the individuals providing the genetic material to the\n",
    "right. The effect of this extra node can be seen in the trees: node 9 is\n",
    "present as a 'unary' node in the left hand tree and node 10 in the right.\n",
    "\n",
    "In this particular case, the first thing that happens to these two lineages\n",
    "as we go up the tree sequence is that they coalesce straight back together again\n",
    "at node 11, forming a (normally undetectable) \"diamond event\" in the ARG, and\n",
    "explaining why the topology of both these trees appears the same. In a full ARG,\n",
    "this is one of many ways that a coalescent event can occur without resulting in a\n",
    "marginal coalescence, and which instead results in an additional unary node in the\n",
    "trees. Such nodes are given a flags value of 262144 ({data}`.NODE_IS_CA_EVENT`).\n",
    "\n",
    "If we wish to reduce these trees down to the minimal representation, we can\n",
    "use {meth}`tskit.TreeSequence.simplify`. The resulting tree sequence will have\n",
    "all of these unary nodes removed and will be equivalent to (but not identical, due to\n",
    "stochastic effects) calling {func}`.sim_ancestry` without the `record_full_arg`\n",
    "argument.\n",
    "\n",
    "Migrations nodes are also recording in the ARG using the\n",
    "{data}`.NODE_IS_MIG_EVENT` flag. See the {ref}`sec_api_node_flags`\n",
    "section for more details.\n",
    "\n",
    "(sec_ancestry_record_migrations)=\n",
    "\n",
    "### Migration events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7fcabdd5",
   "metadata": {
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "    msprime.core.set_seed_rng_seed(42)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c56f7e52",
   "metadata": {},
   "source": [
    "When running simulations with {ref}`demography <sec_demography>`, by\n",
    "default, migration events are not recorded in the resulting tree sequence.\n",
    "While the population associated with each node is determined by the simulated migration\n",
    "events, knowledge of the\n",
    "precise sequence of migrations is useful in a number of situations: for example see\n",
    "the `tskit`\n",
    "[tutorial on introgression.](<https://tskit-dev.github.io/tutorials/introgression.html>)\n",
    "\n",
    "The `record_migrations` parameter allows us to record these events in the\n",
    "simulated tree sequence outputted by {func}`.sim_ancestry`. Briefly, a migration record\n",
    "encodes the node associated with the migration, the time of the event, the source and\n",
    "destination populations (backwards in time), and the left and right coordinates of the\n",
    "migrating segement. For more details on migration records, see the\n",
    "{ref}`migration table definition.<tskit:sec_migration_table_definition>`\n",
    "\n",
    "Here, we provide a simple example of the effect of setting `record_migrations=True`\n",
    "using the {meth}`stepping stone model<msprime.Demography.stepping_stone_model>`\n",
    "where migration is permitted between adjacent populations. Additionally, we\n",
    "set `record_full_arg=True` (see {ref}`previous section <sec_ancestry_full_arg>`)\n",
    "to record nodes corresponding to migration events. This is not necessary, but will be\n",
    "helpful to visualise the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c31f99a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "N = 10\n",
    "demography = msprime.Demography.stepping_stone_model(\n",
    "    [20] * N, migration_rate=0.1, boundaries=True\n",
    ")\n",
    "ts = msprime.sim_ancestry(\n",
    "    {0: 1, N - 1: 1},\n",
    "    demography=demography,\n",
    "    record_migrations=True,\n",
    "    record_full_arg=True,\n",
    "    random_seed=6,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af68e33f",
   "metadata": {},
   "source": [
    "In the following visualisation, we trace the migration history of two samples backwards\n",
    "in time to their most recent common ancestor. Each one of the ten populations in the\n",
    "model is assigned a\n",
    "colour and a position on the x axis. The time of migration events is recorded on the y\n",
    "axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c9452f49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"411.285625pt\" height=\"310.86825pt\" viewBox=\"0 0 411.285625 310.86825\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-10-17T07:24:26.192018</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 310.86825 \n",
       "L 411.285625 310.86825 \n",
       "L 411.285625 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 46.965625 273.312 \n",
       "L 404.085625 273.312 \n",
       "L 404.085625 7.2 \n",
       "L 46.965625 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 63.198352 273.312 \n",
       "L 63.198352 7.2 \n",
       "L 95.663807 7.2 \n",
       "L 95.663807 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #440154; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 95.663807 273.312 \n",
       "L 95.663807 7.2 \n",
       "L 128.129261 7.2 \n",
       "L 128.129261 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #482475; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 128.129261 273.312 \n",
       "L 128.129261 7.2 \n",
       "L 160.594716 7.2 \n",
       "L 160.594716 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #414487; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 160.594716 273.312 \n",
       "L 160.594716 7.2 \n",
       "L 193.06017 7.2 \n",
       "L 193.06017 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #355f8d; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 193.06017 273.312 \n",
       "L 193.06017 7.2 \n",
       "L 225.525625 7.2 \n",
       "L 225.525625 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #2a788e; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 225.525625 273.312 \n",
       "L 225.525625 7.2 \n",
       "L 257.99108 7.2 \n",
       "L 257.99108 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #21918c; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 257.99108 273.312 \n",
       "L 257.99108 7.2 \n",
       "L 290.456534 7.2 \n",
       "L 290.456534 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #22a884; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 290.456534 273.312 \n",
       "L 290.456534 7.2 \n",
       "L 322.921989 7.2 \n",
       "L 322.921989 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #44bf70; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 322.921989 273.312 \n",
       "L 322.921989 7.2 \n",
       "L 355.387443 7.2 \n",
       "L 355.387443 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #7ad151; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 355.387443 273.312 \n",
       "L 355.387443 7.2 \n",
       "L 387.852898 7.2 \n",
       "L 387.852898 273.312 \n",
       "z\n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: #bddf26; opacity: 0.5\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <defs>\n",
       "       <path id=\"m44742be5ae\" d=\"M 0 0 \n",
       "L 0 3.5 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m44742be5ae\" x=\"63.198352\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(60.017102 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m44742be5ae\" x=\"128.129261\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 2 -->\n",
       "      <g transform=\"translate(124.948011 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m44742be5ae\" x=\"193.06017\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 4 -->\n",
       "      <g transform=\"translate(189.87892 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m44742be5ae\" x=\"257.99108\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 6 -->\n",
       "      <g transform=\"translate(254.80983 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_5\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m44742be5ae\" x=\"322.921989\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 8 -->\n",
       "      <g transform=\"translate(319.740739 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_6\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m44742be5ae\" x=\"387.852898\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- 10 -->\n",
       "      <g transform=\"translate(381.490398 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_7\">\n",
       "     <!-- Population -->\n",
       "     <g transform=\"translate(199.259219 301.588562)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-50\" d=\"M 1259 4147 \n",
       "L 1259 2394 \n",
       "L 2053 2394 \n",
       "Q 2494 2394 2734 2622 \n",
       "Q 2975 2850 2975 3272 \n",
       "Q 2975 3691 2734 3919 \n",
       "Q 2494 4147 2053 4147 \n",
       "L 1259 4147 \n",
       "z\n",
       "M 628 4666 \n",
       "L 2053 4666 \n",
       "Q 2838 4666 3239 4311 \n",
       "Q 3641 3956 3641 3272 \n",
       "Q 3641 2581 3239 2228 \n",
       "Q 2838 1875 2053 1875 \n",
       "L 1259 1875 \n",
       "L 1259 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6f\" d=\"M 1959 3097 \n",
       "Q 1497 3097 1228 2736 \n",
       "Q 959 2375 959 1747 \n",
       "Q 959 1119 1226 758 \n",
       "Q 1494 397 1959 397 \n",
       "Q 2419 397 2687 759 \n",
       "Q 2956 1122 2956 1747 \n",
       "Q 2956 2369 2687 2733 \n",
       "Q 2419 3097 1959 3097 \n",
       "z\n",
       "M 1959 3584 \n",
       "Q 2709 3584 3137 3096 \n",
       "Q 3566 2609 3566 1747 \n",
       "Q 3566 888 3137 398 \n",
       "Q 2709 -91 1959 -91 \n",
       "Q 1206 -91 779 398 \n",
       "Q 353 888 353 1747 \n",
       "Q 353 2609 779 3096 \n",
       "Q 1206 3584 1959 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-70\" d=\"M 1159 525 \n",
       "L 1159 -1331 \n",
       "L 581 -1331 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2969 \n",
       "Q 1341 3281 1617 3432 \n",
       "Q 1894 3584 2278 3584 \n",
       "Q 2916 3584 3314 3078 \n",
       "Q 3713 2572 3713 1747 \n",
       "Q 3713 922 3314 415 \n",
       "Q 2916 -91 2278 -91 \n",
       "Q 1894 -91 1617 61 \n",
       "Q 1341 213 1159 525 \n",
       "z\n",
       "M 3116 1747 \n",
       "Q 3116 2381 2855 2742 \n",
       "Q 2594 3103 2138 3103 \n",
       "Q 1681 3103 1420 2742 \n",
       "Q 1159 2381 1159 1747 \n",
       "Q 1159 1113 1420 752 \n",
       "Q 1681 391 2138 391 \n",
       "Q 2594 391 2855 752 \n",
       "Q 3116 1113 3116 1747 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-75\" d=\"M 544 1381 \n",
       "L 544 3500 \n",
       "L 1119 3500 \n",
       "L 1119 1403 \n",
       "Q 1119 906 1312 657 \n",
       "Q 1506 409 1894 409 \n",
       "Q 2359 409 2629 706 \n",
       "Q 2900 1003 2900 1516 \n",
       "L 2900 3500 \n",
       "L 3475 3500 \n",
       "L 3475 0 \n",
       "L 2900 0 \n",
       "L 2900 538 \n",
       "Q 2691 219 2414 64 \n",
       "Q 2138 -91 1772 -91 \n",
       "Q 1169 -91 856 284 \n",
       "Q 544 659 544 1381 \n",
       "z\n",
       "M 1991 3584 \n",
       "L 1991 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6c\" d=\"M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-61\" d=\"M 2194 1759 \n",
       "Q 1497 1759 1228 1600 \n",
       "Q 959 1441 959 1056 \n",
       "Q 959 750 1161 570 \n",
       "Q 1363 391 1709 391 \n",
       "Q 2188 391 2477 730 \n",
       "Q 2766 1069 2766 1631 \n",
       "L 2766 1759 \n",
       "L 2194 1759 \n",
       "z\n",
       "M 3341 1997 \n",
       "L 3341 0 \n",
       "L 2766 0 \n",
       "L 2766 531 \n",
       "Q 2569 213 2275 61 \n",
       "Q 1981 -91 1556 -91 \n",
       "Q 1019 -91 701 211 \n",
       "Q 384 513 384 1019 \n",
       "Q 384 1609 779 1909 \n",
       "Q 1175 2209 1959 2209 \n",
       "L 2766 2209 \n",
       "L 2766 2266 \n",
       "Q 2766 2663 2505 2880 \n",
       "Q 2244 3097 1772 3097 \n",
       "Q 1472 3097 1187 3025 \n",
       "Q 903 2953 641 2809 \n",
       "L 641 3341 \n",
       "Q 956 3463 1253 3523 \n",
       "Q 1550 3584 1831 3584 \n",
       "Q 2591 3584 2966 3190 \n",
       "Q 3341 2797 3341 1997 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-74\" d=\"M 1172 4494 \n",
       "L 1172 3500 \n",
       "L 2356 3500 \n",
       "L 2356 3053 \n",
       "L 1172 3053 \n",
       "L 1172 1153 \n",
       "Q 1172 725 1289 603 \n",
       "Q 1406 481 1766 481 \n",
       "L 2356 481 \n",
       "L 2356 0 \n",
       "L 1766 0 \n",
       "Q 1100 0 847 248 \n",
       "Q 594 497 594 1153 \n",
       "L 594 3053 \n",
       "L 172 3053 \n",
       "L 172 3500 \n",
       "L 594 3500 \n",
       "L 594 4494 \n",
       "L 1172 4494 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-69\" d=\"M 603 3500 \n",
       "L 1178 3500 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 3500 \n",
       "z\n",
       "M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 4134 \n",
       "L 603 4134 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113 \n",
       "L 3513 0 \n",
       "L 2938 0 \n",
       "L 2938 2094 \n",
       "Q 2938 2591 2744 2837 \n",
       "Q 2550 3084 2163 3084 \n",
       "Q 1697 3084 1428 2787 \n",
       "Q 1159 2491 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1366 3272 1645 3428 \n",
       "Q 1925 3584 2291 3584 \n",
       "Q 2894 3584 3203 3211 \n",
       "Q 3513 2838 3513 2113 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-50\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"56.677734\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-70\" x=\"117.859375\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-75\" x=\"181.335938\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"244.714844\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"272.498047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"333.777344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"372.986328\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"400.769531\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"461.951172\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <defs>\n",
       "       <path id=\"m8b354f10bf\" d=\"M 0 0 \n",
       "L -3.5 0 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m8b354f10bf\" x=\"46.965625\" y=\"262.499912\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(33.603125 266.299131)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m8b354f10bf\" x=\"46.965625\" y=\"207.858649\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_9\">\n",
       "      <!-- 50 -->\n",
       "      <g transform=\"translate(27.240625 211.657868)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m8b354f10bf\" x=\"46.965625\" y=\"153.217386\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 100 -->\n",
       "      <g transform=\"translate(20.878125 157.016605)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m8b354f10bf\" x=\"46.965625\" y=\"98.576123\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 150 -->\n",
       "      <g transform=\"translate(20.878125 102.375342)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m8b354f10bf\" x=\"46.965625\" y=\"43.93486\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 200 -->\n",
       "      <g transform=\"translate(20.878125 47.734079)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_13\">\n",
       "     <!-- Time of migration event -->\n",
       "     <g transform=\"translate(14.798438 200.448187)rotate(-90)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-54\" d=\"M -19 4666 \n",
       "L 3928 4666 \n",
       "L 3928 4134 \n",
       "L 2272 4134 \n",
       "L 2272 0 \n",
       "L 1638 0 \n",
       "L 1638 4134 \n",
       "L -19 4134 \n",
       "L -19 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6d\" d=\"M 3328 2828 \n",
       "Q 3544 3216 3844 3400 \n",
       "Q 4144 3584 4550 3584 \n",
       "Q 5097 3584 5394 3201 \n",
       "Q 5691 2819 5691 2113 \n",
       "L 5691 0 \n",
       "L 5113 0 \n",
       "L 5113 2094 \n",
       "Q 5113 2597 4934 2840 \n",
       "Q 4756 3084 4391 3084 \n",
       "Q 3944 3084 3684 2787 \n",
       "Q 3425 2491 3425 1978 \n",
       "L 3425 0 \n",
       "L 2847 0 \n",
       "L 2847 2094 \n",
       "Q 2847 2600 2669 2842 \n",
       "Q 2491 3084 2119 3084 \n",
       "Q 1678 3084 1418 2786 \n",
       "Q 1159 2488 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1356 3278 1631 3431 \n",
       "Q 1906 3584 2284 3584 \n",
       "Q 2666 3584 2933 3390 \n",
       "Q 3200 3197 3328 2828 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-65\" d=\"M 3597 1894 \n",
       "L 3597 1613 \n",
       "L 953 1613 \n",
       "Q 991 1019 1311 708 \n",
       "Q 1631 397 2203 397 \n",
       "Q 2534 397 2845 478 \n",
       "Q 3156 559 3463 722 \n",
       "L 3463 178 \n",
       "Q 3153 47 2828 -22 \n",
       "Q 2503 -91 2169 -91 \n",
       "Q 1331 -91 842 396 \n",
       "Q 353 884 353 1716 \n",
       "Q 353 2575 817 3079 \n",
       "Q 1281 3584 2069 3584 \n",
       "Q 2775 3584 3186 3129 \n",
       "Q 3597 2675 3597 1894 \n",
       "z\n",
       "M 3022 2063 \n",
       "Q 3016 2534 2758 2815 \n",
       "Q 2500 3097 2075 3097 \n",
       "Q 1594 3097 1305 2825 \n",
       "Q 1016 2553 972 2059 \n",
       "L 3022 2063 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-66\" d=\"M 2375 4863 \n",
       "L 2375 4384 \n",
       "L 1825 4384 \n",
       "Q 1516 4384 1395 4259 \n",
       "Q 1275 4134 1275 3809 \n",
       "L 1275 3500 \n",
       "L 2222 3500 \n",
       "L 2222 3053 \n",
       "L 1275 3053 \n",
       "L 1275 0 \n",
       "L 697 0 \n",
       "L 697 3053 \n",
       "L 147 3053 \n",
       "L 147 3500 \n",
       "L 697 3500 \n",
       "L 697 3744 \n",
       "Q 697 4328 969 4595 \n",
       "Q 1241 4863 1831 4863 \n",
       "L 2375 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-67\" d=\"M 2906 1791 \n",
       "Q 2906 2416 2648 2759 \n",
       "Q 2391 3103 1925 3103 \n",
       "Q 1463 3103 1205 2759 \n",
       "Q 947 2416 947 1791 \n",
       "Q 947 1169 1205 825 \n",
       "Q 1463 481 1925 481 \n",
       "Q 2391 481 2648 825 \n",
       "Q 2906 1169 2906 1791 \n",
       "z\n",
       "M 3481 434 \n",
       "Q 3481 -459 3084 -895 \n",
       "Q 2688 -1331 1869 -1331 \n",
       "Q 1566 -1331 1297 -1286 \n",
       "Q 1028 -1241 775 -1147 \n",
       "L 775 -588 \n",
       "Q 1028 -725 1275 -790 \n",
       "Q 1522 -856 1778 -856 \n",
       "Q 2344 -856 2625 -561 \n",
       "Q 2906 -266 2906 331 \n",
       "L 2906 616 \n",
       "Q 2728 306 2450 153 \n",
       "Q 2172 0 1784 0 \n",
       "Q 1141 0 747 490 \n",
       "Q 353 981 353 1791 \n",
       "Q 353 2603 747 3093 \n",
       "Q 1141 3584 1784 3584 \n",
       "Q 2172 3584 2450 3431 \n",
       "Q 2728 3278 2906 2969 \n",
       "L 2906 3500 \n",
       "L 3481 3500 \n",
       "L 3481 434 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-72\" d=\"M 2631 2963 \n",
       "Q 2534 3019 2420 3045 \n",
       "Q 2306 3072 2169 3072 \n",
       "Q 1681 3072 1420 2755 \n",
       "Q 1159 2438 1159 1844 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1341 3275 1631 3429 \n",
       "Q 1922 3584 2338 3584 \n",
       "Q 2397 3584 2469 3576 \n",
       "Q 2541 3569 2628 3553 \n",
       "L 2631 2963 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-76\" d=\"M 191 3500 \n",
       "L 800 3500 \n",
       "L 1894 563 \n",
       "L 2988 3500 \n",
       "L 3597 3500 \n",
       "L 2284 0 \n",
       "L 1503 0 \n",
       "L 191 3500 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-54\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"57.958984\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"85.742188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"183.154297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"244.677734\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"276.464844\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"337.646484\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"372.851562\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"404.638672\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"502.050781\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-67\" x=\"529.833984\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"593.310547\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"634.423828\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"695.703125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"734.912109\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"762.695312\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"823.876953\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"887.255859\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"919.042969\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-76\" x=\"980.566406\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"1039.746094\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"1101.269531\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"1164.648438\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"line2d_12\">\n",
       "    <path d=\"M 111.896534 260.983278 \n",
       "L 144.361989 256.688188 \n",
       "L 111.896534 254.839709 \n",
       "L 144.361989 232.778459 \n",
       "L 176.827443 209.318907 \n",
       "L 144.361989 207.105828 \n",
       "L 176.827443 206.553612 \n",
       "L 144.361989 205.845098 \n",
       "L 176.827443 199.242894 \n",
       "L 209.292898 186.437145 \n",
       "L 176.827443 186.144735 \n",
       "L 144.361989 183.272148 \n",
       "L 111.896534 173.575179 \n",
       "L 144.361989 173.368141 \n",
       "L 176.827443 173.210667 \n",
       "L 209.292898 171.806862 \n",
       "L 241.758352 158.01808 \n",
       "L 209.292898 147.927165 \n",
       "L 176.827443 146.953198 \n",
       "L 144.361989 139.45655 \n",
       "L 111.896534 132.575652 \n",
       "L 79.43108 126.581916 \n",
       "L 111.896534 120.737701 \n",
       "L 144.361989 120.589444 \n",
       "L 111.896534 118.775299 \n",
       "L 144.361989 98.050061 \n",
       "L 111.896534 94.999482 \n",
       "L 144.361989 88.931735 \n",
       "L 176.827443 79.038631 \n",
       "L 209.292898 76.964971 \n",
       "L 241.758352 69.389159 \n",
       "L 274.223807 67.446448 \n",
       "L 241.758352 53.809426 \n",
       "L 209.292898 36.524841 \n",
       "L 241.758352 33.976232 \n",
       "L 274.223807 33.412232 \n",
       "L 241.758352 21.429738 \n",
       "L 274.223807 20.500201 \n",
       "L 306.689261 19.296 \n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: none; stroke-dasharray: 1.5,2.475; stroke-dashoffset: 0; stroke: #000000; stroke-width: 1.5\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_13\">\n",
       "    <path d=\"M 339.154716 261.216 \n",
       "L 306.689261 248.521415 \n",
       "L 274.223807 245.295442 \n",
       "L 306.689261 232.107164 \n",
       "L 274.223807 231.208484 \n",
       "L 306.689261 228.917712 \n",
       "L 339.154716 225.526851 \n",
       "L 306.689261 219.049242 \n",
       "L 274.223807 201.782538 \n",
       "L 241.758352 196.462363 \n",
       "L 274.223807 189.090924 \n",
       "L 306.689261 175.301114 \n",
       "L 339.154716 170.490171 \n",
       "L 371.62017 170.403553 \n",
       "L 339.154716 142.347082 \n",
       "L 371.62017 136.783688 \n",
       "L 339.154716 123.306424 \n",
       "L 371.62017 121.65122 \n",
       "L 339.154716 119.147807 \n",
       "L 306.689261 115.876956 \n",
       "L 274.223807 114.208235 \n",
       "L 306.689261 106.522198 \n",
       "L 339.154716 102.0026 \n",
       "L 371.62017 92.60571 \n",
       "L 339.154716 90.740506 \n",
       "L 371.62017 59.834178 \n",
       "L 339.154716 56.847396 \n",
       "L 371.62017 50.56825 \n",
       "L 339.154716 47.600296 \n",
       "L 371.62017 46.630059 \n",
       "L 339.154716 42.497785 \n",
       "L 306.689261 25.707725 \n",
       "\" clip-path=\"url(#p7f0309709a)\" style=\"fill: none; stroke-dasharray: 9.6,2.4,1.5,2.4; stroke-dashoffset: 0; stroke: #000000; stroke-width: 1.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 46.965625 273.312 \n",
       "L 46.965625 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 404.085625 273.312 \n",
       "L 404.085625 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 46.965625 273.312 \n",
       "L 404.085625 273.312 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_16\">\n",
       "    <path d=\"M 46.965625 7.2 \n",
       "L 404.085625 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p7f0309709a\">\n",
       "   <rect x=\"46.965625\" y=\"7.2\" width=\"357.12\" height=\"266.112\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_70_0.svg"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import tskit\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "\n",
    "cmap = matplotlib.cm.get_cmap('viridis')\n",
    "tree = ts.first()\n",
    "mig_color_dict = {}\n",
    "for u, linestyle in zip(ts.samples()[::2], [\"dotted\", \"dashdot\"]):\n",
    "    mig = ts.tables.migrations\n",
    "    loc = []\n",
    "    time = []\n",
    "    while u != tskit.NULL:\n",
    "        migs = np.where(mig.node == u)[0]\n",
    "        for cur_mig in migs:\n",
    "            cur_mig = mig[cur_mig]\n",
    "            loc.append(cur_mig.dest + 0.5)\n",
    "            time.append(cur_mig.time)\n",
    "        u = tree.parent(u)\n",
    "    plt.plot(loc, time, linestyle=linestyle, color=\"black\")\n",
    "for i in range(N):\n",
    "    rgba = cmap(i / N)\n",
    "    mig_color_dict[i] = np.array(rgba)*255\n",
    "    plt.axvspan(i, i+1, facecolor=rgba, alpha=0.5)\n",
    "plt.ylabel(\"Time of migration event\")\n",
    "plt.xlabel(\"Population\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "642fd960",
   "metadata": {},
   "source": [
    "Note also how many migration events occurred before the two\n",
    "samples coalesced, even in this small example.\n",
    "\n",
    "The next visualisation shows where each of these migrations fall on the edges\n",
    "of the simulated tree. We show the nodes associated with each migration\n",
    "event, colouring nodes and edges by the population with which they are associated\n",
    "(using the same colour scheme as the previous visualisation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "39bc91d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"400\" version=\"1.1\" width=\"500\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.tree .edge {stroke-width:4px}.node.n0> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n0 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n1> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n1 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n2> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n2 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n3> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n3 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n4> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n4 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n5> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n5 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n6> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n6 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n7> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n7 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n8> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n8 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n9> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n9 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n10> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n10 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n11> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n11 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n12> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n12 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n13> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n13 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n14> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n14 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n15> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n15 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n16> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n16 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n17> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n17 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n18> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n18 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n19> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n19 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n20> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n20 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n21> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n21 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n22> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n22 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n23> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n23 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n24> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n24 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n25> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n25 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n26> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n26 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n27> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n27 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n28> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n28 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n29> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n29 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n30> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n30 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n31> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n31 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n32> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n32 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n33> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n33 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n34> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n34 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n35> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n35 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n36> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n36 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n37> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n37 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n38> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n38 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n39> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n39 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n40> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n40 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n41> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n41 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n42> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n42 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n43> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n43 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n44> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n44 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n45> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n45 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n46> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n46 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n47> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n47 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n48> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n48 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n49> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n49 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n50> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n50 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n51> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n51 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n52> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n52 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n53> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n53 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n54> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n54 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n55> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n55 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n56> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n56 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n57> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n57 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n58> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n58 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n59> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n59 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n60> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n60 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n61> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n61 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n62> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n62 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n63> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n63 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n64> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n64 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n65> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n65 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n66> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n66 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n67> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n67 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n68> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n68 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n69> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n69 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n70> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n70 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n71> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n71 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n72> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n72 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n73> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n73 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n74> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n74 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n75> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n75 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n76> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n76 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n77> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n77 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n78> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n78 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n79> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n79 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n80> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n80 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n81> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n81 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n82> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n82 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n83> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n83 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n84> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n84 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n85> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n85 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n86> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n86 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n87> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n87 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n88> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n88 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n89> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n89 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n90> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n90 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n91> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n91 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n92> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n92 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n93> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n93 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n94> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n94 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n95> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n95 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n96> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n96 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n97> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n97 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n98> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n98 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n99> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n99 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n100> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n100 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n101> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n101 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n102> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n102 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n103> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n103 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n104> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n104 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n105> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n105 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n106> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n106 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n107> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n107 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n108> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n108 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n109> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n109 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n110> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n110 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n111> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n111 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n112> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n112 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n113> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n113 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n114> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n114 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n115> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n115 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n116> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n116 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n117> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n117 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n118> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n118 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n4> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n4 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n5> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n5 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n6> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n6 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n7> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n7 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n8> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n8 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n9> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n9 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n10> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n10 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n11> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n11 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n12> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n12 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n13> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n13 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n14> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n14 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n15> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n15 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n16> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n16 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n17> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n17 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n18> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n18 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n19> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n19 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n20> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n20 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n21> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n21 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n22> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n22 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n23> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n23 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n24> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n24 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n25> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n25 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n26> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n26 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n27> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n27 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n28> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n28 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n29> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n29 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n30> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n30 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n31> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n31 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n32> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n32 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n33> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n33 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n34> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n34 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n35> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n35 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n36> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n36 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n37> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n37 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n38> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n38 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n39> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n39 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n40> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n40 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n41> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n41 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n42> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n42 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n43> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n43 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n44> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n44 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n45> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n45 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n46> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n46 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n47> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n47 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n48> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n48 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n50> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n50 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n51> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n51 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n52> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n52 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n53> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n53 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n54> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n54 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n55> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n55 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n56> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n56 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n57> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n57 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n58> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n58 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n59> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n59 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n60> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n60 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n61> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n61 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n62> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n62 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n63> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n63 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n64> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n64 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n65> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n65 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n66> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n66 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n67> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n67 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n68> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n68 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n69> .sym {fill: rgb(68.08602, 1.24287, 84.000825)}.node.n69 > .edge {stroke: rgb(68.08602, 1.24287, 84.000825)}.node.n70> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n70 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n71> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n71 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n72> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n72 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n73> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n73 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n74> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n74 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n75> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n75 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n76> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n76 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n77> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n77 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n78> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n78 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n79> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n79 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n80> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n80 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n81> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n81 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n82> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n82 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n83> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n83 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n84> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n84 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n85> .sym {fill: rgb(72.068865, 35.93613, 116.666835)}.node.n85 > .edge {stroke: rgb(72.068865, 35.93613, 116.666835)}.node.n86> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n86 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n87> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n87 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n88> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n88 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n89> .sym {fill: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n89 > .edge {stroke: rgb(64.75342500000001, 67.63977, 135.145665)}.node.n90> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n90 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n91> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n91 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n92> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n92 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n93> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n93 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n94> .sym {fill: rgb(52.72278, 94.79829, 141.044835)}.node.n94 > .edge {stroke: rgb(52.72278, 94.79829, 141.044835)}.node.n95> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n95 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n96> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n96 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n97> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n97 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n98> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n98 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n99> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n99 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n100> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n100 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n101> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n101 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n102> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n102 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n104> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n104 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n105> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n105 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n106> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n106 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n107> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n107 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n108> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n108 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n109> .sym {fill: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n109 > .edge {stroke: rgb(189.05394, 222.72949500000001, 38.138055)}.node.n110> .sym {fill: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n110 > .edge {stroke: rgb(121.76352, 209.46821999999997, 81.139725)}.node.n111> .sym {fill: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n111 > .edge {stroke: rgb(41.724374999999995, 120.13891500000001, 142.32774)}.node.n112> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n112 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n113> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n113 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n114> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n114 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n115> .sym {fill: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n115 > .edge {stroke: rgb(32.52984, 144.57199500000002, 140.39178)}.node.n116> .sym {fill: rgb(34.34646, 167.95218, 132.000495)}.node.n116 > .edge {stroke: rgb(34.34646, 167.95218, 132.000495)}.node.n117> .sym {fill: rgb(68.069955, 190.93150500000002, 112.346115)}.node.n117 > .edge {stroke: rgb(68.069955, 190.93150500000002, 112.346115)}]]></style></defs><g class=\"tree t0\"><g class=\"plotbox\"><g class=\"c2 node n118 p7 root\" transform=\"translate(321.875 26.8)\"><g class=\"a118 c1 node n114 p7\" transform=\"translate(100.625 9.49463)\"><g class=\"a114 c1 node n110 p8\" transform=\"translate(0.0 23.5342)\"><g class=\"a110 c1 node n109 p9\" transform=\"translate(0.0 5.7921)\"><g class=\"a109 c1 node n108 p8\" transform=\"translate(0.0 1.35996)\"><g class=\"a108 c1 node n107 p9\" transform=\"translate(0.0 4.1601)\"><g class=\"a107 c1 node n105 p8\" transform=\"translate(0.0 8.80131)\"><g class=\"a105 c1 node n104 p9\" transform=\"translate(0.0 4.18649)\"><g class=\"a104 c1 node n88 p8\" transform=\"translate(0.0 43.3206)\"><g class=\"a88 c1 node n87 p9\" transform=\"translate(0.0 2.61441)\"><g class=\"a87 c1 node n80 p8\" transform=\"translate(0.0 13.1714)\"><g class=\"a80 c1 node n79 p7\" transform=\"translate(0.0 6.335)\"><g class=\"a79 c1 node n78 p6\" transform=\"translate(0.0 10.7733)\"><g class=\"a78 c1 node n77 p7\" transform=\"translate(0.0 2.339)\"><g class=\"a77 c1 node n75 p8\" transform=\"translate(0.0 4.58466)\"><g class=\"a75 c1 node n71 p9\" transform=\"translate(0.0 3.50897)\"><g class=\"a71 c1 node n70 p8\" transform=\"translate(0.0 2.32006)\"><g class=\"a70 c1 node n66 p9\" transform=\"translate(0.0 18.8907)\"><g class=\"a66 c1 node n63 p8\" transform=\"translate(0.0 7.79806)\"><g class=\"a63 c1 node n56 p9\" transform=\"translate(0.0 39.326)\"><g class=\"a56 c1 node n55 p8\" transform=\"translate(0.0 0.12141)\"><g class=\"a55 c1 node n50 p7\" transform=\"translate(0.0 6.74337)\"><g class=\"a50 c1 node n42 p6\" transform=\"translate(0.0 19.3288)\"><g class=\"a42 c1 node n41 p5\" transform=\"translate(0.0 10.3324)\"><g class=\"a41 c1 node n39 p6\" transform=\"translate(0.0 7.45715)\"><g class=\"a39 c1 node n25 p7\" transform=\"translate(0.0 24.2023)\"><g class=\"a25 c1 node n20 p8\" transform=\"translate(0.0 9.07949)\"><g class=\"a20 c1 node n19 p7\" transform=\"translate(0.0 4.75288)\"><g class=\"a19 c1 node n18 p6\" transform=\"translate(0.0 3.21092)\"><g class=\"a18 c1 node n17 p7\" transform=\"translate(0.0 1.25966)\"><g class=\"a17 c1 node n13 p6\" transform=\"translate(0.0 18.4857)\"><g class=\"a13 c1 node n12 p7\" transform=\"translate(0.0 4.52176)\"><g class=\"a12 c1 node n4 p8\" transform=\"translate(0.0 17.7937)\"><g class=\"a4 i1 leaf node n2 p9 sample\" transform=\"translate(0.0 1.79963)\"><path class=\"edge\" d=\"M 0 0 V -1.79963 H 0.0\"/><rect class=\"sym\" height=\"10\" width=\"10\" x=\"-5\" y=\"-5\"/><text class=\"lab\" transform=\"translate(0 11)\"/></g><path class=\"edge\" d=\"M 0 0 V -17.7937 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.52176 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -18.4857 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.25966 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.21092 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.75288 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.07949 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -24.2023 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -7.45715 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -10.3324 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -19.3288 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.74337 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.12141 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -39.326 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -7.79806 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -18.8907 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.32006 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.50897 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.58466 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.339 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -10.7733 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.335 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -13.1714 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.61441 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -43.3206 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.18649 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -8.80131 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.1601 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.35996 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -5.7921 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -23.5342 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.49463 H -100.625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\"/></g><g class=\"a118 c1 node n117 p7\" transform=\"translate(-100.625 0.507489)\"><g class=\"a117 c1 node n116 p6\" transform=\"translate(0.0 1.6879)\"><g class=\"a116 c1 node n115 p5\" transform=\"translate(0.0 1.30291)\"><g class=\"a115 c1 node n113 p6\" transform=\"translate(0.0 16.7955)\"><g class=\"a113 c1 node n112 p5\" transform=\"translate(0.0 0.790544)\"><g class=\"a112 c1 node n111 p4\" transform=\"translate(0.0 3.57232)\"><g class=\"a111 c1 node n106 p5\" transform=\"translate(0.0 24.2273)\"><g class=\"a106 c2 node n103 p6\" transform=\"translate(0.0 9.75507)\"><g class=\"a103 c1 node n100 p6\" transform=\"translate(-86.25 9.35959)\"><g class=\"a100 c1 node n98 p5\" transform=\"translate(0.0 2.72305)\"><g class=\"a98 c1 node n95 p4\" transform=\"translate(0.0 10.6188)\"><g class=\"a95 c1 node n94 p3\" transform=\"translate(0.0 2.90659)\"><g class=\"a94 c1 node n89 p2\" transform=\"translate(0.0 13.8669)\"><g class=\"a89 c1 node n85 p1\" transform=\"translate(0.0 8.505)\"><g class=\"a85 c1 node n83 p2\" transform=\"translate(0.0 4.27592)\"><g class=\"a83 c1 node n76 p1\" transform=\"translate(0.0 29.05)\"><g class=\"a76 c1 node n73 p2\" transform=\"translate(0.0 2.54284)\"><g class=\"a73 c1 node n72 p1\" transform=\"translate(0.0 0.207807)\"><g class=\"a72 c1 node n69 p0\" transform=\"translate(0.0 8.19168)\"><g class=\"a69 c1 node n67 p1\" transform=\"translate(0.0 8.40126)\"><g class=\"a67 c1 node n65 p2\" transform=\"translate(0.0 9.64477)\"><g class=\"a65 c1 node n61 p3\" transform=\"translate(0.0 10.5079)\"><g class=\"a61 c1 node n60 p4\" transform=\"translate(0.0 1.36518)\"><g class=\"a60 c1 node n58 p5\" transform=\"translate(0.0 14.1442)\"><g class=\"a58 c1 node n54 p4\" transform=\"translate(0.0 19.3274)\"><g class=\"a54 c1 node n53 p3\" transform=\"translate(0.0 1.96768)\"><g class=\"a53 c1 node n52 p2\" transform=\"translate(0.0 0.220727)\"><g class=\"a52 c1 node n51 p1\" transform=\"translate(0.0 0.2902)\"><g class=\"a51 c2 node n49 p2\" transform=\"translate(0.0 13.1047)\"><g class=\"a49 c1 node n47 p2\" transform=\"translate(57.5 1.39329)\"><g class=\"a47 c1 node n46 p1\" transform=\"translate(0.0 0.768606)\"><g class=\"a46 c1 node n43 p0\" transform=\"translate(0.0 6.18153)\"><g class=\"a43 c1 node n37 p1\" transform=\"translate(0.0 23.4727)\"><g class=\"a37 c1 node n32 p2\" transform=\"translate(0.0 2.47628)\"><g class=\"a32 c1 node n31 p3\" transform=\"translate(0.0 0.697948)\"><g class=\"a31 c1 node n29 p2\" transform=\"translate(0.0 2.19601)\"><g class=\"a29 c1 node n27 p1\" transform=\"translate(0.0 7.33928)\"><g class=\"a27 c1 node n24 p2\" transform=\"translate(0.0 6.75667)\"><g class=\"a24 c1 node n22 p1\" transform=\"translate(0.0 0.941926)\"><g class=\"a22 c1 node n15 p0\" transform=\"translate(0.0 24.614)\"><g class=\"a15 c1 node n10 p1\" transform=\"translate(0.0 19.0225)\"><g class=\"a10 c1 node n9 p0\" transform=\"translate(0.0 0.85802)\"><g class=\"a9 c1 node n6 p1\" transform=\"translate(0.0 11.0891)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 3.73073)\"><path class=\"edge\" d=\"M 0 0 V -3.73073 H 0.0\"/><rect class=\"sym\" height=\"10\" width=\"10\" x=\"-5\" y=\"-5\"/><text class=\"lab\" transform=\"translate(0 11)\"/></g><path class=\"edge\" d=\"M 0 0 V -11.0891 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.85802 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -19.0225 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -24.614 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.941926 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.75667 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -7.33928 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.19601 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.697948 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.47628 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -23.4727 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.18153 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.768606 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.39329 H -57.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\"/></g><g class=\"a49 c1 node n48 p2\" transform=\"translate(-57.5 0.487264)\"><g class=\"a48 c1 node n45 p3\" transform=\"translate(0.0 4.02643)\"><g class=\"a45 c1 node n44 p4\" transform=\"translate(0.0 0.409863)\"><g class=\"a44 c1 node n40 p3\" transform=\"translate(0.0 17.9495)\"><g class=\"a40 c1 node n36 p2\" transform=\"translate(0.0 9.25413)\"><g class=\"a36 c1 node n35 p3\" transform=\"translate(0.0 0.993105)\"><g class=\"a35 c1 node n33 p2\" transform=\"translate(0.0 0.774028)\"><g class=\"a33 c1 node n30 p3\" transform=\"translate(0.0 3.10201)\"><g class=\"a30 c1 node n16 p2\" transform=\"translate(0.0 32.8826)\"><g class=\"a16 c1 node n8 p1\" transform=\"translate(0.0 30.9227)\"><g class=\"a8 c1 node n7 p2\" transform=\"translate(0.0 2.59096)\"><g class=\"a7 c1 node n5 p1\" transform=\"translate(0.0 6.02031)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 2.12583)\"><path class=\"edge\" d=\"M 0 0 V -2.12583 H 0.0\"/><rect class=\"sym\" height=\"10\" width=\"10\" x=\"-5\" y=\"-5\"/><text class=\"lab\" transform=\"translate(0 11)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.02031 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.59096 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -30.9227 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -32.8826 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.10201 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.774028 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.993105 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.25413 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -17.9495 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.409863 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.02643 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.487264 H 57.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -13.1047 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.2902 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.220727 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.96768 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -19.3274 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -14.1442 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.36518 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -10.5079 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.64477 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -8.40126 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -8.19168 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.207807 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.54284 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -29.05 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.27592 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -8.505 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -13.8669 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.90659 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -10.6188 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.72305 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.35959 H 86.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><g class=\"a103 c1 node n102 p6\" transform=\"translate(86.25 0.444834)\"><g class=\"a102 c1 node n101 p5\" transform=\"translate(0.0 3.36635)\"><g class=\"a101 c1 node n99 p6\" transform=\"translate(0.0 5.93912)\"><g class=\"a99 c1 node n97 p7\" transform=\"translate(0.0 9.95915)\"><g class=\"a97 c1 node n96 p6\" transform=\"translate(0.0 0.00795453)\"><g class=\"a96 c1 node n93 p7\" transform=\"translate(0.0 14.4581)\"><g class=\"a93 c1 node n92 p6\" transform=\"translate(0.0 1.02976)\"><g class=\"a92 c1 node n91 p7\" transform=\"translate(0.0 3.09305)\"><g class=\"a91 c1 node n90 p8\" transform=\"translate(0.0 0.384041)\"><g class=\"a90 c1 node n86 p7\" transform=\"translate(0.0 6.66077)\"><g class=\"a86 c1 node n84 p8\" transform=\"translate(0.0 2.79857)\"><g class=\"a84 c1 node n82 p7\" transform=\"translate(0.0 6.88012)\"><g class=\"a82 c1 node n81 p8\" transform=\"translate(0.0 1.35084)\"><g class=\"a81 c1 node n74 p7\" transform=\"translate(0.0 25.519)\"><g class=\"a74 c1 node n68 p8\" transform=\"translate(0.0 14.6431)\"><g class=\"a68 c1 node n64 p9\" transform=\"translate(0.0 15.9208)\"><g class=\"a64 c1 node n62 p8\" transform=\"translate(0.0 2.95471)\"><g class=\"a62 c1 node n59 p9\" transform=\"translate(0.0 19.1138)\"><g class=\"a59 c1 node n57 p8\" transform=\"translate(0.0 10.4149)\"><g class=\"a57 c1 node n38 p9\" transform=\"translate(0.0 57.0374)\"><g class=\"a38 c1 node n34 p8\" transform=\"translate(0.0 2.47112)\"><g class=\"a34 c1 node n28 p9\" transform=\"translate(0.0 9.76538)\"><g class=\"a28 c1 node n26 p8\" transform=\"translate(0.0 4.4469)\"><g class=\"a26 c1 node n23 p7\" transform=\"translate(0.0 3.85779)\"><g class=\"a23 c1 node n21 p6\" transform=\"translate(0.0 5.36391)\"><g class=\"a21 c1 node n14 p7\" transform=\"translate(0.0 24.123)\"><g class=\"a14 c1 node n11 p8\" transform=\"translate(0.0 13.0362)\"><g class=\"a11 i1 leaf node n3 p9 sample\" transform=\"translate(0.0 17.7202)\"><path class=\"edge\" d=\"M 0 0 V -17.7202 H 0.0\"/><rect class=\"sym\" height=\"10\" width=\"10\" x=\"-5\" y=\"-5\"/><text class=\"lab\" transform=\"translate(0 11)\"/></g><path class=\"edge\" d=\"M 0 0 V -13.0362 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -24.123 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -5.36391 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.85779 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -4.4469 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.76538 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.47112 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -57.0374 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -10.4149 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -19.1138 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.95471 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -15.9208 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -14.6431 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -25.519 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.35084 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.88012 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -2.79857 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -6.66077 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.384041 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.09305 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.02976 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -14.4581 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.00795453 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.95915 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -5.93912 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.36635 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.444834 H -86.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -9.75507 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -24.2273 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -3.57232 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.790544 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -16.7955 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.30291 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -1.6879 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><path class=\"edge\" d=\"M 0 0 V -0.507489 H 100.625\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\"/></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"5\"/><text class=\"lab\" transform=\"translate(0 -11)\"/></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 40,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_72_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mig_color_dict = {}\n",
    "cmap = matplotlib.cm.get_cmap(\"viridis\")\n",
    "for i in range(N):\n",
    "    rgba = cmap(i / N)\n",
    "    mig_color_dict[i] = np.array(rgba) * 255\n",
    "\n",
    "edge_colors = {}\n",
    "string = \".tree .edge {stroke-width:4px}\"\n",
    "pop = ts.tables.nodes.population\n",
    "for node in ts.nodes():\n",
    "    node = node.id\n",
    "    edge_colors[node] = mig_color_dict[pop[node]]\n",
    "    string += (\n",
    "        \".node.n\"\n",
    "        + str(node)\n",
    "        + \"> .sym {fill: rgb\"\n",
    "        + str(tuple(mig_color_dict[pop[node]][0:3]))\n",
    "        + \"}\"\n",
    "    )\n",
    "    string += (\n",
    "        \".node.n\"\n",
    "        + str(node)\n",
    "        + \" > .edge {stroke: rgb\"\n",
    "        + str(tuple(mig_color_dict[pop[node]][0:3]))\n",
    "        + \"}\"\n",
    "    )\n",
    "for mig in ts.migrations():\n",
    "    edge_colors[mig.node] = mig_color_dict[mig.source]\n",
    "    string += (\n",
    "        \".node.n\"\n",
    "        + str(mig.node)\n",
    "        + \"> .sym {fill: rgb\"\n",
    "        + str(tuple(mig_color_dict[pop[mig.node]][0:3]))\n",
    "        + \"}\"\n",
    "    )\n",
    "    string += (\n",
    "        \".node.n\"\n",
    "        + str(mig.node)\n",
    "        + \" > .edge {stroke: rgb\"\n",
    "        + str(tuple(mig_color_dict[mig.dest][0:3]))\n",
    "        + \"}\"\n",
    "    )\n",
    "node_labels = {node.id: \"\" for node in ts.nodes()}\n",
    "SVG(\n",
    "    ts.first().draw_svg(\n",
    "        size=(500, 400), style=string, node_labels=node_labels, symbol_size=10\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b03b342",
   "metadata": {},
   "source": [
    "(sec_ancestry_census_events)=\n",
    "\n",
    "### Census events\n",
    "\n",
    "Census events allow you to add a node to each branch of the tree sequence at a given time\n",
    "during the simulation. This can be useful when you wish to study haplotypes that are\n",
    "ancestral to your simulated sample, or when you wish to know which lineages were present in\n",
    "which populations at specified times.\n",
    "\n",
    "For instance, the following example creates a two population island\n",
    "model {class}`.Demography`, with the populations exchanging migrants\n",
    "at rate 0.05. At generation 5000, we add a census event using the\n",
    "{meth}`.Demography.add_census` method\n",
    "to determine where each of the lineages is at that time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5d4d42e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l190,0 l0,138.2 l10.64,25 l0,5 l-200.64,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M210,0 l190,0 l0,138.2 l135.28,25 l0,5 l-314.64,0 l0,-5 l-10.64,-25 l0,-138.2z\"/><path d=\"M400,0 l190,0 l0,138.2 l142.88,25 l0,5 l-197.6,0 l0,-5 l-135.28,-25 l0,-138.2z\"/><path d=\"M590,0 l190,0 l0,138.2 l0,25 l0,5 l-47.12,0 l0,-5 l-142.88,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(400 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(220.64 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">264</text></g></g><g class=\"tick\" transform=\"translate(535.28 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">678</text></g></g><g class=\"tick\" transform=\"translate(732.88 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">938</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1000</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p1 root\" transform=\"translate(95 60.9777)\"><g class=\"a14 c1 node n9 p0\" transform=\"translate(37.5 2.68648)\"><g class=\"a9 c2 node n5 p1\" transform=\"translate(0.0 14.5376)\"><g class=\"a5 i0 leaf node n1 p0 sample\" transform=\"translate(-18.75 43.1983)\"><path class=\"edge\" d=\"M 0 0 V -43.1983 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a5 i1 leaf node n3 p1 sample\" transform=\"translate(18.75 43.1983)\"><path class=\"edge\" d=\"M 0 0 V -43.1983 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -14.5376 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -2.68648 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><g class=\"a14 c1 node n12 p1\" transform=\"translate(-37.5 2.68648)\"><g class=\"a12 c2 node n4 p1\" transform=\"translate(0.0 30.8089)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-18.75 26.927)\"><path class=\"edge\" d=\"M 0 0 V -26.927 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i1 leaf node n2 p1 sample\" transform=\"translate(18.75 26.927)\"><path class=\"edge\" d=\"M 0 0 V -26.927 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -30.8089 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -2.68648 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t1\" transform=\"translate(210 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p1 root\" transform=\"translate(118.438 60.9777)\"><g class=\"a14 c1 node n10 p0\" transform=\"translate(32.8125 2.68648)\"><g class=\"a10 i1 leaf node n3 p1 sample\" transform=\"translate(0.0 57.7359)\"><path class=\"edge\" d=\"M 0 0 V -57.7359 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -2.68648 H -32.8125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a14 c1 node n13 p1\" transform=\"translate(-32.8125 2.68648)\"><g class=\"a13 c2 node n6 p0\" transform=\"translate(0.0 9.0437)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(28.125 48.6922)\"><path class=\"edge\" d=\"M 0 0 V -48.6922 H -28.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 c2 node n4 p1\" transform=\"translate(-28.125 21.7652)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-18.75 26.927)\"><path class=\"edge\" d=\"M 0 0 V -26.927 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a4 i1 leaf node n2 p1 sample\" transform=\"translate(18.75 26.927)\"><path class=\"edge\" d=\"M 0 0 V -26.927 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -21.7652 H 28.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -9.0437 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -2.68648 H 32.8125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t2\" transform=\"translate(400 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p1 root\" transform=\"translate(95 60.9777)\"><g class=\"a14 c1 node n13 p1\" transform=\"translate(-37.5 2.68648)\"><g class=\"a13 c2 node n6 p0\" transform=\"translate(0.0 9.0437)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(18.75 48.6922)\"><path class=\"edge\" d=\"M 0 0 V -48.6922 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-18.75 48.6922)\"><path class=\"edge\" d=\"M 0 0 V -48.6922 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><path class=\"edge\" d=\"M 0 0 V -9.0437 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -2.68648 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><g class=\"a14 c1 node n11 p0\" transform=\"translate(37.5 2.68648)\"><g class=\"a11 c2 node n7 p1\" transform=\"translate(0.0 4.84214)\"><g class=\"a7 i1 leaf node n2 p1 sample\" transform=\"translate(-18.75 52.8937)\"><path class=\"edge\" d=\"M 0 0 V -52.8937 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a7 i1 leaf node n3 p1 sample\" transform=\"translate(18.75 52.8937)\"><path class=\"edge\" d=\"M 0 0 V -52.8937 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -4.84214 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -2.68648 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g><g class=\"tree t3\" transform=\"translate(590 0)\"><g class=\"plotbox\"><g class=\"c2 node n15 p0 root\" transform=\"translate(95 26.8)\"><g class=\"a15 c1 node n8 p0\" transform=\"translate(-37.5 36.8641)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(0.0 9.0437)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(18.75 48.6922)\"><path class=\"edge\" d=\"M 0 0 V -48.6922 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-18.75 48.6922)\"><path class=\"edge\" d=\"M 0 0 V -48.6922 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><path class=\"edge\" d=\"M 0 0 V -9.0437 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -36.8641 H 37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a15 c1 node n11 p0\" transform=\"translate(37.5 36.8641)\"><g class=\"a11 c2 node n7 p1\" transform=\"translate(0.0 4.84214)\"><g class=\"a7 i1 leaf node n2 p1 sample\" transform=\"translate(-18.75 52.8937)\"><path class=\"edge\" d=\"M 0 0 V -52.8937 H 18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a7 i1 leaf node n3 p1 sample\" transform=\"translate(18.75 52.8937)\"><path class=\"edge\" d=\"M 0 0 V -52.8937 H -18.75\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -4.84214 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -36.8641 H -37.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">15</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 41,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_74_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography.island_model([1000, 1000], migration_rate=0.05)\n",
    "demography.add_census(time=5000)\n",
    "ts = msprime.sim_ancestry(\n",
    "    {0: 1, 1: 1},\n",
    "    demography=demography,\n",
    "    recombination_rate=1e-7,\n",
    "    sequence_length=1000,\n",
    "    random_seed=141\n",
    ")\n",
    "SVG(ts.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38a74fb1",
   "metadata": {},
   "source": [
    "The resulting tree sequence has nodes on each tree at the specified census time.\n",
    "These are the nodes with IDs 8, 9, 10, 11, 12 and 13.\n",
    "\n",
    "This tells us that the genetic material ancestral to the present day sample was\n",
    "held within 5 haplotypes at time 5000. The node table shows us that four of\n",
    "these haplotypes (nodes 8, 9, 10 and 11) were in population 0 at this time, and\n",
    "two of these haplotypes (nodes 12 and 13) were in population 1 at this time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7abda506",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "╔══╤═══════╤══════════╤══════════╤═════════════╤════════╗\n",
      "║id│flags  │population│individual│time         │metadata║\n",
      "╠══╪═══════╪══════════╪══════════╪═════════════╪════════╣\n",
      "║0 │      1│         0│         0│   0.00000000│        ║\n",
      "║1 │      1│         0│         0│   0.00000000│        ║\n",
      "║2 │      1│         1│         1│   0.00000000│        ║\n",
      "║3 │      1│         1│         1│   0.00000000│        ║\n",
      "║4 │      0│         1│        -1│2331.91109774│        ║\n",
      "║5 │      0│         1│        -1│3741.02811877│        ║\n",
      "║6 │      0│         0│        -1│4216.80416680│        ║\n",
      "║7 │      0│         1│        -1│4580.66488183│        ║\n",
      "║8 │1048576│         0│        -1│5000.00000000│        ║\n",
      "║9 │1048576│         0│        -1│5000.00000000│        ║\n",
      "║10│1048576│         0│        -1│5000.00000000│        ║\n",
      "║11│1048576│         0│        -1│5000.00000000│        ║\n",
      "║12│1048576│         1│        -1│5000.00000000│        ║\n",
      "║13│1048576│         1│        -1│5000.00000000│        ║\n",
      "║14│      0│         1│        -1│5232.65267392│        ║\n",
      "║15│      0│         0│        -1│8192.48105714│        ║\n",
      "╚══╧═══════╧══════════╧══════════╧═════════════╧════════╝\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(ts.tables.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8dd4431",
   "metadata": {},
   "source": [
    "If we wish to study these ancestral haplotypes further, we can simplify the tree sequence\n",
    "with respect to the census nodes and perform subsequent analyses on this simplified tree\n",
    "sequence.\n",
    "In this example, `ts_anc` is a tree sequence obtained from the original tree sequence\n",
    "`ts` by labelling the census nodes as samples and removing all nodes and edges that are\n",
    "not ancestral to these census nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f924ebff",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = [i.id for i in ts.nodes() if i.flags==msprime.NODE_IS_CEN_EVENT]\n",
    "ts_anc = ts.simplify(samples=nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51dc51f4",
   "metadata": {},
   "source": [
    "(sec_ancestry_census_events_dtwf)=\n",
    "\n",
    "#### Using the DTWF model\n",
    "\n",
    "Census events assume that they are being used as part of a continuous time\n",
    "simulation process, in which multiple events are guaranteed not to happen\n",
    "at the same time. However, the {class}`.DiscreteTimeWrightFisher` model does\n",
    "not meet these assumptions, and allows multiple events (such as coalescences,\n",
    "migrations etc) to all occur at the same time. This includes census events,\n",
    "which can lead to errors when used in conjunction with the DTWF model.\n",
    "\n",
    "Consider the following example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "194b63e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 97.1864)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 72.9727)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3</text></g></g><g class=\"tick\" transform=\"translate(56.8 58.8087)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5</text></g></g><g class=\"tick\" transform=\"translate(56.8 37.6346)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">14</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n10 p0 root\" transform=\"translate(76.3333 26.8)\"><g class=\"a10 i2 leaf node n5 p0 sample\" transform=\"translate(19.9333 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -19.9333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-19.9333 10.8346)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(19.0667 35.3381)\"><g class=\"a7 i0 leaf node n1 p0 sample\" transform=\"translate(-6.93333 48.4273)\"><path class=\"edge\" d=\"M 0 0 V -48.4273 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(6.93333 48.4273)\"><path class=\"edge\" d=\"M 0 0 V -48.4273 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -35.3381 H -19.0667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-19.0667 21.1741)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 62.5913)\"><path class=\"edge\" d=\"M 0 0 V -62.5913 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 c2 node n6 p0\" transform=\"translate(10.4 38.3777)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-6.93333 24.2136)\"><path class=\"edge\" d=\"M 0 0 V -24.2136 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(6.93333 24.2136)\"><path class=\"edge\" d=\"M 0 0 V -24.2136 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -38.3777 H -10.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -21.1741 H 19.0667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -10.8346 H 19.9333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 44,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_80_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", initial_size=10)\n",
    "ts = msprime.sim_ancestry({\"A\": 3}, demography=demography, model=\"dtwf\", random_seed=33)\n",
    "SVG(ts.draw_svg(y_axis=True, time_scale=\"log_time\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aad82a95",
   "metadata": {},
   "source": [
    "We can see that a coalescence happens at time 3, giving node 7. Now, let's\n",
    "try to perform a census at time 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "5525be54",
   "metadata": {
    "tags": [
     "raises-exception",
     "output_scroll"
    ]
   },
   "outputs": [
    {
     "ename": "LibraryError",
     "evalue": "The simulation model supplied resulted in a parent node having a time value <= to its child. This can occur as a result of multiple bottlenecks happening at the same time, multiple census events at the same time or numerical imprecision with very smallpopulation sizes.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLibraryError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [45], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m demography\u001b[38;5;241m.\u001b[39madd_population(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m\"\u001b[39m, initial_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m10\u001b[39m)\n\u001b[1;32m      3\u001b[0m demography\u001b[38;5;241m.\u001b[39madd_census(time\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n\u001b[0;32m----> 4\u001b[0m ts \u001b[38;5;241m=\u001b[39m \u001b[43mmsprime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msim_ancestry\u001b[49m\u001b[43m(\u001b[49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mA\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdtwf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrandom_seed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m33\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1226\u001b[0m, in \u001b[0;36msim_ancestry\u001b[0;34m(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, num_labels, random_seed, num_replicates, replicate_index, record_provenance)\u001b[0m\n\u001b[1;32m   1206\u001b[0m     provenance_dict \u001b[38;5;241m=\u001b[39m provenance\u001b[38;5;241m.\u001b[39mget_provenance_dict(parameters)\n\u001b[1;32m   1207\u001b[0m sim \u001b[38;5;241m=\u001b[39m _parse_sim_ancestry(\n\u001b[1;32m   1208\u001b[0m     samples\u001b[38;5;241m=\u001b[39msamples,\n\u001b[1;32m   1209\u001b[0m     sequence_length\u001b[38;5;241m=\u001b[39msequence_length,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1224\u001b[0m     random_seed\u001b[38;5;241m=\u001b[39mrandom_seed,\n\u001b[1;32m   1225\u001b[0m )\n\u001b[0;32m-> 1226\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_wrap_replicates\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1227\u001b[0m \u001b[43m    \u001b[49m\u001b[43msim\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1228\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_replicates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_replicates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1229\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreplicate_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreplicate_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1230\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprovenance_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprovenance_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1231\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:660\u001b[0m, in \u001b[0;36m_wrap_replicates\u001b[0;34m(simulator, num_replicates, replicate_index, provenance_dict, mutation_rate)\u001b[0m\n\u001b[1;32m    654\u001b[0m iterator \u001b[38;5;241m=\u001b[39m simulator\u001b[38;5;241m.\u001b[39mrun_replicates(\n\u001b[1;32m    655\u001b[0m     num_replicates,\n\u001b[1;32m    656\u001b[0m     mutation_rate\u001b[38;5;241m=\u001b[39mmutation_rate,\n\u001b[1;32m    657\u001b[0m     provenance_dict\u001b[38;5;241m=\u001b[39mprovenance_dict,\n\u001b[1;32m    658\u001b[0m )\n\u001b[1;32m    659\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m replicate_index \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 660\u001b[0m     deque \u001b[38;5;241m=\u001b[39m \u001b[43mcollections\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdeque\u001b[49m\u001b[43m(\u001b[49m\u001b[43miterator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmaxlen\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    661\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m deque\u001b[38;5;241m.\u001b[39mpop()\n\u001b[1;32m    662\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1497\u001b[0m, in \u001b[0;36mSimulator.run_replicates\u001b[0;34m(self, num_replicates, mutation_rate, provenance_dict)\u001b[0m\n\u001b[1;32m   1495\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m replicate_index \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(num_replicates):\n\u001b[1;32m   1496\u001b[0m     logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mStarting replicate \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, replicate_index)\n\u001b[0;32m-> 1497\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1498\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m mutation_rate \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1499\u001b[0m         \u001b[38;5;66;03m# This is only called from simulate() or the ms interface,\u001b[39;00m\n\u001b[1;32m   1500\u001b[0m         \u001b[38;5;66;03m# so does not need any further parameters.\u001b[39;00m\n\u001b[1;32m   1501\u001b[0m         mutations\u001b[38;5;241m.\u001b[39m_simple_mutate(\n\u001b[1;32m   1502\u001b[0m             tables\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtables,\n\u001b[1;32m   1503\u001b[0m             random_generator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrandom_generator,\n\u001b[1;32m   1504\u001b[0m             sequence_length\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msequence_length,\n\u001b[1;32m   1505\u001b[0m             rate\u001b[38;5;241m=\u001b[39mmutation_rate,\n\u001b[1;32m   1506\u001b[0m         )\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1462\u001b[0m, in \u001b[0;36mSimulator.run\u001b[0;34m(self, event_chunk, debug_func)\u001b[0m\n\u001b[1;32m   1460\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mModel durations must be >= 0\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1461\u001b[0m end_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtime \u001b[38;5;241m+\u001b[39m model_duration, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mend_time)\n\u001b[0;32m-> 1462\u001b[0m exit_reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_until\u001b[49m\u001b[43m(\u001b[49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mevent_chunk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdebug_func\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1463\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m exit_reason \u001b[38;5;241m==\u001b[39m ExitReason\u001b[38;5;241m.\u001b[39mCOALESCENCE \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtime \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mend_time:\n\u001b[1;32m   1464\u001b[0m     logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSkipping remaining \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m models\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodels) \u001b[38;5;241m-\u001b[39m j \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1427\u001b[0m, in \u001b[0;36mSimulator._run_until\u001b[0;34m(self, end_time, event_chunk, debug_func)\u001b[0m\n\u001b[1;32m   1425\u001b[0m ret \u001b[38;5;241m=\u001b[39m ExitReason\u001b[38;5;241m.\u001b[39mMAX_EVENTS\n\u001b[1;32m   1426\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m ret \u001b[38;5;241m==\u001b[39m ExitReason\u001b[38;5;241m.\u001b[39mMAX_EVENTS:\n\u001b[0;32m-> 1427\u001b[0m     ret \u001b[38;5;241m=\u001b[39m ExitReason(\u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mevent_chunk\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   1428\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtime \u001b[38;5;241m>\u001b[39m end_time:\n\u001b[1;32m   1429\u001b[0m         \u001b[38;5;66;03m# Currently the Pedigree and Sweeps models are \"non-reentrant\"\u001b[39;00m\n\u001b[1;32m   1430\u001b[0m         \u001b[38;5;66;03m# We can change this to an assertion once these have been fixed.\u001b[39;00m\n\u001b[1;32m   1431\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1432\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mModel \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m does not support interruption. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1433\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPlease open an issue on GitHub\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1434\u001b[0m         )\n",
      "\u001b[0;31mLibraryError\u001b[0m: The simulation model supplied resulted in a parent node having a time value <= to its child. This can occur as a result of multiple bottlenecks happening at the same time, multiple census events at the same time or numerical imprecision with very smallpopulation sizes."
     ]
    }
   ],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", initial_size=10)\n",
    "demography.add_census(time=3)\n",
    "ts = msprime.sim_ancestry({\"A\": 3}, demography=demography, model=\"dtwf\", random_seed=33)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "643bbcb8",
   "metadata": {},
   "source": [
    "We got an error from msprime about a parent node having a time value <= to its child,\n",
    "which occurred because we tried to add set of census nodes at time 3, but node 7 was\n",
    "already at time 3, giving a zero branch length (which is disallowed by tskit).\n",
    "\n",
    "The solution is to use non-integer time values when performing a census\n",
    "under the DTWF:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4142b568",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 97.1864)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 72.9727)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 68.8582)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3.50</text></g></g><g class=\"tick\" transform=\"translate(56.8 58.8087)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 37.6346)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">14.00</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(76.3333 26.8)\"><g class=\"a14 c1 node n9 p0\" transform=\"translate(19.9333 42.0582)\"><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(0.0 52.5418)\"><path class=\"edge\" d=\"M 0 0 V -52.5418 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -42.0582 H -19.9333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(-19.9333 10.8346)\"><g class=\"a13 c1 node n10 p0\" transform=\"translate(19.0667 31.2236)\"><g class=\"a10 c2 node n7 p0\" transform=\"translate(0.0 4.1145)\"><g class=\"a7 i0 leaf node n1 p0 sample\" transform=\"translate(-6.93333 48.4273)\"><path class=\"edge\" d=\"M 0 0 V -48.4273 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a7 i1 leaf node n3 p0 sample\" transform=\"translate(6.93333 48.4273)\"><path class=\"edge\" d=\"M 0 0 V -48.4273 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -4.1145 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -31.2236 H -19.0667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-19.0667 21.1741)\"><g class=\"a12 c1 node n8 p0\" transform=\"translate(10.4 10.0496)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(0.0 28.3281)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-6.93333 24.2136)\"><path class=\"edge\" d=\"M 0 0 V -24.2136 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(6.93333 24.2136)\"><path class=\"edge\" d=\"M 0 0 V -24.2136 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -28.3281 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -10.0496 H -10.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><g class=\"a12 c1 node n11 p0\" transform=\"translate(-10.4 10.0496)\"><g class=\"a11 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 52.5418)\"><path class=\"edge\" d=\"M 0 0 V -52.5418 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><path class=\"edge\" d=\"M 0 0 V -10.0496 H 10.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -21.1741 H 19.0667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -10.8346 H 19.9333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 46,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_84_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", initial_size=10)\n",
    "demography.add_census(time=3.5)\n",
    "ts = msprime.sim_ancestry({\"A\": 3}, demography=demography, model=\"dtwf\", random_seed=33)\n",
    "SVG(ts.draw_svg(y_axis=True, time_scale=\"log_time\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a6142f",
   "metadata": {},
   "source": [
    "## Manipulating simulation time\n",
    "\n",
    "(sec_ancestry_end_time)=\n",
    "\n",
    "### Stopping simulations early\n",
    "\n",
    "In most simulations we want to simulate a complete history for our samples:\n",
    "that is, to when we have a common ancestor at every point along the\n",
    "sequence. For example, here we generate a complete ancestral history\n",
    "for two diploid samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "2b7ed271",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"800\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l180.8,0 l0,138.2 l180.8,25 l0,5 l-361.6,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M237.6,0 l180.8,0 l0,138.2 l216.96,25 l0,5 l-216.96,0 l0,-5 l-180.8,-25 l0,-138.2z\"/><path d=\"M418.4,0 l180.8,0 l0,138.2 l108.48,25 l0,5 l-72.32,0 l0,-5 l-216.96,-25 l0,-138.2z\"/><path d=\"M599.2,0 l180.8,0 l0,138.2 l0,25 l0,5 l-72.32,0 l0,-5 l-108.48,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(418.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"780\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(418.4 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">5</text></g></g><g class=\"tick\" transform=\"translate(635.36 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">8</text></g></g><g class=\"tick\" transform=\"translate(707.68 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">9</text></g></g><g class=\"tick\" transform=\"translate(780 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 99.0584)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 76.7168)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 54.3752)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3</text></g></g><g class=\"tick\" transform=\"translate(56.8 32.0336)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4</text></g></g><g class=\"tick\" transform=\"translate(56.8 9.69196)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n9 p0 root\" transform=\"translate(68.4 26.8)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(-30.8 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 30.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a9 c2 node n6 p0\" transform=\"translate(30.8 84.1936)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-26.4 10.4064)\"><path class=\"edge\" d=\"M 0 0 V -10.4064 H 26.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(26.4 3.02377)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-17.6 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H 17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(17.6 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H -17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -3.02377 H -26.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -84.1936 H -30.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g><g class=\"tree t1\" transform=\"translate(237.6 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(68.4 35.8533)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-30.8 85.5467)\"><path class=\"edge\" d=\"M 0 0 V -85.5467 H 30.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 c2 node n6 p0\" transform=\"translate(30.8 75.1402)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-26.4 10.4064)\"><path class=\"edge\" d=\"M 0 0 V -10.4064 H 26.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(26.4 3.02377)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-17.6 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H 17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(17.6 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H -17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -3.02377 H -26.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -75.1402 H -30.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g><g class=\"tree t2\" transform=\"translate(418.4 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(68.4 95.3192)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-30.8 26.0808)\"><path class=\"edge\" d=\"M 0 0 V -26.0808 H 30.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(30.8 15.6743)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-26.4 10.4064)\"><path class=\"edge\" d=\"M 0 0 V -10.4064 H 26.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(26.4 3.02377)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-17.6 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H 17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(17.6 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H -17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -3.02377 H -26.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -15.6743 H -30.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g><g class=\"tree t3\" transform=\"translate(599.2 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(77.2 95.3192)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-39.6 26.0808)\"><path class=\"edge\" d=\"M 0 0 V -26.0808 H 39.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n5 p0\" transform=\"translate(39.6 18.6981)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(26.4 7.38266)\"><path class=\"edge\" d=\"M 0 0 V -7.38266 H -26.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-26.4 0.47066)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-17.6 6.912)\"><path class=\"edge\" d=\"M 0 0 V -6.912 H 17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(17.6 6.912)\"><path class=\"edge\" d=\"M 0 0 V -6.912 H -17.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -0.47066 H 26.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -18.6981 H -39.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 47,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_86_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, sequence_length=10, recombination_rate=0.1, random_seed=42)\n",
    "SVG(ts.draw_svg(y_axis=True, y_ticks=[0, 1, 2, 3, 4, 5]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b467f597",
   "metadata": {},
   "source": [
    "Sometimes we would like to stop the simulation early, **before** complete\n",
    "coalescence has occurred; perhaps we would like to\n",
    "{ref}`combine several simulations<sec_ancestry_initial_state_combining>`\n",
    "with different properties, or we are really only interested in the\n",
    "recent past and simulating more ancient processes would be a waste\n",
    "of computational resources.\n",
    "\n",
    "We can do this by specifying a ``end_time`` value.\n",
    "Let's repeat the simulation as above with an ``end_time`` of 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "5d6f160e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l174.4,0 l0,138.2 l244.16,25 l0,5 l-418.56,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M231.2,0 l174.4,0 l0,138.2 l122.08,25 l0,5 l-52.32,0 l0,-5 l-244.16,-25 l0,-138.2z\"/><path d=\"M405.6,0 l174.4,0 l0,138.2 l0,25 l0,5 l-52.32,0 l0,-5 l-122.08,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(475.36 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">8</text></g></g><g class=\"tick\" transform=\"translate(527.68 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">9</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 74.1)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c1 node n8 p0 root\" transform=\"translate(62 26.8)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(0.0 72.5683)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-25.2 22.0317)\"><path class=\"edge\" d=\"M 0 0 V -22.0317 H 25.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(25.2 6.40171)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-16.8 15.63)\"><path class=\"edge\" d=\"M 0 0 V -15.63 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(16.8 15.63)\"><path class=\"edge\" d=\"M 0 0 V -15.63 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -6.40171 H -25.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -72.5683 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(137.6 26.8)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g><g class=\"tree t1\" transform=\"translate(231.2 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(66.2 66.1837)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-29.4 55.2163)\"><path class=\"edge\" d=\"M 0 0 V -55.2163 H 29.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(29.4 33.1845)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-25.2 22.0317)\"><path class=\"edge\" d=\"M 0 0 V -22.0317 H 25.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(25.2 6.40171)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-16.8 15.63)\"><path class=\"edge\" d=\"M 0 0 V -15.63 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(16.8 15.63)\"><path class=\"edge\" d=\"M 0 0 V -15.63 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -6.40171 H -25.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -33.1845 H -29.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g><g class=\"tree t2\" transform=\"translate(405.6 0)\"><g class=\"plotbox\"><g class=\"c2 node n7 p0 root\" transform=\"translate(74.6 66.1837)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-37.8 55.2163)\"><path class=\"edge\" d=\"M 0 0 V -55.2163 H 37.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 c2 node n5 p0\" transform=\"translate(37.8 39.5863)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(25.2 15.63)\"><path class=\"edge\" d=\"M 0 0 V -15.63 H -25.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-25.2 0.996447)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-16.8 14.6336)\"><path class=\"edge\" d=\"M 0 0 V -14.6336 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(16.8 14.6336)\"><path class=\"edge\" d=\"M 0 0 V -14.6336 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -0.996447 H 25.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -39.5863 H -37.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 48,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_88_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    2, sequence_length=10, recombination_rate=0.1, end_time=2, random_seed=42)\n",
    "SVG(ts.draw_svg(y_axis=True, y_ticks=[0, 1, 2]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0c02cfe",
   "metadata": {},
   "source": [
    "There are a number of important things to observe about this\n",
    "tree sequence:\n",
    "\n",
    "1. The history up until time 2 is **identical** to the original simulation.\n",
    "2. The first tree has not fully coalesced, and we therefore have **multiple\n",
    "roots**. Trees with multiple roots are fully supported in tskit.\n",
    "3. The nodes 8 and 9 are at time 2 (our ``end_time`` value), and these\n",
    "connect to the uncoalesced portions of the tree. These \"unary\" nodes\n",
    "are very important for ensuring the correct statistical properties\n",
    "when we\n",
    "{ref}`combine multiple simulations<sec_ancestry_initial_state_combining>`.\n",
    "\n",
    ":::{seealso}\n",
    "The ancestry model {ref}`duration<sec_ancestry_models_specifying_duration>`\n",
    "can also be used to stop simulations before coalescence.\n",
    ":::\n",
    "\n",
    "For discrete time models, the exact interpretation\n",
    "of the ``end_time`` is important --- we continue to simulate\n",
    "the process while the time is <= ``end_time``. For example here\n",
    "we use an ``end_time`` of 1 generation in a {ref}`sec_ancestry_models_dtwf`\n",
    "simulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "1277d5ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c3 node n6 p0 root\" transform=\"translate(40.8 26.8)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-13.8667 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 13.8667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(13.8667 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -13.8667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g><g class=\"c1 node n7 p0 root\" transform=\"translate(68.5333 26.8)\"><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g><g class=\"c1 node n8 p0 root\" transform=\"translate(82.4 26.8)\"><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(96.2667 26.8)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 49,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_90_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(3, population_size=10, model=\"dtwf\", end_time=1, random_seed=42)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeeb98e9",
   "metadata": {},
   "source": [
    "We have run 1 generation of the DTWF, resulting in the coalescence\n",
    "at node 9.\n",
    "\n",
    "\n",
    "(sec_ancestry_start_time)=\n",
    "\n",
    "### Setting the start time\n",
    "\n",
    ":::{important}\n",
    "Setting simulation start time is an advanced feature which is rarely\n",
    "needed in practice.\n",
    ":::\n",
    "\n",
    "Sometimes we need control when the simulation process starts. For\n",
    "example, given the initial configuration of at time zero we want to\n",
    "\"jump ahead\" to a later time point before we start generating\n",
    "random events. Consider the following example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "fa9b66b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>3</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>4</td><td>0</td><td>0</td><td>-1</td><td>0.15642269</td><td></td></tr>\n",
       "<tr><td>5</td><td>0</td><td>0</td><td>-1</td><td>0.29152170</td><td></td></tr>\n",
       "<tr><td>6</td><td>0</td><td>0</td><td>-1</td><td>2.11740681</td><td></td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n"
      ],
      "text/plain": [
       "<tskit.tables.NodeTable at 0x7fa2bbcc9f70>"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, random_seed=42)\n",
    "ts.tables.nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "604d1070",
   "metadata": {},
   "source": [
    "Now we run the same simulation with a ``start_time`` of 10:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c8e16574",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>3</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>\n",
       "<tr><td>4</td><td>0</td><td>0</td><td>-1</td><td>10.15642269</td><td></td></tr>\n",
       "<tr><td>5</td><td>0</td><td>0</td><td>-1</td><td>10.29152170</td><td></td></tr>\n",
       "<tr><td>6</td><td>0</td><td>0</td><td>-1</td><td>12.11740681</td><td></td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n"
      ],
      "text/plain": [
       "<tskit.tables.NodeTable at 0x7fa2bbacb850>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(2, start_time=10, random_seed=42)\n",
    "ts.tables.nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d78e73a",
   "metadata": {},
   "source": [
    "We can see that the results are identical except that 10 has\n",
    "been added to all of the internal node times.\n",
    "\n",
    ":::{important}\n",
    "It is important to note that the ``start_time`` only controls the\n",
    "\"random\" events that happen within the simulation model: events\n",
    "that are specified to occur at a given time (such as sampling\n",
    "or demographic events) will occur at that time, even if this\n",
    "is before the ``start_time``.\n",
    ":::\n",
    "\n",
    "Consider the following example in which we create a simple\n",
    "{ref}`population tree<sec_demography_examples_population_tree>`\n",
    "demography, and then take samples at different times all *before*\n",
    "``start_time``:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "beb759ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>1.00000000</td><td></td></tr>\n",
       "<tr><td>1</td><td>1</td><td>1</td><td>1</td><td>2.00000000</td><td></td></tr>\n",
       "<tr><td>2</td><td>1</td><td>2</td><td>2</td><td>6.00000000</td><td></td></tr>\n",
       "<tr><td>3</td><td>0</td><td>2</td><td>-1</td><td>52.47084127</td><td></td></tr>\n",
       "<tr><td>4</td><td>0</td><td>2</td><td>-1</td><td>66.18169933</td><td></td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n"
      ],
      "text/plain": [
       "<tskit.tables.NodeTable at 0x7fa2bbabc0d0>"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", initial_size=100)\n",
    "demography.add_population(name=\"B\", initial_size=100)\n",
    "demography.add_population(name=\"C\", initial_size=100)\n",
    "demography.add_population_split(time=5, derived=[\"A\", \"B\"], ancestral=\"C\")\n",
    "\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=[\n",
    "        msprime.SampleSet(1, population=\"A\", time=1),\n",
    "        msprime.SampleSet(1, population=\"B\", time=2),\n",
    "        msprime.SampleSet(1, population=\"C\", time=6),\n",
    "    ],\n",
    "    demography=demography,\n",
    "    ploidy=1,\n",
    "    start_time=10,\n",
    "    random_seed=1)\n",
    "\n",
    "ts.tables.nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400266cb",
   "metadata": {},
   "source": [
    "We can see that the samples we specify are in the correct populations\n",
    "and their node times are at the sampling time. Note that all the\n",
    "coalescences happen in population ID 2 (\"C\"), indicating that the\n",
    "lineages from \"A\" and \"B\" moved there at the time of the split\n",
    "(which we can observe directly using the\n",
    "{ref}`record_migrations<sec_ancestry_record_migrations>` option if we wish).\n",
    "Note also that we drew a sample from population \"C\" at time 6,\n",
    "which is {ref}`after<sec_demography_direction_of_time>` the split;\n",
    "if we tried to draw a sample before the split, we would get an\n",
    "error because the population would be\n",
    "{ref}`inactive<sec_demography_populations_life_cycle>`.\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_ancestry_event_order` section for details\n",
    "of the precise order in which fixed-time events occur.\n",
    ":::\n",
    "\n",
    ":::{note}\n",
    "The same considerations apply when we are running simulations based\n",
    "on a given {ref}`initial state<sec_ancestry_initial_state>`; in\n",
    "this case, the ``start_time`` is by default set to the time\n",
    "of the youngest root in the input trees.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_event_order)=\n",
    "\n",
    "### Order of event execution\n",
    "\n",
    "There are two different classes of event that happen in ancestry\n",
    "simulations: \"fixed\" and \"random\" events. Random events are\n",
    "generated by the ancestry model, and fixed events are those\n",
    "that occur at a time specified by the user. Fixed events\n",
    "are {ref}`demographic events<sec_demography_events>` or\n",
    "{ref}`sampling events<sec_ancestry_samples>`.\n",
    "If multiple events are specified to occur at exactly the same time, then:\n",
    "\n",
    "1. All demographic events at this time are run, *in the order they\n",
    "were specified*.\n",
    "2. Sampling events occur immediately after all the demographic events\n",
    "have been executed. (See the {ref}`sec_ancestry_samples_sampling_time`\n",
    "section for more details on ancient sampling events.)\n",
    "\n",
    "These details are slightly different for the {ref}`sec_ancestry_models_dtwf`\n",
    "model. See the {class}`API documentation<.DiscreteTimeWrightFisher>` for\n",
    "details.\n",
    "\n",
    "(sec_ancestry_initial_state)=\n",
    "\n",
    "## Specifying the initial state\n",
    "\n",
    "By default `msprime` simulations are initialised by specifying a set of\n",
    "{ref}`samples<sec_ancestry_samples>`,\n",
    "which sets up the simulation with segments of ancestral material covering the\n",
    "whole sequence. Internally, this initial state is implemented by\n",
    "setting up a {class}`tskit.TableCollection` which contains the\n",
    "{ref}`population<tskit:sec_population_table_definition>`,\n",
    "{ref}`node<tskit:sec_node_table_definition>`,\n",
    "and {ref}`individual<tskit:sec_individual_table_definition>` information.\n",
    "The low-level simulation code then reads these tables and sets up the\n",
    "simulation accordingly. We can see what this initial state looks like\n",
    "by setting up a simple simulation and stopping it immediately:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6f46bffd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>1</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>1.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>2</td></tr>\n",
       "          <tr><td>Total Size</td><td>1.3 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>1</td>\n",
       "          <td>52 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>2</td>\n",
       "          <td>64 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>966 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2bbae8700>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "initial_ts = msprime.sim_ancestry(1, end_time=0)\n",
    "initial_ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5317dc0e",
   "metadata": {},
   "source": [
    "We can see that the returned tree sequence has one population,\n",
    "two sample nodes and one individual (since sample individuals are\n",
    "{ref}`diploid<sec_ancestry_ploidy>` by default). There are no\n",
    "{ref}`edges<tskit:sec_edge_table_definition>` because the simulation\n",
    "didn't run for any period of time: we're essentially just returning\n",
    "the initial state used to start the simulation.\n",
    "\n",
    "Then, running a simulation in which the ``initial_state`` is equal to\n",
    "``initial_ts`` is *precisely* the same as running the simulation\n",
    "with the same arguments as we used above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ae02b01a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts1 = msprime.sim_ancestry(initial_state=initial_ts, population_size=100, random_seed=2)\n",
    "ts2 = msprime.sim_ancestry(1, population_size=100, random_seed=2)\n",
    "assert ts1.equals(ts2, ignore_provenance=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d896d784",
   "metadata": {},
   "source": [
    ":::{warning}\n",
    "Note that the ``initial_state`` tree sequence only contains the\n",
    "information about the ancestral histories, and not any of the\n",
    "simulation parameters. In the previous example we used default\n",
    "parameters to keep things simple, but in general the full\n",
    "simulation model will need to be specified as well as the initial state.\n",
    "In particular, see the {ref}`sec_ancestry_initial_state_demography` section.\n",
    ":::\n",
    "\n",
    "While it is possible to specify the initial state manually by\n",
    "constructing the input tree sequence,\n",
    "it is an advanced topic and rarely necessary. More often we are\n",
    "interested in specifying the initial state of an ``msprime``\n",
    "simulation using the output of *another* simulation.\n",
    "\n",
    ":::{note}\n",
    "The ``initial_state`` tree sequence is quite literally the initial\n",
    "state of the tree sequence that is returned at the end of a simulation,\n",
    "and that any information present (including metadata) will not be\n",
    "altered.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_initial_state_combining)=\n",
    "\n",
    "### Combining backward-time simulations\n",
    "\n",
    "Suppose that we wished to simulate a scenario\n",
    "in which we have recombination happening at some rate in the recent past,\n",
    "and in the more distant past there is no recombination.\n",
    "We can do this by combining the simulations,\n",
    "using one as the ``initial_state`` for the other.\n",
    "\n",
    "We first run the simulation of the recent past:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "8365229e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l174.4,0 l0,138.2 l-122.08,25 l0,5 l-52.32,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M231.2,0 l174.4,0 l0,138.2 l69.76,25 l0,5 l-366.24,0 l0,-5 l122.08,-25 l0,-138.2z\"/><path d=\"M405.6,0 l174.4,0 l0,138.2 l0,25 l0,5 l-104.64,0 l0,-5 l-69.76,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(109.12 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(475.36 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">8</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 94.0545)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 78.0584)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 50.7128)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c1 node n7 p0 root\" transform=\"translate(53.6 26.8)\"><g class=\"a7 c2 node n5 p0\" transform=\"translate(0.0 13.6052)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-16.8 80.9948)\"><path class=\"edge\" d=\"M 0 0 V -80.9948 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(16.8 80.9948)\"><path class=\"edge\" d=\"M 0 0 V -80.9948 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -13.6052 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g><g class=\"c1 node n8 p0 root\" transform=\"translate(104 26.8)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(137.6 26.8)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g><g class=\"tree t1\" transform=\"translate(231.2 0)\"><g class=\"plotbox\"><g class=\"c1 node n7 p0 root\" transform=\"translate(53.6 26.8)\"><g class=\"a7 c2 node n5 p0\" transform=\"translate(0.0 13.6052)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-16.8 80.9948)\"><path class=\"edge\" d=\"M 0 0 V -80.9948 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(16.8 80.9948)\"><path class=\"edge\" d=\"M 0 0 V -80.9948 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -13.6052 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(104 26.8)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g><g class=\"c1 node n6 p0 root\" transform=\"translate(137.6 26.8)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g><g class=\"tree t2\" transform=\"translate(405.6 0)\"><g class=\"plotbox\"><g class=\"c1 node n7 p0 root\" transform=\"translate(78.8 26.8)\"><g class=\"a7 c2 node n5 p0\" transform=\"translate(0.0 13.6052)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(25.2 80.9948)\"><path class=\"edge\" d=\"M 0 0 V -80.9948 H -25.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-25.2 44.8272)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-16.8 36.1676)\"><path class=\"edge\" d=\"M 0 0 V -36.1676 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(16.8 36.1676)\"><path class=\"edge\" d=\"M 0 0 V -36.1676 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -44.8272 H 25.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -13.6052 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(137.6 26.8)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 55,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_102_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts1 = msprime.sim_ancestry(\n",
    "    2, recombination_rate=0.01, sequence_length=10, end_time=10,\n",
    "    population_size=100, random_seed=1)\n",
    "SVG(ts1.draw_svg(y_axis=True, time_scale=\"log_time\", y_ticks=[0, 1, 2, 5, 10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d773bb4e",
   "metadata": {},
   "source": [
    "Some recombination and coalescence has happened and we have three\n",
    "marginal trees along the genome. None of these trees has fully coalesced,\n",
    "and so each of the trees has multiple roots.\n",
    "\n",
    ":::{important}\n",
    "The set of nodes at time 10 connecting to all the internal nodes in the\n",
    "trees are very important: without these nodes and unary edges, the\n",
    "statistical properties of the combined simulations would not be correct!\n",
    "(See also section below on\n",
    "{ref}`forward simulations<sec_ancestry_initial_state_forward_simulations>`.)\n",
    ":::\n",
    "\n",
    "We then run the next phase of the simulation in which there is\n",
    "*no* recombination by passing ``ts1`` as the argument to ``initial_state``:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4a0a3a81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l174.4,0 l0,138.2 l-122.08,25 l0,5 l-52.32,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M231.2,0 l174.4,0 l0,138.2 l69.76,25 l0,5 l-366.24,0 l0,-5 l122.08,-25 l0,-138.2z\"/><path d=\"M405.6,0 l174.4,0 l0,138.2 l0,25 l0,5 l-104.64,0 l0,-5 l-69.76,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(109.12 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(475.36 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">8</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 81.6952)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10</text></g></g><g class=\"tick\" transform=\"translate(56.8 70.9882)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">20</text></g></g><g class=\"tick\" transform=\"translate(56.8 56.2961)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">50</text></g></g><g class=\"tick\" transform=\"translate(56.8 44.982)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">100</text></g></g><g class=\"tick\" transform=\"translate(56.8 33.5869)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">200</text></g></g><g class=\"tick\" transform=\"translate(56.8 18.4643)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">500</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n12 p0 root\" transform=\"translate(87.2 26.8)\"><g class=\"a12 c1 node n7 p0\" transform=\"translate(33.6 54.8952)\"><g class=\"a7 c2 node n5 p0\" transform=\"translate(0.0 5.71026)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-16.8 33.9945)\"><path class=\"edge\" d=\"M 0 0 V -33.9945 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(16.8 33.9945)\"><path class=\"edge\" d=\"M 0 0 V -33.9945 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -5.71026 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -54.8952 H -33.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a12 c2 node n11 p0\" transform=\"translate(-33.6 10.5823)\"><g class=\"a11 c1 node n8 p0\" transform=\"translate(16.8 44.3129)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 39.7048)\"><path class=\"edge\" d=\"M 0 0 V -39.7048 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -44.3129 H -16.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><g class=\"a11 c1 node n9 p0\" transform=\"translate(-16.8 44.3129)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 39.7048)\"><path class=\"edge\" d=\"M 0 0 V -39.7048 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><path class=\"edge\" d=\"M 0 0 V -44.3129 H 16.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -10.5823 H 33.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">12</text></g></g></g><g class=\"tree t1\" transform=\"translate(231.2 0)\"><g class=\"plotbox\"><g class=\"c2 node n12 p0 root\" transform=\"translate(87.2 26.8)\"><g class=\"a12 c1 node n7 p0\" transform=\"translate(33.6 54.8952)\"><g class=\"a7 c2 node n5 p0\" transform=\"translate(0.0 5.71026)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-16.8 33.9945)\"><path class=\"edge\" d=\"M 0 0 V -33.9945 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(16.8 33.9945)\"><path class=\"edge\" d=\"M 0 0 V -33.9945 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -5.71026 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -54.8952 H -33.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a12 c2 node n10 p0\" transform=\"translate(-33.6 34.5099)\"><g class=\"a10 c1 node n6 p0\" transform=\"translate(16.8 20.3853)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 39.7048)\"><path class=\"edge\" d=\"M 0 0 V -39.7048 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -20.3853 H -16.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a10 c1 node n9 p0\" transform=\"translate(-16.8 20.3853)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 39.7048)\"><path class=\"edge\" d=\"M 0 0 V -39.7048 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><path class=\"edge\" d=\"M 0 0 V -20.3853 H 16.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -34.5099 H 33.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">12</text></g></g></g><g class=\"tree t2\" transform=\"translate(405.6 0)\"><g class=\"plotbox\"><g class=\"c2 node n12 p0 root\" transform=\"translate(74.6 26.8)\"><g class=\"a12 c1 node n7 p0\" transform=\"translate(37.8 54.8952)\"><g class=\"a7 c2 node n5 p0\" transform=\"translate(0.0 5.71026)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(25.2 33.9945)\"><path class=\"edge\" d=\"M 0 0 V -33.9945 H -25.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-25.2 18.8145)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-16.8 15.18)\"><path class=\"edge\" d=\"M 0 0 V -15.18 H 16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(16.8 15.18)\"><path class=\"edge\" d=\"M 0 0 V -15.18 H -16.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -18.8145 H 25.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -5.71026 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -54.8952 H -37.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a12 c1 node n9 p0\" transform=\"translate(-37.8 54.8952)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 39.7048)\"><path class=\"edge\" d=\"M 0 0 V -39.7048 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><path class=\"edge\" d=\"M 0 0 V -54.8952 H 37.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">12</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 56,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_104_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts2 = msprime.sim_ancestry(initial_state=ts1, population_size=100, random_seed=3)\n",
    "SVG(ts2.draw_svg(y_axis=True, time_scale=\"log_time\", y_ticks=[0, 10, 20, 50, 100, 200, 500]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f28dac12",
   "metadata": {},
   "source": [
    "Since there is no recombination in this more ancient simulation,\n",
    "node ``12`` is the MRCA in all three trees. The unary nodes\n",
    "``6``, ``7``, ``8``, and ``9`` marking the transition between the\n",
    "two simulation regimes are still present in the trees. If you\n",
    "wish to remove these, we can use the {meth}`tskit.TreeSequence.simplify`\n",
    "method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ca1b4a30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l186.667,0 l0,138.2 l-130.667,25 l0,5 l-56,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M206.667,0 l186.667,0 l0,138.2 l74.6667,25 l0,5 l-392,0 l0,-5 l130.667,-25 l0,-138.2z\"/><path d=\"M393.333,0 l186.667,0 l0,138.2 l0,25 l0,5 l-112,0 l0,-5 l-74.6667,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(300 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(76 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g><g class=\"tick\" transform=\"translate(468 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">8</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(93.3333 26.8)\"><g class=\"a8 c2 node n5 p0\" transform=\"translate(36.6667 92.4714)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-18.3333 2.12857)\"><path class=\"edge\" d=\"M 0 0 V -2.12857 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(18.3333 2.12857)\"><path class=\"edge\" d=\"M 0 0 V -2.12857 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -92.4714 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-36.6667 44.8211)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-18.3333 49.7789)\"><path class=\"edge\" d=\"M 0 0 V -49.7789 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i0 leaf node n1 p0 sample\" transform=\"translate(18.3333 49.7789)\"><path class=\"edge\" d=\"M 0 0 V -49.7789 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -44.8211 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g><g class=\"tree t1\" transform=\"translate(206.667 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(93.3333 26.8)\"><g class=\"a8 c2 node n5 p0\" transform=\"translate(36.6667 92.4714)\"><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(-18.3333 2.12857)\"><path class=\"edge\" d=\"M 0 0 V -2.12857 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(18.3333 2.12857)\"><path class=\"edge\" d=\"M 0 0 V -2.12857 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -92.4714 H -36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><g class=\"a8 c2 node n6 p0\" transform=\"translate(-36.6667 83.1051)\"><g class=\"a6 i0 leaf node n0 p0 sample\" transform=\"translate(-18.3333 11.4949)\"><path class=\"edge\" d=\"M 0 0 V -11.4949 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(18.3333 11.4949)\"><path class=\"edge\" d=\"M 0 0 V -11.4949 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><path class=\"edge\" d=\"M 0 0 V -83.1051 H 36.6667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g><g class=\"tree t2\" transform=\"translate(393.333 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(79.5833 26.8)\"><g class=\"a8 c2 node n5 p0\" transform=\"translate(41.25 92.4714)\"><g class=\"a5 i1 leaf node n3 p0 sample\" transform=\"translate(27.5 2.12857)\"><path class=\"edge\" d=\"M 0 0 V -2.12857 H -27.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a5 c2 node n4 p0\" transform=\"translate(-27.5 1.65807)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-18.3333 0.470498)\"><path class=\"edge\" d=\"M 0 0 V -0.470498 H 18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n2 p0 sample\" transform=\"translate(18.3333 0.470498)\"><path class=\"edge\" d=\"M 0 0 V -0.470498 H -18.3333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -1.65807 H 27.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -92.4714 H -41.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">5</text></g><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-41.25 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 41.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 57,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_106_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_simplified = ts2.simplify()\n",
    "SVG(ts_simplified.draw_svg())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "813dc4be",
   "metadata": {},
   "source": [
    "(sec_ancestry_initial_state_demography)=\n",
    "\n",
    "### Interaction with demography\n",
    "\n",
    "In the previous example we saw how to combine two single population simulations.\n",
    "We can also combine more complex simulations involving\n",
    "{ref}`demography<sec_demography>`, but care needs to be taken to ensure that\n",
    "the correct models are simulated in each part of the simulation.\n",
    "Suppose we have simple model with a\n",
    "{ref}`population split<sec_demography_events_population_split>`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e36efb28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td>Contemporary population A</td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Contemporary population B</td><td>200.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td>Ancestral population</td><td>300.0</td><td>0</td><td>1e+02</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=C</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;C&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=100, growth_rate=0, name='A', description='Contemporary population A', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=200, growth_rate=0, name='B', description='Contemporary population B', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1), Population(initial_size=300, growth_rate=0, name='C', description='Ancestral population', extra_metadata={}, default_sampling_time=100, initially_active=False, id=2)], events=[PopulationSplit(time=100, derived=['A', 'B'], ancestral='C')], migration_matrix=array([[0., 0., 0.],\n",
       "       [0., 0., 0.],\n",
       "       [0., 0., 0.]]))"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", description=\"Contemporary population A\", initial_size=100)\n",
    "demography.add_population(name=\"B\", description=\"Contemporary population B\", initial_size=200)\n",
    "demography.add_population(name=\"C\", description=\"Ancestral population\", initial_size=300)\n",
    "demography.add_population_split(time=100, ancestral=\"C\", derived=[\"A\", \"B\"])\n",
    "demography"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13ceedb4",
   "metadata": {},
   "source": [
    "Then, we run a simulation of this model for 50 generations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "ce4eb839",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 40.9527)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">42.52</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">50.00</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c1 node n5 p0 root\" transform=\"translate(40.8 26.8)\"><g class=\"a5 c2 node n4 p0\" transform=\"translate(0.0 14.1527)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 80.4473)\"><path class=\"edge\" d=\"M 0 0 V -80.4473 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0:A</text></g><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(10.4 80.4473)\"><path class=\"edge\" d=\"M 0 0 V -80.4473 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1:A</text></g><path class=\"edge\" d=\"M 0 0 V -14.1527 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4:A</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">5:A</text></g><g class=\"c1 node n6 p1 root\" transform=\"translate(72 26.8)\"><g class=\"a6 i1 leaf node n3 p1 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3:B</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6:B</text></g><g class=\"c1 node n7 p1 root\" transform=\"translate(92.8 26.8)\"><g class=\"a7 i1 leaf node n2 p1 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2:B</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7:B</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 59,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_110_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts1 = msprime.sim_ancestry(\n",
    "    samples={\"A\": 1, \"B\": 1}, demography=demography, end_time=50, random_seed=1234)\n",
    "node_labels = {\n",
    "    node.id: f\"{node.id}:{ts1.population(node.population).metadata['name']}\"\n",
    "    for node in ts1.nodes()}\n",
    "SVG(ts1.draw_svg(node_labels=node_labels, y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70063cc5",
   "metadata": {},
   "source": [
    "After 50 generations we can see there has been a coalescence in population ``A``\n",
    "but none in ``B`` and we therefore have three lineages left when the simulation\n",
    "finishes. We can then continue the simulation based on this initial state:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f3025c06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 114.008)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">42.52</text></g></g><g class=\"tick\" transform=\"translate(56.8 112.708)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">50.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 88.1604)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">191.21</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">544.17</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n9 p2 root\" transform=\"translate(74.6 26.8)\"><g class=\"a9 c1 node n7 p1\" transform=\"translate(18.2 85.9079)\"><g class=\"a7 i1 leaf node n2 p1 sample\" transform=\"translate(0.0 8.69212)\"><path class=\"edge\" d=\"M 0 0 V -8.69212 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2:B</text></g><path class=\"edge\" d=\"M 0 0 V -85.9079 H -18.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7:B</text></g><g class=\"a9 c2 node n8 p2\" transform=\"translate(-18.2 61.3604)\"><g class=\"a8 c1 node n5 p0\" transform=\"translate(-15.6 24.5474)\"><g class=\"a5 c2 node n4 p0\" transform=\"translate(0.0 1.30039)\"><g class=\"a4 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 7.39173)\"><path class=\"edge\" d=\"M 0 0 V -7.39173 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0:A</text></g><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(10.4 7.39173)\"><path class=\"edge\" d=\"M 0 0 V -7.39173 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1:A</text></g><path class=\"edge\" d=\"M 0 0 V -1.30039 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">4:A</text></g><path class=\"edge\" d=\"M 0 0 V -24.5474 H 15.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5:A</text></g><g class=\"a8 c1 node n6 p1\" transform=\"translate(15.6 24.5474)\"><g class=\"a6 i1 leaf node n3 p1 sample\" transform=\"translate(0.0 8.69212)\"><path class=\"edge\" d=\"M 0 0 V -8.69212 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3:B</text></g><path class=\"edge\" d=\"M 0 0 V -24.5474 H -15.6\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6:B</text></g><path class=\"edge\" d=\"M 0 0 V -61.3604 H 18.2\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8:C</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9:C</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 60,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_112_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts2 = msprime.sim_ancestry(\n",
    "    initial_state=ts1, demography=demography, random_seed=5678)\n",
    "node_labels = {\n",
    "    node.id: f\"{node.id}:{ts2.population(node.population).metadata['name']}\"\n",
    "    for node in ts2.nodes()}\n",
    "SVG(ts2.draw_svg(node_labels=node_labels, y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd9080b",
   "metadata": {},
   "source": [
    "Note that we use the **same** demography object, and so the lineages\n",
    "migrate and ultimately coalesce in population ``C``, as we'd expect.\n",
    "This is the simplest case, in which we already have the demography\n",
    "object which we can use to model the entire simulation from end-to-end.\n",
    "\n",
    "In other cases (e.g. when working with simulations from a different\n",
    "program), we don't have the {class}`.Demography` object at hand\n",
    "and we need to create one that is both compatible with the\n",
    "``initial_state`` tree sequence and reflects the demographic\n",
    "model that we are interested in using. The best way to do this\n",
    "is to use the {meth}`.Demography.from_tree_sequence` method to\n",
    "first get a base demography that is based on the\n",
    "tree sequence {ref}`population table<tskit:sec_population_table_definition>`\n",
    "and {ref}`metadata<tskit:sec_metadata>`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c6a1a5c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td>Contemporary population A</td><td>0.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Contemporary population B</td><td>0.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td>Ancestral population</td><td>0.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (0)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=0, growth_rate=0, name='A', description='Contemporary population A', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=0, growth_rate=0, name='B', description='Contemporary population B', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1), Population(initial_size=0, growth_rate=0, name='C', description='Ancestral population', extra_metadata={}, default_sampling_time=None, initially_active=None, id=2)], events=[], migration_matrix=array([[0., 0., 0.],\n",
       "       [0., 0., 0.],\n",
       "       [0., 0., 0.]]))"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography = msprime.Demography.from_tree_sequence(ts1)\n",
    "demography"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f476b39c",
   "metadata": {},
   "source": [
    "We can see that the population names and descriptions have been recovered\n",
    "from the tree sequence metadata but **no other information**: the\n",
    "population sizes are all zero and we no longer have a population split\n",
    "event. If we try to run a simulation using this demography we get an\n",
    "error:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "82f8b219",
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "ename": "InputError",
     "evalue": "Input error in initialise: Attempt to sample lineage in a population with size=0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInputError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [62], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m ts \u001b[38;5;241m=\u001b[39m \u001b[43mmsprime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msim_ancestry\u001b[49m\u001b[43m(\u001b[49m\u001b[43minitial_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mts1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1207\u001b[0m, in \u001b[0;36msim_ancestry\u001b[0;34m(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, num_labels, random_seed, num_replicates, replicate_index, record_provenance)\u001b[0m\n\u001b[1;32m   1184\u001b[0m     parameters \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\n\u001b[1;32m   1185\u001b[0m         command\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msim_ancestry\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   1186\u001b[0m         samples\u001b[38;5;241m=\u001b[39msamples,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1204\u001b[0m         \u001b[38;5;66;03m# replicate index is excluded as it is inserted for each replicate\u001b[39;00m\n\u001b[1;32m   1205\u001b[0m     )\n\u001b[1;32m   1206\u001b[0m     provenance_dict \u001b[38;5;241m=\u001b[39m provenance\u001b[38;5;241m.\u001b[39mget_provenance_dict(parameters)\n\u001b[0;32m-> 1207\u001b[0m sim \u001b[38;5;241m=\u001b[39m \u001b[43m_parse_sim_ancestry\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1208\u001b[0m \u001b[43m    \u001b[49m\u001b[43msamples\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msamples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1209\u001b[0m \u001b[43m    \u001b[49m\u001b[43msequence_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msequence_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1210\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecombination_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecombination_rate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1211\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1212\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1213\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdiscrete_genome\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdiscrete_genome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1214\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpopulation_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpopulation_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1215\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1216\u001b[0m \u001b[43m    \u001b[49m\u001b[43mploidy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mploidy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1217\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1218\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_state\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1219\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstart_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1220\u001b[0m \u001b[43m    \u001b[49m\u001b[43mend_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1221\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecord_migrations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_migrations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1222\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecord_full_arg\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_full_arg\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1223\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_labels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1224\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_seed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_seed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1225\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1226\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _wrap_replicates(\n\u001b[1;32m   1227\u001b[0m     sim,\n\u001b[1;32m   1228\u001b[0m     num_replicates\u001b[38;5;241m=\u001b[39mnum_replicates,\n\u001b[1;32m   1229\u001b[0m     replicate_index\u001b[38;5;241m=\u001b[39mreplicate_index,\n\u001b[1;32m   1230\u001b[0m     provenance_dict\u001b[38;5;241m=\u001b[39mprovenance_dict,\n\u001b[1;32m   1231\u001b[0m )\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1010\u001b[0m, in \u001b[0;36m_parse_sim_ancestry\u001b[0;34m(samples, sequence_length, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, discrete_genome, population_size, demography, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, num_labels, random_seed, init_for_debugger)\u001b[0m\n\u001b[1;32m   1007\u001b[0m random_seed \u001b[38;5;241m=\u001b[39m _parse_random_seed(random_seed)\n\u001b[1;32m   1008\u001b[0m random_generator \u001b[38;5;241m=\u001b[39m _msprime\u001b[38;5;241m.\u001b[39mRandomGenerator(random_seed)\n\u001b[0;32m-> 1010\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSimulator\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1011\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtables\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1012\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecombination_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecombination_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1013\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1014\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1015\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdiscrete_genome\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdiscrete_genome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1016\u001b[0m \u001b[43m    \u001b[49m\u001b[43mploidy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mploidy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1017\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1018\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1019\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstore_migrations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_migrations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1020\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstore_full_arg\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrecord_full_arg\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1021\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstart_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1022\u001b[0m \u001b[43m    \u001b[49m\u001b[43mend_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1023\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_labels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1024\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_generator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_generator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1025\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1318\u001b[0m, in \u001b[0;36mSimulator.__init__\u001b[0;34m(self, tables, recombination_map, gene_conversion_map, gene_conversion_tract_length, discrete_genome, ploidy, demography, random_generator, models, store_migrations, store_full_arg, start_time, end_time, num_labels)\u001b[0m\n\u001b[1;32m   1315\u001b[0m gene_conversion_rate \u001b[38;5;241m=\u001b[39m gene_conversion_map\u001b[38;5;241m.\u001b[39mrate[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1317\u001b[0m start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m start_time \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m start_time\n\u001b[0;32m-> 1318\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1319\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtables\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_tables\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1320\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrecombination_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_recomb_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1321\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstart_time\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart_time\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1322\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrandom_generator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrandom_generator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1323\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmigration_matrix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmigration_matrix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1324\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpopulation_configuration\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_population_configuration\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1325\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdemographic_events\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mll_demographic_events\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1326\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstore_migrations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstore_migrations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1327\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstore_full_arg\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstore_full_arg\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1328\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_labels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1329\u001b[0m \u001b[43m    \u001b[49m\u001b[43msegment_block_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msegment_block_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1330\u001b[0m \u001b[43m    \u001b[49m\u001b[43mavl_node_block_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mavl_node_block_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1331\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnode_mapping_block_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnode_mapping_block_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1332\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_rate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1333\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_conversion_tract_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1334\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdiscrete_genome\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdiscrete_genome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1335\u001b[0m \u001b[43m    \u001b[49m\u001b[43mploidy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mploidy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1336\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1337\u001b[0m \u001b[38;5;66;03m# Highlevel attributes used externally that have no lowlevel equivalent\u001b[39;00m\n\u001b[1;32m   1338\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mend_time \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39minf \u001b[38;5;28;01mif\u001b[39;00m end_time \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m end_time\n",
      "\u001b[0;31mInputError\u001b[0m: Input error in initialise: Attempt to sample lineage in a population with size=0"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(initial_state=ts1, demography=demography)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "345e0fe1",
   "metadata": {},
   "source": [
    "To recover the original model we must update the {class}`.Population`\n",
    "objects in place and add the\n",
    "{ref}`population split<sec_demography_events_population_split>` event:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5d483dcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td>Contemporary population A</td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Contemporary population B</td><td>200.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td>Ancestral population</td><td>300.0</td><td>0</td><td>1e+02</td><td>{}</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Migration matrix (all zero)</b><table border=\"1\" class=\"tskit-table\"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events (1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=C</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;C&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div>"
      ],
      "text/plain": [
       "Demography(populations=[Population(initial_size=100, growth_rate=0, name='A', description='Contemporary population A', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0), Population(initial_size=200, growth_rate=0, name='B', description='Contemporary population B', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1), Population(initial_size=300, growth_rate=0, name='C', description='Ancestral population', extra_metadata={}, default_sampling_time=100, initially_active=False, id=2)], events=[PopulationSplit(time=100, derived=['A', 'B'], ancestral='C')], migration_matrix=array([[0., 0., 0.],\n",
       "       [0., 0., 0.],\n",
       "       [0., 0., 0.]]))"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demography[\"A\"].initial_size = 100\n",
    "demography[\"B\"].initial_size = 200\n",
    "demography[\"C\"].initial_size = 300\n",
    "demography.add_population_split(time=100, ancestral=\"C\", derived=[\"A\", \"B\"])\n",
    "demography"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46b94c1",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "See the {class}`.Demography` documentation for details on how to access\n",
    "{class}`.Population` objects.\n",
    ":::\n",
    "\n",
    "The demography that we obtain from {meth}`.Demography.from_tree_sequence`\n",
    "is really just a template, which we update and elaborate as we need.\n",
    "In particular, we are free to add more populations as needed.\n",
    "\n",
    "(sec_ancestry_initial_state_forward_simulations)=\n",
    "\n",
    "### Continuing forwards-time simulations (\"recapitating\")\n",
    "\n",
    "The most common use for the ``initial_state`` argument is to\n",
    "start the simulation from the output of a forwards-time simulation.\n",
    "Informally, we take an 'unfinished'\n",
    "tree sequence as a parameter to {func}`.sim_ancestry`, initialise the simulation\n",
    "from the state of this tree sequence and then run the simulation until\n",
    "coalescence. The returned tree sequence is then the result of taking the\n",
    "input tree sequence and completing the trees using the coalescent.\n",
    "\n",
    "This is useful for forwards-time simulators such as\n",
    "[SLiM](https://messerlab.org/slim/) and\n",
    "[fwdpy11](https://pypi.org/project/fwdpy11/)\n",
    "that can output tree sequences. By running\n",
    "forward-time simulation for a certain number of generations we obtain a\n",
    "tree sequence, but these trees may not have had sufficient time to\n",
    "reach a most recent common ancestor. By using the `initial_state` argument\n",
    "to {func}`.sim_ancestry` we can combine the best of both forwards- and\n",
    "backwards-time simulators. The recent past can be simulated forwards\n",
    "in time and the ancient past by the coalescent. The coalescent\n",
    "simulation is initialised by the root segments of the\n",
    "input tree sequence, ensuring that the minimal amount of ancestral\n",
    "material possible is simulated.\n",
    "\n",
    "Any tree sequence can be provided as input to this process, but there is a\n",
    "specific topological requirement that must be met for the simulations to be\n",
    "statistically correct. To ensure that ancestral segments are correctly\n",
    "associated within chromosomes when constructing the initial conditions for the\n",
    "coalescent simulation, forward-time simulators **must** retain the nodes\n",
    "corresponding to the initial generation. Furthermore, for every sample in the\n",
    "final generation (i.e. the extant population at the present time) there must be\n",
    "a path to one of the founder population nodes.\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`recapitation tutorial<pyslim:sec_tutorial_recapitation>`\n",
    "from the [pyslim](https://github.com/tskit-dev/pyslim) documentation\n",
    "for a detailed explanation of this process when\n",
    "using [SLiM](https://messerlab.org/slim/).\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models)=\n",
    "\n",
    "## Models\n",
    "\n",
    "The ancestry model determines the model under which the ancestral\n",
    "history of the sample is generated. It is concerned with the\n",
    "basic processes of how (for example) common ancestor and\n",
    "recombination events occur. For example, the default\n",
    "\"standard\" or {ref}`sec_ancestry_models_hudson` is an\n",
    "efficient continuous time model, and the\n",
    "{ref}`sec_ancestry_models_dtwf` model allows\n",
    "for more detailed (if less efficient) simulations by\n",
    "making fewer mathematical approximations.\n",
    "We can combine multiple ancestry models in a simulation\n",
    "using a flexible {ref}`notation<sec_ancestry_models_specifying>`.\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_ancestry_quickref` for a summary of the available\n",
    "ancestry models\n",
    ":::\n",
    "\n",
    ":::{important}\n",
    "The ancestry model is distinct from the {ref}`demographic model<sec_demography>`,\n",
    "which is mostly independent of the chosen ancestry model.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models_specifying)=\n",
    "\n",
    "### Specifying ancestry models\n",
    "\n",
    "The ancestry models used during a simulation are specified using\n",
    "the ``model`` parameter to {func}`.sim_ancestry`. Each model is a\n",
    "subclass of the {class}`.AncestryModel` class (see the following\n",
    "subsections for the available models, and examples of their usage).\n",
    "\n",
    "#### Single models\n",
    "\n",
    "Most of the time we want to run a simulation under a single\n",
    "ancestry model. By default, we run simulations under the\n",
    "{class}`.StandardCoalescent` model. If we wish to run\n",
    "under a different model, we use the ``model`` argument to\n",
    "{func}`.sim_ancestry`. For example, here we use the\n",
    "{class}`SMC<.SmcApproxCoalescent>` model instead of the\n",
    "standard coalescent:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "bd65017f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts1 = msprime.sim_ancestry(\n",
    "    10,\n",
    "    sequence_length=10,\n",
    "    recombination_rate=0.1,\n",
    "    model=msprime.SmcApproxCoalescent(),\n",
    "    random_seed=1234)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02b0f3dc",
   "metadata": {},
   "source": [
    "We specify the model as an instance of one of the ancestry model classes.\n",
    "For nonparametric models, there is also a useful shorthand\n",
    "of the model name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "8eca7461",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts2 = msprime.sim_ancestry(\n",
    "    10,\n",
    "    sequence_length=10,\n",
    "    recombination_rate=0.1,\n",
    "    model=\"smc\",\n",
    "    random_seed=1234)\n",
    "assert ts1.equals(ts2, ignore_provenance=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e80450fd",
   "metadata": {},
   "source": [
    "This string form is useful for single model simulations;\n",
    "however, when working with\n",
    "{ref}`sec_ancestry_models_specifying_multiple` it is better to use\n",
    "the explicit class form so that the\n",
    "{ref}`sec_ancestry_models_specifying_duration` can be set.\n",
    "\n",
    "(sec_ancestry_models_specifying_duration)=\n",
    "\n",
    "#### Model duration\n",
    "\n",
    "Each ancestry model instance has a {attr}`~.AncestryModel.duration` associated\n",
    "with it. This is the maximum amount of time that this model can run for. Thus,\n",
    "if we wanted to run 10 generations of the {class}`.DiscreteTimeWrightFisher`\n",
    "model we could write:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "504c9ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 111.94)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 102.48)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c1 node n8 p0 root\" transform=\"translate(33.8667 26.8)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(0.0 85.14)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-6.93333 9.46)\"><path class=\"edge\" d=\"M 0 0 V -9.46 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(6.93333 9.46)\"><path class=\"edge\" d=\"M 0 0 V -9.46 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -85.14 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(54.6667 26.8)\"><g class=\"a9 i1 leaf node n3 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g><g class=\"c1 node n10 p0 root\" transform=\"translate(75.4667 26.8)\"><g class=\"a10 c2 node n7 p0\" transform=\"translate(0.0 75.68)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-6.93333 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(6.93333 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -75.68 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g><g class=\"c1 node n11 p0 root\" transform=\"translate(96.2667 26.8)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 66,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_124_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3,\n",
    "    population_size=10,\n",
    "    model=msprime.DiscreteTimeWrightFisher(duration=10),\n",
    "    random_seed=1234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5564fc8",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "Using the duration value for a single model like this is identical to\n",
    "specifying an ``end_time`` value: see the {ref}`sec_ancestry_end_time`\n",
    "section for more information.\n",
    ":::\n",
    "\n",
    "It is vital to understand that the ``duration`` value here is **not**\n",
    "an absolute time at which the simulation must stop, but rather\n",
    "the number of generations that the model should run for. To illustrate\n",
    "this we can use the {ref}`start_time<sec_ancestry_start_time>` parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e66c92a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 83.56)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">6</text></g></g><g class=\"tick\" transform=\"translate(56.8 77.2533)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">7</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">15</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c1 node n8 p0 root\" transform=\"translate(33.8667 26.8)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(0.0 56.76)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-6.93333 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(6.93333 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -56.76 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(54.6667 26.8)\"><g class=\"a9 i1 leaf node n3 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g><g class=\"c1 node n10 p0 root\" transform=\"translate(75.4667 26.8)\"><g class=\"a10 c2 node n7 p0\" transform=\"translate(0.0 50.4533)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-6.93333 44.1467)\"><path class=\"edge\" d=\"M 0 0 V -44.1467 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(6.93333 44.1467)\"><path class=\"edge\" d=\"M 0 0 V -44.1467 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -50.4533 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g><g class=\"c1 node n11 p0 root\" transform=\"translate(96.2667 26.8)\"><g class=\"a11 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 67,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_126_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3,\n",
    "    population_size=10,\n",
    "    start_time=5,\n",
    "    model=msprime.DiscreteTimeWrightFisher(duration=10),\n",
    "    random_seed=1234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34bea3ba",
   "metadata": {},
   "source": [
    "We can see that the simulation continued until 15 generations ago,\n",
    "because it didn't start until 5 generations ago and we then specified\n",
    "that the model should run for at most 10 generations.\n",
    "\n",
    "The model activity time is specified as a duration in this way\n",
    "as this is a natural way to specify\n",
    "{ref}`sec_ancestry_models_specifying_multiple`, in which\n",
    "some of the models may run for random periods of time.\n",
    "\n",
    "\n",
    "(sec_ancestry_models_specifying_completion)=\n",
    "\n",
    "#### Model completion\n",
    "\n",
    "Most ancestry models will run until the overall simulation has\n",
    "completed; that is, when we have simulated back to the most\n",
    "recent common ancestor at every position along the sequence.\n",
    "For example, the {class}`.StandardCoalescent` and\n",
    "{class}`.DiscreteTimeWrightFisher` models will simulate\n",
    "until coalescence, unless we explicitly indicate that we wish\n",
    "to stop the simulation early via the\n",
    "{ref}`model duration<sec_ancestry_models_specifying_duration>`\n",
    "or the {ref}`end_time<sec_ancestry_end_time>` parameter.\n",
    "\n",
    "Models such as {class}`.SweepGenicSelection` are different though:\n",
    "they are only defined for a limited amount of time and\n",
    "may not run until coalescence (see the\n",
    "{ref}`sec_ancestry_models_selective_sweeps`\n",
    "for more information on the model itself)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "4d719af0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 104.887)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5.27</text></g></g><g class=\"tick\" transform=\"translate(56.8 76.106)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">14.47</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">30.22</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n8 p0 root\" transform=\"translate(33.8667 26.8)\"><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(-6.93333 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a8 i2 leaf node n5 p0 sample\" transform=\"translate(6.93333 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">8</text></g><g class=\"c1 node n9 p0 root\" transform=\"translate(72 26.8)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(0.0 49.306)\"><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(10.4 45.294)\"><path class=\"edge\" d=\"M 0 0 V -45.294 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-10.4 28.7812)\"><g class=\"a6 i0 leaf node n1 p0 sample\" transform=\"translate(-6.93333 16.5128)\"><path class=\"edge\" d=\"M 0 0 V -16.5128 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(6.93333 16.5128)\"><path class=\"edge\" d=\"M 0 0 V -16.5128 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -28.7812 H 10.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -49.306 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g><g class=\"c1 node n10 p0 root\" transform=\"translate(96.2667 26.8)\"><g class=\"a10 i0 leaf node n0 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 68,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_128_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sweep = msprime.SweepGenicSelection(\n",
    "    position=5,\n",
    "    start_frequency=0.1,\n",
    "    end_frequency=0.9,\n",
    "    s=0.25,\n",
    "    dt=1e-6,\n",
    ")\n",
    "ts = msprime.sim_ancestry(\n",
    "    3,\n",
    "    sequence_length=10,\n",
    "    population_size=100,\n",
    "    model=sweep,\n",
    "    random_seed=1234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8942e078",
   "metadata": {},
   "source": [
    "Here, we see that even though we didn't specify\n",
    "a {ref}`model duration<sec_ancestry_models_specifying_duration>`\n",
    "or {ref}`end_time<sec_ancestry_end_time>` parameter the simulation\n",
    "has ended before we reached coalescence. In addition, the\n",
    "model duration is **random**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "bdd8daef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">10</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 79.6954)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">12.41</text></g></g><g class=\"tick\" transform=\"translate(56.8 44.7757)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">22.80</text></g></g><g class=\"tick\" transform=\"translate(56.8 36.2554)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">25.33</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">28.15</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n9 p0 root\" transform=\"translate(52.9333 26.8)\"><g class=\"a9 c2 node n7 p0\" transform=\"translate(19.0667 17.9757)\"><g class=\"a7 i2 leaf node n4 p0 sample\" transform=\"translate(10.4 76.6243)\"><path class=\"edge\" d=\"M 0 0 V -76.6243 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a7 c2 node n6 p0\" transform=\"translate(-10.4 34.9197)\"><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(-6.93333 41.7046)\"><path class=\"edge\" d=\"M 0 0 V -41.7046 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a6 i2 leaf node n5 p0 sample\" transform=\"translate(6.93333 41.7046)\"><path class=\"edge\" d=\"M 0 0 V -41.7046 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -34.9197 H 10.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -17.9757 H -19.0667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-19.0667 9.45536)\"><g class=\"a8 i0 leaf node n0 p0 sample\" transform=\"translate(-6.93333 85.1446)\"><path class=\"edge\" d=\"M 0 0 V -85.1446 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(6.93333 85.1446)\"><path class=\"edge\" d=\"M 0 0 V -85.1446 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -9.45536 H 19.0667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">9</text></g><g class=\"c1 node n10 p0 root\" transform=\"translate(96.2667 26.8)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 69,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_130_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    3,\n",
    "    sequence_length=10,\n",
    "    population_size=100,\n",
    "    model=sweep,\n",
    "    random_seed=12345)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cd1772",
   "metadata": {},
   "source": [
    "Here we ran the same simulation with a different random seed, and\n",
    "we can see that the time the model completed was different in\n",
    "the two cases (30.22 vs 28.15 generations ago).\n",
    "\n",
    ":::{note}\n",
    "Models such as {class}`.SweepGenicSelection` can also complete due\n",
    "to full coalescence.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models_specifying_multiple)=\n",
    "\n",
    "#### Multiple models\n",
    "\n",
    "Sometimes we are interested in combining different models in a simulation.\n",
    "To do this we provide a **list** of ancestry model instances as the\n",
    "``model`` argument to {func}`.sim_ancestry`. The models\n",
    "are run in the order they are specified. For each model, if its\n",
    "{attr}`~.AncestryModel.duration` is set, then the simulation will run\n",
    "for at most that additional time duration. If the ``duration`` is\n",
    "not set (or ``None``), then the model will run until\n",
    "{ref}`completion<sec_ancestry_models_specifying_completion>`.\n",
    "\n",
    "\n",
    "For example, here we run a hybrid DTWF and coalescent\n",
    "simulation (see the {ref}`sec_ancestry_models_dtwf` section for more\n",
    "details):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "da83228d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    2,\n",
    "    population_size=1000,\n",
    "    model=[\n",
    "        msprime.DiscreteTimeWrightFisher(duration=500),\n",
    "        msprime.StandardCoalescent(),\n",
    "    ],\n",
    "    random_seed=2\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb99e14",
   "metadata": {},
   "source": [
    ":::{warning}\n",
    "It is very imporant to remember to set a ``duration`` value in the first\n",
    "model here! Otherwise, the entire simulation will run under the DTWF model.\n",
    ":::\n",
    "\n",
    "In this example we've run the {class}`.DiscreteTimeWrightFisher`\n",
    "model for the first 500 generations, and then switched to the\n",
    "{class}`.StandardCoalescent`. Because we did not specify a\n",
    "``duration`` in the second model, it will run until coalescence.\n",
    "\n",
    ":::{note}\n",
    "Switching models at a fixed time point like this is equivalent to\n",
    "first running the DTWF phase of the simulation with ``end_time=500``\n",
    "and then using the output as the\n",
    "{ref}`initial state<sec_ancestry_initial_state>` for the simulation\n",
    "of the Hudson phase. The model switching syntax here is a little\n",
    "more convenient and efficient, however.\n",
    ":::\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_ancestry_models_selective_sweeps_multiple` section for\n",
    "an example of running many models with random durations.\n",
    ":::\n",
    "\n",
    ":::{tip}\n",
    "The logging output of msprime can be very useful when working with\n",
    "multiple models. See the {ref}`sec_logging` section for more\n",
    "details.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models_hudson)=\n",
    "\n",
    "### Hudson coalescent\n",
    "\n",
    "The {class}`standard coalescent<.StandardCoalescent>`\n",
    "is the default model in msprime. The algorithm\n",
    "for simulating the coalescent with recombination was developed\n",
    "by [Hudson (1983)](https://doi.org/10.1016/0040-5809(83)90013-8),\n",
    "and so we refer to it as the \"Hudson\" coalescent.\n",
    "\n",
    "Running a simulation without specifying a ``model`` is the\n",
    "same as running with ``model=\"hudson\"``:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1da2a2cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts1 = msprime.sim_ancestry(5, random_seed=2)\n",
    "ts2 = msprime.sim_ancestry(5, model=\"hudson\", random_seed=2)\n",
    "# This is the same simulation so tree sequences are equal\n",
    "assert ts1.equals(ts2, ignore_provenance=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bf48f84",
   "metadata": {},
   "source": [
    "Time is measured in units of generations ago. This is a continuous time\n",
    "model, in which the simulation moves event-by-event backwards\n",
    "in time (contrasting with the {ref}`sec_ancestry_models_dtwf` model, which\n",
    "works generation-by-generation). Thus, we can have fractional\n",
    "generations, as in this example where the MRCA of the sample\n",
    "occurs 10.59 \"generations\" ago:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f9f017bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 121.363)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 121.092)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.03</text></g></g><g class=\"tick\" transform=\"translate(56.8 110.629)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1.21</text></g></g><g class=\"tick\" transform=\"translate(56.8 62.7796)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">6.56</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10.59</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n10 p0 root\" transform=\"translate(68.5333 26.8)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(-20.8 83.8291)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(13.8667 10.7334)\"><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(-6.93333 0.0374811)\"><path class=\"edge\" d=\"M 0 0 V -0.0374811 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(6.93333 0.0374811)\"><path class=\"edge\" d=\"M 0 0 V -0.0374811 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -10.7334 H -13.8667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-13.8667 10.4625)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-6.93333 0.30835)\"><path class=\"edge\" d=\"M 0 0 V -0.30835 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i1 leaf node n2 p0 sample\" transform=\"translate(6.93333 0.30835)\"><path class=\"edge\" d=\"M 0 0 V -0.30835 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -10.4625 H 13.8667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -83.8291 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(20.8 35.9796)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-6.93333 58.6204)\"><path class=\"edge\" d=\"M 0 0 V -58.6204 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(6.93333 58.6204)\"><path class=\"edge\" d=\"M 0 0 V -58.6204 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -35.9796 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 72,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_136_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(3, random_seed=234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "344d3a81",
   "metadata": {},
   "source": [
    "This occurs because time is scaled to be proportional to\n",
    "the population size; if we run the same simulation with a\n",
    "different {ref}`population size<sec_ancestry_demography>`, we can see that\n",
    "the time values are simply scaled up accordingly\n",
    "(the default ``population_size`` is 1):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "78c2984c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 121.363)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.42</text></g></g><g class=\"tick\" transform=\"translate(56.8 121.092)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3.45</text></g></g><g class=\"tick\" transform=\"translate(56.8 110.629)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">120.53</text></g></g><g class=\"tick\" transform=\"translate(56.8 62.7796)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">655.98</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1058.61</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n10 p0 root\" transform=\"translate(68.5333 26.8)\"><g class=\"a10 c2 node n8 p0\" transform=\"translate(-20.8 83.8291)\"><g class=\"a8 c2 node n6 p0\" transform=\"translate(13.8667 10.7334)\"><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(-6.93333 0.0374811)\"><path class=\"edge\" d=\"M 0 0 V -0.0374811 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a6 i2 leaf node n4 p0 sample\" transform=\"translate(6.93333 0.0374811)\"><path class=\"edge\" d=\"M 0 0 V -0.0374811 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -10.7334 H -13.8667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">6</text></g><g class=\"a8 c2 node n7 p0\" transform=\"translate(-13.8667 10.4625)\"><g class=\"a7 i0 leaf node n0 p0 sample\" transform=\"translate(-6.93333 0.30835)\"><path class=\"edge\" d=\"M 0 0 V -0.30835 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a7 i1 leaf node n2 p0 sample\" transform=\"translate(6.93333 0.30835)\"><path class=\"edge\" d=\"M 0 0 V -0.30835 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -10.4625 H 13.8667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -83.8291 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(20.8 35.9796)\"><g class=\"a9 i0 leaf node n1 p0 sample\" transform=\"translate(-6.93333 58.6204)\"><path class=\"edge\" d=\"M 0 0 V -58.6204 H 6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a9 i2 leaf node n5 p0 sample\" transform=\"translate(6.93333 58.6204)\"><path class=\"edge\" d=\"M 0 0 V -58.6204 H -6.93333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -35.9796 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">9</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">10</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 73,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_138_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(3, population_size=100, random_seed=234)\n",
    "SVG(ts.draw_svg(y_axis=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a1c908f",
   "metadata": {},
   "source": [
    ":::{seealso}\n",
    "The {ref}`ploidy<sec_ancestry_ploidy>` parameter also controls\n",
    "the time scale; the\n",
    "default here is the diploid time scale where time is measured\n",
    "in units of 4N generations.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models_smc)=\n",
    "\n",
    "### SMC approximations\n",
    "\n",
    "The {class}`SMC <.SmcApproxCoalescent>` and {class}`SMC' <.SmcPrimeApproxCoalescent>`\n",
    "are approximations of the continuous time\n",
    "{ref}`Hudson coalescent<sec_ancestry_models_hudson>` model. These were originally\n",
    "motivated largely by the need to simulate coalescent processes more efficiently\n",
    "than was possible using the software available at the time; however,\n",
    "[improved algorithms](https://doi.org/10.1371/journal.pcbi.1004842)\n",
    "mean that such approximations are now mostly unnecessary for simulations.\n",
    "\n",
    "The SMC and SMC' are however very important for inference, as the approximations\n",
    "have made many analytical advances possible.\n",
    "\n",
    "Since the SMC approximations are not required for simulation efficiency, these\n",
    "models are implemented using a naive rejection sampling approach in msprime.\n",
    "The implementation is intended to facilitate the study of the\n",
    "SMC approximations, rather than to be used in a general-purpose way.\n",
    "\n",
    "(sec_ancestry_models_dtwf)=\n",
    "\n",
    "### Discrete Time Wright-Fisher\n",
    "\n",
    "Msprime provides the option to perform discrete-time Wright-Fisher (DTWF)\n",
    "simulations for scenarios when the coalescent model is not appropriate, including large\n",
    "sample sizes, multiple chromosomes, or recent migration.\n",
    "Please see the {class}`API documentation<.DiscreteTimeWrightFisher>`\n",
    "for a formal description of the model,\n",
    "and [Nelson et al. 2020](https://doi.org/10.1371/journal.pgen.1008619) for\n",
    "more details and background information.\n",
    "\n",
    "To run DTWF simulations, we need to provide the ``model=\"dtwf\"`` argument\n",
    "to {func}`.sim_ancestry`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "bc33dc0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 103.374)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 74.803)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5</text></g></g><g class=\"tick\" transform=\"translate(56.8 64.2584)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">8</text></g></g><g class=\"tick\" transform=\"translate(56.8 59.0397)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">10</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">37</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n13 p0 root\" transform=\"translate(66.15 26.8)\"><g class=\"a13 c3 node n10 p0\" transform=\"translate(-22.75 48.003)\"><g class=\"a10 i0 leaf node n1 p0 sample\" transform=\"translate(2.6 46.597)\"><path class=\"edge\" d=\"M 0 0 V -46.597 H -2.6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a10 i3 leaf node n7 p0 sample\" transform=\"translate(13 46.597)\"><path class=\"edge\" d=\"M 0 0 V -46.597 H -13\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-13 28.5708)\"><g class=\"a9 i0 leaf node n0 p0 sample\" transform=\"translate(-5.2 18.0262)\"><path class=\"edge\" d=\"M 0 0 V -18.0262 H 5.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a9 i1 leaf node n3 p0 sample\" transform=\"translate(5.2 18.0262)\"><path class=\"edge\" d=\"M 0 0 V -18.0262 H -5.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -28.5708 H 13\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -48.003 H 22.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(22.75 32.2397)\"><g class=\"a12 i2 leaf node n4 p0 sample\" transform=\"translate(9.1 62.3603)\"><path class=\"edge\" d=\"M 0 0 V -62.3603 H -9.1\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a12 c2 node n11 p0\" transform=\"translate(-9.1 5.21869)\"><g class=\"a11 i2 leaf node n5 p0 sample\" transform=\"translate(7.8 57.1416)\"><path class=\"edge\" d=\"M 0 0 V -57.1416 H -7.8\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a11 c2 node n8 p0\" transform=\"translate(-7.8 39.1154)\"><g class=\"a8 i1 leaf node n2 p0 sample\" transform=\"translate(-5.2 18.0262)\"><path class=\"edge\" d=\"M 0 0 V -18.0262 H 5.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a8 i3 leaf node n6 p0 sample\" transform=\"translate(5.2 18.0262)\"><path class=\"edge\" d=\"M 0 0 V -18.0262 H -5.2\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -39.1154 H 7.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -5.21869 H 9.1\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -32.2397 H -22.75\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 74,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_140_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(4, population_size=10, model=\"dtwf\", random_seed=7)\n",
    "SVG(ts.draw_svg(y_axis=True, time_scale=\"log_time\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "898944a9",
   "metadata": {},
   "source": [
    "There are a few important points to note here:\n",
    "\n",
    "1. All node times are integers (this is discrete time counted in generations);\n",
    "2. We can have **non-binary** nodes (see node 10);\n",
    "3. We can have **simultaneous** events (see nodes 8 and 9).\n",
    "\n",
    "Most features can be used with the DTWF model, and an error will be raised if\n",
    "you attempt to use a feature that is not implemented for DTWF. (Please let\n",
    "us know if this feature is important to you!)\n",
    "\n",
    ":::{important}\n",
    "For DTWF simulations with population structure, each row of the migration\n",
    "matrix must sum to one or less.\n",
    ":::\n",
    "\n",
    "Because DTWF simulations trace the history of a sample by going backwards\n",
    "in time generation-by-generation rather than event-by-event like the\n",
    "continuous time models, it is significantly slower to simulate than\n",
    "the standard coalescent.\n",
    "\n",
    "\"Hybrid\" simulations can be a good solution if DTWF simulations are\n",
    "slow. In this case we can use DTWF simulations for the recent past, and using\n",
    "the standard coalescent for the more distant past (where it is an excellent\n",
    "approximation of the discrete time model). Hybrid simulations\n",
    "of this type provide the best of computational efficiency and accuracy.\n",
    "\n",
    ":::{seealso}\n",
    "See [Nelson et al. 2020](https://doi.org/10.1371/journal.pgen.1008619) for\n",
    "more details and examples where the DTWF model is more realistic than\n",
    "the coalescent, and an assessment of the accuracy of hybrid simulations.\n",
    ":::\n",
    "\n",
    "For example, here we simulate with the DTWF model for 500 generations,\n",
    "before switching to the standard (Hudson) coalescent:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "7a36ebec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l123.2,0 l0,138.2 l0,25 l0,5 l-123.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(118.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 69.5098)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">74.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 49.0357)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">411.00</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2619.53</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n6 p0 root\" transform=\"translate(61.6 26.8)\"><g class=\"a6 c2 node n4 p0\" transform=\"translate(20.8 42.7098)\"><g class=\"a4 i0 leaf node n1 p0 sample\" transform=\"translate(-10.4 51.8902)\"><path class=\"edge\" d=\"M 0 0 V -51.8902 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a4 i1 leaf node n3 p0 sample\" transform=\"translate(10.4 51.8902)\"><path class=\"edge\" d=\"M 0 0 V -51.8902 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -42.7098 H -20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4</text></g><g class=\"a6 c2 node n5 p0\" transform=\"translate(-20.8 22.2357)\"><g class=\"a5 i0 leaf node n0 p0 sample\" transform=\"translate(-10.4 72.3643)\"><path class=\"edge\" d=\"M 0 0 V -72.3643 H 10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a5 i1 leaf node n2 p0 sample\" transform=\"translate(10.4 72.3643)\"><path class=\"edge\" d=\"M 0 0 V -72.3643 H -10.4\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -22.2357 H 20.8\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">5</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">6</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 75,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_142_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    2,\n",
    "    population_size=1000,\n",
    "    model=[\n",
    "        msprime.DiscreteTimeWrightFisher(duration=500),\n",
    "        msprime.StandardCoalescent(),\n",
    "    ],\n",
    "    random_seed=4\n",
    ")\n",
    "SVG(ts.draw_svg(y_axis=True, time_scale=\"log_time\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba7d9737",
   "metadata": {},
   "source": [
    "Because of the integer node times, we can see here that most of the coalescent\n",
    "happened during the Wright-Fisher phase of the simulation, and as-of 500\n",
    "generations in the past, there were only two lineages left. The continuous\n",
    "time standard coalescent model was then used to simulate the ancient past of\n",
    "these two lineages.\n",
    "\n",
    "See the {ref}`sec_ancestry_models_specifying` section for more details on\n",
    "how the use the ``model`` argument to {func}`.sim_ancestry` to run\n",
    "multiple models.\n",
    "\n",
    ":::{seealso}\n",
    "See the {ref}`sec_ancestry_multiple_chromosomes` section for information\n",
    "on how to simulate multiple chromosomes using the DTWF.\n",
    ":::\n",
    "\n",
    "\n",
    "(sec_ancestry_models_multiple_mergers)=\n",
    "\n",
    "### Multiple merger coalescents\n",
    "\n",
    "Some evolutionary scenarios, such as a skewed offspring distribution\n",
    "combined with a type III survivorship curve, range expansion, and\n",
    "rapid adaptation, can predict genealogies with up to four simultaneous\n",
    "multiple mergers. Msprime provides the option to simulate from two classes\n",
    "of such genealogical processes: the {class}`.BetaCoalescent` and\n",
    "the {class}`.DiracCoalescent`. Please see the API documentation for formal\n",
    "details of the models.\n",
    "\n",
    "For haploid organisms, both models result in genealogies in which\n",
    "any number of lineages can merge into a common ancestor,\n",
    "but only one merger event can take place at a given time. For {math}`p`-ploids,\n",
    "up to {math}`2p` simultaneous mergers can take place, corresponding to the\n",
    "{math}`2p` available parental chromosome copies.\n",
    "\n",
    "The diploid Beta-Xi-coalescent can be simulated as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "dbdee832",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n16 p0 root\" transform=\"translate(98 26.8)\"><g class=\"a16 c2 node n12 p0\" transform=\"translate(30 71.1024)\"><g class=\"a12 i4 leaf node n8 p0 sample\" transform=\"translate(-6 23.4976)\"><path class=\"edge\" d=\"M 0 0 V -23.4976 H 6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a12 i4 leaf node n9 p0 sample\" transform=\"translate(6 23.4976)\"><path class=\"edge\" d=\"M 0 0 V -23.4976 H -6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -71.1024 H -30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><g class=\"a16 c2 node n15 p0\" transform=\"translate(-30 50.5666)\"><g class=\"a15 c2 node n11 p0\" transform=\"translate(-18 31.832)\"><g class=\"a11 i1 leaf node n2 p0 sample\" transform=\"translate(12 12.2014)\"><path class=\"edge\" d=\"M 0 0 V -12.2014 H -12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a11 c3 node n10 p0\" transform=\"translate(-12 1.6429)\"><g class=\"a10 i0 leaf node n0 p0 sample\" transform=\"translate(-12 10.5585)\"><path class=\"edge\" d=\"M 0 0 V -10.5585 H 12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a10 i2 leaf node n4 p0 sample\" transform=\"translate(0.0 10.5585)\"><path class=\"edge\" d=\"M 0 0 V -10.5585 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a10 i3 leaf node n6 p0 sample\" transform=\"translate(12 10.5585)\"><path class=\"edge\" d=\"M 0 0 V -10.5585 H -12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -1.6429 H 12\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><path class=\"edge\" d=\"M 0 0 V -31.832 H 18\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">11</text></g><g class=\"a15 c2 node n14 p0\" transform=\"translate(18 9.11835)\"><g class=\"a14 i0 leaf node n1 p0 sample\" transform=\"translate(-12 34.9151)\"><path class=\"edge\" d=\"M 0 0 V -34.9151 H 12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a14 c3 node n13 p0\" transform=\"translate(12 11.4175)\"><g class=\"a13 i1 leaf node n3 p0 sample\" transform=\"translate(-12 23.4976)\"><path class=\"edge\" d=\"M 0 0 V -23.4976 H 12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a13 i2 leaf node n5 p0 sample\" transform=\"translate(0.0 23.4976)\"><path class=\"edge\" d=\"M 0 0 V -23.4976 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a13 i3 leaf node n7 p0 sample\" transform=\"translate(12 23.4976)\"><path class=\"edge\" d=\"M 0 0 V -23.4976 H -12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><path class=\"edge\" d=\"M 0 0 V -11.4175 H -12\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -9.11835 H -18\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">14</text></g><path class=\"edge\" d=\"M 0 0 V -50.5666 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">15</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 76,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_144_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=5, ploidy=2, random_seed=1,\n",
    "    model=msprime.BetaCoalescent(alpha=1.001))\n",
    "SVG(ts.draw_svg())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51e5fc38",
   "metadata": {},
   "source": [
    "The specified value of {math}`\\alpha = 1.001` corresponds to a heavily skewed\n",
    "offspring distribution. Values closer to {math}`\\alpha = 2` result in trees\n",
    "whose distribution is closer to that of the standard coalescent, often featuring\n",
    "no multiple mergers for small sample sizes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "89c4bd4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n14 p0 root\" transform=\"translate(102.031 26.8)\"><g class=\"a14 i3 leaf node n6 p0 sample\" transform=\"translate(30.4688 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -30.4688\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(-30.4688 10.6488)\"><g class=\"a13 c2 node n10 p0\" transform=\"translate(32.8125 55.1326)\"><g class=\"a10 i2 leaf node n5 p0 sample\" transform=\"translate(13.125 28.8186)\"><path class=\"edge\" d=\"M 0 0 V -28.8186 H -13.125\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a10 c2 node n9 p0\" transform=\"translate(-13.125 11.6467)\"><g class=\"a9 i3 leaf node n7 p0 sample\" transform=\"translate(11.25 17.1719)\"><path class=\"edge\" d=\"M 0 0 V -17.1719 H -11.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a9 c2 node n8 p0\" transform=\"translate(-11.25 6.03837)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(-7.5 11.1336)\"><path class=\"edge\" d=\"M 0 0 V -11.1336 H 7.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a8 i1 leaf node n3 p0 sample\" transform=\"translate(7.5 11.1336)\"><path class=\"edge\" d=\"M 0 0 V -11.1336 H -7.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -6.03837 H 11.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -11.6467 H 13.125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -55.1326 H -32.8125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10</text></g><g class=\"a13 c2 node n12 p0\" transform=\"translate(-32.8125 7.74681)\"><g class=\"a12 i0 leaf node n0 p0 sample\" transform=\"translate(-11.25 76.2044)\"><path class=\"edge\" d=\"M 0 0 V -76.2044 H 11.25\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a12 c2 node n11 p0\" transform=\"translate(11.25 3.18189)\"><g class=\"a11 i1 leaf node n2 p0 sample\" transform=\"translate(-7.5 73.0225)\"><path class=\"edge\" d=\"M 0 0 V -73.0225 H 7.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a11 i2 leaf node n4 p0 sample\" transform=\"translate(7.5 73.0225)\"><path class=\"edge\" d=\"M 0 0 V -73.0225 H -7.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><path class=\"edge\" d=\"M 0 0 V -3.18189 H -11.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -7.74681 H 32.8125\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -10.6488 H 30.4688\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 77,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_146_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=4, ploidy=2, random_seed=1,\n",
    "    model=msprime.BetaCoalescent(alpha=1.8))\n",
    "SVG(ts.draw_svg())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3890b8c",
   "metadata": {},
   "source": [
    "Multiple mergers still take place in a haploid simulation, but only one merger\n",
    "can take place at a given time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "09c345fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c3 node n14 p0 root\" transform=\"translate(93.5 26.8)\"><g class=\"a14 i1 leaf node n1 p0 sample\" transform=\"translate(28.5 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -28.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a14 i2 leaf node n2 p0 sample\" transform=\"translate(40.5 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -40.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(-40.5 16.9651)\"><g class=\"a13 i0 leaf node n0 p0 sample\" transform=\"translate(-27 77.6349)\"><path class=\"edge\" d=\"M 0 0 V -77.6349 H 27\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a13 c4 node n12 p0\" transform=\"translate(27 0.361523)\"><g class=\"a12 i5 leaf node n5 p0 sample\" transform=\"translate(-6 77.2734)\"><path class=\"edge\" d=\"M 0 0 V -77.2734 H 6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a12 i7 leaf node n7 p0 sample\" transform=\"translate(30 77.2734)\"><path class=\"edge\" d=\"M 0 0 V -77.2734 H -30\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a12 c3 node n10 p0\" transform=\"translate(-30 45.6373)\"><g class=\"a10 i3 leaf node n3 p0 sample\" transform=\"translate(-12 31.6361)\"><path class=\"edge\" d=\"M 0 0 V -31.6361 H 12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a10 i4 leaf node n4 p0 sample\" transform=\"translate(0.0 31.6361)\"><path class=\"edge\" d=\"M 0 0 V -31.6361 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a10 i9 leaf node n9 p0 sample\" transform=\"translate(12 31.6361)\"><path class=\"edge\" d=\"M 0 0 V -31.6361 H -12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -45.6373 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><g class=\"a12 c2 node n11 p0\" transform=\"translate(12 18.6532)\"><g class=\"a11 i6 leaf node n6 p0 sample\" transform=\"translate(-6 58.6202)\"><path class=\"edge\" d=\"M 0 0 V -58.6202 H 6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a11 i8 leaf node n8 p0 sample\" transform=\"translate(6 58.6202)\"><path class=\"edge\" d=\"M 0 0 V -58.6202 H -6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -18.6532 H -12\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -0.361523 H -27\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><path class=\"edge\" d=\"M 0 0 V -16.9651 H 40.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 78,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_148_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=10, ploidy=1, random_seed=1,\n",
    "    model=msprime.BetaCoalescent(alpha=1.001))\n",
    "SVG(ts.draw_svg())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d0a934f",
   "metadata": {},
   "source": [
    "A haploid simulation results in larger individual mergers than a polyploid simulation\n",
    "because large mergers typically get broken up into multiple simultaneous mergers\n",
    "in the polyploid model.\n",
    "\n",
    "The number of generations between merger events in the Beta-coalescent depends\n",
    "nonlinearly on both {math}`\\alpha` and the population size {math}`N`\n",
    "as detailed above.\n",
    "For a fixed {math}`\\alpha`, the number of generations between common ancestor events\n",
    "is proportional to {math}`N^{\\alpha - 1}`, albeit with a complicated constant of\n",
    "proportionality that depends on {math}`\\alpha`. The dependence on {math}`\\alpha`\n",
    "for fixed {math}`N` is not monotone. Thus, branch lengths and the number of\n",
    "generations until a most recent common ancestor depend on both of these parameters.\n",
    "\n",
    "To illustrate, for {math}`\\alpha` close to 2 the relationship between effective\n",
    "population size and number of generations is almost linear:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3924fb79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.14959691919068155\n",
      "14.286394871874865\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=10,\n",
    "    model=msprime.BetaCoalescent(alpha=1.99))\n",
    "tree = ts.first()\n",
    "print(tree.tmrca(0,1))\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=1000,\n",
    "    model=msprime.BetaCoalescent(alpha=1.99))\n",
    "tree = ts.first()\n",
    "print(tree.tmrca(0,1))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53d2cec8",
   "metadata": {},
   "source": [
    "For {math}`\\alpha` close to 1 the effective population size has little effect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5028ab92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16.311807036386615\n",
      "25.85247192870844\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=10,\n",
    "    model=msprime.BetaCoalescent(alpha=1.1))\n",
    "tree = ts.first()\n",
    "print(tree.tmrca(0,1))\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=1000,\n",
    "    model=msprime.BetaCoalescent(alpha=1.1))\n",
    "tree = ts.first()\n",
    "print(tree.tmrca(0,1))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5833a97d",
   "metadata": {},
   "source": [
    "The Dirac-coalescent is simulated similarly in both the diploid case:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "03630133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n15 p0 root\" transform=\"translate(107.75 26.8)\"><g class=\"a15 c2 node n12 p0\" transform=\"translate(20.25 68.1879)\"><g class=\"a12 i0 leaf node n1 p0 sample\" transform=\"translate(-6 26.4121)\"><path class=\"edge\" d=\"M 0 0 V -26.4121 H 6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a12 i1 leaf node n2 p0 sample\" transform=\"translate(6 26.4121)\"><path class=\"edge\" d=\"M 0 0 V -26.4121 H -6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><path class=\"edge\" d=\"M 0 0 V -68.1879 H -20.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">12</text></g><g class=\"a15 c2 node n14 p0\" transform=\"translate(-20.25 35.7213)\"><g class=\"a14 i3 leaf node n6 p0 sample\" transform=\"translate(22.5 58.8787)\"><path class=\"edge\" d=\"M 0 0 V -58.8787 H -22.5\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a14 c2 node n13 p0\" transform=\"translate(-22.5 13.5018)\"><g class=\"a13 c4 node n10 p0\" transform=\"translate(-21 44.817)\"><g class=\"a10 i0 leaf node n0 p0 sample\" transform=\"translate(-18 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H 18\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a10 i2 leaf node n5 p0 sample\" transform=\"translate(-6 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H 6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a10 i3 leaf node n7 p0 sample\" transform=\"translate(6 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H -6\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a10 i4 leaf node n9 p0 sample\" transform=\"translate(18 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H -18\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -44.817 H 21\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><g class=\"a13 c3 node n11 p0\" transform=\"translate(21 44.817)\"><g class=\"a11 i1 leaf node n3 p0 sample\" transform=\"translate(-12 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H 12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a11 i2 leaf node n4 p0 sample\" transform=\"translate(0.0 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a11 i4 leaf node n8 p0 sample\" transform=\"translate(12 0.559851)\"><path class=\"edge\" d=\"M 0 0 V -0.559851 H -12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -44.817 H -21\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">11</text></g><path class=\"edge\" d=\"M 0 0 V -13.5018 H 22.5\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">13</text></g><path class=\"edge\" d=\"M 0 0 V -35.7213 H 20.25\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">14</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">15</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 81,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_154_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=5, ploidy=2, random_seed=1,\n",
    "    model=msprime.DiracCoalescent(psi=0.9, c=10))\n",
    "SVG(ts.draw_svg())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bed30b00",
   "metadata": {},
   "source": [
    "and in the haploid case:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "dcda0567",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M20,0 l160,0 l0,138.2 l0,25 l0,5 l-160,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(100 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"20\" x2=\"180\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(20 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(180 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">1</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(20 0)\"><g class=\"plotbox\"><g class=\"c2 node n11 p0 root\" transform=\"translate(104 26.8)\"><g class=\"a11 i1 leaf node n1 p0 sample\" transform=\"translate(30 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H -30\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a11 c9 node n10 p0\" transform=\"translate(-30 77.5503)\"><g class=\"a10 i0 leaf node n0 p0 sample\" transform=\"translate(-48 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H 48\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a10 i2 leaf node n2 p0 sample\" transform=\"translate(-36 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H 36\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a10 i3 leaf node n3 p0 sample\" transform=\"translate(-24 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H 24\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><g class=\"a10 i4 leaf node n4 p0 sample\" transform=\"translate(-12 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H 12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a10 i5 leaf node n5 p0 sample\" transform=\"translate(0.0 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><g class=\"a10 i6 leaf node n6 p0 sample\" transform=\"translate(12 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H -12\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6</text></g><g class=\"a10 i7 leaf node n7 p0 sample\" transform=\"translate(24 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H -24\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7</text></g><g class=\"a10 i8 leaf node n8 p0 sample\" transform=\"translate(36 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H -36\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8</text></g><g class=\"a10 i9 leaf node n9 p0 sample\" transform=\"translate(48 17.0497)\"><path class=\"edge\" d=\"M 0 0 V -17.0497 H -48\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9</text></g><path class=\"edge\" d=\"M 0 0 V -77.5503 H 30\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 82,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_156_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=10, ploidy=1, random_seed=1,\n",
    "    model=msprime.DiracCoalescent(psi=0.9, c=10))\n",
    "SVG(ts.draw_svg())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6960f587",
   "metadata": {},
   "source": [
    "As with the Beta-coalescent, a haploid simulation results in larger individual\n",
    "mergers than a polyploid simulation because large mergers typically get broken\n",
    "up into multiple simultaneous mergers in the polyploid model. Larger values\n",
    "of the parameter {math}`c > 0` result in more frequent multiple mergers,\n",
    "while larger values of {math}`0 < \\psi \\leq 1` result in multiple mergers\n",
    "with more participating lineages. Setting either parameter to 0 would correspond\n",
    "to the standard coalescent.\n",
    "\n",
    "The Dirac-coalescent is obtained as the infinite population scaling limit of\n",
    "Moran models, and therefore branch lengths are proportional to {math}`N^2`\n",
    "generations, as opposed to {math}`N` generations under the standard coalescent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "af8aa2f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64.7216732891741\n",
      "6472.167328917411\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=10,\n",
    "    model=msprime.DiracCoalescent(psi=0.1, c=1))\n",
    "tree = ts.first()\n",
    "print(tree.tmrca(0,1))\n",
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=100,\n",
    "    model=msprime.DiracCoalescent(psi=0.1, c=1))\n",
    "tree = ts.first()\n",
    "print(tree.tmrca(0,1))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfd58cf6",
   "metadata": {},
   "source": [
    "Because branch lengths are proportional to {math}`N^2` generations, the number\n",
    "of events simulated along branches (such as mutations or recombinations) will\n",
    "also be proportional to {math}`N^2`. For example, if a tree sequence generated\n",
    "using a given census size {math}`N` is used as input for simulating segregating\n",
    "sites (see {ref}`Mutation simulations<sec_mutations>` for details on how to do\n",
    "this), the expected number of segregating sites obtained will be proportional\n",
    "to {math}`N^2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "6be69dce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6652\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=10000,recombination_rate=1e-8,\n",
    "    sequence_length=10**4,model=msprime.DiracCoalescent(psi=0.1, c=0))\n",
    "mts = msprime.sim_mutations(ts, rate=1e-8)\n",
    "print(str(mts.get_num_sites()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "102c05d9",
   "metadata": {},
   "source": [
    "If the desired amount of diversity (as measured by the number of segregating\n",
    "sites) is proportional to {math}`N`, then {math}`N^{1/2}` should be used for\n",
    "the population size under the Dirac-coalescent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "4b08ae6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    samples=1, ploidy=2, random_seed=1, population_size=100,recombination_rate=1e-8,\n",
    "    sequence_length=10**4,model=msprime.DiracCoalescent(psi=0.1, c=0))\n",
    "mts = msprime.sim_mutations(ts, rate=1e-8)\n",
    "print(str(mts.get_num_sites()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be26c35d",
   "metadata": {},
   "source": [
    "Therefore the population size must be carefully chosen (and\n",
    "potentially rescaled) to obtain the desired simulated data in these models.\n",
    "\n",
    "(sec_ancestry_models_selective_sweeps)=\n",
    "\n",
    "### Selective sweeps\n",
    "\n",
    "Although the coalescent assumes that lineages are exchangable, (and therefore\n",
    "evolving neutrally), approximations to some forms of selective sweep\n",
    "(beneficial mutation moves through the population), have been derived. This is\n",
    "done through the use of a structured coalescent model in the spirit of\n",
    "[Braverman et al. (1995)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1206652/).\n",
    "\n",
    "In the {class}`.SweepGenicSelection` model,\n",
    "the population is split between two classes of selective backgrounds,\n",
    "those linked to the beneficial allele, call it {math}`B`, and those not,\n",
    "{math}`b`. In this model we track competing\n",
    "rates of coalescence and recombination on the {math}`B` and {math}`b`\n",
    "backgrounds.\n",
    "The user supplies a final allele frequency and a starting\n",
    "allele frequency, between which msprime simulates a stochastic\n",
    "sweep trajectory according to a conditional diffusion model [(Coop\n",
    "and Griffiths, 2004](https://pubmed.ncbi.nlm.nih.gov/15465123/);\n",
    "[Kim and Stephan 2002](https://pubmed.ncbi.nlm.nih.gov/11861577/)).\n",
    "\n",
    "Beyond the start and end frequencies of the sweep trajectory, the user\n",
    "must specify the selection coefficient of the beneficial mutation\n",
    "{math}`s` the selective\n",
    "advantage of the {math}`B` homozygote over the {math}`b` homozygote\n",
    "and {math}`h=0.5`. The position represents the location along\n",
    "the chromosome where the beneficial allele occurs.\n",
    "All other parameters can be set as usual.\n",
    "\n",
    ":::{warning}\n",
    "The implementation of the structured coalescent is quite general, but there are\n",
    "some limitations in the current implementation of the sweeps model (e.g., no\n",
    "change of population size during a sweep). Please let us know if there are\n",
    "related features you would like to see (or if you would be interested in\n",
    "helping to create this functionality!)\n",
    ":::\n",
    "\n",
    "\n",
    "#### Hard sweeps\n",
    "\n",
    ":::{todo}\n",
    "What's a hard sweep? Some intro sentences here.\n",
    ":::\n",
    "\n",
    "In this example we run some replicate hard sweep simulations\n",
    "and plot the mean\n",
    "{meth}`pairwise diversity<tskit.TreeSequence.diversity>` in windows\n",
    "across the simulated region, showing the characteristic\n",
    "valley of diversity.\n",
    "\n",
    "First we set up some basic parameters, and define the sweep model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "9cc62768",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ne = 1e3\n",
    "L = 1e6  # Length of simulated region\n",
    "num_reps = 100\n",
    "\n",
    "# define hard sweep model\n",
    "sweep_model = msprime.SweepGenicSelection(\n",
    "    position=L / 2,  # middle of chrom\n",
    "    start_frequency=1.0 / (2 * Ne),\n",
    "    end_frequency=1.0 - (1.0 / (2 * Ne)),\n",
    "    s=0.25,\n",
    "    dt=1e-6,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb1b7780",
   "metadata": {},
   "source": [
    ":::{todo}\n",
    "Explain why these parameters give a hard sweep.\n",
    ":::\n",
    "\n",
    "Next we set up the replicate simulations. As described\n",
    "in the {ref}`sec_ancestry_models_specifying_multiple` section,\n",
    "the ``model`` parameter is a list when we want to run multiple\n",
    "ancestry models. In this example, we have a sweep that occurs\n",
    "in the immediate past, and is then followed by the\n",
    "standard coalescent for the rest of time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "352339d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reps = msprime.sim_ancestry(\n",
    "    5,\n",
    "    model=[sweep_model, msprime.StandardCoalescent()],\n",
    "    population_size=Ne,\n",
    "    recombination_rate=1e-7,\n",
    "    sequence_length=L,\n",
    "    num_replicates=num_reps,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8c58885",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "Because the {class}`.SweepGenicSelection` model has a random\n",
    "{ref}`duration<sec_ancestry_models_specifying_duration>` we do\n",
    "not set this value in the model. Please see the\n",
    "{ref}`sec_ancestry_models_specifying_completion` section for\n",
    "more discussion on this important point.\n",
    ":::\n",
    "\n",
    "Once we've set up the replicate simulations we can compute the\n",
    "windows for plotting, run the actual simulations\n",
    "(see the {ref}`sec_randomness_replication` section for more details)\n",
    "and compute the\n",
    "{meth}`pairwise diversity<tskit.TreeSequence.diversity>`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "ac76bb5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "wins = np.linspace(0, L, 21)\n",
    "mids = (wins[1:] + wins[:-1]) / 2\n",
    "diversity = np.zeros((num_reps, mids.shape[0]))\n",
    "for j, ts in enumerate(reps):\n",
    "    diversity[j] = ts.diversity(windows=wins, mode=\"branch\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b201571",
   "metadata": {},
   "source": [
    "Finally, we can plot the observed mean diversity across the replicates\n",
    "and compare it to the neutral expectation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "90c5068a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"417.910501pt\" height=\"310.86825pt\" viewBox=\"0 0 417.910501 310.86825\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-10-17T07:24:33.377157</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 310.86825 \n",
       "L 417.910501 310.86825 \n",
       "L 417.910501 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 53.328125 273.312 \n",
       "L 410.448125 273.312 \n",
       "L 410.448125 7.2 \n",
       "L 53.328125 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <defs>\n",
       "       <path id=\"m0bfc4866d5\" d=\"M 0 0 \n",
       "L 0 3.5 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0bfc4866d5\" x=\"61.017312\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0.0 -->\n",
       "      <g transform=\"translate(53.065749 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-2e\" d=\"M 684 794 \n",
       "L 1344 794 \n",
       "L 1344 0 \n",
       "L 684 0 \n",
       "L 684 794 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0bfc4866d5\" x=\"129.365637\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 0.2 -->\n",
       "      <g transform=\"translate(121.414074 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0bfc4866d5\" x=\"197.713962\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 0.4 -->\n",
       "      <g transform=\"translate(189.7624 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-34\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0bfc4866d5\" x=\"266.062288\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 0.6 -->\n",
       "      <g transform=\"translate(258.110725 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-36\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_5\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0bfc4866d5\" x=\"334.410613\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 0.8 -->\n",
       "      <g transform=\"translate(326.459051 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-38\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_6\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0bfc4866d5\" x=\"402.758938\" y=\"273.312\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- 1.0 -->\n",
       "      <g transform=\"translate(394.807376 287.910437)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_7\">\n",
       "     <!-- Position (bp) -->\n",
       "     <g transform=\"translate(200.585 301.588562)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-50\" d=\"M 1259 4147 \n",
       "L 1259 2394 \n",
       "L 2053 2394 \n",
       "Q 2494 2394 2734 2622 \n",
       "Q 2975 2850 2975 3272 \n",
       "Q 2975 3691 2734 3919 \n",
       "Q 2494 4147 2053 4147 \n",
       "L 1259 4147 \n",
       "z\n",
       "M 628 4666 \n",
       "L 2053 4666 \n",
       "Q 2838 4666 3239 4311 \n",
       "Q 3641 3956 3641 3272 \n",
       "Q 3641 2581 3239 2228 \n",
       "Q 2838 1875 2053 1875 \n",
       "L 1259 1875 \n",
       "L 1259 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6f\" d=\"M 1959 3097 \n",
       "Q 1497 3097 1228 2736 \n",
       "Q 959 2375 959 1747 \n",
       "Q 959 1119 1226 758 \n",
       "Q 1494 397 1959 397 \n",
       "Q 2419 397 2687 759 \n",
       "Q 2956 1122 2956 1747 \n",
       "Q 2956 2369 2687 2733 \n",
       "Q 2419 3097 1959 3097 \n",
       "z\n",
       "M 1959 3584 \n",
       "Q 2709 3584 3137 3096 \n",
       "Q 3566 2609 3566 1747 \n",
       "Q 3566 888 3137 398 \n",
       "Q 2709 -91 1959 -91 \n",
       "Q 1206 -91 779 398 \n",
       "Q 353 888 353 1747 \n",
       "Q 353 2609 779 3096 \n",
       "Q 1206 3584 1959 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-73\" d=\"M 2834 3397 \n",
       "L 2834 2853 \n",
       "Q 2591 2978 2328 3040 \n",
       "Q 2066 3103 1784 3103 \n",
       "Q 1356 3103 1142 2972 \n",
       "Q 928 2841 928 2578 \n",
       "Q 928 2378 1081 2264 \n",
       "Q 1234 2150 1697 2047 \n",
       "L 1894 2003 \n",
       "Q 2506 1872 2764 1633 \n",
       "Q 3022 1394 3022 966 \n",
       "Q 3022 478 2636 193 \n",
       "Q 2250 -91 1575 -91 \n",
       "Q 1294 -91 989 -36 \n",
       "Q 684 19 347 128 \n",
       "L 347 722 \n",
       "Q 666 556 975 473 \n",
       "Q 1284 391 1588 391 \n",
       "Q 1994 391 2212 530 \n",
       "Q 2431 669 2431 922 \n",
       "Q 2431 1156 2273 1281 \n",
       "Q 2116 1406 1581 1522 \n",
       "L 1381 1569 \n",
       "Q 847 1681 609 1914 \n",
       "Q 372 2147 372 2553 \n",
       "Q 372 3047 722 3315 \n",
       "Q 1072 3584 1716 3584 \n",
       "Q 2034 3584 2315 3537 \n",
       "Q 2597 3491 2834 3397 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-69\" d=\"M 603 3500 \n",
       "L 1178 3500 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 3500 \n",
       "z\n",
       "M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 4134 \n",
       "L 603 4134 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-74\" d=\"M 1172 4494 \n",
       "L 1172 3500 \n",
       "L 2356 3500 \n",
       "L 2356 3053 \n",
       "L 1172 3053 \n",
       "L 1172 1153 \n",
       "Q 1172 725 1289 603 \n",
       "Q 1406 481 1766 481 \n",
       "L 2356 481 \n",
       "L 2356 0 \n",
       "L 1766 0 \n",
       "Q 1100 0 847 248 \n",
       "Q 594 497 594 1153 \n",
       "L 594 3053 \n",
       "L 172 3053 \n",
       "L 172 3500 \n",
       "L 594 3500 \n",
       "L 594 4494 \n",
       "L 1172 4494 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113 \n",
       "L 3513 0 \n",
       "L 2938 0 \n",
       "L 2938 2094 \n",
       "Q 2938 2591 2744 2837 \n",
       "Q 2550 3084 2163 3084 \n",
       "Q 1697 3084 1428 2787 \n",
       "Q 1159 2491 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1366 3272 1645 3428 \n",
       "Q 1925 3584 2291 3584 \n",
       "Q 2894 3584 3203 3211 \n",
       "Q 3513 2838 3513 2113 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-28\" d=\"M 1984 4856 \n",
       "Q 1566 4138 1362 3434 \n",
       "Q 1159 2731 1159 2009 \n",
       "Q 1159 1288 1364 580 \n",
       "Q 1569 -128 1984 -844 \n",
       "L 1484 -844 \n",
       "Q 1016 -109 783 600 \n",
       "Q 550 1309 550 2009 \n",
       "Q 550 2706 781 3412 \n",
       "Q 1013 4119 1484 4856 \n",
       "L 1984 4856 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-62\" d=\"M 3116 1747 \n",
       "Q 3116 2381 2855 2742 \n",
       "Q 2594 3103 2138 3103 \n",
       "Q 1681 3103 1420 2742 \n",
       "Q 1159 2381 1159 1747 \n",
       "Q 1159 1113 1420 752 \n",
       "Q 1681 391 2138 391 \n",
       "Q 2594 391 2855 752 \n",
       "Q 3116 1113 3116 1747 \n",
       "z\n",
       "M 1159 2969 \n",
       "Q 1341 3281 1617 3432 \n",
       "Q 1894 3584 2278 3584 \n",
       "Q 2916 3584 3314 3078 \n",
       "Q 3713 2572 3713 1747 \n",
       "Q 3713 922 3314 415 \n",
       "Q 2916 -91 2278 -91 \n",
       "Q 1894 -91 1617 61 \n",
       "Q 1341 213 1159 525 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 4863 \n",
       "L 1159 4863 \n",
       "L 1159 2969 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-70\" d=\"M 1159 525 \n",
       "L 1159 -1331 \n",
       "L 581 -1331 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2969 \n",
       "Q 1341 3281 1617 3432 \n",
       "Q 1894 3584 2278 3584 \n",
       "Q 2916 3584 3314 3078 \n",
       "Q 3713 2572 3713 1747 \n",
       "Q 3713 922 3314 415 \n",
       "Q 2916 -91 2278 -91 \n",
       "Q 1894 -91 1617 61 \n",
       "Q 1341 213 1159 525 \n",
       "z\n",
       "M 3116 1747 \n",
       "Q 3116 2381 2855 2742 \n",
       "Q 2594 3103 2138 3103 \n",
       "Q 1681 3103 1420 2742 \n",
       "Q 1159 2381 1159 1747 \n",
       "Q 1159 1113 1420 752 \n",
       "Q 1681 391 2138 391 \n",
       "Q 2594 391 2855 752 \n",
       "Q 3116 1113 3116 1747 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-29\" d=\"M 513 4856 \n",
       "L 1013 4856 \n",
       "Q 1481 4119 1714 3412 \n",
       "Q 1947 2706 1947 2009 \n",
       "Q 1947 1309 1714 600 \n",
       "Q 1481 -109 1013 -844 \n",
       "L 513 -844 \n",
       "Q 928 -128 1133 580 \n",
       "Q 1338 1288 1338 2009 \n",
       "Q 1338 2731 1133 3434 \n",
       "Q 928 4138 513 4856 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-50\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"56.677734\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"117.859375\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"169.958984\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"197.742188\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"236.951172\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"264.734375\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"325.916016\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"389.294922\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-28\" x=\"421.082031\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-62\" x=\"460.095703\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-70\" x=\"523.572266\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-29\" x=\"587.048828\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_8\">\n",
       "     <!-- 1e6 -->\n",
       "     <g transform=\"translate(391.57 300.588562)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-65\" d=\"M 3597 1894 \n",
       "L 3597 1613 \n",
       "L 953 1613 \n",
       "Q 991 1019 1311 708 \n",
       "Q 1631 397 2203 397 \n",
       "Q 2534 397 2845 478 \n",
       "Q 3156 559 3463 722 \n",
       "L 3463 178 \n",
       "Q 3153 47 2828 -22 \n",
       "Q 2503 -91 2169 -91 \n",
       "Q 1331 -91 842 396 \n",
       "Q 353 884 353 1716 \n",
       "Q 353 2575 817 3079 \n",
       "Q 1281 3584 2069 3584 \n",
       "Q 2775 3584 3186 3129 \n",
       "Q 3597 2675 3597 1894 \n",
       "z\n",
       "M 3022 2063 \n",
       "Q 3016 2534 2758 2815 \n",
       "Q 2500 3097 2075 3097 \n",
       "Q 1594 3097 1305 2825 \n",
       "Q 1016 2553 972 2059 \n",
       "L 3022 2063 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-36\" x=\"125.146484\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <defs>\n",
       "       <path id=\"m0f50121b17\" d=\"M 0 0 \n",
       "L -3.5 0 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"262.188155\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_9\">\n",
       "      <!-- 1000 -->\n",
       "      <g transform=\"translate(20.878125 265.987374)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"221.706129\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 1500 -->\n",
       "      <g transform=\"translate(20.878125 225.505348)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"181.224103\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 2000 -->\n",
       "      <g transform=\"translate(20.878125 185.023322)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"140.742077\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 2500 -->\n",
       "      <g transform=\"translate(20.878125 144.541296)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"100.260052\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 3000 -->\n",
       "      <g transform=\"translate(20.878125 104.05927)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-33\" d=\"M 2597 2516 \n",
       "Q 3050 2419 3304 2112 \n",
       "Q 3559 1806 3559 1356 \n",
       "Q 3559 666 3084 287 \n",
       "Q 2609 -91 1734 -91 \n",
       "Q 1441 -91 1130 -33 \n",
       "Q 819 25 488 141 \n",
       "L 488 750 \n",
       "Q 750 597 1062 519 \n",
       "Q 1375 441 1716 441 \n",
       "Q 2309 441 2620 675 \n",
       "Q 2931 909 2931 1356 \n",
       "Q 2931 1769 2642 2001 \n",
       "Q 2353 2234 1838 2234 \n",
       "L 1294 2234 \n",
       "L 1294 2753 \n",
       "L 1863 2753 \n",
       "Q 2328 2753 2575 2939 \n",
       "Q 2822 3125 2822 3475 \n",
       "Q 2822 3834 2567 4026 \n",
       "Q 2313 4219 1838 4219 \n",
       "Q 1578 4219 1281 4162 \n",
       "Q 984 4106 628 3988 \n",
       "L 628 4550 \n",
       "Q 988 4650 1302 4700 \n",
       "Q 1616 4750 1894 4750 \n",
       "Q 2613 4750 3031 4423 \n",
       "Q 3450 4097 3450 3541 \n",
       "Q 3450 3153 3228 2886 \n",
       "Q 3006 2619 2597 2516 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\">\n",
       "     <g id=\"line2d_12\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"59.778026\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_14\">\n",
       "      <!-- 3500 -->\n",
       "      <g transform=\"translate(20.878125 63.577245)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_7\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0f50121b17\" x=\"53.328125\" y=\"19.296\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_15\">\n",
       "      <!-- 4000 -->\n",
       "      <g transform=\"translate(20.878125 23.095219)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"190.869141\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_16\">\n",
       "     <!-- Branch $\\pi$ -->\n",
       "     <g transform=\"translate(14.798438 162.506)rotate(-90)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-42\" d=\"M 1259 2228 \n",
       "L 1259 519 \n",
       "L 2272 519 \n",
       "Q 2781 519 3026 730 \n",
       "Q 3272 941 3272 1375 \n",
       "Q 3272 1813 3026 2020 \n",
       "Q 2781 2228 2272 2228 \n",
       "L 1259 2228 \n",
       "z\n",
       "M 1259 4147 \n",
       "L 1259 2741 \n",
       "L 2194 2741 \n",
       "Q 2656 2741 2882 2914 \n",
       "Q 3109 3088 3109 3444 \n",
       "Q 3109 3797 2882 3972 \n",
       "Q 2656 4147 2194 4147 \n",
       "L 1259 4147 \n",
       "z\n",
       "M 628 4666 \n",
       "L 2241 4666 \n",
       "Q 2963 4666 3353 4366 \n",
       "Q 3744 4066 3744 3513 \n",
       "Q 3744 3084 3544 2831 \n",
       "Q 3344 2578 2956 2516 \n",
       "Q 3422 2416 3680 2098 \n",
       "Q 3938 1781 3938 1306 \n",
       "Q 3938 681 3513 340 \n",
       "Q 3088 0 2303 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-72\" d=\"M 2631 2963 \n",
       "Q 2534 3019 2420 3045 \n",
       "Q 2306 3072 2169 3072 \n",
       "Q 1681 3072 1420 2755 \n",
       "Q 1159 2438 1159 1844 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1341 3275 1631 3429 \n",
       "Q 1922 3584 2338 3584 \n",
       "Q 2397 3584 2469 3576 \n",
       "Q 2541 3569 2628 3553 \n",
       "L 2631 2963 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-61\" d=\"M 2194 1759 \n",
       "Q 1497 1759 1228 1600 \n",
       "Q 959 1441 959 1056 \n",
       "Q 959 750 1161 570 \n",
       "Q 1363 391 1709 391 \n",
       "Q 2188 391 2477 730 \n",
       "Q 2766 1069 2766 1631 \n",
       "L 2766 1759 \n",
       "L 2194 1759 \n",
       "z\n",
       "M 3341 1997 \n",
       "L 3341 0 \n",
       "L 2766 0 \n",
       "L 2766 531 \n",
       "Q 2569 213 2275 61 \n",
       "Q 1981 -91 1556 -91 \n",
       "Q 1019 -91 701 211 \n",
       "Q 384 513 384 1019 \n",
       "Q 384 1609 779 1909 \n",
       "Q 1175 2209 1959 2209 \n",
       "L 2766 2209 \n",
       "L 2766 2266 \n",
       "Q 2766 2663 2505 2880 \n",
       "Q 2244 3097 1772 3097 \n",
       "Q 1472 3097 1187 3025 \n",
       "Q 903 2953 641 2809 \n",
       "L 641 3341 \n",
       "Q 956 3463 1253 3523 \n",
       "Q 1550 3584 1831 3584 \n",
       "Q 2591 3584 2966 3190 \n",
       "Q 3341 2797 3341 1997 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-63\" d=\"M 3122 3366 \n",
       "L 3122 2828 \n",
       "Q 2878 2963 2633 3030 \n",
       "Q 2388 3097 2138 3097 \n",
       "Q 1578 3097 1268 2742 \n",
       "Q 959 2388 959 1747 \n",
       "Q 959 1106 1268 751 \n",
       "Q 1578 397 2138 397 \n",
       "Q 2388 397 2633 464 \n",
       "Q 2878 531 3122 666 \n",
       "L 3122 134 \n",
       "Q 2881 22 2623 -34 \n",
       "Q 2366 -91 2075 -91 \n",
       "Q 1284 -91 818 406 \n",
       "Q 353 903 353 1747 \n",
       "Q 353 2603 823 3093 \n",
       "Q 1294 3584 2113 3584 \n",
       "Q 2378 3584 2631 3529 \n",
       "Q 2884 3475 3122 3366 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-68\" d=\"M 3513 2113 \n",
       "L 3513 0 \n",
       "L 2938 0 \n",
       "L 2938 2094 \n",
       "Q 2938 2591 2744 2837 \n",
       "Q 2550 3084 2163 3084 \n",
       "Q 1697 3084 1428 2787 \n",
       "Q 1159 2491 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 4863 \n",
       "L 1159 4863 \n",
       "L 1159 2956 \n",
       "Q 1366 3272 1645 3428 \n",
       "Q 1925 3584 2291 3584 \n",
       "Q 2894 3584 3203 3211 \n",
       "Q 3513 2838 3513 2113 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-Oblique-3c0\" d=\"M 584 3500 \n",
       "L 3938 3500 \n",
       "L 3825 2925 \n",
       "L 3384 2925 \n",
       "L 2966 775 \n",
       "Q 2922 550 2981 450 \n",
       "Q 3038 353 3209 353 \n",
       "Q 3256 353 3325 363 \n",
       "Q 3397 369 3419 372 \n",
       "L 3338 -44 \n",
       "Q 3222 -84 3103 -103 \n",
       "Q 2981 -122 2866 -122 \n",
       "Q 2491 -122 2388 81 \n",
       "Q 2284 288 2391 838 \n",
       "L 2797 2925 \n",
       "L 1506 2925 \n",
       "L 938 0 \n",
       "L 350 0 \n",
       "L 919 2925 \n",
       "L 472 2925 \n",
       "L 584 3500 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-42\" transform=\"translate(0 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" transform=\"translate(68.603516 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" transform=\"translate(109.716797 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" transform=\"translate(170.996094 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" transform=\"translate(234.375 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-68\" transform=\"translate(289.355469 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" transform=\"translate(352.734375 0.015625)\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-Oblique-3c0\" transform=\"translate(384.521484 0.015625)\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"line2d_14\">\n",
       "    <path d=\"M 69.560852 49.579994 \n",
       "L 86.647934 47.506533 \n",
       "L 103.735015 62.340583 \n",
       "L 120.822096 78.16267 \n",
       "L 137.909178 95.826118 \n",
       "L 154.996259 104.517865 \n",
       "L 172.08334 130.976376 \n",
       "L 189.170422 150.02897 \n",
       "L 206.257503 181.139052 \n",
       "L 223.344584 252.800164 \n",
       "L 240.431666 261.216 \n",
       "L 257.518747 179.993625 \n",
       "L 274.605828 131.726639 \n",
       "L 291.69291 109.849384 \n",
       "L 308.779991 80.711797 \n",
       "L 325.867072 73.324219 \n",
       "L 342.954154 61.201073 \n",
       "L 360.041235 43.248984 \n",
       "L 377.128316 39.571725 \n",
       "L 394.215398 46.645479 \n",
       "\" clip-path=\"url(#p580de37101)\" style=\"fill: none; stroke: #1f77b4; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_15\">\n",
       "    <path d=\"M 53.328125 19.296 \n",
       "L 410.448125 19.296 \n",
       "\" clip-path=\"url(#p580de37101)\" style=\"fill: none; stroke-dasharray: 1.5,2.475; stroke-dashoffset: 0; stroke: #1f77b4; stroke-width: 1.5\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 53.328125 273.312 \n",
       "L 53.328125 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 410.448125 273.312 \n",
       "L 410.448125 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 53.328125 273.312 \n",
       "L 410.448125 273.312 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 53.328125 7.2 \n",
       "L 410.448125 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"legend_1\">\n",
       "    <g id=\"patch_7\">\n",
       "     <path d=\"M 60.328125 268.312 \n",
       "L 191.515625 268.312 \n",
       "Q 193.515625 268.312 193.515625 266.312 \n",
       "L 193.515625 237.95575 \n",
       "Q 193.515625 235.95575 191.515625 235.95575 \n",
       "L 60.328125 235.95575 \n",
       "Q 58.328125 235.95575 58.328125 237.95575 \n",
       "L 58.328125 266.312 \n",
       "Q 58.328125 268.312 60.328125 268.312 \n",
       "z\n",
       "\" style=\"fill: #ffffff; opacity: 0.8; stroke: #cccccc; stroke-linejoin: miter\"/>\n",
       "    </g>\n",
       "    <g id=\"line2d_16\">\n",
       "     <path d=\"M 62.328125 244.054187 \n",
       "L 72.328125 244.054187 \n",
       "L 82.328125 244.054187 \n",
       "\" style=\"fill: none; stroke: #1f77b4; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "    </g>\n",
       "    <g id=\"text_17\">\n",
       "     <!-- Simulations -->\n",
       "     <g transform=\"translate(90.328125 247.554187)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-53\" d=\"M 3425 4513 \n",
       "L 3425 3897 \n",
       "Q 3066 4069 2747 4153 \n",
       "Q 2428 4238 2131 4238 \n",
       "Q 1616 4238 1336 4038 \n",
       "Q 1056 3838 1056 3469 \n",
       "Q 1056 3159 1242 3001 \n",
       "Q 1428 2844 1947 2747 \n",
       "L 2328 2669 \n",
       "Q 3034 2534 3370 2195 \n",
       "Q 3706 1856 3706 1288 \n",
       "Q 3706 609 3251 259 \n",
       "Q 2797 -91 1919 -91 \n",
       "Q 1588 -91 1214 -16 \n",
       "Q 841 59 441 206 \n",
       "L 441 856 \n",
       "Q 825 641 1194 531 \n",
       "Q 1563 422 1919 422 \n",
       "Q 2459 422 2753 634 \n",
       "Q 3047 847 3047 1241 \n",
       "Q 3047 1584 2836 1778 \n",
       "Q 2625 1972 2144 2069 \n",
       "L 1759 2144 \n",
       "Q 1053 2284 737 2584 \n",
       "Q 422 2884 422 3419 \n",
       "Q 422 4038 858 4394 \n",
       "Q 1294 4750 2059 4750 \n",
       "Q 2388 4750 2728 4690 \n",
       "Q 3069 4631 3425 4513 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6d\" d=\"M 3328 2828 \n",
       "Q 3544 3216 3844 3400 \n",
       "Q 4144 3584 4550 3584 \n",
       "Q 5097 3584 5394 3201 \n",
       "Q 5691 2819 5691 2113 \n",
       "L 5691 0 \n",
       "L 5113 0 \n",
       "L 5113 2094 \n",
       "Q 5113 2597 4934 2840 \n",
       "Q 4756 3084 4391 3084 \n",
       "Q 3944 3084 3684 2787 \n",
       "Q 3425 2491 3425 1978 \n",
       "L 3425 0 \n",
       "L 2847 0 \n",
       "L 2847 2094 \n",
       "Q 2847 2600 2669 2842 \n",
       "Q 2491 3084 2119 3084 \n",
       "Q 1678 3084 1418 2786 \n",
       "Q 1159 2488 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1356 3278 1631 3431 \n",
       "Q 1906 3584 2284 3584 \n",
       "Q 2666 3584 2933 3390 \n",
       "Q 3200 3197 3328 2828 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-75\" d=\"M 544 1381 \n",
       "L 544 3500 \n",
       "L 1119 3500 \n",
       "L 1119 1403 \n",
       "Q 1119 906 1312 657 \n",
       "Q 1506 409 1894 409 \n",
       "Q 2359 409 2629 706 \n",
       "Q 2900 1003 2900 1516 \n",
       "L 2900 3500 \n",
       "L 3475 3500 \n",
       "L 3475 0 \n",
       "L 2900 0 \n",
       "L 2900 538 \n",
       "Q 2691 219 2414 64 \n",
       "Q 2138 -91 1772 -91 \n",
       "Q 1169 -91 856 284 \n",
       "Q 544 659 544 1381 \n",
       "z\n",
       "M 1991 3584 \n",
       "L 1991 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6c\" d=\"M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-53\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"63.476562\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"91.259766\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-75\" x=\"188.671875\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"252.050781\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"279.833984\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"341.113281\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"380.322266\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"408.105469\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"469.287109\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"532.666016\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"line2d_17\">\n",
       "     <path d=\"M 62.328125 258.732312 \n",
       "L 72.328125 258.732312 \n",
       "L 82.328125 258.732312 \n",
       "\" style=\"fill: none; stroke-dasharray: 1.5,2.475; stroke-dashoffset: 0; stroke: #1f77b4; stroke-width: 1.5\"/>\n",
       "    </g>\n",
       "    <g id=\"text_18\">\n",
       "     <!-- Neutral expectation -->\n",
       "     <g transform=\"translate(90.328125 262.232312)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-4e\" d=\"M 628 4666 \n",
       "L 1478 4666 \n",
       "L 3547 763 \n",
       "L 3547 4666 \n",
       "L 4159 4666 \n",
       "L 4159 0 \n",
       "L 3309 0 \n",
       "L 1241 3903 \n",
       "L 1241 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-78\" d=\"M 3513 3500 \n",
       "L 2247 1797 \n",
       "L 3578 0 \n",
       "L 2900 0 \n",
       "L 1881 1375 \n",
       "L 863 0 \n",
       "L 184 0 \n",
       "L 1544 1831 \n",
       "L 300 3500 \n",
       "L 978 3500 \n",
       "L 1906 2253 \n",
       "L 2834 3500 \n",
       "L 3513 3500 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-4e\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"74.804688\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-75\" x=\"136.328125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"199.707031\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"238.916016\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"280.029297\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"341.308594\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"369.091797\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"400.878906\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-78\" x=\"460.652344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-70\" x=\"519.832031\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"583.308594\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"644.832031\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"699.8125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"739.021484\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"800.300781\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"839.509766\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"867.292969\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"928.474609\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p580de37101\">\n",
       "   <rect x=\"53.328125\" y=\"7.2\" width=\"357.12\" height=\"266.112\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_170_0.svg"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "\n",
    "plt.plot(mids, diversity.mean(axis=0), label=\"Simulations\")\n",
    "plt.axhline(4 * Ne, linestyle=\":\", label=r'Neutral expectation')\n",
    "plt.ylabel(r'Branch $\\pi$');\n",
    "plt.xlabel('Position (bp)')\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "323a1887",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "We use the \"branch\" measure of pairwise diversity which is\n",
    "defined in terms of the trees rather than nucleotide sequences. See\n",
    "[Ralph et al. 2020](https://doi.org/10.1534/genetics.120.303253)\n",
    "for more information.\n",
    ":::\n",
    "\n",
    "As we can see, the selective sweep has reduced variation in the region\n",
    "most closely linked to the beneficial allele and then diversity\n",
    "increases with distance to each side.\n",
    "\n",
    "(sec_ancestry_models_selective_sweeps_multiple)=\n",
    "\n",
    "#### Multiple sweeps\n",
    "\n",
    ":::{todo}\n",
    "This example is very artificial; it would be much better to illustrate\n",
    "with a meaningful example from the literature. Pull requests welcome!\n",
    ":::\n",
    "\n",
    "To illustrate the idea that we can simulate large numbers of different\n",
    "models, in this example we create a simulation in which we have\n",
    "sweeps happening at random points along the genome, which\n",
    "are separated by a random duration of the standard coalescent.\n",
    "\n",
    "First we set up our models:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "d58fd2bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[StandardCoalescent(duration=0.9664535356921388),\n",
       " SweepGenicSelection(duration=None, position=57, start_frequency=5e-07, end_frequency=0.9999995, s=0.01, dt=1e-06),\n",
       " StandardCoalescent(duration=0.11685051774599753),\n",
       " SweepGenicSelection(duration=None, position=12, start_frequency=5e-07, end_frequency=0.9999995, s=0.01, dt=1e-06),\n",
       " StandardCoalescent(duration=0.9109759624491242)]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "random.seed(1234)\n",
    "L = 100\n",
    "Ne = 1e6\n",
    "\n",
    "models = []\n",
    "for _ in range(100):\n",
    "    models.append(msprime.StandardCoalescent(duration=random.uniform(0, 1)))\n",
    "    models.append(\n",
    "        msprime.SweepGenicSelection(\n",
    "            position=random.randint(1, L - 1),\n",
    "            start_frequency=1 / (2 * Ne),\n",
    "            end_frequency=1.0 - 1 / (2 * Ne),\n",
    "            s=0.01,\n",
    "            dt=1e-6,\n",
    "        )\n",
    "    )\n",
    "models.append(msprime.StandardCoalescent())\n",
    "models[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe59f2f",
   "metadata": {},
   "source": [
    "We've created a list of models that have random parameters using the\n",
    "standard Python tools. There are lots of sweeps (many more than we\n",
    "will actually run, in all probability), each at a random position\n",
    "on the genome and each separated by a random amount of time running\n",
    "the standard coalescent. Finally, we add an instance of the standard\n",
    "coalescent without a {ref}`duration<sec_ancestry_models_specifying_duration>`\n",
    "to guarantee that our simulation will always coalesce fully\n",
    "(See the {ref}`sec_ancestry_models_specifying_completion` section\n",
    "for more information.) Finally, we print out the first 5 models\n",
    "so we can inspect them.\n",
    "\n",
    "Then, we can run our simulation as usual:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b2421464",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>100</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>100.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>20</td></tr>\n",
       "          <tr><td>Total Size</td><td>102.7 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>1588</td>\n",
       "          <td>49.6 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>10</td>\n",
       "          <td>304 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>523</td>\n",
       "          <td>14.3 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>25.7 KiB</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2bdf199a0>"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.sim_ancestry(\n",
    "    10,\n",
    "    population_size=100,\n",
    "    sequence_length=L,\n",
    "    recombination_rate=0.01,\n",
    "    model=models,\n",
    "    random_seed=6789\n",
    ")\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09e6650d",
   "metadata": {},
   "source": [
    ":::{tip}\n",
    "The logging output of msprime can be very useful when working with\n",
    "multiple models. See the {ref}`sec_logging` section for more\n",
    "details.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models_fixed_pedigree)=\n",
    "\n",
    "### Fixed pedigree\n",
    "\n",
    "A pedigree describes parent-offspring relationships between individuals, and\n",
    "these relationships constrain how genomes are inherited. We can simulate\n",
    "ancestry conditioned on a fixed pedigree using the {class}`.FixedPedigree`\n",
    "ancestry model. In this model the transfer of genetic material from parent to\n",
    "offspring is determined by the pedigree. Recombination occurs randomly such that\n",
    "each of a diploid individual's two genomes is a mixture of two corresponding\n",
    "parental genomes, but everything else is determined by the pedigree.\n",
    "\n",
    ":::{warning}\n",
    "It is important to note that there is no concept of a population genetic\n",
    "model \"outside\" of the pedigree in this model, and there\n",
    "are significant caveats when dealing with incomplete pedigree information\n",
    "and the treatment of founders.\n",
    "Please see the {ref}`sec_ancestry_models_fixed_pedigree_completing`\n",
    "section for more information.\n",
    ":::\n",
    "\n",
    "#### Simple example\n",
    "\n",
    "First we define a small pedigree, which follows a sample of three\n",
    "individuals through four generations:\n",
    "\n",
    "```{seealso}\n",
    "See the {ref}`sec_pedigrees` section for details on this pedigree\n",
    "text format and information on how pedigrees are encoded in msprime\n",
    "via the individual and node tables.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "17f4cc26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"371.52pt\" height=\"280.512pt\" viewBox=\"0 0 371.52 280.512\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-10-17T07:24:34.431334</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 280.512 \n",
       "L 371.52 280.512 \n",
       "L 371.52 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "L 7.2 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 333.330248 241.558255 \n",
       "Q 333.330248 222.727855 333.330248 205.015489 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 331.330248 209.015489 \n",
       "L 333.330248 205.015489 \n",
       "L 335.330248 209.015489 \n",
       "L 331.330248 209.015489 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 325.210462 247.194361 \n",
       "Q 259.543032 222.727948 194.92328 198.651879 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 197.9733 201.922564 \n",
       "L 194.92328 198.651879 \n",
       "L 199.36984 198.174274 \n",
       "L 197.9733 201.922564 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 325.210462 192.212543 \n",
       "Q 259.543032 167.74613 194.92328 143.670061 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 197.9733 146.940746 \n",
       "L 194.92328 143.670061 \n",
       "L 199.36984 143.192456 \n",
       "L 197.9733 146.940746 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 324.814155 193.651352 \n",
       "Q 185.759596 167.746834 47.804162 142.047071 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 51.37023 144.745804 \n",
       "L 47.804162 142.047071 \n",
       "L 52.102788 140.813456 \n",
       "L 51.37023 144.745804 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 185.76 186.576437 \n",
       "Q 185.76 167.746037 185.76 150.033671 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 183.76 154.033671 \n",
       "L 185.76 150.033671 \n",
       "L 187.76 154.033671 \n",
       "L 183.76 154.033671 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 177.640214 247.194361 \n",
       "Q 111.972784 222.727948 47.353032 198.651879 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 50.403052 201.922564 \n",
       "L 47.353032 198.651879 \n",
       "L 51.799592 198.174274 \n",
       "L 50.403052 201.922564 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 46.705845 193.651352 \n",
       "Q 185.760404 167.746834 323.715838 142.047071 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 319.417212 140.813456 \n",
       "L 323.715838 142.047071 \n",
       "L 320.14977 144.745804 \n",
       "L 319.417212 140.813456 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 38.189752 241.558255 \n",
       "Q 38.189752 222.727855 38.189752 205.015489 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 36.189752 209.015489 \n",
       "L 38.189752 205.015489 \n",
       "L 40.189752 209.015489 \n",
       "L 36.189752 209.015489 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 46.309538 247.194361 \n",
       "Q 111.976968 222.727948 176.59672 198.651879 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 172.15016 198.174274 \n",
       "L 176.59672 198.651879 \n",
       "L 173.5467 201.922564 \n",
       "L 172.15016 198.174274 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 324.923364 138.167839 \n",
       "Q 222.652136 112.764985 121.465971 87.631647 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 124.865886 90.536915 \n",
       "L 121.465971 87.631647 \n",
       "L 125.830134 86.654876 \n",
       "L 124.865886 90.536915 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 192.702125 135.082998 \n",
       "Q 222.650718 112.766465 251.702805 91.117974 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 247.300344 91.904313 \n",
       "L 251.702805 91.117974 \n",
       "L 249.690399 95.111747 \n",
       "L 247.300344 91.904313 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 46.596636 138.167839 \n",
       "Q 148.867864 112.764985 250.054029 87.631647 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 245.689866 86.654876 \n",
       "L 250.054029 87.631647 \n",
       "L 246.654114 90.536915 \n",
       "L 245.689866 86.654876 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 45.131877 135.082998 \n",
       "Q 75.08047 112.766465 104.132558 91.117974 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 99.730096 91.904313 \n",
       "L 104.132558 91.117974 \n",
       "L 102.120151 95.111747 \n",
       "L 99.730096 91.904313 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_16\">\n",
       "    <path d=\"M 259.545124 76.6128 \n",
       "Q 259.545124 57.782401 259.545124 40.070035 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 257.545124 44.070035 \n",
       "L 259.545124 40.070035 \n",
       "L 261.545124 44.070035 \n",
       "L 257.545124 44.070035 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_17\">\n",
       "    <path d=\"M 251.425338 82.248907 \n",
       "Q 185.757908 57.782493 121.138156 33.706425 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 124.188176 36.97711 \n",
       "L 121.138156 33.706425 \n",
       "L 125.584716 33.22882 \n",
       "L 124.188176 36.97711 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_18\">\n",
       "    <path d=\"M 120.094662 82.248907 \n",
       "Q 185.762092 57.782493 250.381844 33.706425 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 245.935284 33.22882 \n",
       "L 250.381844 33.706425 \n",
       "L 247.331824 36.97711 \n",
       "L 245.935284 33.22882 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_19\">\n",
       "    <path d=\"M 111.974876 76.6128 \n",
       "Q 111.974876 57.782401 111.974876 40.070035 \n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 109.974876 44.070035 \n",
       "L 111.974876 40.070035 \n",
       "L 113.974876 44.070035 \n",
       "L 109.974876 44.070035 \n",
       "z\n",
       "\" clip-path=\"url(#pc0c9c37939)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\"/>\n",
       "    <g id=\"ytick_2\"/>\n",
       "    <g id=\"ytick_3\"/>\n",
       "    <g id=\"ytick_4\"/>\n",
       "    <g id=\"ytick_5\"/>\n",
       "    <g id=\"ytick_6\"/>\n",
       "    <g id=\"ytick_7\"/>\n",
       "    <g id=\"ytick_8\"/>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_1\">\n",
       "    <defs>\n",
       "     <path id=\"C0_0_00f24c0979\" d=\"M 0 8.660254 \n",
       "C 2.296726 8.660254 4.499694 7.747755 6.123724 6.123724 \n",
       "C 7.747755 4.499694 8.660254 2.296726 8.660254 -0 \n",
       "C 8.660254 -2.296726 7.747755 -4.499694 6.123724 -6.123724 \n",
       "C 4.499694 -7.747755 2.296726 -8.660254 0 -8.660254 \n",
       "C -2.296726 -8.660254 -4.499694 -7.747755 -6.123724 -6.123724 \n",
       "C -7.747755 -4.499694 -8.660254 -2.296726 -8.660254 0 \n",
       "C -8.660254 2.296726 -7.747755 4.499694 -6.123724 6.123724 \n",
       "C -4.499694 7.747755 -2.296726 8.660254 0 8.660254 \n",
       "z\n",
       "\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"333.330248\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"333.330248\" y=\"195.237818\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"185.76\" y=\"195.237818\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"185.76\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"38.189752\" y=\"195.237818\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"38.189752\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"333.330248\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"185.76\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"38.189752\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"259.545124\" y=\"85.274182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"111.974876\" y=\"85.274182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"259.545124\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <use xlink:href=\"#C0_0_00f24c0979\" x=\"111.974876\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_20\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 7.2 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_21\">\n",
       "    <path d=\"M 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_22\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_23\">\n",
       "    <path d=\"M 7.2 7.2 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 0 -->\n",
       "     <g transform=\"translate(329.512748 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 5 -->\n",
       "     <g transform=\"translate(329.512748 198.549068)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 4 -->\n",
       "     <g transform=\"translate(181.9425 198.549068)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 1 -->\n",
       "     <g transform=\"translate(181.9425 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 3 -->\n",
       "     <g transform=\"translate(34.372252 198.549068)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-33\" d=\"M 2597 2516 \n",
       "Q 3050 2419 3304 2112 \n",
       "Q 3559 1806 3559 1356 \n",
       "Q 3559 666 3084 287 \n",
       "Q 2609 -91 1734 -91 \n",
       "Q 1441 -91 1130 -33 \n",
       "Q 819 25 488 141 \n",
       "L 488 750 \n",
       "Q 750 597 1062 519 \n",
       "Q 1375 441 1716 441 \n",
       "Q 2309 441 2620 675 \n",
       "Q 2931 909 2931 1356 \n",
       "Q 2931 1769 2642 2001 \n",
       "Q 2353 2234 1838 2234 \n",
       "L 1294 2234 \n",
       "L 1294 2753 \n",
       "L 1863 2753 \n",
       "Q 2328 2753 2575 2939 \n",
       "Q 2822 3125 2822 3475 \n",
       "Q 2822 3834 2567 4026 \n",
       "Q 2313 4219 1838 4219 \n",
       "Q 1578 4219 1281 4162 \n",
       "Q 984 4106 628 3988 \n",
       "L 628 4550 \n",
       "Q 988 4650 1302 4700 \n",
       "Q 1616 4750 1894 4750 \n",
       "Q 2613 4750 3031 4423 \n",
       "Q 3450 4097 3450 3541 \n",
       "Q 3450 3153 3228 2886 \n",
       "Q 3006 2619 2597 2516 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 2 -->\n",
       "     <g transform=\"translate(34.372252 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_7\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 7 -->\n",
       "     <g transform=\"translate(329.512748 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-37\" d=\"M 525 4666 \n",
       "L 3525 4666 \n",
       "L 3525 4397 \n",
       "L 1831 0 \n",
       "L 1172 0 \n",
       "L 2766 4134 \n",
       "L 525 4134 \n",
       "L 525 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-37\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_8\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 8 -->\n",
       "     <g transform=\"translate(181.9425 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 6 -->\n",
       "     <g transform=\"translate(34.372252 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_10\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 9 -->\n",
       "     <g transform=\"translate(255.727624 88.585432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-39\" d=\"M 703 97 \n",
       "L 703 672 \n",
       "Q 941 559 1184 500 \n",
       "Q 1428 441 1663 441 \n",
       "Q 2288 441 2617 861 \n",
       "Q 2947 1281 2994 2138 \n",
       "Q 2813 1869 2534 1725 \n",
       "Q 2256 1581 1919 1581 \n",
       "Q 1219 1581 811 2004 \n",
       "Q 403 2428 403 3163 \n",
       "Q 403 3881 828 4315 \n",
       "Q 1253 4750 1959 4750 \n",
       "Q 2769 4750 3195 4129 \n",
       "Q 3622 3509 3622 2328 \n",
       "Q 3622 1225 3098 567 \n",
       "Q 2575 -91 1691 -91 \n",
       "Q 1453 -91 1209 -44 \n",
       "Q 966 3 703 97 \n",
       "z\n",
       "M 1959 2075 \n",
       "Q 2384 2075 2632 2365 \n",
       "Q 2881 2656 2881 3163 \n",
       "Q 2881 3666 2632 3958 \n",
       "Q 2384 4250 1959 4250 \n",
       "Q 1534 4250 1286 3958 \n",
       "Q 1038 3666 1038 3163 \n",
       "Q 1038 2656 1286 2365 \n",
       "Q 1534 2075 1959 2075 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-39\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_11\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 10 -->\n",
       "     <g transform=\"translate(104.339876 88.585432)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 11 -->\n",
       "     <g transform=\"translate(251.910124 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_13\">\n",
       "    <g clip-path=\"url(#pc0c9c37939)\">\n",
       "     <!-- 12 -->\n",
       "     <g transform=\"translate(104.339876 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pc0c9c37939\">\n",
       "   <rect x=\"7.2\" y=\"7.2\" width=\"357.12\" height=\"266.112\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_176_0.svg"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import io\n",
    "ped_txt = \"\"\"\\\n",
    "# id parent0 parent1 time is_sample\n",
    "0   5   4   0.0 1\n",
    "1   3   3   0.0 1\n",
    "2   3   4   0.0 1\n",
    "3   7   7   1.0 0\n",
    "4   8   8   1.0 0\n",
    "5   8   6   1.0 0\n",
    "6   9   10  2.0 0\n",
    "7   10  10  2.0 0\n",
    "8   9   9   2.0 0\n",
    "9   11  12  3.0 0\n",
    "10  11  12  3.0 0\n",
    "11  .   .   4.0 0\n",
    "12  .   .   4.0 0\n",
    "\"\"\"\n",
    "pedigree = msprime.parse_pedigree(io.StringIO(ped_txt), sequence_length=100)\n",
    "\n",
    "draw_pedigree(pedigree.tree_sequence())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a41889e",
   "metadata": {},
   "source": [
    ":::{note} See the {ref}`sec_pedigrees_visualisation` section for a\n",
    "definition of the ``draw_pedigree`` function used here.\n",
    ":::\n",
    "\n",
    "We then use this pedigree information as the\n",
    "{ref}`initial state<sec_ancestry_initial_state>` for the simulation\n",
    "(note that we set the sequence length on the pedigree tables\n",
    "when we called {func}`.parse_pedigree`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "4e4f4a66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l261.6,0 l0,138.2 l-10.464,25 l0,5 l-251.136,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M318.4,0 l261.6,0 l0,138.2 l0,25 l0,5 l-272.064,0 l0,-5 l10.464,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(307.936 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">48</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 97.75)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 74.1)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 50.45)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c1 i11 node n22 p0 root\" transform=\"translate(66.1667 26.8)\"><g class=\"a22 c2 i9 node n18 p0\" transform=\"translate(0.0 23.65)\"><g class=\"a18 i0 leaf node n0 p0 sample\" transform=\"translate(-27.7 70.95)\"><path class=\"edge\" d=\"M 0 0 V -70.95 H 27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(0)</text></g><g class=\"a18 c2 i4 node n8 p0\" transform=\"translate(27.7 47.3)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(-18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(1)</text></g><g class=\"a8 i2 leaf node n5 p0 sample\" transform=\"translate(18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(5)</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H -27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4(8)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9(18)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11(22)</text></g><g class=\"c1 i11 node n23 p0 root\" transform=\"translate(195.433 26.8)\"><g class=\"a23 c2 i7 node n15 p0\" transform=\"translate(0.0 47.3)\"><g class=\"a15 i2 leaf node n4 p0 sample\" transform=\"translate(27.7 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(4)</text></g><g class=\"a15 c2 i3 node n6 p0\" transform=\"translate(-27.7 23.65)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(2)</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(3)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">3(6)</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7(15)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11(23)</text></g></g></g><g class=\"tree t1\" transform=\"translate(318.4 0)\"><g class=\"plotbox\"><g class=\"c1 i11 node n22 p0 root\" transform=\"translate(66.1667 26.8)\"><g class=\"a22 c2 i9 node n18 p0\" transform=\"translate(0.0 23.65)\"><g class=\"a18 i0 leaf node n0 p0 sample\" transform=\"translate(-27.7 70.95)\"><path class=\"edge\" d=\"M 0 0 V -70.95 H 27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(0)</text></g><g class=\"a18 c2 i8 node n16 p0\" transform=\"translate(27.7 23.65)\"><g class=\"a16 i0 leaf node n1 p0 sample\" transform=\"translate(-18.4667 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(1)</text></g><g class=\"a16 i2 leaf node n5 p0 sample\" transform=\"translate(18.4667 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(5)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H -27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8(16)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9(18)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11(22)</text></g><g class=\"c1 i11 node n23 p0 root\" transform=\"translate(195.433 26.8)\"><g class=\"a23 c2 i7 node n15 p0\" transform=\"translate(0.0 47.3)\"><g class=\"a15 i2 leaf node n4 p0 sample\" transform=\"translate(27.7 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(4)</text></g><g class=\"a15 c2 i3 node n6 p0\" transform=\"translate(-27.7 23.65)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(2)</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(3)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">3(6)</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7(15)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11(23)</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 93,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_178_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ped_ts = msprime.sim_ancestry(\n",
    "    initial_state=pedigree, model=\"fixed_pedigree\", random_seed=41,\n",
    "    recombination_rate=0.001)\n",
    "node_labels = {node.id: f\"{node.individual}({node.id})\" for node in ped_ts.nodes()}\n",
    "SVG(ped_ts.draw_svg(y_axis=True,  node_labels=node_labels, size=(600,200)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed4b7c40",
   "metadata": {},
   "source": [
    "The output tree sequence contains the simulated genetic ancestry conditioned\n",
    "on this input pedigree, as shown in the trees above. Each node is labelled\n",
    "with it the individual ID it's associated with, and its ID (as individual(node)).\n",
    "The simulation stops at time 4, and we are left with multiple\n",
    "roots, which correspond to the founder individuals. If you only\n",
    "need simulations of the genetic ancestry through the pedigree,\n",
    "then these trees can be used perfectly well in downstream\n",
    "applications (such as {ref}`sec_mutations`).\n",
    "See the {ref}`sec_ancestry_models_fixed_pedigree_completing`\n",
    "section for information (and caveats about) completing these simulations.\n",
    "\n",
    ":::{note}\n",
    "Because the input pedigree fully describes the simulation many features such as\n",
    "{ref}`demography<sec_demography>` are not available when we use the\n",
    "{class}`.FixedPedigree` model. The {class}`.FixedPedigree` model also cannot be\n",
    "combined with other {ref}`ancestry models<sec_ancestry_models>`.\n",
    ":::\n",
    "\n",
    "#### Censoring pedigree information\n",
    "\n",
    ":::{warning}\n",
    "The output tree sequence also contains the full pedigree information\n",
    "provided as input. This may be important if you are performing\n",
    "simulations on pedigrees that are not freely available.\n",
    ":::\n",
    "\n",
    "It is straightforward to remove the pedigree information from an\n",
    "output tree sequence, if required:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "4b5b1d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "  <style>\n",
       "    .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "    .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "    .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "    .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "    .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "    details {display: inline-block;}\n",
       "    summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "  </style>\n",
       "  <div class=\"tskit-table-set\">\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"padding:0;line-height:21px;\">\n",
       "              <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "              <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "            </th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "          <tr><td>Trees</td><td>2</td></tr>\n",
       "          <tr><td>Sequence Length</td><td>100.0</td></tr>\n",
       "          <tr><td>Time Units</td><td>generations</td></tr>\n",
       "          <tr><td>Sample Nodes</td><td>6</td></tr>\n",
       "          <tr><td>Total Size</td><td>2.5 KiB</td></tr>\n",
       "          <tr>\n",
       "            <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "    <div class=\"tskit-table-set-table\">\n",
       "      <table class=\"tskit-table\">\n",
       "        <thead>\n",
       "          <tr>\n",
       "            <th style=\"line-height:21px;\">Table</th>\n",
       "            <th>Rows</th>\n",
       "            <th>Size</th>\n",
       "            <th>Has Metadata</th>\n",
       "          </tr>\n",
       "        </thead>\n",
       "        <tbody>\n",
       "\n",
       "      <tr>\n",
       "        <td>Edges</td>\n",
       "          <td>13</td>\n",
       "          <td>424 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Individuals</td>\n",
       "          <td>0</td>\n",
       "          <td>40 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Migrations</td>\n",
       "          <td>0</td>\n",
       "          <td>8 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Mutations</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Nodes</td>\n",
       "          <td>26</td>\n",
       "          <td>736 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Populations</td>\n",
       "          <td>1</td>\n",
       "          <td>224 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "            ✅\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Provenances</td>\n",
       "          <td>1</td>\n",
       "          <td>1007 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "      <tr>\n",
       "        <td>Sites</td>\n",
       "          <td>0</td>\n",
       "          <td>16 Bytes</td>\n",
       "          <td style=\"text-align: center;\">\n",
       "\n",
       "          </td>\n",
       "        </tr>\n",
       "\n",
       "        </tbody>\n",
       "      </table>\n",
       "    </div>\n",
       "  </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fa2bbb45c40>"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables = ped_ts.dump_tables()\n",
    "tables.individuals.clear()\n",
    "tables.nodes.individual = np.full_like(tables.nodes.individual, tskit.NULL)\n",
    "censored_ts = tables.tree_sequence()\n",
    "censored_ts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09dac116",
   "metadata": {},
   "source": [
    "(sec_ancestry_models_fixed_pedigree_completing)=\n",
    "\n",
    "#### Completing fixed pedigree simulations\n",
    "\n",
    "If you wish to simulate ancestry older than what is defined in the\n",
    "pedigree, you can \"complete\" (or \"recapitate\") the within-pedigree\n",
    "simulation using the standard methods described in the\n",
    "{ref}`sec_ancestry_initial_state` section.\n",
    "\n",
    "Continuing the example in the previous section, we provide the simulated\n",
    "ancestry in ``ped_ts`` as the initial state for a {class}`.DiscreteTimeWrightFisher`\n",
    "simulation with a population size of 10:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "1df6ac25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l261.6,0 l0,138.2 l-10.464,25 l0,5 l-251.136,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M318.4,0 l261.6,0 l0,138.2 l0,25 l0,5 l-272.064,0 l0,-5 l10.464,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(307.936 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">48</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 105.633)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 89.8667)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 74.1)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3</text></g></g><g class=\"tick\" transform=\"translate(56.8 58.3333)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">6</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n26 p0 root\" transform=\"translate(130.8 26.8)\"><g class=\"a26 c1 i11 node n22 p0\" transform=\"translate(-64.6333 31.5333)\"><g class=\"a22 c2 i9 node n18 p0\" transform=\"translate(0.0 15.7667)\"><g class=\"a18 i0 leaf node n0 p0 sample\" transform=\"translate(-27.7 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a18 c2 i4 node n8 p0\" transform=\"translate(27.7 31.5333)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(-18.4667 15.7667)\"><path class=\"edge\" d=\"M 0 0 V -15.7667 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a8 i2 leaf node n5 p0 sample\" transform=\"translate(18.4667 15.7667)\"><path class=\"edge\" d=\"M 0 0 V -15.7667 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H -27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8</text></g><path class=\"edge\" d=\"M 0 0 V -15.7667 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H 64.6333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">22</text></g><g class=\"a26 c1 i11 node n23 p0\" transform=\"translate(64.6333 31.5333)\"><g class=\"a23 c2 i7 node n15 p0\" transform=\"translate(0.0 31.5333)\"><g class=\"a15 i2 leaf node n4 p0 sample\" transform=\"translate(27.7 31.5333)\"><path class=\"edge\" d=\"M 0 0 V -31.5333 H -27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a15 c2 i3 node n6 p0\" transform=\"translate(-27.7 15.7667)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-18.4667 15.7667)\"><path class=\"edge\" d=\"M 0 0 V -15.7667 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(18.4667 15.7667)\"><path class=\"edge\" d=\"M 0 0 V -15.7667 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -15.7667 H 27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H -64.6333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">23</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">26</text></g></g></g><g class=\"tree t1\" transform=\"translate(318.4 0)\"><g class=\"plotbox\"><g class=\"c2 node n26 p0 root\" transform=\"translate(130.8 26.8)\"><g class=\"a26 c1 i11 node n22 p0\" transform=\"translate(-64.6333 31.5333)\"><g class=\"a22 c2 i9 node n18 p0\" transform=\"translate(0.0 15.7667)\"><g class=\"a18 i0 leaf node n0 p0 sample\" transform=\"translate(-27.7 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0</text></g><g class=\"a18 c2 i8 node n16 p0\" transform=\"translate(27.7 15.7667)\"><g class=\"a16 i0 leaf node n1 p0 sample\" transform=\"translate(-18.4667 31.5333)\"><path class=\"edge\" d=\"M 0 0 V -31.5333 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1</text></g><g class=\"a16 i2 leaf node n5 p0 sample\" transform=\"translate(18.4667 31.5333)\"><path class=\"edge\" d=\"M 0 0 V -31.5333 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5</text></g><path class=\"edge\" d=\"M 0 0 V -15.7667 H -27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">16</text></g><path class=\"edge\" d=\"M 0 0 V -15.7667 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">18</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H 64.6333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">22</text></g><g class=\"a26 c1 i11 node n23 p0\" transform=\"translate(64.6333 31.5333)\"><g class=\"a23 c2 i7 node n15 p0\" transform=\"translate(0.0 31.5333)\"><g class=\"a15 i2 leaf node n4 p0 sample\" transform=\"translate(27.7 31.5333)\"><path class=\"edge\" d=\"M 0 0 V -31.5333 H -27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4</text></g><g class=\"a15 c2 i3 node n6 p0\" transform=\"translate(-27.7 15.7667)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-18.4667 15.7667)\"><path class=\"edge\" d=\"M 0 0 V -15.7667 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(18.4667 15.7667)\"><path class=\"edge\" d=\"M 0 0 V -15.7667 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3</text></g><path class=\"edge\" d=\"M 0 0 V -15.7667 H 27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">6</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">15</text></g><path class=\"edge\" d=\"M 0 0 V -31.5333 H -64.6333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">23</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">26</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 95,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_182_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_ts = msprime.sim_ancestry(\n",
    "    initial_state=ped_ts, model=\"dtwf\", population_size=10, random_seed=42)\n",
    "SVG(final_ts.draw_svg(y_axis=True, size=(600,200)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f94f004b",
   "metadata": {},
   "source": [
    "We can see that the nodes 22 and 23 representing the founder genomes\n",
    "from the input pedigree are used as the starting point for the DTWF simulation,\n",
    "which then finishes at generation 6 when the MRCA (node 26) is found.\n",
    "\n",
    "(sec_ancestry_models_fixed_pedigree_missing_data)=\n",
    "\n",
    "#### Missing pedigree data\n",
    "\n",
    "The previous example is very straightforward because the pedigree is\n",
    "complete and we can trace all individuals back to the same\n",
    "generation in the pedigree. The pedigrees we have in practice\n",
    "are usually not like this, and there is a great deal of missing data.\n",
    "The consequence of this for the {class}`.FixedPedigree` simulations\n",
    "is that genetic material will often be at a \"dead end\" at many\n",
    "levels of the pedigree during a simulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "3575871c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"371.52pt\" height=\"280.512pt\" viewBox=\"0 0 371.52 280.512\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-10-17T07:24:34.793939</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 280.512 \n",
       "L 371.52 280.512 \n",
       "L 371.52 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "L 7.2 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 333.330248 241.558255 \n",
       "Q 333.330248 222.727855 333.330248 205.015489 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 331.330248 209.015489 \n",
       "L 333.330248 205.015489 \n",
       "L 335.330248 209.015489 \n",
       "L 331.330248 209.015489 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 325.210462 247.194361 \n",
       "Q 259.543032 222.727948 194.92328 198.651879 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 197.9733 201.922564 \n",
       "L 194.92328 198.651879 \n",
       "L 199.36984 198.174274 \n",
       "L 197.9733 201.922564 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 325.210462 192.212543 \n",
       "Q 259.543032 167.74613 194.92328 143.670061 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 197.9733 146.940746 \n",
       "L 194.92328 143.670061 \n",
       "L 199.36984 143.192456 \n",
       "L 197.9733 146.940746 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 324.814155 193.651352 \n",
       "Q 185.759596 167.746834 47.804162 142.047071 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 51.37023 144.745804 \n",
       "L 47.804162 142.047071 \n",
       "L 52.102788 140.813456 \n",
       "L 51.37023 144.745804 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 185.76 186.576437 \n",
       "Q 185.76 167.746037 185.76 150.033671 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 183.76 154.033671 \n",
       "L 185.76 150.033671 \n",
       "L 187.76 154.033671 \n",
       "L 183.76 154.033671 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 177.640214 247.194361 \n",
       "Q 111.972784 222.727948 47.353032 198.651879 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 50.403052 201.922564 \n",
       "L 47.353032 198.651879 \n",
       "L 51.799592 198.174274 \n",
       "L 50.403052 201.922564 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 46.705845 193.651352 \n",
       "Q 185.760404 167.746834 323.715838 142.047071 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 319.417212 140.813456 \n",
       "L 323.715838 142.047071 \n",
       "L 320.14977 144.745804 \n",
       "L 319.417212 140.813456 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 38.189752 241.558255 \n",
       "Q 38.189752 222.727855 38.189752 205.015489 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 36.189752 209.015489 \n",
       "L 38.189752 205.015489 \n",
       "L 40.189752 209.015489 \n",
       "L 36.189752 209.015489 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 46.309538 247.194361 \n",
       "Q 111.976968 222.727948 176.59672 198.651879 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 172.15016 198.174274 \n",
       "L 176.59672 198.651879 \n",
       "L 173.5467 201.922564 \n",
       "L 172.15016 198.174274 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 192.702125 135.082998 \n",
       "Q 222.650718 112.766465 251.702805 91.117974 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 247.300344 91.904313 \n",
       "L 251.702805 91.117974 \n",
       "L 249.690399 95.111747 \n",
       "L 247.300344 91.904313 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 46.596636 138.167839 \n",
       "Q 148.867864 112.764985 250.054029 87.631647 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 245.689866 86.654876 \n",
       "L 250.054029 87.631647 \n",
       "L 246.654114 90.536915 \n",
       "L 245.689866 86.654876 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 45.131877 135.082998 \n",
       "Q 75.08047 112.766465 104.132558 91.117974 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 99.730096 91.904313 \n",
       "L 104.132558 91.117974 \n",
       "L 102.120151 95.111747 \n",
       "L 99.730096 91.904313 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 259.545124 76.6128 \n",
       "Q 259.545124 57.782401 259.545124 40.070035 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 257.545124 44.070035 \n",
       "L 259.545124 40.070035 \n",
       "L 261.545124 44.070035 \n",
       "L 257.545124 44.070035 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_16\">\n",
       "    <path d=\"M 251.425338 82.248907 \n",
       "Q 185.757908 57.782493 121.138156 33.706425 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 124.188176 36.97711 \n",
       "L 121.138156 33.706425 \n",
       "L 125.584716 33.22882 \n",
       "L 124.188176 36.97711 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_17\">\n",
       "    <path d=\"M 120.094662 82.248907 \n",
       "Q 185.762092 57.782493 250.381844 33.706425 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 245.935284 33.22882 \n",
       "L 250.381844 33.706425 \n",
       "L 247.331824 36.97711 \n",
       "L 245.935284 33.22882 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_18\">\n",
       "    <path d=\"M 111.974876 76.6128 \n",
       "Q 111.974876 57.782401 111.974876 40.070035 \n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 109.974876 44.070035 \n",
       "L 111.974876 40.070035 \n",
       "L 113.974876 44.070035 \n",
       "L 109.974876 44.070035 \n",
       "z\n",
       "\" clip-path=\"url(#pacf297f015)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\"/>\n",
       "    <g id=\"ytick_2\"/>\n",
       "    <g id=\"ytick_3\"/>\n",
       "    <g id=\"ytick_4\"/>\n",
       "    <g id=\"ytick_5\"/>\n",
       "    <g id=\"ytick_6\"/>\n",
       "    <g id=\"ytick_7\"/>\n",
       "    <g id=\"ytick_8\"/>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_1\">\n",
       "    <defs>\n",
       "     <path id=\"C0_0_b88d1be8a2\" d=\"M 0 8.660254 \n",
       "C 2.296726 8.660254 4.499694 7.747755 6.123724 6.123724 \n",
       "C 7.747755 4.499694 8.660254 2.296726 8.660254 -0 \n",
       "C 8.660254 -2.296726 7.747755 -4.499694 6.123724 -6.123724 \n",
       "C 4.499694 -7.747755 2.296726 -8.660254 0 -8.660254 \n",
       "C -2.296726 -8.660254 -4.499694 -7.747755 -6.123724 -6.123724 \n",
       "C -7.747755 -4.499694 -8.660254 -2.296726 -8.660254 0 \n",
       "C -8.660254 2.296726 -7.747755 4.499694 -6.123724 6.123724 \n",
       "C -4.499694 7.747755 -2.296726 8.660254 0 8.660254 \n",
       "z\n",
       "\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"333.330248\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"333.330248\" y=\"195.237818\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"185.76\" y=\"195.237818\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"185.76\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"38.189752\" y=\"195.237818\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"38.189752\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"333.330248\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"185.76\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"38.189752\" y=\"140.256\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"259.545124\" y=\"85.274182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"111.974876\" y=\"85.274182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"259.545124\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <use xlink:href=\"#C0_0_b88d1be8a2\" x=\"111.974876\" y=\"30.292364\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_19\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 7.2 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_20\">\n",
       "    <path d=\"M 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_21\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_22\">\n",
       "    <path d=\"M 7.2 7.2 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 0 -->\n",
       "     <g transform=\"translate(329.512748 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 5 -->\n",
       "     <g transform=\"translate(329.512748 198.549068)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 4 -->\n",
       "     <g transform=\"translate(181.9425 198.549068)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 1 -->\n",
       "     <g transform=\"translate(181.9425 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 3 -->\n",
       "     <g transform=\"translate(34.372252 198.549068)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-33\" d=\"M 2597 2516 \n",
       "Q 3050 2419 3304 2112 \n",
       "Q 3559 1806 3559 1356 \n",
       "Q 3559 666 3084 287 \n",
       "Q 2609 -91 1734 -91 \n",
       "Q 1441 -91 1130 -33 \n",
       "Q 819 25 488 141 \n",
       "L 488 750 \n",
       "Q 750 597 1062 519 \n",
       "Q 1375 441 1716 441 \n",
       "Q 2309 441 2620 675 \n",
       "Q 2931 909 2931 1356 \n",
       "Q 2931 1769 2642 2001 \n",
       "Q 2353 2234 1838 2234 \n",
       "L 1294 2234 \n",
       "L 1294 2753 \n",
       "L 1863 2753 \n",
       "Q 2328 2753 2575 2939 \n",
       "Q 2822 3125 2822 3475 \n",
       "Q 2822 3834 2567 4026 \n",
       "Q 2313 4219 1838 4219 \n",
       "Q 1578 4219 1281 4162 \n",
       "Q 984 4106 628 3988 \n",
       "L 628 4550 \n",
       "Q 988 4650 1302 4700 \n",
       "Q 1616 4750 1894 4750 \n",
       "Q 2613 4750 3031 4423 \n",
       "Q 3450 4097 3450 3541 \n",
       "Q 3450 3153 3228 2886 \n",
       "Q 3006 2619 2597 2516 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 2 -->\n",
       "     <g transform=\"translate(34.372252 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_7\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 7 -->\n",
       "     <g transform=\"translate(329.512748 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-37\" d=\"M 525 4666 \n",
       "L 3525 4666 \n",
       "L 3525 4397 \n",
       "L 1831 0 \n",
       "L 1172 0 \n",
       "L 2766 4134 \n",
       "L 525 4134 \n",
       "L 525 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-37\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_8\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 8 -->\n",
       "     <g transform=\"translate(181.9425 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 6 -->\n",
       "     <g transform=\"translate(34.372252 143.56725)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_10\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 9 -->\n",
       "     <g transform=\"translate(255.727624 88.585432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-39\" d=\"M 703 97 \n",
       "L 703 672 \n",
       "Q 941 559 1184 500 \n",
       "Q 1428 441 1663 441 \n",
       "Q 2288 441 2617 861 \n",
       "Q 2947 1281 2994 2138 \n",
       "Q 2813 1869 2534 1725 \n",
       "Q 2256 1581 1919 1581 \n",
       "Q 1219 1581 811 2004 \n",
       "Q 403 2428 403 3163 \n",
       "Q 403 3881 828 4315 \n",
       "Q 1253 4750 1959 4750 \n",
       "Q 2769 4750 3195 4129 \n",
       "Q 3622 3509 3622 2328 \n",
       "Q 3622 1225 3098 567 \n",
       "Q 2575 -91 1691 -91 \n",
       "Q 1453 -91 1209 -44 \n",
       "Q 966 3 703 97 \n",
       "z\n",
       "M 1959 2075 \n",
       "Q 2384 2075 2632 2365 \n",
       "Q 2881 2656 2881 3163 \n",
       "Q 2881 3666 2632 3958 \n",
       "Q 2384 4250 1959 4250 \n",
       "Q 1534 4250 1286 3958 \n",
       "Q 1038 3666 1038 3163 \n",
       "Q 1038 2656 1286 2365 \n",
       "Q 1534 2075 1959 2075 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-39\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_11\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 10 -->\n",
       "     <g transform=\"translate(104.339876 88.585432)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 11 -->\n",
       "     <g transform=\"translate(251.910124 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_13\">\n",
       "    <g clip-path=\"url(#pacf297f015)\">\n",
       "     <!-- 12 -->\n",
       "     <g transform=\"translate(104.339876 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pacf297f015\">\n",
       "   <rect x=\"7.2\" y=\"7.2\" width=\"357.12\" height=\"266.112\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_184_0.svg"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ped_txt = \"\"\"\\\n",
    "# id parent0 parent1 time is_sample\n",
    "0   5   4   0.0 1\n",
    "1   3   3   0.0 1\n",
    "2   3   4   0.0 1\n",
    "3   7   7   1.0 0\n",
    "4   8   8   1.0 0\n",
    "5   8   6   1.0 0\n",
    "6   9   10  2.0 0\n",
    "7   .   .   2.0 0\n",
    "8   9   9   2.0 0\n",
    "9   11  12  3.0 0\n",
    "10  11  12  3.0 0\n",
    "11  .   .   4.0 0\n",
    "12  .   .   4.0 0\n",
    "\"\"\"\n",
    "pedigree = msprime.parse_pedigree(io.StringIO(ped_txt), sequence_length=100)\n",
    "\n",
    "draw_pedigree(pedigree.tree_sequence())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25ae13d7",
   "metadata": {},
   "source": [
    "This pedigree has a \"dead end\" for individual 7, which is\n",
    "reflected in the trees that we output from the {class}`.FixedPedigree` simulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "2abb26e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l261.6,0 l0,138.2 l-10.464,25 l0,5 l-251.136,0 l0,-5 l0,-25 l0,-138.2z\"/><path d=\"M318.4,0 l261.6,0 l0,138.2 l0,25 l0,5 l-272.064,0 l0,-5 l10.464,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(307.936 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">48</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 97.75)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 74.1)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 50.45)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 i7 node n15 p0 root\" transform=\"translate(84.6333 74.1)\"><g class=\"a15 i2 leaf node n4 p0 sample\" transform=\"translate(27.7 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(4)</text></g><g class=\"a15 c2 i3 node n6 p0\" transform=\"translate(-27.7 23.65)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(2)</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(3)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">3(6)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7(15)</text></g><g class=\"c1 i11 node n22 p0 root\" transform=\"translate(176.967 26.8)\"><g class=\"a22 c2 i9 node n18 p0\" transform=\"translate(0.0 23.65)\"><g class=\"a18 i0 leaf node n0 p0 sample\" transform=\"translate(-27.7 70.95)\"><path class=\"edge\" d=\"M 0 0 V -70.95 H 27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(0)</text></g><g class=\"a18 c2 i4 node n8 p0\" transform=\"translate(27.7 47.3)\"><g class=\"a8 i0 leaf node n1 p0 sample\" transform=\"translate(-18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(1)</text></g><g class=\"a8 i2 leaf node n5 p0 sample\" transform=\"translate(18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(5)</text></g><path class=\"edge\" d=\"M 0 0 V -47.3 H -27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">4(8)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9(18)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11(22)</text></g></g></g><g class=\"tree t1\" transform=\"translate(318.4 0)\"><g class=\"plotbox\"><g class=\"c2 i7 node n15 p0 root\" transform=\"translate(84.6333 74.1)\"><g class=\"a15 i2 leaf node n4 p0 sample\" transform=\"translate(27.7 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(4)</text></g><g class=\"a15 c2 i3 node n6 p0\" transform=\"translate(-27.7 23.65)\"><g class=\"a6 i1 leaf node n2 p0 sample\" transform=\"translate(-18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(2)</text></g><g class=\"a6 i1 leaf node n3 p0 sample\" transform=\"translate(18.4667 23.65)\"><path class=\"edge\" d=\"M 0 0 V -23.65 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(3)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">3(6)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">7(15)</text></g><g class=\"c1 i11 node n22 p0 root\" transform=\"translate(176.967 26.8)\"><g class=\"a22 c2 i9 node n18 p0\" transform=\"translate(0.0 23.65)\"><g class=\"a18 i0 leaf node n0 p0 sample\" transform=\"translate(-27.7 70.95)\"><path class=\"edge\" d=\"M 0 0 V -70.95 H 27.7\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(0)</text></g><g class=\"a18 c2 i8 node n16 p0\" transform=\"translate(27.7 23.65)\"><g class=\"a16 i0 leaf node n1 p0 sample\" transform=\"translate(-18.4667 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H 18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(1)</text></g><g class=\"a16 i2 leaf node n5 p0 sample\" transform=\"translate(18.4667 47.3)\"><path class=\"edge\" d=\"M 0 0 V -47.3 H -18.4667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(5)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H -27.7\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">8(16)</text></g><path class=\"edge\" d=\"M 0 0 V -23.65 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">9(18)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">11(22)</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 97,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_186_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ped_ts = msprime.sim_ancestry(\n",
    "    initial_state=pedigree, model=\"fixed_pedigree\", random_seed=41,\n",
    "    recombination_rate=0.001)\n",
    "node_labels = {node.id: f\"{node.individual}({node.id})\" for node in ped_ts.nodes()}\n",
    "SVG(ped_ts.draw_svg(y_axis=True,  node_labels=node_labels, size=(600,200)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b2293d8",
   "metadata": {},
   "source": [
    "When we provide this tree sequence as the\n",
    "{ref}`initial state<sec_ancestry_initial_state>` of another simulation, we\n",
    "then effectively start the simulation at the earlier time point corresponding to\n",
    "individual 7. Ancestral material in older parts of the pedigree is then\n",
    "brought into the simulation in the same manner as\n",
    "{ref}`ancient samples<sec_ancestry_samples_sampling_time>` as we proceed backwards\n",
    "in time.\n",
    "\n",
    ":::{warning}\n",
    "It is important to note that although ancestral material for pedigree\n",
    "founders is introduced into the recapitation simulation at the correct time,\n",
    "no account is taken of the number of lineages present in the pedigree\n",
    "when calculating population sizes. Thus, the pedigree must be seen as\n",
    "entirely **external** to the population model simulated during recapitation.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_models_fixed_pedigree_demography)=\n",
    "\n",
    "#### Pedigrees and demography\n",
    "\n",
    "For complex simulations in which our pedigree individuals (and\n",
    "founders, in particular) are drawn from different populations,\n",
    "we will want our follow-up recapitation simulations to use\n",
    "a {ref}`demographic model<sec_demography>`. In order to do this,\n",
    "we must define the demographic model *before* creating\n",
    "the input pedigree. For example, consider the following\n",
    "simple demography:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "9ed68062",
   "metadata": {},
   "outputs": [],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", initial_size=10)\n",
    "demography.add_population(name=\"B\", initial_size=20)\n",
    "demography.add_population(name=\"C\", initial_size=100)\n",
    "demography.add_population_split(time=10, derived=[\"A\", \"B\"], ancestral=\"C\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71c7fe3e",
   "metadata": {},
   "source": [
    "We define a pedigree in which the individuals belong\n",
    "to the two \"leaf\" populations, A and B:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "bc125f11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"371.52pt\" height=\"280.512pt\" viewBox=\"0 0 371.52 280.512\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-10-17T07:24:35.359659</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 280.512 \n",
       "L 371.52 280.512 \n",
       "L 371.52 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "L 7.2 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 333.330248 241.561611 \n",
       "Q 333.330248 228.229733 333.330248 216.015888 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 331.330248 220.015888 \n",
       "L 333.330248 216.015888 \n",
       "L 335.330248 220.015888 \n",
       "L 331.330248 220.015888 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 327.76619 243.585844 \n",
       "Q 314.885058 228.22821 302.722406 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 303.760561 218.077159 \n",
       "L 302.722406 213.72719 \n",
       "L 306.825274 215.506647 \n",
       "L 303.760561 218.077159 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 326.551375 200.846071 \n",
       "Q 305.66057 184.241251 285.645002 168.332103 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 287.531889 172.386687 \n",
       "L 285.645002 168.332103 \n",
       "L 290.020802 169.255344 \n",
       "L 287.531889 172.386687 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 324.957985 204.015982 \n",
       "Q 250.323759 184.241925 176.770278 164.754207 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 180.124649 167.711941 \n",
       "L 176.770278 164.754207 \n",
       "L 181.149088 163.84535 \n",
       "L 180.124649 167.711941 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 288.065423 204.015982 \n",
       "Q 213.431197 184.241925 139.877716 164.754207 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 143.232087 167.711941 \n",
       "L 139.877716 164.754207 \n",
       "L 144.256526 163.84535 \n",
       "L 143.232087 167.711941 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 289.658813 200.846071 \n",
       "Q 268.768008 184.241251 248.75244 168.332103 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 250.639327 172.386687 \n",
       "L 248.75244 168.332103 \n",
       "L 253.12824 169.255344 \n",
       "L 250.639327 172.386687 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 290.873628 243.585844 \n",
       "Q 277.992496 228.22821 265.829844 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 266.867999 218.077159 \n",
       "L 265.829844 213.72719 \n",
       "L 269.932712 215.506647 \n",
       "L 266.867999 218.077159 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 266.323997 200.846071 \n",
       "Q 287.214802 184.241251 307.23037 168.332103 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 302.85457 169.255344 \n",
       "L 307.23037 168.332103 \n",
       "L 305.343483 172.386687 \n",
       "L 302.85457 169.255344 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 262.894592 198.247321 \n",
       "Q 268.768749 184.2403 274.210516 171.264317 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 270.819178 174.179591 \n",
       "L 274.210516 171.264317 \n",
       "L 274.507932 175.726552 \n",
       "L 270.819178 174.179591 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 253.981066 243.585844 \n",
       "Q 241.099934 228.22821 228.937282 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 229.975437 218.077159 \n",
       "L 228.937282 213.72719 \n",
       "L 233.04015 215.506647 \n",
       "L 229.975437 218.077159 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 252.107616 245.785922 \n",
       "Q 222.653308 228.227354 194.159342 211.241273 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 196.57107 215.007381 \n",
       "L 194.159342 211.241273 \n",
       "L 198.619265 211.571556 \n",
       "L 196.57107 215.007381 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 230.468926 202.506533 \n",
       "Q 268.76715 184.241986 306.056225 166.458706 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 301.584866 166.375316 \n",
       "L 306.056225 166.458706 \n",
       "L 303.306698 169.985759 \n",
       "L 301.584866 166.375316 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 193.576364 202.506533 \n",
       "Q 231.874588 184.241986 269.163663 166.458706 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 264.692304 166.375316 \n",
       "L 269.163663 166.458706 \n",
       "L 266.414136 169.985759 \n",
       "L 264.692304 166.375316 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_16\">\n",
       "    <path d=\"M 217.088504 243.585844 \n",
       "Q 204.207372 228.22821 192.04472 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 193.082875 218.077159 \n",
       "L 192.04472 213.72719 \n",
       "L 196.147589 215.506647 \n",
       "L 193.082875 218.077159 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_17\">\n",
       "    <path d=\"M 180.195942 243.585844 \n",
       "Q 167.31481 228.22821 155.152158 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 156.190313 218.077159 \n",
       "L 155.152158 213.72719 \n",
       "L 159.255027 215.506647 \n",
       "L 156.190313 218.077159 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_18\">\n",
       "    <path d=\"M 194.059926 247.745723 \n",
       "Q 259.544773 228.227014 323.95817 209.027666 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 319.553537 208.253578 \n",
       "L 323.95817 209.027666 \n",
       "L 320.69612 212.086919 \n",
       "L 319.553537 208.253578 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_19\">\n",
       "    <path d=\"M 156.683802 202.506533 \n",
       "Q 194.982026 184.241986 232.271101 166.458706 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 227.799742 166.375316 \n",
       "L 232.271101 166.458706 \n",
       "L 229.521574 169.985759 \n",
       "L 227.799742 166.375316 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_20\">\n",
       "    <path d=\"M 155.646311 200.846071 \n",
       "Q 176.537116 184.241251 196.552684 168.332103 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 192.176884 169.255344 \n",
       "L 196.552684 168.332103 \n",
       "L 194.665797 172.386687 \n",
       "L 192.176884 169.255344 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_21\">\n",
       "    <path d=\"M 156.916294 247.020865 \n",
       "Q 204.207677 228.226354 250.460068 209.844759 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 246.004218 209.463445 \n",
       "L 250.460068 209.844759 \n",
       "L 247.481508 213.180651 \n",
       "L 246.004218 209.463445 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_22\">\n",
       "    <path d=\"M 111.974876 241.561611 \n",
       "Q 111.974876 228.229733 111.974876 216.015888 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 109.974876 220.015888 \n",
       "L 111.974876 216.015888 \n",
       "L 113.974876 220.015888 \n",
       "L 109.974876 220.015888 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_23\">\n",
       "    <path d=\"M 106.410818 243.585844 \n",
       "Q 93.529686 228.22821 81.367034 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 82.405189 218.077159 \n",
       "L 81.367034 213.72719 \n",
       "L 85.469903 215.506647 \n",
       "L 82.405189 218.077159 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_24\">\n",
       "    <path d=\"M 105.196003 200.846071 \n",
       "Q 84.305198 184.241251 64.28963 168.332103 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 66.176517 172.386687 \n",
       "L 64.28963 168.332103 \n",
       "L 68.66543 169.255344 \n",
       "L 66.176517 172.386687 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_25\">\n",
       "    <path d=\"M 71.732846 198.247321 \n",
       "Q 65.858689 184.2403 60.416922 171.264317 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 60.119506 175.726552 \n",
       "L 60.416922 171.264317 \n",
       "L 63.80826 174.179591 \n",
       "L 60.119506 175.726552 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_26\">\n",
       "    <path d=\"M 78.431782 198.247321 \n",
       "Q 84.305939 184.2403 89.747707 171.264317 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 86.356368 174.179591 \n",
       "L 89.747707 171.264317 \n",
       "L 90.045122 175.726552 \n",
       "L 86.356368 174.179591 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_27\">\n",
       "    <path d=\"M 75.082314 241.561611 \n",
       "Q 75.082314 228.229733 75.082314 216.015888 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 73.082314 220.015888 \n",
       "L 75.082314 216.015888 \n",
       "L 77.082314 220.015888 \n",
       "L 73.082314 220.015888 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_28\">\n",
       "    <path d=\"M 38.189752 241.561611 \n",
       "Q 38.189752 228.229733 38.189752 216.015888 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 36.189752 220.015888 \n",
       "L 38.189752 216.015888 \n",
       "L 40.189752 220.015888 \n",
       "L 36.189752 220.015888 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_29\">\n",
       "    <path d=\"M 43.75381 243.585844 \n",
       "Q 56.634942 228.22821 68.797594 213.72719 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 64.694726 215.506647 \n",
       "L 68.797594 213.72719 \n",
       "L 67.759439 218.077159 \n",
       "L 64.694726 215.506647 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_30\">\n",
       "    <path d=\"M 44.968625 200.846071 \n",
       "Q 65.85943 184.241251 85.874998 168.332103 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 81.499198 169.255344 \n",
       "L 85.874998 168.332103 \n",
       "L 83.988111 172.386687 \n",
       "L 81.499198 169.255344 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_31\">\n",
       "    <path d=\"M 101.577451 159.049956 \n",
       "Q 148.868834 140.255445 195.121225 121.87385 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 190.665375 121.492536 \n",
       "L 195.121225 121.87385 \n",
       "L 192.142665 125.209742 \n",
       "L 190.665375 121.492536 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_32\">\n",
       "    <path d=\"M 100.966103 157.815013 \n",
       "Q 130.420411 140.256445 158.914377 123.270364 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 154.454454 123.600647 \n",
       "L 158.914377 123.270364 \n",
       "L 156.502649 127.036472 \n",
       "L 154.454454 123.600647 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_33\">\n",
       "    <path d=\"M 64.935959 159.774814 \n",
       "Q 130.420806 140.256104 194.834203 121.056757 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 190.42957 120.282669 \n",
       "L 194.834203 121.056757 \n",
       "L 191.572153 124.11601 \n",
       "L 190.42957 120.282669 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_34\">\n",
       "    <path d=\"M 64.684889 159.049956 \n",
       "Q 111.976272 140.255445 158.228663 121.87385 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 153.772813 121.492536 \n",
       "L 158.228663 121.87385 \n",
       "L 155.250103 125.209742 \n",
       "L 153.772813 121.492536 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_35\">\n",
       "    <path d=\"M 207.555749 110.276412 \n",
       "Q 213.429906 96.269391 218.871673 83.293408 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 215.480335 86.208682 \n",
       "L 218.871673 83.293408 \n",
       "L 219.169089 87.755643 \n",
       "L 215.480335 86.208682 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_36\">\n",
       "    <path d=\"M 197.427408 112.875162 \n",
       "Q 176.536603 96.270342 156.521035 80.361193 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 158.407922 84.415778 \n",
       "L 156.521035 80.361193 \n",
       "L 160.896835 81.284435 \n",
       "L 158.407922 84.415778 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_37\">\n",
       "    <path d=\"M 174.092592 112.875162 \n",
       "Q 194.983397 96.270342 214.998965 80.361193 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 210.623165 81.284435 \n",
       "L 214.998965 80.361193 \n",
       "L 213.112078 84.415778 \n",
       "L 210.623165 81.284435 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_38\">\n",
       "    <path d=\"M 170.663187 110.276412 \n",
       "Q 176.537344 96.269391 181.979111 83.293408 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 178.587773 86.208682 \n",
       "L 181.979111 83.293408 \n",
       "L 182.276527 87.755643 \n",
       "L 178.587773 86.208682 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_39\">\n",
       "    <path d=\"M 215.873689 68.889707 \n",
       "Q 194.982884 52.284887 174.967316 36.375739 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 176.854203 40.430324 \n",
       "L 174.967316 36.375739 \n",
       "L 179.343116 37.29898 \n",
       "L 176.854203 40.430324 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_40\">\n",
       "    <path d=\"M 219.303094 66.290957 \n",
       "Q 213.428937 52.283936 207.987169 39.307953 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 207.689754 43.770189 \n",
       "L 207.987169 39.307953 \n",
       "L 211.378508 42.223227 \n",
       "L 207.689754 43.770189 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_41\">\n",
       "    <path d=\"M 189.109468 66.290957 \n",
       "Q 194.983625 52.283936 200.425392 39.307953 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 197.034054 42.223227 \n",
       "L 200.425392 39.307953 \n",
       "L 200.722808 43.770189 \n",
       "L 197.034054 42.223227 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_42\">\n",
       "    <path d=\"M 182.410532 66.290957 \n",
       "Q 176.536375 52.283936 171.094608 39.307953 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 170.797192 43.770189 \n",
       "L 171.094608 39.307953 \n",
       "L 174.485946 42.223227 \n",
       "L 170.797192 43.770189 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_43\">\n",
       "    <path d=\"M 155.646311 68.889707 \n",
       "Q 176.537116 52.284887 196.552684 36.375739 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 192.176884 37.29898 \n",
       "L 196.552684 36.375739 \n",
       "L 194.665797 40.430324 \n",
       "L 192.176884 37.29898 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_44\">\n",
       "    <path d=\"M 152.216906 66.290957 \n",
       "Q 158.091063 52.283936 163.532831 39.307953 \n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"fill: none; stroke: #000000; stroke-linecap: round\"/>\n",
       "    <path d=\"M 160.141492 42.223227 \n",
       "L 163.532831 39.307953 \n",
       "L 163.830246 43.770189 \n",
       "L 160.141492 42.223227 \n",
       "z\n",
       "\" clip-path=\"url(#p867621ae02)\" style=\"stroke: #000000; stroke-linecap: round\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\"/>\n",
       "    <g id=\"ytick_2\"/>\n",
       "    <g id=\"ytick_3\"/>\n",
       "    <g id=\"ytick_4\"/>\n",
       "    <g id=\"ytick_5\"/>\n",
       "    <g id=\"ytick_6\"/>\n",
       "    <g id=\"ytick_7\"/>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_1\">\n",
       "    <defs>\n",
       "     <path id=\"C0_0_766d4faf8b\" d=\"M 0 8.660254 \n",
       "C 2.296726 8.660254 4.499694 7.747755 6.123724 6.123724 \n",
       "C 7.747755 4.499694 8.660254 2.296726 8.660254 -0 \n",
       "C 8.660254 -2.296726 7.747755 -4.499694 6.123724 -6.123724 \n",
       "C 4.499694 -7.747755 2.296726 -8.660254 0 -8.660254 \n",
       "C -2.296726 -8.660254 -4.499694 -7.747755 -6.123724 -6.123724 \n",
       "C -7.747755 -4.499694 -8.660254 -2.296726 -8.660254 0 \n",
       "C -8.660254 2.296726 -7.747755 4.499694 -6.123724 6.123724 \n",
       "C -4.499694 7.747755 -2.296726 8.660254 0 8.660254 \n",
       "z\n",
       "\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"333.330248\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"333.330248\" y=\"206.234182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"296.437686\" y=\"206.234182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"296.437686\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"259.545124\" y=\"206.234182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"259.545124\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"222.652562\" y=\"206.234182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"185.76\" y=\"206.234182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"222.652562\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"185.76\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"148.867438\" y=\"206.234182\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"148.867438\" y=\"250.219636\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"314.883967\" y=\"162.248727\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"277.991405\" y=\"162.248727\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"241.098843\" y=\"162.248727\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"204.206281\" y=\"162.248727\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"167.313719\" y=\"162.248727\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"130.421157\" y=\"162.248727\" style=\"fill: #1f77b4; stroke: #1f77b4\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"111.974876\" y=\"250.219636\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"111.974876\" y=\"206.234182\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"75.082314\" y=\"206.234182\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"75.082314\" y=\"250.219636\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"38.189752\" y=\"250.219636\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"38.189752\" y=\"206.234182\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"93.528595\" y=\"162.248727\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"56.636033\" y=\"162.248727\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"204.206281\" y=\"118.263273\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"167.313719\" y=\"118.263273\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"222.652562\" y=\"74.277818\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"185.76\" y=\"74.277818\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"148.867438\" y=\"74.277818\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"204.206281\" y=\"30.292364\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <use xlink:href=\"#C0_0_766d4faf8b\" x=\"167.313719\" y=\"30.292364\" style=\"fill: #ff7f0e; stroke: #ff7f0e\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_45\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 7.2 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_46\">\n",
       "    <path d=\"M 364.32 273.312 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_47\">\n",
       "    <path d=\"M 7.2 273.312 \n",
       "L 364.32 273.312 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_48\">\n",
       "    <path d=\"M 7.2 7.2 \n",
       "L 364.32 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 0 -->\n",
       "     <g transform=\"translate(329.512748 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 9 -->\n",
       "     <g transform=\"translate(329.512748 209.545432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-39\" d=\"M 703 97 \n",
       "L 703 672 \n",
       "Q 941 559 1184 500 \n",
       "Q 1428 441 1663 441 \n",
       "Q 2288 441 2617 861 \n",
       "Q 2947 1281 2994 2138 \n",
       "Q 2813 1869 2534 1725 \n",
       "Q 2256 1581 1919 1581 \n",
       "Q 1219 1581 811 2004 \n",
       "Q 403 2428 403 3163 \n",
       "Q 403 3881 828 4315 \n",
       "Q 1253 4750 1959 4750 \n",
       "Q 2769 4750 3195 4129 \n",
       "Q 3622 3509 3622 2328 \n",
       "Q 3622 1225 3098 567 \n",
       "Q 2575 -91 1691 -91 \n",
       "Q 1453 -91 1209 -44 \n",
       "Q 966 3 703 97 \n",
       "z\n",
       "M 1959 2075 \n",
       "Q 2384 2075 2632 2365 \n",
       "Q 2881 2656 2881 3163 \n",
       "Q 2881 3666 2632 3958 \n",
       "Q 2384 4250 1959 4250 \n",
       "Q 1534 4250 1286 3958 \n",
       "Q 1038 3666 1038 3163 \n",
       "Q 1038 2656 1286 2365 \n",
       "Q 1534 2075 1959 2075 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-39\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 11 -->\n",
       "     <g transform=\"translate(288.802686 209.545432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 1 -->\n",
       "     <g transform=\"translate(292.620186 253.530886)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 10 -->\n",
       "     <g transform=\"translate(251.910124 209.545432)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 2 -->\n",
       "     <g transform=\"translate(255.727624 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_7\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 6 -->\n",
       "     <g transform=\"translate(218.835062 209.545432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-36\" d=\"M 2113 2584 \n",
       "Q 1688 2584 1439 2293 \n",
       "Q 1191 2003 1191 1497 \n",
       "Q 1191 994 1439 701 \n",
       "Q 1688 409 2113 409 \n",
       "Q 2538 409 2786 701 \n",
       "Q 3034 994 3034 1497 \n",
       "Q 3034 2003 2786 2293 \n",
       "Q 2538 2584 2113 2584 \n",
       "z\n",
       "M 3366 4563 \n",
       "L 3366 3988 \n",
       "Q 3128 4100 2886 4159 \n",
       "Q 2644 4219 2406 4219 \n",
       "Q 1781 4219 1451 3797 \n",
       "Q 1122 3375 1075 2522 \n",
       "Q 1259 2794 1537 2939 \n",
       "Q 1816 3084 2150 3084 \n",
       "Q 2853 3084 3261 2657 \n",
       "Q 3669 2231 3669 1497 \n",
       "Q 3669 778 3244 343 \n",
       "Q 2819 -91 2113 -91 \n",
       "Q 1303 -91 875 529 \n",
       "Q 447 1150 447 2328 \n",
       "Q 447 3434 972 4092 \n",
       "Q 1497 4750 2381 4750 \n",
       "Q 2619 4750 2861 4703 \n",
       "Q 3103 4656 3366 4563 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_8\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 7 -->\n",
       "     <g transform=\"translate(181.9425 209.545432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-37\" d=\"M 525 4666 \n",
       "L 3525 4666 \n",
       "L 3525 4397 \n",
       "L 1831 0 \n",
       "L 1172 0 \n",
       "L 2766 4134 \n",
       "L 525 4134 \n",
       "L 525 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-37\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 3 -->\n",
       "     <g transform=\"translate(218.835062 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-33\" d=\"M 2597 2516 \n",
       "Q 3050 2419 3304 2112 \n",
       "Q 3559 1806 3559 1356 \n",
       "Q 3559 666 3084 287 \n",
       "Q 2609 -91 1734 -91 \n",
       "Q 1441 -91 1130 -33 \n",
       "Q 819 25 488 141 \n",
       "L 488 750 \n",
       "Q 750 597 1062 519 \n",
       "Q 1375 441 1716 441 \n",
       "Q 2309 441 2620 675 \n",
       "Q 2931 909 2931 1356 \n",
       "Q 2931 1769 2642 2001 \n",
       "Q 2353 2234 1838 2234 \n",
       "L 1294 2234 \n",
       "L 1294 2753 \n",
       "L 1863 2753 \n",
       "Q 2328 2753 2575 2939 \n",
       "Q 2822 3125 2822 3475 \n",
       "Q 2822 3834 2567 4026 \n",
       "Q 2313 4219 1838 4219 \n",
       "Q 1578 4219 1281 4162 \n",
       "Q 984 4106 628 3988 \n",
       "L 628 4550 \n",
       "Q 988 4650 1302 4700 \n",
       "Q 1616 4750 1894 4750 \n",
       "Q 2613 4750 3031 4423 \n",
       "Q 3450 4097 3450 3541 \n",
       "Q 3450 3153 3228 2886 \n",
       "Q 3006 2619 2597 2516 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_10\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 4 -->\n",
       "     <g transform=\"translate(181.9425 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-34\" d=\"M 2419 4116 \n",
       "L 825 1625 \n",
       "L 2419 1625 \n",
       "L 2419 4116 \n",
       "z\n",
       "M 2253 4666 \n",
       "L 3047 4666 \n",
       "L 3047 1625 \n",
       "L 3713 1625 \n",
       "L 3713 1100 \n",
       "L 3047 1100 \n",
       "L 3047 0 \n",
       "L 2419 0 \n",
       "L 2419 1100 \n",
       "L 313 1100 \n",
       "L 313 1709 \n",
       "L 2253 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_11\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 8 -->\n",
       "     <g transform=\"translate(145.049938 209.545432)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-38\" d=\"M 2034 2216 \n",
       "Q 1584 2216 1326 1975 \n",
       "Q 1069 1734 1069 1313 \n",
       "Q 1069 891 1326 650 \n",
       "Q 1584 409 2034 409 \n",
       "Q 2484 409 2743 651 \n",
       "Q 3003 894 3003 1313 \n",
       "Q 3003 1734 2745 1975 \n",
       "Q 2488 2216 2034 2216 \n",
       "z\n",
       "M 1403 2484 \n",
       "Q 997 2584 770 2862 \n",
       "Q 544 3141 544 3541 \n",
       "Q 544 4100 942 4425 \n",
       "Q 1341 4750 2034 4750 \n",
       "Q 2731 4750 3128 4425 \n",
       "Q 3525 4100 3525 3541 \n",
       "Q 3525 3141 3298 2862 \n",
       "Q 3072 2584 2669 2484 \n",
       "Q 3125 2378 3379 2068 \n",
       "Q 3634 1759 3634 1313 \n",
       "Q 3634 634 3220 271 \n",
       "Q 2806 -91 2034 -91 \n",
       "Q 1263 -91 848 271 \n",
       "Q 434 634 434 1313 \n",
       "Q 434 1759 690 2068 \n",
       "Q 947 2378 1403 2484 \n",
       "z\n",
       "M 1172 3481 \n",
       "Q 1172 3119 1398 2916 \n",
       "Q 1625 2713 2034 2713 \n",
       "Q 2441 2713 2670 2916 \n",
       "Q 2900 3119 2900 3481 \n",
       "Q 2900 3844 2670 4047 \n",
       "Q 2441 4250 2034 4250 \n",
       "Q 1625 4250 1398 4047 \n",
       "Q 1172 3844 1172 3481 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 5 -->\n",
       "     <g transform=\"translate(145.049938 253.530886)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_13\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 14 -->\n",
       "     <g transform=\"translate(307.248967 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_14\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 12 -->\n",
       "     <g transform=\"translate(270.356405 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_15\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 16 -->\n",
       "     <g transform=\"translate(233.463843 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-36\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_16\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 17 -->\n",
       "     <g transform=\"translate(196.571281 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-37\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_17\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 13 -->\n",
       "     <g transform=\"translate(159.678719 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-33\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_18\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 15 -->\n",
       "     <g transform=\"translate(122.786157 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_19\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 18 -->\n",
       "     <g transform=\"translate(104.339876 253.530886)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-38\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_20\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 23 -->\n",
       "     <g transform=\"translate(104.339876 209.545432)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-33\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_21\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 22 -->\n",
       "     <g transform=\"translate(67.447314 209.545432)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_22\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 19 -->\n",
       "     <g transform=\"translate(67.447314 253.530886)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-39\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_23\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 20 -->\n",
       "     <g transform=\"translate(30.554752 253.530886)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_24\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 21 -->\n",
       "     <g transform=\"translate(30.554752 209.545432)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_25\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 25 -->\n",
       "     <g transform=\"translate(85.893595 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_26\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 24 -->\n",
       "     <g transform=\"translate(49.001033 165.559977)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_27\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 27 -->\n",
       "     <g transform=\"translate(196.571281 121.574523)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-37\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_28\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 26 -->\n",
       "     <g transform=\"translate(159.678719 121.574523)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-36\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_29\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 30 -->\n",
       "     <g transform=\"translate(215.017562 77.589068)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_30\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 28 -->\n",
       "     <g transform=\"translate(178.125 77.589068)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-38\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_31\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 29 -->\n",
       "     <g transform=\"translate(141.232438 77.589068)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-39\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_32\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 31 -->\n",
       "     <g transform=\"translate(196.571281 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_33\">\n",
       "    <g clip-path=\"url(#p867621ae02)\">\n",
       "     <!-- 32 -->\n",
       "     <g transform=\"translate(159.678719 33.603614)scale(0.12 -0.12)\">\n",
       "      <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"63.623047\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p867621ae02\">\n",
       "   <rect x=\"7.2\" y=\"7.2\" width=\"357.12\" height=\"266.112\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_190_0.svg"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ped_txt = \"\"\"\\\n",
    "# id    parent0 parent1 time    is_sample   population\n",
    "0   9   11  0.0 1   A\n",
    "1   10  10  0.0 1   A\n",
    "2   6   7   0.0 1   A\n",
    "3   7   7   0.0 1   A\n",
    "4   8   9   0.0 1   A\n",
    "5   10  10  0.0 1   A\n",
    "6   14  14  1.0 0   A\n",
    "7   12  12  1.0 0   A\n",
    "8   16  17  1.0 0   A\n",
    "9   12  13  1.0 0   A\n",
    "10  14  12  1.0 0   A\n",
    "11  15  16  1.0 0   A\n",
    "12  .   .   2.0 0   A\n",
    "13  .   .   2.0 0   A\n",
    "14  .   .   2.0 0   A\n",
    "15  .   .   2.0 0   A\n",
    "16  .   .   2.0 0   A\n",
    "17  .   .   2.0 0   A\n",
    "18  23  22  0.0 1   B\n",
    "19  22  22  0.0 1   B\n",
    "20  21  22  0.0 1   B\n",
    "21  25  25  1.0 0   B\n",
    "22  24  25  1.0 0   B\n",
    "23  24  24  1.0 0   B\n",
    "24  27  26  2.0 0   B\n",
    "25  27  26  2.0 0   B\n",
    "26  30  28  3.0 0   B\n",
    "27  30  29  3.0 0   B\n",
    "28  31  32  4.0 0   B\n",
    "29  31  32  4.0 0   B\n",
    "30  32  31  4.0 0   B\n",
    "31  .   .   5.0 0   B\n",
    "32  .   .   5.0 0   B\n",
    "\"\"\"\n",
    "pedigree = msprime.parse_pedigree(\n",
    "    io.StringIO(ped_txt), demography=demography, sequence_length=100)\n",
    "\n",
    "draw_pedigree(pedigree.tree_sequence())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a33b9798",
   "metadata": {},
   "source": [
    "We can see that the pedigree is split into two sub-pedigrees for the\n",
    "two populations, and the founders for the two populations exist\n",
    "at different times (2 and 5 generations ago for A and B, respectively).\n",
    "Running the {class}`.FixedPedigree` simulation, we can see that the\n",
    "trees reflect this structure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "e669c690",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l523.2,0 l0,138.2 l0,25 l0,5 l-523.2,0 l0,-5 l0,-25 l0,-138.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 200)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"163.2\" y2=\"163.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(580 163.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 65.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"121.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 121.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 102.48)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 83.56)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 45.72)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 i12 node n24 p0 root\" transform=\"translate(46.8444 83.56)\"><g class=\"a24 i0 leaf node n0 p0 sample\" transform=\"translate(-13.4222 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(0)</text></g><g class=\"a24 i3 leaf node n7 p0 sample\" transform=\"translate(13.4222 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3(7)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">12(24)</text></g><g class=\"c2 i12 node n25 p0 root\" transform=\"translate(127.378 83.56)\"><g class=\"a25 c2 i7 node n14 p0\" transform=\"translate(-26.8444 18.92)\"><g class=\"a14 i2 leaf node n5 p0 sample\" transform=\"translate(-13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(5)</text></g><g class=\"a14 i3 leaf node n6 p0 sample\" transform=\"translate(13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3(6)</text></g><path class=\"edge\" d=\"M 0 0 V -18.92 H 26.8444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">7(14)</text></g><g class=\"a25 c2 i10 node n21 p0\" transform=\"translate(26.8444 18.92)\"><g class=\"a21 i5 leaf node n10 p0 sample\" transform=\"translate(-13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5(10)</text></g><g class=\"a21 i5 leaf node n11 p0 sample\" transform=\"translate(13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5(11)</text></g><path class=\"edge\" d=\"M 0 0 V -18.92 H -26.8444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">10(21)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">12(25)</text></g><g class=\"c1 i13 node n26 p0 root\" transform=\"translate(194.489 83.56)\"><g class=\"a26 i4 leaf node n9 p0 sample\" transform=\"translate(0.0 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4(9)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">13(26)</text></g><g class=\"c2 i14 node n28 p0 root\" transform=\"translate(254.889 83.56)\"><g class=\"a28 i2 leaf node n4 p0 sample\" transform=\"translate(20.1333 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H -20.1333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2(4)</text></g><g class=\"a28 c2 i10 node n20 p0\" transform=\"translate(-20.1333 18.92)\"><g class=\"a20 i1 leaf node n2 p0 sample\" transform=\"translate(-13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(2)</text></g><g class=\"a20 i1 leaf node n3 p0 sample\" transform=\"translate(13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1(3)</text></g><path class=\"edge\" d=\"M 0 0 V -18.92 H 20.1333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">10(20)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">14(28)</text></g><g class=\"c2 i16 node n32 p0 root\" transform=\"translate(315.289 83.56)\"><g class=\"a32 i0 leaf node n1 p0 sample\" transform=\"translate(-13.4222 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0(1)</text></g><g class=\"a32 i4 leaf node n8 p0 sample\" transform=\"translate(13.4222 37.84)\"><path class=\"edge\" d=\"M 0 0 V -37.84 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4(8)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">16(32)</text></g><g class=\"c1 i31 node n63 p1 root\" transform=\"translate(355.556 26.8)\"><g class=\"a63 i20 leaf node n40 p1 sample\" transform=\"translate(0.0 94.6)\"><path class=\"edge\" d=\"M 0 0 V -94.6 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">20(40)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">31(63)</text></g><g class=\"c1 i32 node n64 p1 root\" transform=\"translate(415.956 26.8)\"><g class=\"a64 c2 i30 node n60 p1\" transform=\"translate(0.0 18.92)\"><g class=\"a60 i18 leaf node n36 p1 sample\" transform=\"translate(-33.5556 75.68)\"><path class=\"edge\" d=\"M 0 0 V -75.68 H 33.5556\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">18(36)</text></g><g class=\"a60 c4 i22 node n45 p1\" transform=\"translate(33.5556 56.76)\"><g class=\"a45 i18 leaf node n37 p1 sample\" transform=\"translate(-40.2667 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H 40.2667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">18(37)</text></g><g class=\"a45 i19 leaf node n38 p1 sample\" transform=\"translate(-13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">19(38)</text></g><g class=\"a45 i19 leaf node n39 p1 sample\" transform=\"translate(13.4222 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">19(39)</text></g><g class=\"a45 i20 leaf node n41 p1 sample\" transform=\"translate(40.2667 18.92)\"><path class=\"edge\" d=\"M 0 0 V -18.92 H -40.2667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">20(41)</text></g><path class=\"edge\" d=\"M 0 0 V -56.76 H -33.5556\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">22(45)</text></g><path class=\"edge\" d=\"M 0 0 V -18.92 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">30(60)</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">32(64)</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 100,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_192_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ped_ts = msprime.sim_ancestry(\n",
    "    initial_state=pedigree, model=\"fixed_pedigree\", random_seed=41)\n",
    "node_labels = {node.id: f\"{node.individual}({node.id})\" for node in ped_ts.nodes()}\n",
    "SVG(ped_ts.draw_svg(y_axis=True,  node_labels=node_labels, size=(600,200)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d613ec6",
   "metadata": {},
   "source": [
    "When we complete this simulation we must use a {class}`.Demography` object\n",
    "that matches the populations defined in the original pedigree tables. The\n",
    "simplest way to do this is define the demography first, and use the same\n",
    "object throughout as we have done here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "3aa87c44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" baseProfile=\"full\" height=\"300\" version=\"1.1\" width=\"600\"><defs><style type=\"text/css\"><![CDATA[.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}]]></style></defs><g class=\"tree-sequence\"><g class=\"background\"><path d=\"M56.8,0 l523.2,0 l0,238.2 l0,25 l0,5 l-523.2,0 l0,-5 l0,-25 l0,-238.2z\"/></g><g class=\"axes\"><g class=\"x-axis\"><g transform=\"translate(318.4 300)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(0 -11)\">Genome position</text></g><line x1=\"56.8\" x2=\"580\" y1=\"263.2\" y2=\"263.2\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 263.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">0</text></g></g><g class=\"tick\" transform=\"translate(580 263.2)\"><line x1=\"0\" x2=\"0\" y1=\"0\" y2=\"5\"/><g transform=\"translate(0 6)\"><text class=\"lab\">100</text></g></g></g></g><g class=\"y-axis\"><g transform=\"translate(0 115.7)\"><text class=\"lab\" text-anchor=\"middle\" transform=\"translate(11) rotate(-90)\">Node time (generations)</text></g><line x1=\"56.8\" x2=\"56.8\" y1=\"221.4\" y2=\"10\"/><g class=\"ticks\"><g class=\"tick\" transform=\"translate(56.8 221.4)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">0</text></g></g><g class=\"tick\" transform=\"translate(56.8 201.94)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">1</text></g></g><g class=\"tick\" transform=\"translate(56.8 182.48)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">2</text></g></g><g class=\"tick\" transform=\"translate(56.8 163.02)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">3</text></g></g><g class=\"tick\" transform=\"translate(56.8 143.56)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">4</text></g></g><g class=\"tick\" transform=\"translate(56.8 124.1)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">5</text></g></g><g class=\"tick\" transform=\"translate(56.8 104.64)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">6</text></g></g><g class=\"tick\" transform=\"translate(56.8 85.18)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">43</text></g></g><g class=\"tick\" transform=\"translate(56.8 65.72)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">52</text></g></g><g class=\"tick\" transform=\"translate(56.8 46.26)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">64</text></g></g><g class=\"tick\" transform=\"translate(56.8 26.8)\"><line x1=\"0\" x2=\"-5\" y1=\"0\" y2=\"0\"/><g transform=\"translate(-6 0)\"><text class=\"lab\" text-anchor=\"end\">80</text></g></g></g></g></g><g class=\"plotbox trees\"><g class=\"tree t0\" transform=\"translate(56.8 0)\"><g class=\"plotbox\"><g class=\"c2 node n71 p2 root\" transform=\"translate(301.867 26.8)\"><g class=\"a71 c2 node n67 p0\" transform=\"translate(134.222 77.84)\"><g class=\"a67 c1 i13 node n26 p0\" transform=\"translate(53.6889 77.84)\"><g class=\"a26 i4 leaf node n9 p0 sample\" transform=\"translate(0.0 38.92)\"><path class=\"edge\" d=\"M 0 0 V -38.92 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">9:A</text></g><path class=\"edge\" d=\"M 0 0 V -77.84 H -53.6889\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">26:A</text></g><g class=\"a67 c2 node n66 p0\" transform=\"translate(-53.6889 58.38)\"><g class=\"a66 c2 i12 node n25 p0\" transform=\"translate(40.2667 19.46)\"><g class=\"a25 c2 i7 node n14 p0\" transform=\"translate(-26.8444 19.46)\"><g class=\"a14 i2 leaf node n5 p0 sample\" transform=\"translate(-13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">5:A</text></g><g class=\"a14 i3 leaf node n6 p0 sample\" transform=\"translate(13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">6:A</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 26.8444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">14:A</text></g><g class=\"a25 c2 i10 node n21 p0\" transform=\"translate(26.8444 19.46)\"><g class=\"a21 i5 leaf node n10 p0 sample\" transform=\"translate(-13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">10:A</text></g><g class=\"a21 i5 leaf node n11 p0 sample\" transform=\"translate(13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">11:A</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H -26.8444\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">21:A</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H -40.2667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">25:A</text></g><g class=\"a66 c2 i16 node n32 p0\" transform=\"translate(-40.2667 19.46)\"><g class=\"a32 i0 leaf node n1 p0 sample\" transform=\"translate(-13.4222 38.92)\"><path class=\"edge\" d=\"M 0 0 V -38.92 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">1:A</text></g><g class=\"a32 i4 leaf node n8 p0 sample\" transform=\"translate(13.4222 38.92)\"><path class=\"edge\" d=\"M 0 0 V -38.92 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">8:A</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 40.2667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">32:A</text></g><path class=\"edge\" d=\"M 0 0 V -58.38 H 53.6889\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">66:A</text></g><path class=\"edge\" d=\"M 0 0 V -77.84 H -134.222\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">67:A</text></g><g class=\"a71 c2 node n70 p2\" transform=\"translate(-134.222 19.46)\"><g class=\"a70 c1 i32 node n64 p1\" transform=\"translate(60.4 77.84)\"><g class=\"a64 c2 i30 node n60 p1\" transform=\"translate(0.0 19.46)\"><g class=\"a60 i18 leaf node n36 p1 sample\" transform=\"translate(-33.5556 77.84)\"><path class=\"edge\" d=\"M 0 0 V -77.84 H 33.5556\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">36:B</text></g><g class=\"a60 c4 i22 node n45 p1\" transform=\"translate(33.5556 58.38)\"><g class=\"a45 i18 leaf node n37 p1 sample\" transform=\"translate(-40.2667 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H 40.2667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">37:B</text></g><g class=\"a45 i19 leaf node n38 p1 sample\" transform=\"translate(-13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">38:B</text></g><g class=\"a45 i19 leaf node n39 p1 sample\" transform=\"translate(13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">39:B</text></g><g class=\"a45 i20 leaf node n41 p1 sample\" transform=\"translate(40.2667 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H -40.2667\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">41:B</text></g><path class=\"edge\" d=\"M 0 0 V -58.38 H -33.5556\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">45:B</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 0.0\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">60:B</text></g><path class=\"edge\" d=\"M 0 0 V -77.84 H -60.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">64:B</text></g><g class=\"a70 c2 node n69 p2\" transform=\"translate(-60.4 19.46)\"><g class=\"a69 c2 i14 node n28 p0\" transform=\"translate(40.2667 116.76)\"><g class=\"a28 i2 leaf node n4 p0 sample\" transform=\"translate(20.1333 38.92)\"><path class=\"edge\" d=\"M 0 0 V -38.92 H -20.1333\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">4:A</text></g><g class=\"a28 c2 i10 node n20 p0\" transform=\"translate(-20.1333 19.46)\"><g class=\"a20 i1 leaf node n2 p0 sample\" transform=\"translate(-13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">2:A</text></g><g class=\"a20 i1 leaf node n3 p0 sample\" transform=\"translate(13.4222 19.46)\"><path class=\"edge\" d=\"M 0 0 V -19.46 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">3:A</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 20.1333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">20:A</text></g><path class=\"edge\" d=\"M 0 0 V -116.76 H -40.2667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">28:A</text></g><g class=\"a69 c2 node n68 p2\" transform=\"translate(-40.2667 19.46)\"><g class=\"a68 c2 i12 node n24 p0\" transform=\"translate(-20.1333 97.3)\"><g class=\"a24 i0 leaf node n0 p0 sample\" transform=\"translate(-13.4222 38.92)\"><path class=\"edge\" d=\"M 0 0 V -38.92 H 13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">0:A</text></g><g class=\"a24 i3 leaf node n7 p0 sample\" transform=\"translate(13.4222 38.92)\"><path class=\"edge\" d=\"M 0 0 V -38.92 H -13.4222\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">7:A</text></g><path class=\"edge\" d=\"M 0 0 V -97.3 H 20.1333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">24:A</text></g><g class=\"a68 c1 i31 node n63 p1\" transform=\"translate(20.1333 38.92)\"><g class=\"a63 i20 leaf node n40 p1 sample\" transform=\"translate(0.0 97.3)\"><path class=\"edge\" d=\"M 0 0 V -97.3 H 0.0\"/><rect class=\"sym\" height=\"6\" width=\"6\" x=\"-3\" y=\"-3\"/><text class=\"lab\" transform=\"translate(0 11)\">40:B</text></g><path class=\"edge\" d=\"M 0 0 V -38.92 H -20.1333\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab rgt\" transform=\"translate(3 -7.0)\">63:B</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 40.2667\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">68:C</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 60.4\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">69:C</text></g><path class=\"edge\" d=\"M 0 0 V -19.46 H 134.222\"/><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab lft\" transform=\"translate(-3 -7.0)\">70:C</text></g><circle class=\"sym\" cx=\"0\" cy=\"0\" r=\"3\"/><text class=\"lab\" transform=\"translate(0 -11)\">71:C</text></g></g></g></g></g></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 101,
     "metadata": {
      "filenames": {
       "image/svg+xml": "/home/runner/work/tskit-site/tskit-site/docs/_build/jupyter_execute/ancestry_194_0.svg"
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_ts = msprime.sim_ancestry(\n",
    "    initial_state=ped_ts, demography=demography, random_seed=42, model=\"dtwf\")\n",
    "node_labels = {\n",
    "    node.id: f\"{node.id}:{full_ts.population(node.population).metadata['name']}\"\n",
    "    for node in full_ts.nodes()}\n",
    "SVG(full_ts.draw_svg(\n",
    "    y_axis=True,  node_labels=node_labels, size=(600,300), time_scale=\"rank\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bebbb5b4",
   "metadata": {},
   "source": [
    "Here we have changed the node labelling in the trees to show the name of the\n",
    "population of each node. We can see that all of the early nodes are in the\n",
    "populations defined from the pedigree. When the recapitation simulation begins,\n",
    "all nodes are assigned into the relevant populations and the simulation proceeds\n",
    "from there under the defined demography. In this demographic model, populations\n",
    "A and B split from C 10 generations ago, and so the remaining lineages in populations\n",
    "A and B evolve independently until this time. We can see that older coalescences\n",
    "all happen within population C, as stipulated by the demographic model.\n",
    "\n",
    ":::{warning}\n",
    "The same caveats that we noted\n",
    "{ref}`above<sec_ancestry_models_fixed_pedigree_missing_data>` about missing\n",
    "data in the pedigree apply: although the remaining ancestral lineages\n",
    "will join the simulation at the correct times and in the\n",
    "specified population, *no account* is taken\n",
    "of the number of extant lineages from the pedigree when computing\n",
    "population sizes.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_missing_data)=\n",
    "\n",
    "## Missing data\n",
    "\n",
    ":::{todo}\n",
    "Give an example of doing a simulation with unknown rates on the flanks\n",
    "and show that the resulting tree sequence has missing data in these\n",
    "regions.\n",
    ":::\n",
    "\n",
    "(sec_ancestry_errors)=\n",
    "\n",
    "## Common errors\n",
    "\n",
    "(sec_ancestry_errors_infinite_waiting_time)=\n",
    "\n",
    "### Infinite waiting time\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: explain this, why it happens and give examples of when\n",
    "   we don't detect it. Mention the possible_lineage_locations method.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "b2758939",
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "ename": "LibraryError",
     "evalue": "Infinite waiting time until next simulation event.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLibraryError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [102], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m demography\u001b[38;5;241m.\u001b[39madd_population(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m\"\u001b[39m, initial_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      3\u001b[0m demography\u001b[38;5;241m.\u001b[39madd_population(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mB\u001b[39m\u001b[38;5;124m\"\u001b[39m, initial_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m----> 4\u001b[0m \u001b[43mmsprime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msim_ancestry\u001b[49m\u001b[43m(\u001b[49m\u001b[43msamples\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mA\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mB\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdemography\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdemography\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1226\u001b[0m, in \u001b[0;36msim_ancestry\u001b[0;34m(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, num_labels, random_seed, num_replicates, replicate_index, record_provenance)\u001b[0m\n\u001b[1;32m   1206\u001b[0m     provenance_dict \u001b[38;5;241m=\u001b[39m provenance\u001b[38;5;241m.\u001b[39mget_provenance_dict(parameters)\n\u001b[1;32m   1207\u001b[0m sim \u001b[38;5;241m=\u001b[39m _parse_sim_ancestry(\n\u001b[1;32m   1208\u001b[0m     samples\u001b[38;5;241m=\u001b[39msamples,\n\u001b[1;32m   1209\u001b[0m     sequence_length\u001b[38;5;241m=\u001b[39msequence_length,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1224\u001b[0m     random_seed\u001b[38;5;241m=\u001b[39mrandom_seed,\n\u001b[1;32m   1225\u001b[0m )\n\u001b[0;32m-> 1226\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_wrap_replicates\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1227\u001b[0m \u001b[43m    \u001b[49m\u001b[43msim\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1228\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_replicates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_replicates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1229\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreplicate_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreplicate_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1230\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprovenance_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprovenance_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1231\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:660\u001b[0m, in \u001b[0;36m_wrap_replicates\u001b[0;34m(simulator, num_replicates, replicate_index, provenance_dict, mutation_rate)\u001b[0m\n\u001b[1;32m    654\u001b[0m iterator \u001b[38;5;241m=\u001b[39m simulator\u001b[38;5;241m.\u001b[39mrun_replicates(\n\u001b[1;32m    655\u001b[0m     num_replicates,\n\u001b[1;32m    656\u001b[0m     mutation_rate\u001b[38;5;241m=\u001b[39mmutation_rate,\n\u001b[1;32m    657\u001b[0m     provenance_dict\u001b[38;5;241m=\u001b[39mprovenance_dict,\n\u001b[1;32m    658\u001b[0m )\n\u001b[1;32m    659\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m replicate_index \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 660\u001b[0m     deque \u001b[38;5;241m=\u001b[39m \u001b[43mcollections\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdeque\u001b[49m\u001b[43m(\u001b[49m\u001b[43miterator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmaxlen\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    661\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m deque\u001b[38;5;241m.\u001b[39mpop()\n\u001b[1;32m    662\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1497\u001b[0m, in \u001b[0;36mSimulator.run_replicates\u001b[0;34m(self, num_replicates, mutation_rate, provenance_dict)\u001b[0m\n\u001b[1;32m   1495\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m replicate_index \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(num_replicates):\n\u001b[1;32m   1496\u001b[0m     logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mStarting replicate \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, replicate_index)\n\u001b[0;32m-> 1497\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1498\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m mutation_rate \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1499\u001b[0m         \u001b[38;5;66;03m# This is only called from simulate() or the ms interface,\u001b[39;00m\n\u001b[1;32m   1500\u001b[0m         \u001b[38;5;66;03m# so does not need any further parameters.\u001b[39;00m\n\u001b[1;32m   1501\u001b[0m         mutations\u001b[38;5;241m.\u001b[39m_simple_mutate(\n\u001b[1;32m   1502\u001b[0m             tables\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtables,\n\u001b[1;32m   1503\u001b[0m             random_generator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrandom_generator,\n\u001b[1;32m   1504\u001b[0m             sequence_length\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msequence_length,\n\u001b[1;32m   1505\u001b[0m             rate\u001b[38;5;241m=\u001b[39mmutation_rate,\n\u001b[1;32m   1506\u001b[0m         )\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1462\u001b[0m, in \u001b[0;36mSimulator.run\u001b[0;34m(self, event_chunk, debug_func)\u001b[0m\n\u001b[1;32m   1460\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mModel durations must be >= 0\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1461\u001b[0m end_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtime \u001b[38;5;241m+\u001b[39m model_duration, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mend_time)\n\u001b[0;32m-> 1462\u001b[0m exit_reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_until\u001b[49m\u001b[43m(\u001b[49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mevent_chunk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdebug_func\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1463\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m exit_reason \u001b[38;5;241m==\u001b[39m ExitReason\u001b[38;5;241m.\u001b[39mCOALESCENCE \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtime \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mend_time:\n\u001b[1;32m   1464\u001b[0m     logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSkipping remaining \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m models\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodels) \u001b[38;5;241m-\u001b[39m j \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m~/work/tskit-site/tskit-site/msprime/ancestry.py:1427\u001b[0m, in \u001b[0;36mSimulator._run_until\u001b[0;34m(self, end_time, event_chunk, debug_func)\u001b[0m\n\u001b[1;32m   1425\u001b[0m ret \u001b[38;5;241m=\u001b[39m ExitReason\u001b[38;5;241m.\u001b[39mMAX_EVENTS\n\u001b[1;32m   1426\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m ret \u001b[38;5;241m==\u001b[39m ExitReason\u001b[38;5;241m.\u001b[39mMAX_EVENTS:\n\u001b[0;32m-> 1427\u001b[0m     ret \u001b[38;5;241m=\u001b[39m ExitReason(\u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mend_time\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mevent_chunk\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   1428\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtime \u001b[38;5;241m>\u001b[39m end_time:\n\u001b[1;32m   1429\u001b[0m         \u001b[38;5;66;03m# Currently the Pedigree and Sweeps models are \"non-reentrant\"\u001b[39;00m\n\u001b[1;32m   1430\u001b[0m         \u001b[38;5;66;03m# We can change this to an assertion once these have been fixed.\u001b[39;00m\n\u001b[1;32m   1431\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1432\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mModel \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m does not support interruption. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1433\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPlease open an issue on GitHub\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1434\u001b[0m         )\n",
      "\u001b[0;31mLibraryError\u001b[0m: Infinite waiting time until next simulation event."
     ]
    }
   ],
   "source": [
    "demography = msprime.Demography()\n",
    "demography.add_population(name=\"A\", initial_size=1)\n",
    "demography.add_population(name=\"B\", initial_size=1)\n",
    "msprime.sim_ancestry(samples={\"A\": 1, \"B\": 1}, demography=demography)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.14"
  },
  "source_map": [
   15,
   44,
   62,
   65,
   138,
   141,
   169,
   177,
   185,
   193,
   226,
   236,
   250,
   260,
   282,
   291,
   353,
   358,
   361,
   369,
   372,
   406,
   410,
   440,
   450,
   458,
   487,
   490,
   495,
   498,
   520,
   524,
   528,
   531,
   535,
   539,
   556,
   560,
   586,
   591,
   598,
   602,
   610,
   617,
   620,
   626,
   634,
   637,
   655,
   661,
   664,
   669,
   676,
   679,
   795,
   816,
   822,
   841,
   853,
   869,
   887,
   899,
   909,
   919,
   962,
   968,
   1009,
   1013,
   1037,
   1049,
   1057,
   1084,
   1094,
   1143,
   1162,
   1173,
   1183,
   1185,
   1194,
   1197,
   1212,
   1217,
   1222,
   1228,
   1237,
   1243,
   1256,
   1259,
   1271,
   1275,
   1300,
   1303,
   1323,
   1326,
   1330,
   1333,
   1351,
   1370,
   1434,
   1437,
   1448,
   1452,
   1488,
   1493,
   1510,
   1513,
   1522,
   1525,
   1539,
   1546,
   1548,
   1555,
   1560,
   1567,
   1583,
   1586,
   1594,
   1598,
   1604,
   1610,
   1708,
   1715,
   1721,
   1729,
   1746,
   1753,
   1766,
   1774,
   1806,
   1821,
   1828,
   1836,
   1865,
   1875,
   1920,
   1925,
   1934,
   1937,
   1945,
   1948,
   1992,
   1995,
   2032,
   2043,
   2081,
   2088,
   2095,
   2102,
   2107,
   2114,
   2132,
   2145,
   2149,
   2162,
   2166,
   2173,
   2177,
   2184,
   2198,
   2211,
   2221,
   2227,
   2234,
   2240,
   2298,
   2311,
   2324,
   2333,
   2349,
   2355,
   2360,
   2368,
   2397,
   2417,
   2431,
   2441,
   2481,
   2502,
   2512,
   2518,
   2549,
   2555,
   2570,
   2574,
   2592,
   2612,
   2617,
   2623,
   2653,
   2659,
   2665,
   2706,
   2713,
   2718,
   2725,
   2733,
   2776
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}