<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Inference overview &mdash; tsinfer 0.2.3 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API Documentation" href="api.html" />
    <link rel="prev" title="Tutorial" href="tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> tsinfer
          </a>
              <div class="version">
                0.2
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Inference overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-requirements">Data requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#data-model">Data model</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#individual">Individual</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sample">Sample</a></li>
<li class="toctree-l3"><a class="reference internal" href="#population">Population</a></li>
<li class="toctree-l3"><a class="reference internal" href="#metadata">Metadata</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#import-samples-data">Import samples data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generate-ancestors">Generate ancestors</a></li>
<li class="toctree-l2"><a class="reference internal" href="#matching-ancestors-samples">Matching ancestors &amp; samples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#match-ancestors">Match ancestors</a></li>
<li class="toctree-l3"><a class="reference internal" href="#match-samples">Match samples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="file_formats.html">File formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Developer documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">tsinfer</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Inference overview</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/inference.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="inference-overview">
<span id="sec-inference"></span><h1>Inference overview<a class="headerlink" href="#inference-overview" title="Permalink to this headline"></a></h1>
<p>The process of inferring a tree sequence from variation data is split into a
number of different steps. We first review the
<a class="reference internal" href="#sec-inference-data-requirements"><span class="std std-ref">requirements</span></a> for input data
and discuss how such data is <a class="reference internal" href="#sec-inference-import-samples"><span class="std std-ref">imported</span></a>
into <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>’s sample data format. We then outline the three
basic steps for inference:
<a class="reference internal" href="#sec-inference-generate-ancestors"><span class="std std-ref">generating ancestors</span></a>,
<a class="reference internal" href="#sec-inference-match-ancestors"><span class="std std-ref">matching ancestors</span></a> and
<a class="reference internal" href="#sec-inference-match-samples"><span class="std std-ref">matching samples</span></a>.</p>
<section id="data-requirements">
<span id="sec-inference-data-requirements"></span><h2>Data requirements<a class="headerlink" href="#data-requirements" title="Permalink to this headline"></a></h2>
<p>Input haplotype data for tskit must satisfy the following requirements:</p>
<ul class="simple">
<li><p>Data must be <em>phased</em>.</p></li>
<li><p>For each site used for inference we must know the <em>ancestral state</em>. Note that
this is not necessarily the same as the REF column from VCF, and ancestral
states must be computed using some other method.</p></li>
<li><p>Only biallelic sites can be used for inference.</p></li>
<li><p>Missing data can be included in the haplotypes using the value
<code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code> (-1). The inferred tree sequence will have values
imputed for the haplotypes at these sites.</p></li>
</ul>
</section>
<section id="data-model">
<span id="sec-inference-data-model"></span><h2>Data model<a class="headerlink" href="#data-model" title="Permalink to this headline"></a></h2>
<p>The data model for <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> is tightly integrated with
<code class="docutils literal notranslate"><span class="pre">tskit</span></code>’s <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-data-model" title="(in Python)"><span class="xref std std-ref">data model</span></a>
and uses the same concepts throughout. The intermediate file formats and APIs
described here provide a bridge between this model and existing data sources. For
convenience, we provide a brief description of concepts needed for importing
data into <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> here. Please see the <a class="reference external" href="https://tskit.readthedocs.io/">tskit documentation</a> for more detailed information.</p>
<section id="individual">
<span id="sec-inference-data-model-individual"></span><h3>Individual<a class="headerlink" href="#individual" title="Permalink to this headline"></a></h3>
<p>In <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> an individual defines one of the subjects for which we have
genotype data. Individuals may have arbitrary ploidy levels (i.e. haploid,
diploid, tetraploid, etc.). Different individuals within a dataset can have
different ploidy levels. You can add an individual, and specify arbitrary
<a class="reference internal" href="#sec-inference-data-model-metadata"><span class="std std-ref">Metadata</span></a> for it, using the
<a class="reference internal" href="api.html#tsinfer.SampleData.add_individual" title="tsinfer.SampleData.add_individual"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SampleData.add_individual()</span></code></a> method; if you do not do so <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>
will create a default set of haploid individuals, one for each sampled genome.</p>
<p>The tree sequence that you infer from your data will contain sample nodes
(i.e. genomes) that are linked to these individuals: the <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
documentation provides more detail on the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/glossary.html#sec-nodes-or-individuals" title="(in Python)"><span class="xref std std-ref">distinction between individuals and the genomes they contain</span></a>.</p>
</section>
<section id="sample">
<span id="sec-inference-data-model-sample"></span><h3>Sample<a class="headerlink" href="#sample" title="Permalink to this headline"></a></h3>
<p>Each individual is composed of one or more <code class="docutils literal notranslate"><span class="pre">sample</span></code> genomes, depending on their
ploidy. If an individual is diploid they have two samples, one each for the
maternal and paternal chromosome copies. More generally, a <code class="docutils literal notranslate"><span class="pre">node</span></code> refers
to a maternal or paternal chromosome that is either a sample, or an
ancestor of our samples.</p>
<p>When we add an individual with ploidy <code class="docutils literal notranslate"><span class="pre">k</span></code> using
<a class="reference internal" href="api.html#tsinfer.SampleData.add_individual" title="tsinfer.SampleData.add_individual"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SampleData.add_individual()</span></code></a>, <code class="docutils literal notranslate"><span class="pre">k</span></code> new samples are also added
which refer to this new individual. When adding genotype information using the
<a class="reference internal" href="api.html#tsinfer.SampleData.add_site" title="tsinfer.SampleData.add_site"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SampleData.add_site()</span></code></a> method, the user must ensure that the observed
genotypes are in this same order.</p>
</section>
<section id="population">
<span id="sec-inference-data-model-population"></span><h3>Population<a class="headerlink" href="#population" title="Permalink to this headline"></a></h3>
<p>A population is some grouping of individuals. Populations principally
exist to allow us define metadata for groups of individuals (although
technically, population IDs are associated with samples).
Populations are added using the <a class="reference internal" href="api.html#tsinfer.SampleData.add_population" title="tsinfer.SampleData.add_population"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SampleData.add_population()</span></code></a>
method.</p>
</section>
<section id="metadata">
<span id="sec-inference-data-model-metadata"></span><h3>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline"></a></h3>
<p>Metadata can be associated with populations and individuals in <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>,
which results in this information being available in the final tree
sequence. Metadata allows us to incorporate extra information
that is not part of the basic data model; for example, we can record
the upstream ID of a given individual and their family relationships
with other individuals.</p>
<p>In <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>, metadata can be stored by providing a JSON encodable
mapping. This information is then stored as JSON, and embedded in the
final tree sequence object and can be recovered using the <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
APIs.</p>
</section>
</section>
<section id="import-samples-data">
<span id="sec-inference-import-samples"></span><h2>Import samples data<a class="headerlink" href="#import-samples-data" title="Permalink to this headline"></a></h2>
<p>In <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> we make several passes through the input sample haplotypes
in order to construct ancestors and to find copying paths for samples. To
do this efficiently we store the data using the <a class="reference external" href="http://zarr.readthedocs.io">zarr library</a>, which provides very fast access to
large arrays of numerical data compressed using cutting-edge
<a class="reference external" href="http://numcodecs.readthedocs.io">compression methods</a>. As a result, we
can store the input sample haplotypes and related metadata in a
fraction of the size of a compressed VCF as well as process it efficiently.</p>
<p>Rather than require the user to understand the internal structure of this
file format, we provide a simple <a class="reference internal" href="api.html#sec-api-file-formats"><span class="std std-ref">Python API</span></a>
to allow the user to efficiently construct it from their own data.
An example of how to use this API is given in the <a class="reference internal" href="tutorial.html#sec-tutorial"><span class="std std-ref">Tutorial</span></a>.</p>
<p>We do not provide an automatic means of importing data from VCF (or any
other format) intentionally, as we believe that this would be extremely difficult to do.
As there is no universally accepted way of encoding ancestral state
information in VCF, in practise the user would most often have to write
a new VCF file with ancestral state and metadata information in a specific
form that we would require. Thus, it is more efficient to skip this intermediate
step and to directly produce a <a class="reference internal" href="file_formats.html#sec-file-formats-samples"><span class="std std-ref">format</span></a>
that is both compact and very efficient to process.</p>
</section>
<section id="generate-ancestors">
<span id="sec-inference-generate-ancestors"></span><h2>Generate ancestors<a class="headerlink" href="#generate-ancestors" title="Permalink to this headline"></a></h2>
<p>The first step in a <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> inference process is to generate a large
number of potential ancestors and to store these in an
<a class="reference internal" href="file_formats.html#sec-file-formats-ancestors"><span class="std std-ref">ancestors file</span></a>. The ancestors
file conventionally ends with <code class="docutils literal notranslate"><span class="pre">.ancestors</span></code>.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Describe the ancestor generation algorithm.</p>
</div>
</section>
<section id="matching-ancestors-samples">
<span id="sec-inference-match-ancestors-and-samples"></span><h2>Matching ancestors &amp; samples<a class="headerlink" href="#matching-ancestors-samples" title="Permalink to this headline"></a></h2>
<p>After we have generated a set of potential ancestors and stored them in
an <a class="reference internal" href="file_formats.html#sec-file-formats-ancestors"><span class="std std-ref">ancestors file</span></a>, we then
run two matching steps. First we match the ancestors against each
other to generate an “ancestors tree sequence”, then we match the samples
against this ancestors tree sequence to generate the final result.</p>
<p>In both matching stages, we can set parameters that adjust the
behaviour of the matching algorithm, in particular the <code class="docutils literal notranslate"><span class="pre">path_compression</span></code>
setting, and the <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> and <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> parameters.
The latter two only need to be specified if you wish to allow multiple
mutations to occur at a single site (i.e. breaking the infinite sites model
of mutation). Note, however, that multiple mutations are useful not only to
show true recurrent or back mutations in the evolutionary history of a
site, but also to represent errors in sequencing etc. which cause the
distribution of variation to fit poorly to the marginal tree at that site.
Hence, if there is error in your dataset, you may wish to experiment with
these settings to obtain optimal results.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> parameter is either a floating point value giving a
single rate (<span class="math notranslate nohighlight">\(\rho\)</span>) per unit length of genome, used to calculate the
genetic distance between adjacent sites, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a> object
(whose <code class="docutils literal notranslate"><span class="pre">.get_cumulative_mass</span></code> method provides the genetic distances instead).
The genetic distances are then used to derive an array of probabilities of
recombination between adjacent sites, (<span class="math notranslate nohighlight">\(r\)</span>) used when assessing the relative
likelihood that a mismatch (and hence an extra mutation) may be responsible
for some patterns of variation at a site.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> parameter is only relevant if a recombination rate has
been provided. It is used to adjust the balance of recombination to multiple
mutations at a site. More specifically, a single probability of mismatch is
used for all sites, which is calculated from the median genetic distance between
inference sites, such that for conventional (small) distances, the probability of
a mismatch is approximately equal to the probability of recombination multiplied
by the <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code>. In other words, a mismatch ratio of 2 makes a recurrent
mutation two times more likely than a recombination event to explain why an
otherwise closely matching ancestral haplotype does not match at a particular site.
Setting a high <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> therefore results in tree sequences with more
recurrent mutations and fewer recombinations (and edges). Setting a low value
results in tree sequences with more recombination events and edges, and fewer
mutations. In the limit, as the mismatch_ratio tends to zero, only one mutation
will be inferred per variable site. This is the default behaviour if no
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is given or if there is only one inference site.
Alternatively, if <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set, <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> defaults to
1, which has been shown to give reasonable results in simulated inference of
human-like data with error.  As a rough guide, such simulations recommend
mismatch ratios between 1e-3 and 1e3.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">path_compression</span></code> setting is used to further minimise recombination events
by looking for <em>shared recombination breakpoints</em>. A shared
breakpoint exists if a set of children share a breakpoint in the same position,
and they also have identical parents to the left of the breakpoint and identical
parents to the right of the breakpoint. Rather than supposing that these
children experienced multiple identical recombination events in parallel, we can
reduce the number of ancestral recombination events by postulating a “synthetic
ancestor” with this breakpoint, existing at a slightly older point
in time, from whom all the children are descended at this genomic position. We
call the algorithm used to implement this addition to the ancestral copying
paths, “path compression”.</p>
<section id="match-ancestors">
<span id="sec-inference-match-ancestors"></span><h3>Match ancestors<a class="headerlink" href="#match-ancestors" title="Permalink to this headline"></a></h3>
<p>Matching ancestors is dependent on the time allocated to each ancestor; an
ancestor can only copy from any older ancestor. For each ancestor,
we find the most likely path through older ancestors: that is the path that
maximises the product of the probabilities of recombination and mismatch
over all sites.</p>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>Schematic of the ancestors copying process.</p>
</div>
<p>The copying path for each ancestor then describes its ancestry at every
point in the sequence: from a genealogical perspective, we know its
parent node. This information is encoded precisely as an <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-table-definition" title="(in Python)"><span class="xref std std-ref">edge</span></a> in a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-data-model" title="(in Python)"><span class="xref std std-ref">tree sequence</span></a>.
Thus, we refer to the output of this step as the “ancestors tree sequence”,
which is conventionally stored in a file ending with <code class="docutils literal notranslate"><span class="pre">.ancestors.trees</span></code>.</p>
</section>
<section id="match-samples">
<span id="sec-inference-match-samples"></span><h3>Match samples<a class="headerlink" href="#match-samples" title="Permalink to this headline"></a></h3>
<p>The final phase of a <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> inference consists of a number steps:</p>
<ol class="arabic simple">
<li><p>The first (and usually most time-consuming) is to find copying paths
for our sample haplotypes through the ancestors. Each copying path
corresponds to a set of tree sequence edges in precisely the same
way as for ancestors, and the path compression algorithm can be equally
applied here.</p></li>
<li><p>As we only use a subset of the available sites for inference
(excluding by default any sites that are fixed or singletons)
we then place mutations on the inferred trees in order to
represent the information at these sites. This is done using
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.map_mutations" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.Tree.map_mutations()</span></code></a>.</p></li>
<li><p>Remove ancestral paths that do not lead to any of the samples by
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplifying</span></code></a>
the final tree sequence. When simplifying, we keep non-branching (“unary”)
nodes, as they indicate ancestors which we have actively inferred, and
for technical reasons keeping unary ancestors can also lead to better
compression. Note that this means that not every internal node in the
inferred tree sequence will correspond to a coalescent event.</p></li>
</ol>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<ol class="arabic simple">
<li><p>Describe path compression here and above in the ancestors
section</p></li>
<li><p>Describe the structure of the output tree sequences; how the
nodes are mapped, what the time values mean, etc.</p></li>
</ol>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial.html" class="btn btn-neutral float-left" title="Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api.html" class="btn btn-neutral float-right" title="API Documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2018, Jerome Kelleher.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>