
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vignette: Parallelizing SLiM simulations in a phylogenetic tree &#8212; PySLiM manual</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/tskit-book-theme.css?digest=4e121f05f62280197eeb3e8760a1903920f0778f" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/scripts/tskit-book-theme.js?digest=9a6366feaade5ec470560844be9ff19788eb9b85"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Time units" href="time_units.html" />
    <link rel="prev" title="Vignette: Starting with diversity generated by coalescent simulation" href="vignette_coalescent_diversity.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">PySLiM manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction.html">
   Introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Using pyslim
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_space.html">
   Vignette: A spatial simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_continuing.html">
   Vignette: Following up with more coalescent simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_coalescent_diversity.html">
   Vignette: Starting with diversity generated by coalescent simulation
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Vignette: Parallelizing SLiM simulations in a phylogenetic tree
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="time_units.html">
   Time units
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="metadata.html">
   Metadata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="previous_versions.html">
   Migrating from previous versions of pyslim
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  pyslim reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="python_api.html">
   Python API
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Development
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            pyslim 1.0.1
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right"><label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#simulating-the-branches">
   Simulating the branches
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#putting-it-all-together-unioning-the-tree-sequences">
   Putting it all together: unioning the tree sequences
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Vignette: Parallelizing SLiM simulations in a phylogenetic tree</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#simulating-the-branches">
   Simulating the branches
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#putting-it-all-together-unioning-the-tree-sequences">
   Putting it all together: unioning the tree sequences
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="vignette-parallelizing-slim-simulations-in-a-phylogenetic-tree">
<span id="sec-vignette-parallel"></span><h1>Vignette: Parallelizing SLiM simulations in a phylogenetic tree<a class="headerlink" href="#vignette-parallelizing-slim-simulations-in-a-phylogenetic-tree" title="Permalink to this headline">#</a></h1>
<p>Imagine you want to simulate the evolutionary history of a group. If there is
no migration between any of the branches in your tree, any branches stemming
from the same node can be simulated in parallel (see <a class="reference internal" href="#phylo"><span class="std std-numref">Fig. 7</span></a>).</p>
<div class="figure align-default" id="phylo">
<a class="reference internal image-reference" href="_images/phylo.png"><img alt="_images/phylo.png" src="_images/phylo.png" style="height: 200px;" /></a>
<p class="caption"><span class="caption-number">Fig. 7 </span><span class="caption-text">Example of phylogeny we might want to simulate. Note how branches with the same color can be simulated in parallel when there is no migration.</span><a class="headerlink" href="#phylo" title="Permalink to this image">#</a></p>
</div>
<p>To do this, we’ll need to do two things:
(1) be able to <em>simulate</em> branches in parallel, and
(2) glue the resulting simulations (one per tip) back together.</p>
<div class="section" id="simulating-the-branches">
<h2>Simulating the branches<a class="headerlink" href="#simulating-the-branches" title="Permalink to this headline">#</a></h2>
<p>First, we need to write a SLiM script that will be used for simulating the
history of each branch in our phylogeny.
We will perform a simple simulation, in which each branch can have a different
(but fixed) population size and length (number of ticks).
Also, we will allow deleterious mutations to happen across the entire chromosome
at a fixed rate.</p>
<p>Here is a SLiM script that would do this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">//</span> <span class="n">The</span> <span class="n">following</span> <span class="n">constants</span> <span class="n">need</span> <span class="n">to</span> <span class="n">be</span> <span class="n">defined</span><span class="p">:</span>
<span class="o">//</span> <span class="o">-</span> <span class="n">outfile</span><span class="p">:</span> <span class="n">path</span> <span class="n">to</span> <span class="n">save</span> <span class="n">the</span> <span class="n">tree</span> <span class="n">sequence</span> <span class="n">of</span> <span class="n">the</span> <span class="n">simulation</span><span class="o">.</span>
<span class="o">//</span> <span class="o">-</span> <span class="n">popsize</span><span class="p">:</span> <span class="n">population</span> <span class="n">size</span><span class="o">.</span>
<span class="o">//</span> <span class="o">-</span> <span class="n">num_gens</span><span class="p">:</span> <span class="n">run</span> <span class="n">simulation</span> <span class="k">for</span> <span class="n">num_gens</span> <span class="n">ticks</span><span class="o">.</span>
<span class="o">//</span> <span class="o">-</span> <span class="n">infile</span><span class="p">:</span> <span class="n">a</span> <span class="n">string</span> <span class="k">with</span> <span class="n">path</span> <span class="n">to</span> <span class="n">tree</span> <span class="n">sequence</span> <span class="n">to</span> <span class="n">start</span> <span class="n">from</span><span class="p">;</span> <span class="n">optional</span><span class="o">.</span>
<span class="n">initialize</span><span class="p">()</span>
<span class="p">{</span>
    <span class="n">initializeSLiMModelType</span><span class="p">(</span><span class="s2">&quot;WF&quot;</span><span class="p">);</span>
    <span class="n">initializeTreeSeq</span><span class="p">();</span>
    <span class="n">initializeMutationRate</span><span class="p">(</span><span class="mf">1e-8</span><span class="p">);</span>
    <span class="n">initializeMutationType</span><span class="p">(</span><span class="s2">&quot;m1&quot;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.01</span><span class="p">);</span>
    <span class="n">initializeGenomicElementType</span><span class="p">(</span><span class="s2">&quot;g1&quot;</span><span class="p">,</span> <span class="n">m1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">);</span>
    <span class="n">initializeGenomicElement</span><span class="p">(</span><span class="n">g1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">1e6</span><span class="o">-</span><span class="mi">1</span><span class="p">);</span>
    <span class="n">initializeRecombinationRate</span><span class="p">(</span><span class="mf">1e-9</span><span class="p">);</span>
<span class="p">}</span>

<span class="mi">1</span> <span class="n">late</span><span class="p">()</span> <span class="p">{</span>
    <span class="o">//</span> <span class="k">if</span> <span class="n">no</span> <span class="nb">input</span> <span class="n">tree</span> <span class="n">sequence</span> <span class="ow">is</span> <span class="n">provided</span><span class="p">,</span> <span class="n">then</span> <span class="n">start</span> <span class="n">a</span> <span class="n">subpopulation</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">infile</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">addSubpop</span><span class="p">(</span><span class="s2">&quot;p1&quot;</span><span class="p">,</span> <span class="n">popsize</span><span class="p">);</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="o">//</span> <span class="n">reloading</span> <span class="n">must</span> <span class="n">happen</span> <span class="ow">in</span> <span class="n">late</span><span class="p">()</span>
        <span class="n">sim</span><span class="o">.</span><span class="n">readFromPopulationFile</span><span class="p">(</span><span class="n">infile</span><span class="p">);</span>
        <span class="n">parent</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">subpopulations</span><span class="p">[</span><span class="mi">0</span><span class="p">];</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">addSubpopSplit</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">subpopulations</span><span class="o">.</span><span class="n">id</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">popsize</span><span class="p">,</span> <span class="n">parent</span><span class="p">);</span>
        <span class="n">parent</span><span class="o">.</span><span class="n">setSubpopulationSize</span><span class="p">(</span><span class="mi">0</span><span class="p">);</span>
    <span class="p">}</span>
    <span class="n">p</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">popname</span><span class="p">;</span>
<span class="p">}</span>

<span class="o">//</span> <span class="n">schedule</span> <span class="n">the</span> <span class="n">end</span> <span class="n">of</span> <span class="n">the</span> <span class="n">simulation</span>
<span class="mi">1</span> <span class="n">late</span><span class="p">()</span> <span class="p">{</span>
    <span class="n">finaltick</span> <span class="o">=</span> <span class="n">num_gens</span> <span class="o">+</span> <span class="n">community</span><span class="o">.</span><span class="n">tick</span><span class="p">;</span>
    <span class="n">community</span><span class="o">.</span><span class="n">rescheduleScriptBlock</span><span class="p">(</span><span class="n">s0</span><span class="p">,</span> <span class="n">ticks</span><span class="o">=</span><span class="n">finaltick</span><span class="p">);</span>
<span class="p">}</span>

<span class="o">//</span> <span class="n">event</span> <span class="n">that</span> <span class="n">saves</span> <span class="n">the</span> <span class="n">tree</span> <span class="n">sequence</span>
<span class="n">s0</span> <span class="mi">1000</span> <span class="n">late</span><span class="p">()</span> <span class="p">{</span>
    <span class="n">sim</span><span class="o">.</span><span class="n">treeSeqRememberIndividuals</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">subpopulations</span><span class="o">.</span><span class="n">individuals</span><span class="p">);</span>
    <span class="n">sim</span><span class="o">.</span><span class="n">treeSeqOutput</span><span class="p">(</span><span class="n">outfile</span><span class="p">);</span>
<span class="p">}</span>
</pre></div>
</div>
<p>For each branch, the presence or absence of <code class="docutils literal notranslate"><span class="pre">infile</span></code> tells SLiM
whether we want to start it from a previous branch or not.
If so, SLiM will read the previous tree sequence and change the
population size accordingly.
Note that when you read a tree sequence into SLiM, the tick counter will
be updated with the time encoded in the tree sequence, so we need to set the end
of the simulation as the length of the branch (<code class="docutils literal notranslate"><span class="pre">num_gens</span></code>) plus the current
“time” at the end of the loaded tree sequence.
At the end of the simulation, we call <code class="docutils literal notranslate"><span class="pre">sim.treeSeqRememberIndividuals</span></code> right
before saving the resulting tree sequence. This is necessary because we need to
ensure the individuals in the final generation are never dropped from the tree
sequence in future runs of SLiM which are started from the output of the
simulation, as they will later be used to glue the tree sequences together.</p>
<p>I encoded the phylogeny we will simulate in a simple table,
which we’ll use as <code class="docutils literal notranslate"><span class="pre">df</span></code> in the code below:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;_static/phylo.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;infile&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">parent</span> <span class="o">+</span> <span class="s2">&quot;.trees&quot;</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;outfile&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">child</span> <span class="o">+</span> <span class="s2">&quot;.trees&quot;</span>
<span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;infile&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;.trees&quot;</span><span class="p">,</span> <span class="s2">&quot;infile&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;is_leaf&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">child</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">parent</span><span class="p">)</span>
<span class="n">df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>child</th>
      <th>parent</th>
      <th>popsize</th>
      <th>edgelen</th>
      <th>infile</th>
      <th>outfile</th>
      <th>is_leaf</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>root</td>
      <td></td>
      <td>500</td>
      <td>2000</td>
      <td></td>
      <td>root.trees</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C</td>
      <td>root</td>
      <td>50</td>
      <td>250</td>
      <td>root.trees</td>
      <td>C.trees</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>I</td>
      <td>root</td>
      <td>100</td>
      <td>200</td>
      <td>root.trees</td>
      <td>I.trees</td>
      <td>False</td>
    </tr>
    <tr>
      <th>3</th>
      <td>B</td>
      <td>I</td>
      <td>70</td>
      <td>50</td>
      <td>I.trees</td>
      <td>B.trees</td>
      <td>True</td>
    </tr>
    <tr>
      <th>4</th>
      <td>A</td>
      <td>I</td>
      <td>40</td>
      <td>50</td>
      <td>I.trees</td>
      <td>A.trees</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>With our phylogeny and the simulation parameters, we are ready to run our
simulations.
One way to parallelize the simulation of sister branches is to use <code class="docutils literal notranslate"><span class="pre">make</span></code>.
You do not need to know much about this tool (though it is totally worthwhile
to check it out).
The main idea here is that you can specify dependency between files and <code class="docutils literal notranslate"><span class="pre">make</span></code>
works its magic to run the simulations in the right order.
Here is python code that will write out a makefile from the information in <code class="docutils literal notranslate"><span class="pre">df</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;sims.make&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;all: </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">outfile</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">outfile</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">infile</span><span class="si">}</span><span class="s2"> phylo_bgs.slim&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">slim -d </span><span class="se">\&quot;</span><span class="s2">infile=&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">infile</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\&quot;</span><span class="s2"> -d popsize=</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">popsize</span><span class="si">}</span><span class="s2"> &quot;</span>
          <span class="sa">f</span><span class="s2">&quot;-d </span><span class="se">\&quot;</span><span class="s2">popname=</span><span class="se">\&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">child</span><span class="si">}</span><span class="se">\&#39;\&quot;</span><span class="s2"> &quot;</span>
          <span class="sa">f</span><span class="s2">&quot;-d num_gens=</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">edgelen</span><span class="si">}</span><span class="s2"> &quot;</span> <span class="sa">f</span><span class="s2">&quot;-d </span><span class="se">\&quot;</span><span class="s2">outfile=&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">child</span><span class="si">}</span><span class="s2">.trees&#39;</span><span class="se">\&quot;</span><span class="s2"> &quot;</span>
          <span class="s2">&quot;phylo_bgs.slim</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
          <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>
<span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Here’s the result. Again, don’t worry about the details,
but you can see that the file encodes the phylogeny
through a bunch of <code class="docutils literal notranslate"><span class="pre">child</span> <span class="pre">:</span> <span class="pre">parent</span></code> “rules”:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">bash</span>
cat sims.make
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>all: root.trees C.trees I.trees B.trees A.trees
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>root.trees:  phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	slim -d &quot;infile=&#39;&#39;&quot; -d popsize=500 -d &quot;popname=&#39;root&#39;&quot; -d num_gens=2000 -d &quot;outfile=&#39;root.trees&#39;&quot; phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C.trees: root.trees phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	slim -d &quot;infile=&#39;root.trees&#39;&quot; -d popsize=50 -d &quot;popname=&#39;C&#39;&quot; -d num_gens=250 -d &quot;outfile=&#39;C.trees&#39;&quot; phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>I.trees: root.trees phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	slim -d &quot;infile=&#39;root.trees&#39;&quot; -d popsize=100 -d &quot;popname=&#39;I&#39;&quot; -d num_gens=200 -d &quot;outfile=&#39;I.trees&#39;&quot; phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>B.trees: I.trees phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	slim -d &quot;infile=&#39;I.trees&#39;&quot; -d popsize=70 -d &quot;popname=&#39;B&#39;&quot; -d num_gens=50 -d &quot;outfile=&#39;B.trees&#39;&quot; phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>A.trees: I.trees phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>	slim -d &quot;infile=&#39;I.trees&#39;&quot; -d popsize=40 -d &quot;popname=&#39;A&#39;&quot; -d num_gens=50 -d &quot;outfile=&#39;A.trees&#39;&quot; phylo_bgs.slim
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>With the makefile in hand,
we can now run make, specifying the maximum number of simulations
to be run simultaneously the <code class="docutils literal notranslate"><span class="pre">-j</span></code>.
(Click on the “+” icon to see SLiM’s output.)</p>
<div class="cell tag_hide-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">bash</span>
make -f sims.make -j 3
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>make[1]: Entering directory &#39;/home/runner/work/tskit-site/tskit-site/docs&#39;
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>make[1]: Nothing to be done for &#39;all&#39;.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>make[1]: Leaving directory &#39;/home/runner/work/tskit-site/tskit-site/docs&#39;
</pre></div>
</div>
</div>
</div>
<details class="sphinx-bs dropdown card mb-3">
<summary class="summary-title card-header">
Click here for how to use python instead of make<div class="summary-down docutils">
<svg version="1.1" width="24" height="24" class="octicon octicon-chevron-down" viewBox="0 0 24 24" aria-hidden="true"><path fill-rule="evenodd" d="M5.22 8.72a.75.75 0 000 1.06l6.25 6.25a.75.75 0 001.06 0l6.25-6.25a.75.75 0 00-1.06-1.06L12 14.44 6.28 8.72a.75.75 0 00-1.06 0z"></path></svg></div>
<div class="summary-up docutils">
<svg version="1.1" width="24" height="24" class="octicon octicon-chevron-up" viewBox="0 0 24 24" aria-hidden="true"><path fill-rule="evenodd" d="M18.78 15.28a.75.75 0 000-1.06l-6.25-6.25a.75.75 0 00-1.06 0l-6.25 6.25a.75.75 0 101.06 1.06L12 9.56l5.72 5.72a.75.75 0 001.06 0z"></path></svg></div>
</summary><div class="summary-content card-body docutils">
<p class="card-text">You would have to write a recursion over the branches in your tree (starting
from the root) and then parallelize the runs of sister branches somehow.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">phylo_recursion</span><span class="p">(</span><span class="n">parent</span><span class="p">,</span> <span class="n">df</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">parent</span><span class="p">)</span>
    <span class="n">childs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">parent</span><span class="o">==</span><span class="n">parent</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">childs</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">childs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span>
    <span class="c1"># you could parallelize this loop over childs with same parent</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">childs</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">outfile</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;slim -d </span><span class="se">\&quot;</span><span class="s2">infile=&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">infile</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\&quot;</span><span class="s2"> -d popsize=</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">popsize</span><span class="si">}</span><span class="s2"> -d num_gens=</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">edgelen</span><span class="si">}</span><span class="s2"> -d </span><span class="se">\&quot;</span><span class="s2">outfile=&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">child</span><span class="si">}</span><span class="s2">.trees&#39;</span><span class="se">\&quot;</span><span class="s2"> phylo_bgs.slim&quot;</span><span class="p">)</span>
            <span class="n">phylo_recursion</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">child</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span>

<span class="n">phylo_recursion</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</div>
</details></div>
<div class="section" id="putting-it-all-together-unioning-the-tree-sequences">
<h2>Putting it all together: unioning the tree sequences<a class="headerlink" href="#putting-it-all-together-unioning-the-tree-sequences" title="Permalink to this headline">#</a></h2>
<p>With the tree sequences in hand, we now need to glue them together.
This can be done using
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.union"><strong>union</strong></a>
from tskit.
For two tree sequences which share some of its past history is shared, <strong>union</strong>
works by copying the non-shared parts of one of the tree sequence onto the other.
The trickiest part of this operation is defining the parts that are equivalent
in the two tree sequences. For that, you will have to create an array that serves
as a map of node IDs between the two tree sequences.</p>
<p>Here is a function that will construct a map of the node IDs of two SLiM tree sequences
that correspond to the same chromosomes in SLiM
at any time older than the given time ago at which the two populations split.
Given two tree sequences <code class="docutils literal notranslate"><span class="pre">other</span></code> and <code class="docutils literal notranslate"><span class="pre">ts</span></code>,
the goal here is to find,
for each node born before <code class="docutils literal notranslate"><span class="pre">split_time</span></code> ago in <code class="docutils literal notranslate"><span class="pre">other</span></code>,
the matching node in <code class="docutils literal notranslate"><span class="pre">ts</span></code>, where we can identify matching using the SLiM ID in metadata.
The code could be made easier to read by iterating over nodes,
but the following numpy-based version is much faster:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">match_nodes</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">ts</span><span class="p">,</span> <span class="n">split_time</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given SLiM tree sequences `other` and `ts`, builds a numpy array with length</span>
<span class="sd">    `other.num_nodes` in which the indexes represent the node id in `other` and the</span>
<span class="sd">    entries represent the equivalent node id in `ts`. If a node in `other` has no</span>
<span class="sd">    equivalent in `ts`, then the entry takes the value `tskit.NULL` (-1). The</span>
<span class="sd">    matching is done by comparing the IDs assigned by SLiM which are kept in</span>
<span class="sd">    node metadata. This matching of SLiM IDs is *only* done for nodes with time</span>
<span class="sd">    older than the specified `split_time`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">node_mapping</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">,</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">)</span>
    <span class="n">sids0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">n</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;slim_id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">()])</span>
    <span class="n">sids1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">n</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;slim_id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">other</span><span class="o">.</span><span class="n">nodes</span><span class="p">()])</span>
    <span class="n">alive_before_split1</span> <span class="o">=</span> <span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span> <span class="o">&gt;=</span> <span class="n">split_time</span><span class="p">)</span>
    <span class="n">is_1in0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sids1</span><span class="p">,</span> <span class="n">sids0</span><span class="p">)</span>
    <span class="n">both</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">alive_before_split1</span><span class="p">,</span> <span class="n">is_1in0</span><span class="p">)</span>
    <span class="n">sorted_ids0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">sids0</span><span class="p">)</span>
    <span class="n">matches</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span>
        <span class="n">sids0</span><span class="p">,</span>
        <span class="n">sids1</span><span class="p">[</span><span class="n">both</span><span class="p">],</span>
        <span class="n">side</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span>
        <span class="n">sorter</span><span class="o">=</span><span class="n">sorted_ids0</span>
    <span class="p">)</span>
    <span class="n">node_mapping</span><span class="p">[</span><span class="n">both</span><span class="p">]</span> <span class="o">=</span> <span class="n">sorted_ids0</span><span class="p">[</span><span class="n">matches</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">node_mapping</span>
</pre></div>
</div>
</div>
</div>
<p>Now we are finally ready to <strong>union</strong> our tree sequences. For that, I wrote a
recursive function that goes through our data frame with the phylogeny and
returns a dictionary with the merged tree sequences from the tip to the root.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">merged</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">row</span><span class="o">.</span><span class="n">child</span> <span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;ts&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">outfile</span><span class="p">),</span>
        <span class="s2">&quot;depth&quot;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">edgelen</span><span class="p">,</span>
        <span class="s2">&quot;children&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">child</span><span class="p">]</span>
    <span class="p">}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span>
<span class="p">}</span>

<span class="k">def</span> <span class="nf">union_children</span><span class="p">(</span><span class="n">parent</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">merged</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Going in: </span><span class="si">{</span><span class="n">parent</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">child_rows</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">parent</span> <span class="o">==</span> <span class="n">parent</span><span class="p">]</span>
    <span class="k">assert</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">child_rows</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">childs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">child_rows</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">children</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">child</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">child_rows</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
        <span class="k">for</span> <span class="n">child</span> <span class="ow">in</span> <span class="n">children</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">child</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merged</span><span class="p">:</span>
                <span class="n">union_children</span><span class="p">(</span><span class="n">child</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">merged</span><span class="p">)</span>
        <span class="n">split_time</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="n">children</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span>
        <span class="k">assert</span> <span class="n">split_time</span> <span class="o">==</span> <span class="n">merged</span><span class="p">[</span><span class="n">children</span><span class="p">[</span><span class="mi">1</span><span class="p">]][</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span> <span class="c1"># ultrametric</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Unioning: </span><span class="si">{</span><span class="n">children</span><span class="si">}</span><span class="s1">, Split time: </span><span class="si">{</span><span class="n">split_time</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">ts0</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="n">children</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;ts&quot;</span><span class="p">]</span>
        <span class="n">ts1</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="n">children</span><span class="p">[</span><span class="mi">1</span><span class="p">]][</span><span class="s2">&quot;ts&quot;</span><span class="p">]</span>
        <span class="n">node_map</span> <span class="o">=</span> <span class="n">match_nodes</span><span class="p">(</span><span class="n">ts1</span><span class="p">,</span> <span class="n">ts0</span><span class="p">,</span> <span class="n">split_time</span><span class="p">)</span>
        <span class="n">tsu</span> <span class="o">=</span> <span class="n">ts0</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">ts1</span><span class="p">,</span> <span class="n">node_map</span><span class="p">,</span> <span class="n">check_shared_equality</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># the time from tip to start of simulation is split_time plus the</span>
        <span class="c1"># length of the edge</span>
        <span class="n">parent_edgelength</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">child</span><span class="o">==</span><span class="n">parent</span><span class="p">]</span><span class="o">.</span><span class="n">edgelen</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
        <span class="n">merged</span><span class="p">[</span><span class="n">parent</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ts&quot;</span><span class="p">:</span> <span class="n">tsu</span><span class="p">,</span>
            <span class="s2">&quot;depth&quot;</span><span class="p">:</span> <span class="n">split_time</span> <span class="o">+</span> <span class="n">parent_edgelength</span><span class="p">,</span>
            <span class="s2">&quot;children&quot;</span><span class="p">:</span> <span class="n">merged</span><span class="p">[</span><span class="n">children</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;children&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">merged</span><span class="p">[</span><span class="n">children</span><span class="p">[</span><span class="mi">1</span><span class="p">]][</span><span class="s2">&quot;children&quot;</span><span class="p">]</span>
        <span class="p">}</span>

<span class="n">union_children</span><span class="p">(</span><span class="s2">&quot;root&quot;</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">merged</span><span class="p">)</span>
<span class="c1"># union of all three species tree sequences is in the root.</span>
<span class="n">tsu</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="s2">&quot;root&quot;</span><span class="p">][</span><span class="s2">&quot;ts&quot;</span><span class="p">]</span>
<span class="n">pops</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="s2">&quot;root&quot;</span><span class="p">][</span><span class="s2">&quot;children&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Going in: root
Going in: I
Unioning: [&#39;B&#39;, &#39;A&#39;], Split time: 50
Unioning: [&#39;C&#39;, &#39;I&#39;], Split time: 250
</pre></div>
</div>
</div>
</div>
<p>A slightly tricky thing we had to do there was to make sure we kept track of
which population in the union’ed tree sequence corresponds to
which population in our phylogeny.
Happily, we’ve stored each population’s name in its metadata field,
so it’s easy to match populations in the tree sequence up to what they’re supposed to be.</p>
<p>Let’s make sure we have the right number of present-day samples
in each of the populations. To do this we need to make sure to get
“alive” samples, because recall that we have saved the state of the
population at each species split time.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">alive</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">tsu</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span><span class="p">,</span> <span class="mi">0</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">pop_ids</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">tsu</span><span class="o">.</span><span class="n">populations</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">pop</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pop_ids</span><span class="p">[</span><span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">id</span>

<span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
    <span class="n">pop_samples</span> <span class="o">=</span> <span class="n">tsu</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">pop_ids</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>
    <span class="n">n_samples</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">pop_samples</span><span class="p">,</span> <span class="n">alive</span><span class="p">))</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Union-ed tree sequence has </span><span class="si">{</span><span class="n">n_samples</span><span class="si">}</span><span class="s2"> samples in population </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">,</span><span class="se">\n</span><span class="s2">&quot;</span>
          <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">and we specified </span><span class="si">{</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">child</span><span class="o">==</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">popsize</span><span class="o">.</span><span class="n">item</span><span class="p">()</span><span class="si">}</span><span class="s2"> individuals in our simulations.&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">n_samples</span> <span class="o">==</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">child</span><span class="o">==</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">popsize</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Union-ed tree sequence has 50 samples in population C,
	and we specified 50 individuals in our simulations.
Union-ed tree sequence has 70 samples in population B,
	and we specified 70 individuals in our simulations.
Union-ed tree sequence has 40 samples in population A,
	and we specified 40 individuals in our simulations.
</pre></div>
</div>
</div>
</div>
<p>Let’s do an additional consistency check now, to see if we need to recapitate
(i.e., if some trees haven’t coalesced),
and to make sure that all roots are in the root population,
as they should be:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># TODO: fix up</span>
<span class="c1"># for t in tsu.trees():</span>
<span class="c1">#     for r in t.roots:</span>
<span class="c1">#         assert tsu.node(r).population == pop_ids[&quot;root&quot;]</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Max number of roots: </span><span class="si">{</span><span class="nb">max</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">num_roots</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tsu</span><span class="o">.</span><span class="n">trees</span><span class="p">()])</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Max number of roots: 1.
</pre></div>
</div>
</div>
</div>
<p>Finally, we will recapitate the result with a small population size of 100,
in case some trees on the root branch haven’t coalesced,
and write out the result:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tsu</span> <span class="o">=</span> <span class="n">pyslim</span><span class="o">.</span><span class="n">recapitate</span><span class="p">(</span><span class="n">tsu</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">ancestral_Ne</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">tsu</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="s2">&quot;final.trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/msprime/ancestry.py:831: TimeUnitsMismatchWarning: The initial_state has time_units=ticks but time is measured in generations in msprime. This may lead to significant discrepancies between the timescales. If you wish to suppress this warning, you can use, e.g., warnings.simplefilter(&#39;ignore&#39;, msprime.TimeUnitsMismatchWarning)
  warnings.warn(message, TimeUnitsMismatchWarning)
</pre></div>
</div>
</div>
</div>
<p>Now we’re done, and can analyse the final tree sequence!
Just for fun, I’ll look at the trees produced by the simulation.
For instance, we might be curious how often there are
disagreements between the species tree and the simulated gene trees
(also called incomplete lineage sorting, or ILS).</p>
<p>To make it possible to look at the trees,
I will first simplify the union-ed tree sequence to keep only two diploid
samples per population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span>
<span class="n">ind_alive</span> <span class="o">=</span> <span class="n">pyslim</span><span class="o">.</span><span class="n">individuals_alive_at</span><span class="p">(</span><span class="n">tsu</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="c1"># TODO: this will work in the next tskit</span>
<span class="c1"># ind_pops = tsu.individuals_population[ind_alive]</span>
<span class="n">ind_pops</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">tsu</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">tsu</span><span class="o">.</span><span class="n">individual</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">population</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ind_alive</span><span class="p">])</span>
<span class="n">subsample_indivs</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">ind_alive</span><span class="p">[</span><span class="n">ind_pops</span> <span class="o">==</span> <span class="n">pop_ids</span><span class="p">[</span><span class="n">name</span><span class="p">]],</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">pops</span>
<span class="p">]</span>
<span class="n">subsample_nodes</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">tsu</span><span class="o">.</span><span class="n">individual</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">nodes</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">subsample_indivs</span>
<span class="p">]</span>
<span class="n">tsus</span> <span class="o">=</span> <span class="n">tsu</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">subsample_nodes</span><span class="p">),</span>
        <span class="n">filter_populations</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">pop_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">v</span><span class="p">:</span> <span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">pop_ids</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">tsus</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="p">{</span>
        <span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">pop_labels</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">population</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tsus</span><span class="o">.</span><span class="n">nodes</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">time</span> <span class="o">&gt;</span> <span class="mf">0.0</span>
    <span class="p">},</span>
    <span class="n">x_lim</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2200</span><span class="p">],</span>
    <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">800</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span>
<span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/vignette_parallel_phylo_20_0.svg" src="_images/vignette_parallel_phylo_20_0.svg" /></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A possible gotcha in the code above lies in getting the time units to work out.
Note that in the SLiM script we both save and reload .trees files in the
<code class="docutils literal notranslate"><span class="pre">late()</span></code> stage of the SLiM life cycle. This is important: if we had reloaded the
files in <code class="docutils literal notranslate"><span class="pre">early()</span></code>, then each time we did so the “tskit time” and “SLiM time”
would become one step out of sync. This leads to errors either in union (since
if the time units in the two tree sequences do not match, union will raise an error)
or in recapitate (since recapitate assumes that the “top” of the trees are at
the number of generations ago recorded by SLiM in metadata).</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="vignette_coalescent_diversity.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Vignette: Starting with diversity generated by coalescent simulation</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="time_units.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Time units</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Tskit Developers<br/>
  
      &copy; Copyright 2021.<br/>
    <div class="extra_footer">
      pyslim 1.0.1
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>