
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vignette: Following up with more coalescent simulation &#8212; PySLiM manual</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/tskit-book-theme.css?digest=4e121f05f62280197eeb3e8760a1903920f0778f" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/scripts/tskit-book-theme.js?digest=9a6366feaade5ec470560844be9ff19788eb9b85"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Vignette: Starting with diversity generated by coalescent simulation" href="vignette_coalescent_diversity.html" />
    <link rel="prev" title="Vignette: A spatial simulation" href="vignette_space.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">PySLiM manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction.html">
   Introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Using pyslim
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_space.html">
   Vignette: A spatial simulation
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Vignette: Following up with more coalescent simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_coalescent_diversity.html">
   Vignette: Starting with diversity generated by coalescent simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vignette_parallel_phylo.html">
   Vignette: Parallelizing SLiM simulations in a phylogenetic tree
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="time_units.html">
   Time units
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="metadata.html">
   Metadata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="previous_versions.html">
   Migrating from previous versions of pyslim
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  pyslim reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="python_api.html">
   Python API
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Development
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            pyslim 1.0.1
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right"><label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#positive-selection">
   Positive selection
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#recapitation">
   Recapitation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#continuing-the-simulation">
   Continuing the simulation
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Vignette: Following up with more coalescent simulation</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#positive-selection">
   Positive selection
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#recapitation">
   Recapitation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#continuing-the-simulation">
   Continuing the simulation
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="vignette-following-up-with-more-coalescent-simulation">
<span id="sec-vignette-continuing"></span><h1>Vignette: Following up with more coalescent simulation<a class="headerlink" href="#vignette-following-up-with-more-coalescent-simulation" title="Permalink to this headline">#</a></h1>
<p>Previously, we saw how to use recapitation
to simulate the period <em>before</em> a SLiM simulation
with the coalescent simulator, msprime.
We can do the same thing <em>after</em> a period of SLiM simulation.
To demonstrate this,
below we’ll run a simulation in which</p>
<ol class="simple">
<li><p>A population evolves neutrally for a long time, but then</p></li>
<li><p>it experiences strong positive selection on new mutations for 100 generations, and</p></li>
<li><p>evolves neutrally for another 1000 generations.</p></li>
</ol>
<p>To do this, we’ll simulate step (2) first, with SLiM,
then recapitate to add step (1), and then “continue” the simulation using msprime to add in (3).</p>
<div class="section" id="positive-selection">
<h2>Positive selection<a class="headerlink" href="#positive-selection" title="Permalink to this headline">#</a></h2>
<p>Here’s a SLiM script that has rapid, strong selection acting genome-wide for 20 generations.
It is perhaps not very realistic, but it’s dramatic.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">initialize</span><span class="p">()</span> <span class="p">{</span>
	<span class="n">initializeTreeSeq</span><span class="p">();</span>
	<span class="n">initializeMutationRate</span><span class="p">(</span><span class="mf">1e-8</span><span class="p">);</span>
	<span class="n">initializeMutationType</span><span class="p">(</span><span class="s2">&quot;m1&quot;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;e&quot;</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">);</span>
	<span class="n">initializeGenomicElementType</span><span class="p">(</span><span class="s2">&quot;g1&quot;</span><span class="p">,</span> <span class="n">m1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">);</span>
	<span class="n">initializeGenomicElement</span><span class="p">(</span><span class="n">g1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">99999999</span><span class="p">);</span>
	<span class="n">initializeRecombinationRate</span><span class="p">(</span><span class="mf">1e-8</span><span class="p">);</span>
<span class="p">}</span>

<span class="mi">1</span> <span class="n">early</span><span class="p">()</span> <span class="p">{</span>
	<span class="n">sim</span><span class="o">.</span><span class="n">addSubpop</span><span class="p">(</span><span class="s2">&quot;p1&quot;</span><span class="p">,</span> <span class="mi">10000</span><span class="p">);</span>
<span class="p">}</span>

<span class="mi">20</span> <span class="n">late</span><span class="p">()</span> <span class="p">{</span>
	<span class="n">sim</span><span class="o">.</span><span class="n">treeSeqOutput</span><span class="p">(</span><span class="s2">&quot;rapid_adaptation.trees&quot;</span><span class="p">);</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">bash</span>
slim -s 5 rapid_adaptation.slim
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>// Initial random seed:
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>// RunInitializeCallbacks():
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>initializeTreeSeq();
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>initializeMutationRate(1e-08);
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>initializeMutationType(1, 0.5, &quot;e&quot;, 0.1);
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>initializeGenomicElementType(1, m1, 1);
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>initializeGenomicElement(g1, 0, 99999999);
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>initializeRecombinationRate(1e-08);
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>// Starting run at tick &lt;start&gt;:
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>We can see what happened in the GUI,
but let’s pull some more statistics out of the tree sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;rapid_adaptation.trees&quot;</span><span class="p">)</span>

<span class="c1"># allele frequencies</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">sample_count_stat</span><span class="p">(</span>
                <span class="p">[</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()],</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">/</span><span class="mi">20000</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="s1">&#39;sites&#39;</span><span class="p">,</span>
                <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> segregating sites, of which </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.25</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;are at frequency above 25%, and </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.05</span><span class="p">)</span><span class="si">}</span><span class="s2"> are above 5%.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 66664 segregating sites, of which 4
are at frequency above 25%, and 98 are above 5%.
</pre></div>
</div>
</div>
</div>
<p>The selection was, indeed, strong.</p>
</div>
<div class="section" id="recapitation">
<h2>Recapitation<a class="headerlink" href="#recapitation" title="Permalink to this headline">#</a></h2>
<p>Ok, now let’s do phase (1), recapitating and mutating the result.
We’ll add SLiM mutations with “mutation type” 0
(so in SLiM these would be called type <code class="docutils literal notranslate"><span class="pre">m0</span></code>),
so first we check that all the existing mutations are of a different type.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rts</span> <span class="o">=</span> <span class="n">pyslim</span><span class="o">.</span><span class="n">recapitate</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">ancestral_Ne</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>

<span class="c1"># check type m0 is not used:</span>
<span class="n">mut_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">md</span><span class="p">[</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">md</span> <span class="ow">in</span> <span class="n">mut</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;mutation_list&#39;</span><span class="p">]])</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Keeping </span><span class="si">{</span><span class="n">rts</span><span class="o">.</span><span class="n">num_mutations</span><span class="si">}</span><span class="s2"> existing mutations of type(s) </span><span class="si">{</span><span class="n">mut_types</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
<span class="k">assert</span> <span class="mi">0</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mut_types</span>

<span class="c1"># add type m0 mutations</span>
<span class="n">next_id</span> <span class="o">=</span> <span class="n">pyslim</span><span class="o">.</span><span class="n">next_slim_mutation_id</span><span class="p">(</span><span class="n">rts</span><span class="p">)</span>
<span class="n">rts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span>
            <span class="n">rts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">SLiMMutationModel</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">next_id</span><span class="o">=</span><span class="n">next_id</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">rts</span><span class="o">.</span><span class="n">sample_count_stat</span><span class="p">(</span>
                <span class="p">[</span><span class="n">rts</span><span class="o">.</span><span class="n">samples</span><span class="p">()],</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">/</span><span class="mi">20000</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="s1">&#39;sites&#39;</span><span class="p">,</span>
                <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;After mutation, there are </span><span class="si">{</span><span class="n">rts</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> segregating sites, of which </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.25</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;are at frequency above 25%, and </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.05</span><span class="p">)</span><span class="si">}</span><span class="s2"> are above 5%.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/msprime/ancestry.py:831: TimeUnitsMismatchWarning: The initial_state has time_units=ticks but time is measured in generations in msprime. This may lead to significant discrepancies between the timescales. If you wish to suppress this warning, you can use, e.g., warnings.simplefilter(&#39;ignore&#39;, msprime.TimeUnitsMismatchWarning)
  warnings.warn(message, TimeUnitsMismatchWarning)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Keeping 66693 existing mutations of type(s) {1}.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>After mutation, there are 153526 segregating sites, of which 5824
are at frequency above 25%, and 11643 are above 5%.
</pre></div>
</div>
</div>
</div>
<p>Now, there are more segregating sites - neutral ones.</p>
</div>
<div class="section" id="continuing-the-simulation">
<h2>Continuing the simulation<a class="headerlink" href="#continuing-the-simulation" title="Permalink to this headline">#</a></h2>
<p>To “continue” the simulation neutrally, we’ll</p>
<ol class="simple">
<li><p>simulate the desired period of time in msprime</p></li>
<li><p>randomly match the initial ancestors in the msprime simulation
with the final individuals of the SLim simulation, and</p></li>
<li><p>merge the two together, using the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.union" title="(in Python)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.TreeSequence.union()</span></code></a> method.</p></li>
</ol>
<p><strong>(1)</strong> Simulating for a given period of time in msprime requires the <code class="docutils literal notranslate"><span class="pre">end_time</span></code> argument
(remembering that this is <em>time ago</em>);
we’ll do this to simulate an additional 1000 generations.</p>
<p>This is almost what we need, but there is one more detail:
if complete coalescence occurs on any region of the genome,
msprime will stop simulating the history of that region.
This is a problem, since we need all lineages to extend back to <code class="docutils literal notranslate"><span class="pre">end_time</span></code>.
To make sure all lineages trace back to <code class="docutils literal notranslate"><span class="pre">end_time</span></code>,
we’ll add one “fake” sample from a separate population, that <em>can’t</em> coalesce with the rest,
then remove it before the next step, using the <code class="docutils literal notranslate"><span class="pre">keep_input_roots=True</span></code> argument to <code class="docutils literal notranslate"><span class="pre">simplify()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_time</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">demog_model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demog_model</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">initial_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;real&#39;</span><span class="p">)</span>
<span class="n">demog_model</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">initial_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;fake&#39;</span><span class="p">)</span>
<span class="n">new_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
              <span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;real&#39;</span> <span class="p">:</span> <span class="mi">10000</span><span class="p">,</span> <span class="s1">&#39;fake&#39;</span> <span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
              <span class="n">demography</span><span class="o">=</span><span class="n">demog_model</span><span class="p">,</span>
              <span class="n">end_time</span><span class="o">=</span><span class="n">new_time</span><span class="p">,</span>
              <span class="n">sequence_length</span><span class="o">=</span><span class="n">rts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
              <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
              <span class="n">random_seed</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
<span class="n">new_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span>
                 <span class="n">new_ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                 <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">SLiMMutationModel</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="p">)</span>
<span class="n">new_tables</span> <span class="o">=</span> <span class="n">new_ts</span><span class="o">.</span><span class="n">tables</span>
<span class="c1"># check that the spurious samples are 20000 and 20001</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">20000</span><span class="p">,</span> <span class="mi">20001</span><span class="p">):</span>
   <span class="k">assert</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">new_ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()</span>
   <span class="k">assert</span> <span class="n">new_ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">population</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">new_tables</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">20000</span><span class="p">),</span> <span class="n">keep_input_roots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remaining number of populations: </span><span class="si">{</span><span class="n">new_tables</span><span class="o">.</span><span class="n">populations</span><span class="o">.</span><span class="n">num_rows</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Remaining number of populations: 1
</pre></div>
</div>
</div>
</div>
<p><strong>(2)</strong> Now we’ll pull out the IDs of the nodes from 1000 generations ago,
shift the times in the SLiM tree sequence back 1000 generations,
randomly assign each to a node at the end of the SLiM simulation,
and merge them.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_nodes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">new_tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span> <span class="o">==</span> <span class="n">new_time</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">new_nodes</span><span class="p">)</span><span class="si">}</span><span class="s2"> nodes from the start of the new simulation.&quot;</span><span class="p">)</span>
<span class="c1"># There are 4425 nodes from the start of the new simulation.</span>

<span class="n">slim_nodes</span> <span class="o">=</span> <span class="n">rts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="k">assert</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">slim_nodes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">20000</span><span class="p">)</span>

<span class="c1"># randomly give new_nodes IDs in rts</span>
<span class="n">node_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span> <span class="n">new_tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">num_rows</span><span class="p">)</span>
<span class="n">node_map</span><span class="p">[</span><span class="n">new_nodes</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">slim_nodes</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_nodes</span><span class="p">),</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># shift times: in nodes and mutations</span>
<span class="c1"># since tree sequences are not mutable, we do this in the tables directly</span>
<span class="c1"># also, unmark the nodes at the end of the SLiM simulation as samples</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">rts</span><span class="o">.</span><span class="n">tables</span>
<span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">flags</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">flags</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">(</span><span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span> <span class="o">+</span> <span class="n">new_time</span>
<span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">time</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">time</span> <span class="o">+</span> <span class="n">new_time</span>

<span class="c1"># merge the two sets of tables</span>
<span class="n">tables</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">new_tables</span><span class="p">,</span> <span class="n">node_map</span><span class="p">,</span>
            <span class="n">add_populations</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">check_shared_equality</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># get back the tree sequence</span>
<span class="n">full_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">full_ts</span><span class="o">.</span><span class="n">sample_count_stat</span><span class="p">(</span>
                <span class="p">[</span><span class="n">full_ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()],</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">/</span><span class="mi">20000</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">windows</span><span class="o">=</span><span class="s1">&#39;sites&#39;</span><span class="p">,</span> <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="n">full_ts</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> segregating sites, of which </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.25</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;are at frequency above 25%, and </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">p</span> <span class="o">&gt;</span> <span class="mf">0.05</span><span class="p">)</span><span class="si">}</span><span class="s2"> are above 5%.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 4390 nodes from the start of the new simulation.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>There are 402099 segregating sites, of which 5888
are at frequency above 25%, and 13711 are above 5%.
</pre></div>
</div>
</div>
</div>
<p>Well, allele frequencies have drifted.
Don’t worry, we’ll explain what happened there in a minute.</p>
<p>Let’s do a consistency check.
First, here’s the root of the first tree in the recapitated SLiM simulation:</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span> <span class="o">=</span> <span class="n">rts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="k">assert</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">num_roots</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">rts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">root</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node(id=66051, flags=0, time=2693.221890795138, population=2, individual=-1, metadata=None)
</pre></div>
</div>
</div>
</div>
<p>Now, here’s the root of the first tree <em>after</em> continuing
for 1000 generations, which should be the same:</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ft</span> <span class="o">=</span> <span class="n">full_ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="k">assert</span><span class="p">(</span><span class="n">ft</span><span class="o">.</span><span class="n">num_roots</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">fr</span> <span class="o">=</span> <span class="n">full_ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">ft</span><span class="o">.</span><span class="n">root</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">fr</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node(id=66051, flags=0, time=3693.221890795138, population=2, individual=-1, metadata=None)
</pre></div>
</div>
</div>
</div>
<p>That matches up - the time of what should be the same node in the “continued” tree sequence
is 1000 generations earlier.</p>
<p>So, what happened with <code class="docutils literal notranslate"><span class="pre">union</span></code> back there?
Well, the basic usage is <code class="docutils literal notranslate"><span class="pre">tables.union(other,</span> <span class="pre">node_map)</span></code>,
where <code class="docutils literal notranslate"><span class="pre">node_map</span></code> is an array of length equal to the number of nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code>,
whose entries are either <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> or the ID of a node in <code class="docutils literal notranslate"><span class="pre">tables</span></code>.
The entries that <em>aren’t</em> NULL indicate that
<code class="docutils literal notranslate"><span class="pre">union</span></code> should glue together <code class="docutils literal notranslate"><span class="pre">tables</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code> by saying that that pair of nodes are the same.
(So, e.g., if <code class="docutils literal notranslate"><span class="pre">node_map[3]</span></code> is equal to <code class="docutils literal notranslate"><span class="pre">25</span></code>, then it says that node 25 in <code class="docutils literal notranslate"><span class="pre">tables</span></code>
is actually the same, really, as node 3 in <code class="docutils literal notranslate"><span class="pre">other</span></code>.)
We then asked <code class="docutils literal notranslate"><span class="pre">union</span></code> to please not create new populations,
since otherwise it would have assigned all the new nodes to a new population.
We also asked it to not “check for overlap equality”:
sometimes, when unioning together two tree sequences,
we really expect everything having to do with the set of nodes we’re saying are identical
to be identical in the two tree sequences, so <code class="docutils literal notranslate"><span class="pre">union</span></code> by default throws an error if it’s not.
We don’t expect that in this case, because, for instance,
there could be a mutation above one of the terminal nodes in the SLiM tree sequence;
this would clearly not be present in the new tree sequence.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="vignette_space.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Vignette: A spatial simulation</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="vignette_coalescent_diversity.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Vignette: Starting with diversity generated by coalescent simulation</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Tskit Developers<br/>
  
      &copy; Copyright 2021.<br/>
    <div class="extra_footer">
      pyslim 1.0.1
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>